{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mattiazza/University/biological_data/bd_project/.venv/lib/python3.13/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Align\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import substitution_matrices\n",
    "from Bio.Data import IUPACData\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "'''import cafaeval\n",
    "from cafaeval.evaluation import cafa_eval\n",
    "from cafaeval.parser import obo_parser, gt_parser'''\n",
    "\n",
    "from pathlib import Path\n",
    "import os\n",
    "import ast\n",
    "import h5py\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating the path to the main folder we are going to use to create our datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to train data\n",
    "training_data_path = Path('../data/train')\n",
    "\n",
    "# Path to test data\n",
    "test_data_path = Path('../data/test')\n",
    "\n",
    "# Path to baseline data\n",
    "baseline_data_path = Path('../data/baseline')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training Set\n",
    "\n",
    "In this section we are **creating the Training Set**, fusing togheter the data from:\n",
    "- _train_set.tsv_\n",
    "- _train_ids.txt_\n",
    "- _train.fasta_\n",
    "- _train_embeddings.h5_\n",
    "- _train_protein2ipr.dat_\n",
    "- _go-basic.obo_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_set.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sub_ontology</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0110165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0043226</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID        sub_ontology     GO_term\n",
       "0  P91124  cellular_component  GO:0005575\n",
       "1  P91124  cellular_component  GO:0110165\n",
       "2  P91124  cellular_component  GO:0005737\n",
       "3  P91124  cellular_component  GO:0005622\n",
       "4  P91124  cellular_component  GO:0043226"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_set = pd.read_csv(training_data_path / 'train_set.tsv', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "train_set.rename(columns={'Protein_ID': 'ID', 'aspect' : 'sub_ontology'}, inplace=True)\n",
    "\n",
    "# Display the first few rows of the train set\n",
    "train_set.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_ids.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['P91124', 'Q55DL5', 'O81027', 'Q04418', 'Q8IXT2']\n"
     ]
    }
   ],
   "source": [
    "# Extracting train_ids.txt\n",
    "with open(training_data_path / 'train_ids.txt', 'r') as file:\n",
    "    train_ids = file.read().splitlines()\n",
    "\n",
    "# Display the first few IDs to verify\n",
    "print(train_ids[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set (123969).\n",
      "Proceeding with the analysis.\n"
     ]
    }
   ],
   "source": [
    "if len(train_ids) == len(train_set['ID'].unique()):\n",
    "    print(f\"The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set ({len(train_ids)}).\\n\"\n",
    "          \"Proceeding with the analysis.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train.fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: P91124\n",
      "Name: P91124\n",
      "Description: P91124\n",
      "Number of features: 0\n",
      "Seq('MAAQFYERNTSGMNADRFMTRLTDESTVNTMQRHYWTARQFIRTKLGKKEDEHL...NLL')\n"
     ]
    }
   ],
   "source": [
    "train_fasta_list = list(SeqIO.parse(training_data_path / 'train.fasta', 'fasta'))\n",
    "\n",
    "# Print the first sequence to verify\n",
    "print(train_fasta_list[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's transform it into a DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>num_features</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>P91124</td>\n",
       "      <td>P91124</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O81027</td>\n",
       "      <td>O81027</td>\n",
       "      <td>O81027</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q04418</td>\n",
       "      <td>Q04418</td>\n",
       "      <td>Q04418</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID    name description  num_features  \\\n",
       "0  P91124  P91124      P91124             0   \n",
       "1  Q55DL5  Q55DL5      Q55DL5             0   \n",
       "2  O81027  O81027      O81027             0   \n",
       "3  Q04418  Q04418      Q04418             0   \n",
       "4  Q8IXT2  Q8IXT2      Q8IXT2             0   \n",
       "\n",
       "                                            sequence  \n",
       "0  (M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...  \n",
       "1  (M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...  \n",
       "2  (M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...  \n",
       "3  (M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...  \n",
       "4  (M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract relevant information from SeqRecord\n",
    "train_fasta_dict = [{\n",
    "    'ID': record.id,\n",
    "    'name': record.name,\n",
    "    'description': record.description,\n",
    "    'num_features': len(record.features),\n",
    "    'sequence': record.seq,\n",
    "} for record in train_fasta_list]\n",
    "\n",
    "# Create a DataFrame from the extracted data\n",
    "train_fasta = pd.DataFrame(train_fasta_dict)\n",
    "\n",
    "# Display the DataFrame\n",
    "train_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking if `ID`, `name` and `description` have the same information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 differences between the ID and name columns.\n",
      "We have a total of 0 differences between the ID and description columns.\n"
     ]
    }
   ],
   "source": [
    "# Checking for differences between the ID and name columns\n",
    "diff_id_name = sum(train_fasta['ID'] != train_fasta['name'])\n",
    "\n",
    "# Checking for differences between the ID and description columns\n",
    "diff_id_description = sum(train_fasta['ID'] != train_fasta['description'])\n",
    "\n",
    "print(f\"We have a total of {diff_id_name} differences between the ID and name columns.\\nWe have a total of {diff_id_description} differences between the ID and description columns.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking if `num_features` has value different from 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 sequences with features.\n"
     ]
    }
   ],
   "source": [
    "num_features_values = sum(train_fasta['num_features'] != 0)\n",
    "\n",
    "print(f\"We have a total of {num_features_values} sequences with features.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Nice!**\n",
    "\n",
    "We found that `ID`, `name` and `description` columns are the same. We can remove two of them (`name`, `descritpion`). \n",
    "\n",
    "In the same way we saw that `num_features` is not very informative since it has only 0, let's remove it as well\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>(M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>(M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O81027</td>\n",
       "      <td>(M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q04418</td>\n",
       "      <td>(M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>(M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID                                           sequence\n",
       "0  P91124  (M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...\n",
       "1  Q55DL5  (M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...\n",
       "2  O81027  (M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...\n",
       "3  Q04418  (M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...\n",
       "4  Q8IXT2  (M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ..."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_fasta.drop(columns=['name', 'description', 'num_features'], inplace=True)\n",
    "\n",
    "\n",
    "train_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_embeddings.h5`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>[0.0682, -0.04648, 0.001752, -0.00858, 0.00376...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW4</td>\n",
       "      <td>[0.002447, 0.007053, 0.06445, 0.00727, -0.0335...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>[0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings\n",
       "0  A0A009IHW8  [0.0682, -0.04648, 0.001752, -0.00858, 0.00376...\n",
       "1  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...\n",
       "2  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...\n",
       "3  A0A023FBW4  [0.002447, 0.007053, 0.06445, 0.00727, -0.0335...\n",
       "4  A0A023FBW7  [0.04932, 0.02069, 0.10864, 0.01634, -0.05106,..."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"train_embeddings.h5\"\n",
    "\n",
    "data_list = []\n",
    "\n",
    "with h5py.File(training_data_path / filename, \"r\") as f:\n",
    "    for dataset_name in f.keys():\n",
    "        dataset = f[dataset_name][:]\n",
    "        data_list.append([dataset_name, dataset])\n",
    "\n",
    "train_embeddings = pd.DataFrame(data_list, columns=[\"ID\", \"embeddings\"])\n",
    "\n",
    "train_embeddings.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_protein2ipr.dat`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR000157</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) domain</td>\n",
       "      <td>PS50104</td>\n",
       "      <td>133</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR000157</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) domain</td>\n",
       "      <td>SM00255</td>\n",
       "      <td>134</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR035897</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) dom...</td>\n",
       "      <td>G3DSA:3.40.50.10140</td>\n",
       "      <td>80</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR035897</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) dom...</td>\n",
       "      <td>SSF52200</td>\n",
       "      <td>128</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>IPR006910</td>\n",
       "      <td>Rad21/Rec8-like protein, N-terminal</td>\n",
       "      <td>PF04825</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        ipr                                             domain  \\\n",
       "0  A0A009IHW8  IPR000157  Toll/interleukin-1 receptor homology (TIR) domain   \n",
       "1  A0A009IHW8  IPR000157  Toll/interleukin-1 receptor homology (TIR) domain   \n",
       "2  A0A009IHW8  IPR035897  Toll/interleukin-1 receptor homology (TIR) dom...   \n",
       "3  A0A009IHW8  IPR035897  Toll/interleukin-1 receptor homology (TIR) dom...   \n",
       "4  A0A021WW32  IPR006910                Rad21/Rec8-like protein, N-terminal   \n",
       "\n",
       "              familyID  start  end  \n",
       "0              PS50104    133  266  \n",
       "1              SM00255    134  258  \n",
       "2  G3DSA:3.40.50.10140     80  266  \n",
       "3             SSF52200    128  249  \n",
       "4              PF04825      1  103  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_protein2ipr = pd.read_csv(training_data_path / 'train_protein2ipr.dat', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "train_protein2ipr.columns = ['ID', 'ipr', 'domain', 'familyID', 'start', 'end']\n",
    "\n",
    "# Display the first few rows of the train set\n",
    "train_protein2ipr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train protein2ipr ((1103545, 6)):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>(IPR000157, IPR000157, IPR035897, IPR035897)</td>\n",
       "      <td>(Toll/interleukin-1 receptor homology (TIR) do...</td>\n",
       "      <td>(PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)</td>\n",
       "      <td>(133, 134, 80, 128)</td>\n",
       "      <td>(266, 258, 266, 249)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FDY8</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                                ipr  \\\n",
       "0  A0A009IHW8       (IPR000157, IPR000157, IPR035897, IPR035897)   \n",
       "1  A0A021WW32                  (IPR006910, IPR039781, IPR049589)   \n",
       "2  A0A021WZA4  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "3  A0A023FBW7                                       (IPR045797,)   \n",
       "4  A0A023FDY8                                       (IPR045797,)   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Toll/interleukin-1 receptor homology (TIR) do...   \n",
       "1  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "2  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "3                                 (Evasins Class A,)   \n",
       "4                                 (Evasins Class A,)   \n",
       "\n",
       "                                            familyID  \\\n",
       "0  (PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)   \n",
       "1                      (PF04825, PTHR12585, cd21792)   \n",
       "2  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "3                                         (PF19429,)   \n",
       "4                                         (PF19429,)   \n",
       "\n",
       "                       start                             end  \n",
       "0        (133, 134, 80, 128)            (266, 258, 266, 249)  \n",
       "1                (1, 1, 327)                 (103, 607, 404)  \n",
       "2  (6, 31, 33, 329, 64, 286)  (496, 476, 175, 480, 258, 482)  \n",
       "3                      (38,)                          (110,)  \n",
       "4                      (38,)                          (110,)  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by 'ID' and aggregate other columns into lists\n",
    "train_protein2ipr_grouped = train_protein2ipr.groupby('ID').agg(lambda x: tuple(x)).reset_index()\n",
    "\n",
    "print(f\"Train protein2ipr ({train_protein2ipr.shape}):\")\n",
    "train_protein2ipr_grouped.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Still have to understand what *tizio, caio, sempronio* are."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `go-basic.obo`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All IDs with namespaces:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0048308, GO:0048311]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0007005]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0019952</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0050876</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                              name           namespace  \\\n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "1  GO:0000002  mitochondrial genome maintenance  biological_process   \n",
       "2  GO:0000003                      reproduction  biological_process   \n",
       "3  GO:0019952                      reproduction  biological_process   \n",
       "4  GO:0050876                      reproduction  biological_process   \n",
       "\n",
       "                       is_a part_of  \n",
       "0  [GO:0048308, GO:0048311]      []  \n",
       "1              [GO:0007005]      []  \n",
       "2              [GO:0008150]      []  \n",
       "3              [GO:0008150]      []  \n",
       "4              [GO:0008150]      []  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "file_path = \"../data/train/go-basic.obo\"  # Replace with your file path\n",
    "\n",
    "# Step 1: Initialize storage for GO terms\n",
    "go_terms = []\n",
    "\n",
    "# Step 2: Parse the .obo file\n",
    "with open(file_path, 'r') as file:\n",
    "    current_term = {}\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "        \n",
    "        # Start of a new term\n",
    "        if line == \"[Term]\":\n",
    "            if current_term:  # Save the previous term\n",
    "                go_terms.append(current_term)\n",
    "            current_term = {}  # Start a new term\n",
    "            \n",
    "        elif line.startswith(\"id:\"):\n",
    "            current_term['ID'] = line.split(\"id: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"alt_id:\"):\n",
    "            alt_id = line.split(\"alt_id: \")[1]\n",
    "            current_term.setdefault('alt_ids', []).append(alt_id)\n",
    "            \n",
    "        elif line.startswith(\"name:\"):\n",
    "            current_term['name'] = line.split(\"name: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"namespace:\"):\n",
    "            current_term['namespace'] = line.split(\"namespace: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"is_a\"):\n",
    "            match = re.search(r\"GO:\\d+\", line)  # Search for GO ID\n",
    "            if match:  # Check if a match was found\n",
    "                is_a_id = match.group()\n",
    "                current_term.setdefault('is_a', []).append(is_a_id)\n",
    "                \n",
    "        elif line.startswith(\"relationship: part_of\"):\n",
    "            match = re.search(r\"GO:\\d+\", line)  # Search for GO ID\n",
    "            if match:  # Check if a match was found\n",
    "                part_of_id = match.group()\n",
    "                current_term.setdefault('part_of', []).append(part_of_id)\n",
    "\n",
    "            \n",
    "    # Add the last term\n",
    "    if current_term:\n",
    "        go_terms.append(current_term)\n",
    "\n",
    "# Step 3: Create a unified list of all IDs (primary and alt_ids)\n",
    "expanded_terms = []\n",
    "for term in go_terms:\n",
    "    primary_id = term['ID']\n",
    "    alt_ids = term.get('alt_ids', [])\n",
    "    all_ids = [primary_id] + alt_ids\n",
    "    \n",
    "    for term_id in all_ids:\n",
    "        expanded_terms.append({\n",
    "            'ID': term_id,\n",
    "            'name': term.get('name'),\n",
    "            'namespace': term.get('namespace'),\n",
    "            'is_a': term.get('is_a', []),\n",
    "            'part_of': term.get('part_of', [])\n",
    "        })\n",
    "\n",
    "# Step 4: Convert to a DataFrame\n",
    "df = pd.DataFrame(expanded_terms)\n",
    "\n",
    "# Step 5: Filter by namespace and explode relationships\n",
    "df_is_a = df.explode('is_a').dropna(subset=['is_a'])\n",
    "df_part_of = df.explode('part_of').dropna(subset=['part_of'])\n",
    "\n",
    "\n",
    "print(\"All IDs with namespaces:\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'is_a' relationships:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0048308</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0048311</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0007005</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0019952</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                              name           namespace  \\\n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "1  GO:0000002  mitochondrial genome maintenance  biological_process   \n",
       "2  GO:0000003                      reproduction  biological_process   \n",
       "3  GO:0019952                      reproduction  biological_process   \n",
       "\n",
       "         is_a part_of  \n",
       "0  GO:0048308      []  \n",
       "0  GO:0048311      []  \n",
       "1  GO:0007005      []  \n",
       "2  GO:0008150      []  \n",
       "3  GO:0008150      []  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"'is_a' relationships:\")\n",
    "df_is_a.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'part_of' relationships:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GO:0000015</td>\n",
       "      <td>phosphopyruvate hydratase complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>[GO:1902494]</td>\n",
       "      <td>GO:0005829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GO:0000022</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GO:0000022</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0007052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GO:1905121</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GO:1905121</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0007052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID                               name           namespace  \\\n",
       "16  GO:0000015  phosphopyruvate hydratase complex  cellular_component   \n",
       "22  GO:0000022         mitotic spindle elongation  biological_process   \n",
       "22  GO:0000022         mitotic spindle elongation  biological_process   \n",
       "23  GO:1905121         mitotic spindle elongation  biological_process   \n",
       "23  GO:1905121         mitotic spindle elongation  biological_process   \n",
       "\n",
       "                        is_a     part_of  \n",
       "16              [GO:1902494]  GO:0005829  \n",
       "22  [GO:0051231, GO:1903047]  GO:0000070  \n",
       "22  [GO:0051231, GO:1903047]  GO:0007052  \n",
       "23  [GO:0051231, GO:1903047]  GO:0000070  \n",
       "23  [GO:0051231, GO:1903047]  GO:0007052  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"'part_of' relationships:\")\n",
    "df_part_of.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging all previous extractions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us collect the dataframes and check lengths. Noting that:\n",
    "- _train_set.tsv_ contains the proteins, their GO annotations and their corresponding aspects which will be used to subdivde it into three separate datasets\n",
    "- _train.fasta_ and _train_embeddings.h5_ can be grouped together since they refers to the whole input\n",
    "- _train_protein2ipr.dat_\n",
    "- _go-basic.obo_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: CC (84638, 2), MF (55698, 2), BP (83064, 2)\n",
      "Cellular Component (84638):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0000785, GO:003299...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0071944, GO:000588...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005622, GO:000582...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005576, GO:0005615)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>(GO:0005575, GO:0005737, GO:0042175, GO:003299...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                            GO_term\n",
       "0  A0A021WW32  (GO:0005575, GO:0110165, GO:0000785, GO:003299...\n",
       "1  A0A021WZA4  (GO:0005575, GO:0110165, GO:0071944, GO:000588...\n",
       "2  A0A023GPJ3  (GO:0005575, GO:0110165, GO:0005622, GO:000582...\n",
       "3  A0A023GUT0   (GO:0005575, GO:0110165, GO:0005576, GO:0005615)\n",
       "4  A0A023IM54  (GO:0005575, GO:0005737, GO:0042175, GO:003299..."
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def group_and_combine(df, sub_ontology_value):\n",
    "    return df[df['sub_ontology'] == sub_ontology_value].groupby('ID')['GO_term'].apply(tuple).reset_index()\n",
    "\n",
    "# Create three dataframes for each sub_ontology value\n",
    "df_CC = group_and_combine(train_set, 'cellular_component')\n",
    "df_MF = group_and_combine(train_set, 'molecular_function')\n",
    "df_BP = group_and_combine(train_set, 'biological_process')\n",
    "\n",
    "print(f\"Shapes: CC {df_CC.shape}, MF {df_MF.shape}, BP {df_BP.shape}\")\n",
    "\n",
    "print(f\"Cellular Component ({df_CC.shape[0]}):\")\n",
    "df_CC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows missing from train_protein2ipr_grouped: 6522\n",
      "Combined DataFrame shape: (123969, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>[0.0682, -0.04648, 0.001752, -0.00858, 0.00376...</td>\n",
       "      <td>(M, S, L, E, Q, K, K, G, A, D, I, I, S, K, I, ...</td>\n",
       "      <td>(IPR000157, IPR000157, IPR035897, IPR035897)</td>\n",
       "      <td>(Toll/interleukin-1 receptor homology (TIR) do...</td>\n",
       "      <td>(PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)</td>\n",
       "      <td>(133, 134, 80, 128)</td>\n",
       "      <td>(266, 258, 266, 249)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW4</td>\n",
       "      <td>[0.002447, 0.007053, 0.06445, 0.00727, -0.0335...</td>\n",
       "      <td>(M, T, S, H, G, A, V, K, I, A, I, F, A, V, I, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>[0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...</td>\n",
       "      <td>(M, K, V, L, L, Y, I, A, A, S, C, L, M, L, L, ...</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A009IHW8  [0.0682, -0.04648, 0.001752, -0.00858, 0.00376...   \n",
       "1  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...   \n",
       "2  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...   \n",
       "3  A0A023FBW4  [0.002447, 0.007053, 0.06445, 0.00727, -0.0335...   \n",
       "4  A0A023FBW7  [0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, S, L, E, Q, K, K, G, A, D, I, I, S, K, I, ...   \n",
       "1  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "2  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "3  (M, T, S, H, G, A, V, K, I, A, I, F, A, V, I, ...   \n",
       "4  (M, K, V, L, L, Y, I, A, A, S, C, L, M, L, L, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0       (IPR000157, IPR000157, IPR035897, IPR035897)   \n",
       "1                  (IPR006910, IPR039781, IPR049589)   \n",
       "2  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "3                                                NaN   \n",
       "4                                       (IPR045797,)   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Toll/interleukin-1 receptor homology (TIR) do...   \n",
       "1  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "2  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "3                                                NaN   \n",
       "4                                 (Evasins Class A,)   \n",
       "\n",
       "                                            familyID  \\\n",
       "0  (PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)   \n",
       "1                      (PF04825, PTHR12585, cd21792)   \n",
       "2  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "3                                                NaN   \n",
       "4                                         (PF19429,)   \n",
       "\n",
       "                       start                             end  \n",
       "0        (133, 134, 80, 128)            (266, 258, 266, 249)  \n",
       "1                (1, 1, 327)                 (103, 607, 404)  \n",
       "2  (6, 31, 33, 329, 64, 286)  (496, 476, 175, 480, 258, 482)  \n",
       "3                        NaN                             NaN  \n",
       "4                      (38,)                          (110,)  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_train = pd.merge(train_embeddings, train_fasta, on='ID')\n",
    "combined_train = pd.merge(combined_train, train_protein2ipr_grouped, on='ID', how='left')\n",
    "\n",
    "missing_rows = combined_train[combined_train['ipr'].isna()].shape[0]\n",
    "print(f\"Number of rows missing from train_protein2ipr_grouped: {missing_rows}\")\n",
    "\n",
    "# combined_train = combined_train.drop('domain', axis=1)\n",
    "\n",
    "print(f\"Combined DataFrame shape: {combined_train.shape}\")\n",
    "combined_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: CC (84638, 9), MF (55698, 9), BP (83064, 9)\n",
      "Full df CC (84638, 9):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0000785, GO:003299...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0071944, GO:000588...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>[0.01512, 0.01102, 0.0217, -0.02512, 0.0396, 0...</td>\n",
       "      <td>(M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...</td>\n",
       "      <td>(IPR013087, IPR013087, IPR013087, IPR013087, I...</td>\n",
       "      <td>(Zinc finger C2H2-type, Zinc finger C2H2-type,...</td>\n",
       "      <td>(PF00096, PF00096, PF00096, PF00096, PS00028, ...</td>\n",
       "      <td>(644, 956, 983, 1013, 197, 646, 674, 957, 985,...</td>\n",
       "      <td>(666, 977, 1005, 1034, 218, 666, 694, 977, 100...</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005622, GO:000582...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>[-0.00414, -0.01288, 0.0716, 0.01605, -0.03983...</td>\n",
       "      <td>(M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...</td>\n",
       "      <td>(IPR000065, IPR000065, IPR009079, IPR009079)</td>\n",
       "      <td>(Leptin, Leptin, Four-helical cytokine-like, c...</td>\n",
       "      <td>(PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...</td>\n",
       "      <td>(28, 7, 24, 29)</td>\n",
       "      <td>(172, 172, 172, 172)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005576, GO:0005615)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>[-0.01651, 0.02525, 0.04333, 0.01558, -0.01678...</td>\n",
       "      <td>(M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...</td>\n",
       "      <td>(IPR003006, IPR003597, IPR003597, IPR007110, I...</td>\n",
       "      <td>(Immunoglobulin/major histocompatibility compl...</td>\n",
       "      <td>(PS00290, PF07654, SM00407, PS50835, PR01669, ...</td>\n",
       "      <td>(382, 312, 312, 294, 55, 135, 199, 247, 271, 3...</td>\n",
       "      <td>(388, 391, 394, 401, 63, 146, 216, 261, 287, 3...</td>\n",
       "      <td>(GO:0005575, GO:0005737, GO:0042175, GO:003299...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...   \n",
       "1  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...   \n",
       "2  A0A023GPJ3  [0.01512, 0.01102, 0.0217, -0.02512, 0.0396, 0...   \n",
       "3  A0A023GUT0  [-0.00414, -0.01288, 0.0716, 0.01605, -0.03983...   \n",
       "4  A0A023IM54  [-0.01651, 0.02525, 0.04333, 0.01558, -0.01678...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "1  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "2  (M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...   \n",
       "3  (M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...   \n",
       "4  (M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0                  (IPR006910, IPR039781, IPR049589)   \n",
       "1  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "2  (IPR013087, IPR013087, IPR013087, IPR013087, I...   \n",
       "3       (IPR000065, IPR000065, IPR009079, IPR009079)   \n",
       "4  (IPR003006, IPR003597, IPR003597, IPR007110, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "1  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "2  (Zinc finger C2H2-type, Zinc finger C2H2-type,...   \n",
       "3  (Leptin, Leptin, Four-helical cytokine-like, c...   \n",
       "4  (Immunoglobulin/major histocompatibility compl...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                      (PF04825, PTHR12585, cd21792)   \n",
       "1  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "2  (PF00096, PF00096, PF00096, PF00096, PS00028, ...   \n",
       "3  (PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...   \n",
       "4  (PS00290, PF07654, SM00407, PS50835, PR01669, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                        (1, 1, 327)   \n",
       "1                          (6, 31, 33, 329, 64, 286)   \n",
       "2  (644, 956, 983, 1013, 197, 646, 674, 957, 985,...   \n",
       "3                                    (28, 7, 24, 29)   \n",
       "4  (382, 312, 312, 294, 55, 135, 199, 247, 271, 3...   \n",
       "\n",
       "                                                 end  \\\n",
       "0                                    (103, 607, 404)   \n",
       "1                     (496, 476, 175, 480, 258, 482)   \n",
       "2  (666, 977, 1005, 1034, 218, 666, 694, 977, 100...   \n",
       "3                               (172, 172, 172, 172)   \n",
       "4  (388, 391, 394, 401, 63, 146, 216, 261, 287, 3...   \n",
       "\n",
       "                                             GO_term  \n",
       "0  (GO:0005575, GO:0110165, GO:0000785, GO:003299...  \n",
       "1  (GO:0005575, GO:0110165, GO:0071944, GO:000588...  \n",
       "2  (GO:0005575, GO:0110165, GO:0005622, GO:000582...  \n",
       "3   (GO:0005575, GO:0110165, GO:0005576, GO:0005615)  \n",
       "4  (GO:0005575, GO:0005737, GO:0042175, GO:003299...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge combined_train with df_ab\n",
    "df_CC_full= pd.merge(combined_train, df_CC, on='ID', how='right')\n",
    "df_MF_full= pd.merge(combined_train, df_MF, on='ID', how='right')\n",
    "df_BP_full= pd.merge(combined_train, df_BP, on='ID', how='right')\n",
    "\n",
    "print(f\"Shapes: CC {df_CC_full.shape}, MF {df_MF_full.shape}, BP {df_BP_full.shape}\")\n",
    "\n",
    "print(f\"Full df CC {df_CC_full.shape}:\")\n",
    "df_CC_full.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: x_df_CC (84638, 8), y_df_CC (84638,), x_df_MF (55698, 8), y_df_MF (55698,), x_df_BP (83064, 8), y_df_BP (83064,)\n",
      "CC Input:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>[0.01512, 0.01102, 0.0217, -0.02512, 0.0396, 0...</td>\n",
       "      <td>(M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...</td>\n",
       "      <td>(IPR013087, IPR013087, IPR013087, IPR013087, I...</td>\n",
       "      <td>(Zinc finger C2H2-type, Zinc finger C2H2-type,...</td>\n",
       "      <td>(PF00096, PF00096, PF00096, PF00096, PS00028, ...</td>\n",
       "      <td>(644, 956, 983, 1013, 197, 646, 674, 957, 985,...</td>\n",
       "      <td>(666, 977, 1005, 1034, 218, 666, 694, 977, 100...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>[-0.00414, -0.01288, 0.0716, 0.01605, -0.03983...</td>\n",
       "      <td>(M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...</td>\n",
       "      <td>(IPR000065, IPR000065, IPR009079, IPR009079)</td>\n",
       "      <td>(Leptin, Leptin, Four-helical cytokine-like, c...</td>\n",
       "      <td>(PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...</td>\n",
       "      <td>(28, 7, 24, 29)</td>\n",
       "      <td>(172, 172, 172, 172)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>[-0.01651, 0.02525, 0.04333, 0.01558, -0.01678...</td>\n",
       "      <td>(M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...</td>\n",
       "      <td>(IPR003006, IPR003597, IPR003597, IPR007110, I...</td>\n",
       "      <td>(Immunoglobulin/major histocompatibility compl...</td>\n",
       "      <td>(PS00290, PF07654, SM00407, PS50835, PR01669, ...</td>\n",
       "      <td>(382, 312, 312, 294, 55, 135, 199, 247, 271, 3...</td>\n",
       "      <td>(388, 391, 394, 401, 63, 146, 216, 261, 287, 3...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...   \n",
       "1  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...   \n",
       "2  A0A023GPJ3  [0.01512, 0.01102, 0.0217, -0.02512, 0.0396, 0...   \n",
       "3  A0A023GUT0  [-0.00414, -0.01288, 0.0716, 0.01605, -0.03983...   \n",
       "4  A0A023IM54  [-0.01651, 0.02525, 0.04333, 0.01558, -0.01678...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "1  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "2  (M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...   \n",
       "3  (M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...   \n",
       "4  (M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0                  (IPR006910, IPR039781, IPR049589)   \n",
       "1  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "2  (IPR013087, IPR013087, IPR013087, IPR013087, I...   \n",
       "3       (IPR000065, IPR000065, IPR009079, IPR009079)   \n",
       "4  (IPR003006, IPR003597, IPR003597, IPR007110, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "1  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "2  (Zinc finger C2H2-type, Zinc finger C2H2-type,...   \n",
       "3  (Leptin, Leptin, Four-helical cytokine-like, c...   \n",
       "4  (Immunoglobulin/major histocompatibility compl...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                      (PF04825, PTHR12585, cd21792)   \n",
       "1  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "2  (PF00096, PF00096, PF00096, PF00096, PS00028, ...   \n",
       "3  (PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...   \n",
       "4  (PS00290, PF07654, SM00407, PS50835, PR01669, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                        (1, 1, 327)   \n",
       "1                          (6, 31, 33, 329, 64, 286)   \n",
       "2  (644, 956, 983, 1013, 197, 646, 674, 957, 985,...   \n",
       "3                                    (28, 7, 24, 29)   \n",
       "4  (382, 312, 312, 294, 55, 135, 199, 247, 271, 3...   \n",
       "\n",
       "                                                 end  \n",
       "0                                    (103, 607, 404)  \n",
       "1                     (496, 476, 175, 480, 258, 482)  \n",
       "2  (666, 977, 1005, 1034, 218, 666, 694, 977, 100...  \n",
       "3                               (172, 172, 172, 172)  \n",
       "4  (388, 391, 394, 401, 63, 146, 216, 261, 287, 3...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_df_CC = df_CC_full.iloc[:, :-1]\n",
    "y_df_CC = df_CC_full.iloc[:, -1]\n",
    "\n",
    "x_df_MF = df_MF_full.iloc[:, :-1]\n",
    "y_df_MF = df_MF_full.iloc[:, -1]\n",
    "\n",
    "X_df_BP = df_BP_full.iloc[:, :-1]\n",
    "y_df_BP = df_BP_full.iloc[:, -1]\n",
    "\n",
    "print(f\"Shapes: x_df_CC {x_df_CC.shape}, y_df_CC {y_df_CC.shape}, x_df_MF {x_df_MF.shape}, y_df_MF {y_df_MF.shape}, x_df_BP {X_df_BP.shape}, y_df_BP {y_df_BP.shape}\")\n",
    "\n",
    "print(\"CC Input:\")\n",
    "x_df_CC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CC Target:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    (GO:0005575, GO:0110165, GO:0000785, GO:003299...\n",
       "1    (GO:0005575, GO:0110165, GO:0071944, GO:000588...\n",
       "2    (GO:0005575, GO:0110165, GO:0005622, GO:000582...\n",
       "3     (GO:0005575, GO:0110165, GO:0005576, GO:0005615)\n",
       "4    (GO:0005575, GO:0005737, GO:0042175, GO:003299...\n",
       "Name: GO_term, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"CC Target:\")\n",
    "y_df_CC.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test set\n",
    "\n",
    "In this section we are **creating the Test set**, fusing togheter the data from:\n",
    "- _test_ids.txt_\n",
    "- _test.fasta_\n",
    "- _test_embeddings.h5_\n",
    "- _test_protein2ipr.dat_\n",
    "- _blast_test_results.tsv_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_ids.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['O43747', 'Q969H0', 'Q9JMA2', 'P18065', 'A0A8I6AN32']\n"
     ]
    }
   ],
   "source": [
    "# Extracting test_ids.txt\n",
    "with open(test_data_path / 'test_ids.txt', 'r') as file:\n",
    "    test_ids = file.read().splitlines()\n",
    "\n",
    "# Display the first few IDs to verify\n",
    "print(test_ids[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test.fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: O43747\n",
      "Name: O43747\n",
      "Description: O43747\n",
      "Number of features: 0\n",
      "Seq('MPAPIRLRELIRTIRTARTQAEEREMIQKECAAIRSSFREEDNTYRCRNVAKLL...SWQ')\n"
     ]
    }
   ],
   "source": [
    "test_fasta_list = list(SeqIO.parse(test_data_path / 'test.fasta', 'fasta'))\n",
    "\n",
    "# Print the first sequence to verify\n",
    "print(test_fasta_list[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>num_features</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O43747</td>\n",
       "      <td>O43747</td>\n",
       "      <td>O43747</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q969H0</td>\n",
       "      <td>Q969H0</td>\n",
       "      <td>Q969H0</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P18065</td>\n",
       "      <td>P18065</td>\n",
       "      <td>P18065</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        name description  num_features  \\\n",
       "0      O43747      O43747      O43747             0   \n",
       "1      Q969H0      Q969H0      Q969H0             0   \n",
       "2      Q9JMA2      Q9JMA2      Q9JMA2             0   \n",
       "3      P18065      P18065      P18065             0   \n",
       "4  A0A8I6AN32  A0A8I6AN32  A0A8I6AN32             0   \n",
       "\n",
       "                                            sequence  \n",
       "0  (M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...  \n",
       "1  (M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...  \n",
       "2  (M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...  \n",
       "3  (M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...  \n",
       "4  (M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract relevant information from SeqRecord\n",
    "test_fasta_dict = [{\n",
    "    'ID': record.id,\n",
    "    'name': record.name,\n",
    "    'description': record.description,\n",
    "    'num_features': len(record.features),\n",
    "    'sequence': record.seq,\n",
    "} for record in test_fasta_list]\n",
    "\n",
    "# Create a DataFrame from the extracted data\n",
    "test_fasta = pd.DataFrame(test_fasta_dict)\n",
    "\n",
    "# Display the DataFrame\n",
    "test_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check if ID, name and description are the same thing, as well as seeing if num_features has relevant information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 differences between the ID and name columns.\n",
      "We have a total of 0 differences between the ID and description columns.\n"
     ]
    }
   ],
   "source": [
    "# Checking for differences between the ID and name columns\n",
    "diff_id_name = sum(test_fasta['ID'] != test_fasta['name'])\n",
    "\n",
    "# Checking for differences between the ID and description columns\n",
    "diff_id_description = sum(test_fasta['ID'] != test_fasta['description'])\n",
    "\n",
    "print(f\"We have a total of {diff_id_name} differences between the ID and name columns.\\nWe have a total of {diff_id_description} differences between the ID and description columns.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 sequences with features.\n"
     ]
    }
   ],
   "source": [
    "num_features_values = sum(test_fasta['num_features'] != 0)\n",
    "\n",
    "print(f\"We have a total of {num_features_values} sequences with features.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's drop the useless columns, as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O43747</td>\n",
       "      <td>(M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q969H0</td>\n",
       "      <td>(M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>(M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P18065</td>\n",
       "      <td>(M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>(M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                           sequence\n",
       "0      O43747  (M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...\n",
       "1      Q969H0  (M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...\n",
       "2      Q9JMA2  (M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...\n",
       "3      P18065  (M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...\n",
       "4  A0A8I6AN32  (M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ..."
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_fasta.drop(columns=['name', 'description', 'num_features'], inplace=True)\n",
    "\n",
    "\n",
    "test_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's also check if the IDs are the same as the ones in the txt file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set (1000).\n",
      "Proceeding with the analysis.\n"
     ]
    }
   ],
   "source": [
    "len_ID = len(test_fasta['ID'].unique()) # assigned because gave problem on else statement print\n",
    "\n",
    "if len(test_ids) == len_ID:\n",
    "    print(f\"The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set ({len(test_ids)}).\\n\"\n",
    "          \"Proceeding with the analysis.\")\n",
    "else:\n",
    "    print(f'The numbers are not the same: test_ids are {len(test_ids)}, while the length of the fasta file is {len_ID})')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_embeddings.h5`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>[0.00979, -0.03973, 0.03653, -0.006447, -0.040...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>[0.02786, -0.01154, 0.008865, -0.01765, 0.0073...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4P506</td>\n",
       "      <td>[0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K1A2</td>\n",
       "      <td>[0.00882, 0.0835, -0.001374, -0.0003645, -0.06...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2K1V4</td>\n",
       "      <td>[0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings\n",
       "0  A0A0B4JCV4  [0.00979, -0.03973, 0.03653, -0.006447, -0.040...\n",
       "1  A0A0B4KHT0  [0.02786, -0.01154, 0.008865, -0.01765, 0.0073...\n",
       "2  A0A0B4P506  [0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...\n",
       "3  A0A0G2K1A2  [0.00882, 0.0835, -0.001374, -0.0003645, -0.06...\n",
       "4  A0A0G2K1V4  [0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0...."
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_list = []\n",
    "\n",
    "with h5py.File(test_data_path / \"test_embeddings.h5\", \"r\") as f:\n",
    "    for dataset_name in f.keys():\n",
    "        dataset = f[dataset_name][:]\n",
    "        data_list.append([dataset_name, dataset])\n",
    "\n",
    "test_embeddings = pd.DataFrame(data_list, columns=[\"ID\", \"embeddings\"])\n",
    "\n",
    "test_embeddings.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_protein2ipr.dat`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>IPR039915</td>\n",
       "      <td>TACC family</td>\n",
       "      <td>PTHR13924</td>\n",
       "      <td>38</td>\n",
       "      <td>1206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PF00643</td>\n",
       "      <td>177</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PF00643</td>\n",
       "      <td>236</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PS50119</td>\n",
       "      <td>173</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PS50119</td>\n",
       "      <td>235</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        ipr                  domain   familyID  start   end\n",
       "0  A0A0B4JCV4  IPR039915             TACC family  PTHR13924     38  1206\n",
       "1  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PF00643    177   219\n",
       "2  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PF00643    236   274\n",
       "3  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PS50119    173   220\n",
       "4  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PS50119    235   282"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_protein2ipr = pd.read_csv(test_data_path / 'test_protein2ipr.dat', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "test_protein2ipr.columns = ['ID', 'ipr', 'domain', 'familyID', 'start', 'end']\n",
    "\n",
    "# Remove 'domain' that is useless\n",
    "test_protein2ipr.drop('domain', axis=1)\n",
    "\n",
    "test_protein2ipr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test protein2ipr ((11263, 6)):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>(IPR039915,)</td>\n",
       "      <td>(TACC family,)</td>\n",
       "      <td>(PTHR13924,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(1206,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>(IPR000315, IPR000315, IPR000315, IPR000315, I...</td>\n",
       "      <td>(B-box-type zinc finger, B-box-type zinc finge...</td>\n",
       "      <td>(PF00643, PF00643, PS50119, PS50119, SM00336, ...</td>\n",
       "      <td>(177, 236, 173, 235, 173, 235, 976, 826, 988, ...</td>\n",
       "      <td>(219, 274, 220, 282, 220, 276, 1048, 839, 1004...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4P506</td>\n",
       "      <td>(IPR003417, IPR003417, IPR036552, IPR036552)</td>\n",
       "      <td>(Core-binding factor, beta subunit, Core-bindi...</td>\n",
       "      <td>(PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)</td>\n",
       "      <td>(1, 1, 1, 4)</td>\n",
       "      <td>(164, 168, 142, 140)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K1A2</td>\n",
       "      <td>(IPR010255, IPR019791, IPR019791, IPR019791, I...</td>\n",
       "      <td>(Haem peroxidase superfamily, Haem peroxidase,...</td>\n",
       "      <td>(SSF48113, PF03098, PR00457, PR00457, PR00457,...</td>\n",
       "      <td>(142, 148, 172, 226, 374, 392, 417, 470, 598, ...</td>\n",
       "      <td>(718, 692, 183, 241, 392, 412, 443, 480, 618, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2K1V4</td>\n",
       "      <td>(IPR000048, IPR000048, IPR001609, IPR001609, I...</td>\n",
       "      <td>(IQ motif, EF-hand binding site, IQ motif, EF-...</td>\n",
       "      <td>(PS50096, SM00015, PF00063, PR00193, PR00193, ...</td>\n",
       "      <td>(789, 788, 89, 116, 172, 228, 459, 513, 86, 80...</td>\n",
       "      <td>(818, 810, 774, 135, 197, 255, 487, 541, 786, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                                ipr  \\\n",
       "0  A0A0B4JCV4                                       (IPR039915,)   \n",
       "1  A0A0B4KHT0  (IPR000315, IPR000315, IPR000315, IPR000315, I...   \n",
       "2  A0A0B4P506       (IPR003417, IPR003417, IPR036552, IPR036552)   \n",
       "3  A0A0G2K1A2  (IPR010255, IPR019791, IPR019791, IPR019791, I...   \n",
       "4  A0A0G2K1V4  (IPR000048, IPR000048, IPR001609, IPR001609, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0                                     (TACC family,)   \n",
       "1  (B-box-type zinc finger, B-box-type zinc finge...   \n",
       "2  (Core-binding factor, beta subunit, Core-bindi...   \n",
       "3  (Haem peroxidase superfamily, Haem peroxidase,...   \n",
       "4  (IQ motif, EF-hand binding site, IQ motif, EF-...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                                       (PTHR13924,)   \n",
       "1  (PF00643, PF00643, PS50119, PS50119, SM00336, ...   \n",
       "2  (PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)   \n",
       "3  (SSF48113, PF03098, PR00457, PR00457, PR00457,...   \n",
       "4  (PS50096, SM00015, PF00063, PR00193, PR00193, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                              (38,)   \n",
       "1  (177, 236, 173, 235, 173, 235, 976, 826, 988, ...   \n",
       "2                                       (1, 1, 1, 4)   \n",
       "3  (142, 148, 172, 226, 374, 392, 417, 470, 598, ...   \n",
       "4  (789, 788, 89, 116, 172, 228, 459, 513, 86, 80...   \n",
       "\n",
       "                                                 end  \n",
       "0                                            (1206,)  \n",
       "1  (219, 274, 220, 282, 220, 276, 1048, 839, 1004...  \n",
       "2                               (164, 168, 142, 140)  \n",
       "3  (718, 692, 183, 241, 392, 412, 443, 480, 618, ...  \n",
       "4  (818, 810, 774, 135, 197, 255, 487, 541, 786, ...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by 'ID' and aggregate other columns into lists\n",
    "test_protein2ipr_grouped = test_protein2ipr.groupby('ID').agg(lambda x: tuple(x)).reset_index()\n",
    "\n",
    "print(f\"Test protein2ipr ({test_protein2ipr.shape}):\")\n",
    "test_protein2ipr_grouped.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `blast_test_results.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows missing from train_protein2ipr_grouped: 19\n",
      "Combined DataFrame shape: (1000, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>[0.00979, -0.03973, 0.03653, -0.006447, -0.040...</td>\n",
       "      <td>(M, E, F, D, D, A, E, N, G, L, G, M, G, F, G, ...</td>\n",
       "      <td>(IPR039915,)</td>\n",
       "      <td>(TACC family,)</td>\n",
       "      <td>(PTHR13924,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(1206,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>[0.02786, -0.01154, 0.008865, -0.01765, 0.0073...</td>\n",
       "      <td>(M, D, M, D, L, E, Q, L, K, N, D, F, L, P, L, ...</td>\n",
       "      <td>(IPR000315, IPR000315, IPR000315, IPR000315, I...</td>\n",
       "      <td>(B-box-type zinc finger, B-box-type zinc finge...</td>\n",
       "      <td>(PF00643, PF00643, PS50119, PS50119, SM00336, ...</td>\n",
       "      <td>(177, 236, 173, 235, 173, 235, 976, 826, 988, ...</td>\n",
       "      <td>(219, 274, 220, 282, 220, 276, 1048, 839, 1004...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4P506</td>\n",
       "      <td>[0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...</td>\n",
       "      <td>(M, P, R, V, V, P, D, Q, R, S, K, F, E, N, E, ...</td>\n",
       "      <td>(IPR003417, IPR003417, IPR036552, IPR036552)</td>\n",
       "      <td>(Core-binding factor, beta subunit, Core-bindi...</td>\n",
       "      <td>(PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)</td>\n",
       "      <td>(1, 1, 1, 4)</td>\n",
       "      <td>(164, 168, 142, 140)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K1A2</td>\n",
       "      <td>[0.00882, 0.0835, -0.001374, -0.0003645, -0.06...</td>\n",
       "      <td>(M, K, L, F, L, A, L, A, G, L, L, A, P, L, A, ...</td>\n",
       "      <td>(IPR010255, IPR019791, IPR019791, IPR019791, I...</td>\n",
       "      <td>(Haem peroxidase superfamily, Haem peroxidase,...</td>\n",
       "      <td>(SSF48113, PF03098, PR00457, PR00457, PR00457,...</td>\n",
       "      <td>(142, 148, 172, 226, 374, 392, 417, 470, 598, ...</td>\n",
       "      <td>(718, 692, 183, 241, 392, 412, 443, 480, 618, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2K1V4</td>\n",
       "      <td>[0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0....</td>\n",
       "      <td>(M, S, S, D, A, E, M, A, V, F, G, E, A, A, P, ...</td>\n",
       "      <td>(IPR000048, IPR000048, IPR001609, IPR001609, I...</td>\n",
       "      <td>(IQ motif, EF-hand binding site, IQ motif, EF-...</td>\n",
       "      <td>(PS50096, SM00015, PF00063, PR00193, PR00193, ...</td>\n",
       "      <td>(789, 788, 89, 116, 172, 228, 459, 513, 86, 80...</td>\n",
       "      <td>(818, 810, 774, 135, 197, 255, 487, 541, 786, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A0B4JCV4  [0.00979, -0.03973, 0.03653, -0.006447, -0.040...   \n",
       "1  A0A0B4KHT0  [0.02786, -0.01154, 0.008865, -0.01765, 0.0073...   \n",
       "2  A0A0B4P506  [0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...   \n",
       "3  A0A0G2K1A2  [0.00882, 0.0835, -0.001374, -0.0003645, -0.06...   \n",
       "4  A0A0G2K1V4  [0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0....   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, E, F, D, D, A, E, N, G, L, G, M, G, F, G, ...   \n",
       "1  (M, D, M, D, L, E, Q, L, K, N, D, F, L, P, L, ...   \n",
       "2  (M, P, R, V, V, P, D, Q, R, S, K, F, E, N, E, ...   \n",
       "3  (M, K, L, F, L, A, L, A, G, L, L, A, P, L, A, ...   \n",
       "4  (M, S, S, D, A, E, M, A, V, F, G, E, A, A, P, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0                                       (IPR039915,)   \n",
       "1  (IPR000315, IPR000315, IPR000315, IPR000315, I...   \n",
       "2       (IPR003417, IPR003417, IPR036552, IPR036552)   \n",
       "3  (IPR010255, IPR019791, IPR019791, IPR019791, I...   \n",
       "4  (IPR000048, IPR000048, IPR001609, IPR001609, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0                                     (TACC family,)   \n",
       "1  (B-box-type zinc finger, B-box-type zinc finge...   \n",
       "2  (Core-binding factor, beta subunit, Core-bindi...   \n",
       "3  (Haem peroxidase superfamily, Haem peroxidase,...   \n",
       "4  (IQ motif, EF-hand binding site, IQ motif, EF-...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                                       (PTHR13924,)   \n",
       "1  (PF00643, PF00643, PS50119, PS50119, SM00336, ...   \n",
       "2  (PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)   \n",
       "3  (SSF48113, PF03098, PR00457, PR00457, PR00457,...   \n",
       "4  (PS50096, SM00015, PF00063, PR00193, PR00193, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                              (38,)   \n",
       "1  (177, 236, 173, 235, 173, 235, 976, 826, 988, ...   \n",
       "2                                       (1, 1, 1, 4)   \n",
       "3  (142, 148, 172, 226, 374, 392, 417, 470, 598, ...   \n",
       "4  (789, 788, 89, 116, 172, 228, 459, 513, 86, 80...   \n",
       "\n",
       "                                                 end  \n",
       "0                                            (1206,)  \n",
       "1  (219, 274, 220, 282, 220, 276, 1048, 839, 1004...  \n",
       "2                               (164, 168, 142, 140)  \n",
       "3  (718, 692, 183, 241, 392, 412, 443, 480, 618, ...  \n",
       "4  (818, 810, 774, 135, 197, 255, 487, 541, 786, ...  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_test = pd.merge(test_embeddings, test_fasta, on='ID')\n",
    "combined_test = pd.merge(combined_test, test_protein2ipr_grouped, on='ID', how='left')\n",
    "\n",
    "missing_rows = combined_test[combined_test['ipr'].isna()].shape[0]\n",
    "print(f\"Number of rows missing from train_protein2ipr_grouped: {missing_rows}\")\n",
    "\n",
    "print(f\"Combined DataFrame shape: {combined_test.shape}\")\n",
    "combined_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combined_train and combined_test have the same columns: True\n"
     ]
    }
   ],
   "source": [
    "# Check if combined_train and combined_test have the same columns\n",
    "print(\"Combined_train and combined_test have the same columns:\" , set(combined_train.columns) == set(combined_test.columns))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PCA for embeddings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll see that considering each df separately doesn't change much."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArMAAAGwCAYAAAC6m+0oAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAcqpJREFUeJzt3Qd4U9X7B/C3ewBl771kyt4ylD1EUJTxQ6aiLEURBygCDkBUFJWhIOBAhgqCgMgGkb33krIpmxZauu//+Z7+b0jSpCRt0iTN9/M8gZub2+TknpubN+ee8x4fTdM0ISIiIiLyQL6uLgARERERUXoxmCUiIiIij8VgloiIiIg8FoNZIiIiIvJYDGaJiIiIyGMxmCUiIiIij8VgloiIiIg8FoNZIiIiIvJYDGaJiIiIyGMxmKUsoW/fvlKqVCmHPufcuXPFx8dHzp49K94iI/sRf4e/zyr1nxXLlB54D08++aSri0Fkt59++kkqVqwoAQEBkitXLlcXh5yIwSwZ/Pfff/Lyyy9LmTJlJDg4WMLCwuSxxx6TKVOmyP379yWrGj9+vPzxxx/iLvQg2tpt+/btri6ix7l27Zr4+/vL888/b3Wbu3fvSkhIiDzzzDOZWjYS9YPR+Bj38/OTEiVKyNNPPy379+9PtX1sbKx88cUXUr9+fcmZM6c6Xz3yyCMydOhQOXnypMXXeOutt9Rzd+vWze7yJSUlyZw5c+Txxx+XPHnySFBQkAry+/XrJ7t3707Xe6YHLl++LGPHjrVY1+l1/Phx9YOybNmyMnPmTPnuu+/EmVB+42PY19dXChcurH4Imp+z7T3e6eH8bdiGvMCKFSvkueeeUyfp3r17S9WqVSU+Pl62bNkib775phw5csTpJwNXBrPPPvusdO7c2WR9r169pHv37mqfuMIHH3wgpUuXTrW+XLly4o5OnDihTuDuqECBAtKqVStZunSpxMTESGhoaKptFi9erIKktAJee+ALNDk52SHP5S169Ogh7du3V8HjsWPHZPr06fLXX3+pYKBGjRpqmxs3bkjbtm1lz549KlD43//+J9mzZ1fH34IFC9R5CucuY5qmyfz581UA+ueff6ofLjly5LCpTPghjx84q1atkqZNm8qoUaNUQIuAZNGiRfLDDz/I+fPnpVixYk7ZJ94SzI4bN07Vj17PGbVx40b1+UNjTGaeM3HM4njEa1+4cEGdB3Dc7Ny5M9V7s+V4J9swmCUJDw9XQVvJkiVl/fr16tekbsiQIXL69GkV7Hob/FrGzVXatWsnderUEU/hqqDfVj179lQBybJly9Txbu6XX35RrXwdOnTI0OtER0dLtmzZ1KVNsk+tWrVMfkzgytBTTz2lvuS//fZbtQ6tbfv27ZPffvtNunTpYvL3H374obz77rsWA5uLFy+q81ubNm3UD5c+ffrYVCb8mMdxg5bg1157zeSxMWPGqPXknldjwJHdC6z9EDaGhpF8+fIZ7qORBI1Dv/76a6oA1ZbjnWzjns0olKkmTZok9+7dk++//94kkNXhV+2wYcNMLo/gUrg5rMelFvPLLrjshw8sAoX8+fPL6NGjVUsJfrV26tRJdWcoVKiQfP755zb1WcUXE9bj/7R89tln0qhRI8mbN6+6fFy7dm31BWheZgQfaF3RL/no/T7NXx+tQOiCYUnDhg1TBZ4///yzek28NlpyEEDhPTsKvkjRErpu3TqT9S+99JIEBgbKgQMHTPbXwoULVasS9jWCLZw0bSmPLfvRUp9Zff/9+++/Mnz4cFX3eF1cSrt+/Xqqv0eLRJMmTdQ2aDVDUIkrAubQJQRfDri0jP+XLFli0/7C6+K5EbRa+uLDfsQXEYLyf/75R12pwKU/3C9evLi8/vrrqbrb4P2iFQZddNDCgnIjaLbWZ9bWfYn9hkvm+ntFGapUqaKCKnOXLl2SF154QYoUKaK2Q2v+oEGDTFon79y5owIxvA9sg8/0J598YlfL8erVq9WXMfZ75cqVVUCoO3PmjCqzpcBu69at6jG0jNqrefPmhh/csGPHDvXDGu/XPJAFvDfsY3Pz5s1TZX7iiSekZcuW6r4tEAAjqECrvnkgC/ixO2LECJNWWQTa+CGK8xqOjRYtWqS6zKx/NnDl69VXX1WfDQRd6OaFekN94QpZ7ty51Q1dJHDO1OnnYbxX7HM0ROB4atasmRw+fDhVORHE658tvA7Ou2gJNKafr9F4gWMX2+Gcja4UCOLM2XJ+Q7cMHL9Hjx5V+x6BYNGiRdV3jg7np7p166plvJZ+Hta/Y06dOqXqGuctHHvY13ityMhIq/WGzx3Oj4B9a/7dNG3aNPV5wvGCzw0abbDPLZUdVwDQsoqy4/xpL5Qb0M3J3uOdbMdgltRlNwRp+JJ1BvRRw5fmxIkTVR+3jz76SL788kv1BYETG75U8eWKL4XNmzc77HVxealmzZrqcj26EuBkggDFuJUZAwRwQsOJHsu44QvF2vvASWbXrl0m68+dO6e+rIxb+z7++GP1ZVS+fHmZPHmy+iJEsISTovlJ0xqcrHFJ1fh28+ZNw+PvvfeeCi7wxY7LpvD333+ry1rvv/++VK9e3eT5UCa897ffflt9ga5Zs0Z9sT+sP7Qt+zEtr7zyigqs8eWCIAvHGwI1Y9jvCF7x5Y/jAT948AXYuHFjkx8zCKjwxYYvpwkTJqhWD1v7LeKLHF/i2Ee3bt0yeQyBPi716YEoWlHwBY7yfv3116o1D/+jTs0lJiaqx9GVAcGFpSArPfsSgc7gwYPVcYUvf3SBwHMbHwO4PFuvXj11eR3H51dffaW6x2zatMkQgOB/BDkIPlB+bIMWoJEjR6ofGbZAQIHnR5CG/a6XG8cQ4PyB57QUJGIdgnzse3vhRwIg+Ae0qgPeo63i4uLk999/V5d0Af8juIuIiHjo3+IHFurX1tfDjy+cS3C8IwDFcYxzBgIjBOKWPhvYt7jEjh+X6CKBv+nYsaM6HnGM4DPw6aefqs+IuR9//FHVJ4Ix1CcCWQREV69eNWyzdu1adXziBxsCOtQ5fmCgviwNbu3atas6n6CesYygEuUzZs/57fbt26pbCM5HaLDAgCycg7BvoVKlSurzoP8Q18/DeC4E9ig7zq/YV1OnTlXb4MdTWudRfL/gxyuglRPPp/eFxz7A/kIQi/LgM4UfLK1bt5aEhAST58FnDcc8zrN4TgTkD4NzC87V2N/4YTNgwAAVhGNf2nu8kx008mqRkZH4ua916tTJpu3Dw8PV9nPmzEn1GNaPGTPGcB/LWPfSSy8Z1iUmJmrFihXTfHx8tIkTJxrW3759WwsJCdH69OljWIfXwN/jNY1t2LBBrcf/OvxdyZIlTbaLiYkxuR8fH69VrVpVa968ucn6bNmymbyutdfHvgoKCtLeeOMNk+0mTZqk3s+5c+fU/bNnz2p+fn7axx9/bLLdoUOHNH9//1Trrb2upRte3/w5AwMDtRdffFHtw6JFi2p16tTREhISUu0vPBYVFWVYv2jRIrV+ypQpDtmP+DtL9deyZUstOTnZsP71119X++fOnTvq/t27d7VcuXJpAwYMMHm+iIgILWfOnCbra9SooRUuXNjwt7B69Wr1OubltmTFihVq22+//dZkfYMGDdT+SUpKsvieYcKECSb1rO8vPN8777yTavuM7Es8J+r19OnThnUHDhxQ67/++mvDut69e2u+vr7arl27Ur2+vs8//PBDdYyfPHnS5HGUGfVw/vx5LS14D3jd33//3bAOnwXUQ82aNQ3rsE+x3bFjx0zeX758+Sx+viydV8aNG6ddv35d1f3GjRvV8xu/9tNPP63u41i31W+//ab+5tSpU+o+PgPBwcHaF1988dC/xbGKv923b59Nr9W5c2dVb//9959h3eXLl7UcOXJoTZs2TfXZaNOmjclno2HDhuoYGzhwYKpzZrNmzVLtL5wzL168aFi/Y8cOtR7lNv7MFChQQLt586bJsYTjBseP+fm6f//+Ju8J+zxv3ryG+/ac31BmPOePP/5oWBcXF6cVKlRI69Kli2Edjl9L3yvY71j/66+/avbS3w+OJ921a9dU/bRu3drwWYdvvvlGbTt79uxUZZ8xY4Zdr2d+w7lt1apV6TreyXZsmfVyUVFR6n9bB0Okx4svvmhyWQ6X4/F9jRZFHS5pVahQQf3idhRc/jJuHUBLJ1pN9u7dm67nw2VD/ErHoA/jS35o1WvQoIG6JA24/IqWaPwSN25VxeUmtGRs2LDBptdDKwRavoxvemuGDpfB0Goya9Ys1YKB10GXCUuXtNCSYlzPuKSObiUrV6506n5ESwpaUnX4W7Q6oUUb8L7QyoIWM+P9hWMFLfn6/rpy5Yoa6Yu+jrj8qUMLPy4h2wKtL7jsaNzVAC1naPnB6+sD2IzfM7qhoDy4coF6R2uLObTg2sKefYlWc4zE1lWrVk0dg/pnBMcYuiGgFc9S32p9n6OVGa+By9XG+xfPj3qw5WoIWrH0li5AOXA8YV/oLZw43tECZdw6i1ZwvJatg+rQeo/6wWcFrZloqUJLvd6qlp7zFcqD/aMPAtK7sNjS1cCe18O+xJUDXC0w7o6EzxgGqaGlXX8+Hc6Bxp8NHO/m50b9nGnp3IjXwtUtHVrp8Rz6Z1r/zKDbALoCGB9L+NxY+uwPHDjQ5D6OHbRQ6mW39/yGqy3G9Y8uUCinLed6/XOO48hSVwd7oZUarb1oSTYerIrWUxzT5ldIcNUOV37sgasAOKfhWEAGDGTZQOsvWsPtPd7JdhwA5uXwAQb9MrUz6EGeTk+lY9xJXl9vfAk1o5YvX666NOBkjkuNOuMvD3vhUisCiG3btqngBicf9KnCJSgdLhviCwkndktsHRiEE74tA8AwQAWXmTFaFpclrQV25uXBfsAX/MPy6GZ0P5rXP4IqPZjT95dxfzFrx6ge/Frar/ghZEtwjSAfdYg+c+hrikBAD2z1LgaA0enoqoHL2no5deZ99fCcto5kt2dfmu83fd/p5UG/YwQY+EGTFuzfgwcPqi/NtAbKpAXHiXkZ8SUNOH7wZYwfpAissT8xEAsQMGIfW6tbSz980H0BgQaeT+/XaOl8ZcvAHvxIQsCGbi3oC6rDJXYEHejPr7+PjJ4fUR8IuHAsmsOldH10O95TWudGQN9m8/Xmx6G1zwLeD35wG39mrJUJQaI+YNGWzyv2h73nN3w2zI8dPCeOyYdB/290i0BXBhxLCKzRHUMfg2Eva/sDATZ+gOiP63Ds4jF7oHuE8XcbGg2wr9BNAt8V9hzvZDsGs14OJye0ulgaNGCJtQAGrRLWWMoIYC1LgHGLZ3peS4cBPDjp4cSCwAWtIzjJ4peypQFAtsKXNQYC4MsCwSz+x4kIJyQdvrRQdrSiWnqfaKlwJLRw6AHhoUOHHPrcjtiPD6trfRAS+rXpgyWM2TJwwh74Ivzmm2/UgCT008b/+AGgjzTG8YVWK/R9Q98+9PHDlz2CX7RwmQ+awpePLSnJ7N2XtnxGbIHy4v2gD6claQVz9kJrLVqC0Qr16KOPqh8D6Pdra8o2fOmjxdga1IV+nCOweRiUBT8a0DfSfIApIEAy7w9q7fWckSrJWh1bWm9vvaeXLZ9Xe85vGT2OUW/43CGtHlo70d8f/XlxNcXZ6dCMr6SkF/YHWstRfvMfDg873sl2DGZJjdLHwAO0NmJUflr0X+nmne/Nf9E6QkZeC60uaP1Fy4PxL10EDubsaanFiQj7C1+SaC1AFwN8qeIHgQ6XhnGiRquCIwMFS/DFghM9fpTg0pmeM9fSZSo94NWhjGitwiVHR+zH9NIvpWMAVVondozYtvQ+ADlGbYUvFrwmAkgEeRi0gwEtOgQuaLFDdw3jAV/6YKf0cvS+REsr6v1hP0TxXpGtJCNfmjhOcLwYf1b0yQmMMzZgoA/KhSAR+xktlfYM1rLlxyQCGQxmsyWYRTnQcq2PbDeGQT84BtIKZtGtCMEYXu9h7wPvGz90LR2LSOCPgN68xTWjLH0WUC96neifGWtlQguicXBlC2ec3x52DsYPI9ww6FUfvDZjxgx1lcMexvvDuCsIuh6gu5GzAksMIgR8Du3d32Qb9pkl1WKDDxj6thqPgtXhUjpGYQO+PHECNO9nh5YmR9ODHOPXQquZLZM34AsIJ0jjVlxcDrU00xfeu60ZBgCXqTGKHP1UMWrZfEYhBJJ4fXxJmrc+4L4ju1IgoMbJHfsEl3bRWoz+m+jDZmnks/HlUqSEQp86fGE7Yj+mF/r64rhCIG4+mhj0NF5oyUTrGIJM40v9CDKR+cAe6FKA/p4IcvD+0KfRvCXJuO6wrH8G0svR+xLBEfpMIjuEpWwOevnRtxE/VBFEm8Nxr3/RpgXHu3EKNHRvwPGE+jBuTUcrOvoe44oFRsEjAEnrx5K98GMbATM+e5b2G4IStLYDLunj3IH3jx945jf0hUSQbinLgA7BJ/pTokUQ2Sws/ZhEyyFSeKF+0ScbLXDGXXdwTkXQjKwEercFR8E+wBUDHboa4f3on2njz4zxOQ4/gPCekE7OXs44v+kBnvl5GMeZ+fGJYwrHvnE3HVshWEW3AWSAMC470lLinJLRHNOW4AoPztH4nOAHOzkHW2bJ0EqFoAz9qIxnAMOHEK2QxvlDEfQizRb+R59OfGFYm0IyI9B/CAOrkHIGJwQMYEDfUFu+fHFSQqCHLz4EKugXiAFV6Ptn3lcLuRIxMADbo4UVLQ5oVbJGzyeKL02c1M1TMWF/osUA5caXGgIObI9f/ggI0E9K/8JNCy7jofXEHAJWtCogTyTS+KBu0GIFCCDw5YVLu3q/OR32H75Q8SWOL1j088X+wJe1I/ZjeuELHulz0PKFJOJIRYVWLvRbxYAMtMKgWwCgVQ5lwvvo37+/Oi4QZOBYQauHPV0NkA4IgQee37h1EZeWUYeoIwQKKB9aVS31WbSHM/YlfgAgKEHqLRxX+PziBwo+sxhwhH546FONy/24ooBjBcc7LneiBRo/aHCMmvdfN4cWOAxKQlq6ggULyuzZs9UxZKlVWU//hYFAGMziaAiiETQiqMJxjzyuCIbQSonzA94/UqThnIaABV07rH2OEXzrrcjWIFjFD3pc3sbgJ+xHXDXC8Yn9jM+onpYPn3v8uMLxic8gnh8twAi8jHOrOgqOHbwWfsDiNfCZRlon4y4lSOuF4BY/BFCHSMWHzwz6nBrnXrWVo85v5s+JYxWtrXgu1CfqBI0F6O+Mblw4BnHuR3ckS+ddW+C8gnIjEMfnEMcGWmnRGINct46Y/Q+fKXQtwLGHH4EIlHHuwHvLyHgNegg7Mh9QFofUPUiDVKpUKZW+BOlkHnvsMZUKKDY21iS90AsvvKDSJmGbrl27qpQn1lJzGadGAaTpQaogc0iFUqVKFZN1SHGD1E5ISVWwYEFt1KhR2po1a2xKzfX9999r5cuXV39bsWJFlfZFL5Ox48ePq7Q5SHODx/Q0QtZSg0HPnj0NaaesQXqVxo0bq/eKG8owZMgQ7cSJE1b/xvh1rd3wONL11K1bV6XsMU5TBUi1he0WLlxokppr/vz52siRI1WaHrzXDh06mKSZyuh+tJaayzxtlKXUavp6pCrCcYXUSWXLltX69u2r7d69O9V+rVSpkipP5cqVtcWLF1ss98Ng/6Ec06ZNS/XY0aNHVd1mz55dpZbC50JPjWWcPsjasZzRfYn7OFbMme9jQB0ixVL+/PnV85YpU0b9LVIg6ZD+DHVfrlw59dnGe2rUqJH22WefqfRZacFr4lj5+++/tWrVqhnKnla6JHyOkfrJOG1UWvRURZ9++qlN2+MchLKjDlFHeE/Yr6+88oohndmjjz6qlShRIs3nefzxx9XnwTiVnSX4vM2aNUtr0qSJOj4DAgLUfunXr1+qtF179+5VxzHKFRoaqj3xxBPa1q1bTbax9tmw9ZxpvL8+//xzrXjx4qpeUD4cp+bWrl2rzuX43IeFhWkdO3ZUx7gtr23tPGjL+c3SOd3aZ2Pp0qXq84z0Xvrn7MyZMypVGM4FOCfkyZNH7U+8n4ex9n70VFwoL+oR3yuDBg1Kle7NWtkf9nrGN+wXpFtDGsSMHO/0cD7452EBLxF5Lsywg2TfaEXC5VUiZ8PEELgSYD47HTkGWkRxBQmtrva2ghJlRewzS0REDoP+u0g9Zmm2NCIiZ2CfWSIiyjAMKkIeTfQxxcAj84GRRETOwpZZIiJyyMAXDC5ERgrk7kUaMiKizMA+s0RERETksdgyS0REREQei8EsEREREXksrxsAhhlbkMgYiZmZwJiIiIjI/aAXLGatxGRGmPUtLV4XzCKQdfT82ERERETkeJiaulixYmlu43XBLFpk9Z3j6HmyLcHIXkw3iekXAwICnP565B11Fx0fLUU+L6KWL79xWbIFpsxtTu5db2Qd680zsd48U4IH1FtUVJRqfNTjtrR4XTCrdy1AIJtZwWxoaKh6LXc9YMjz6s4v3k/k/zMfoXwMZj2j3sg61ptnYr15pgQPqjdbuoRyABgREREReSyva5klygr8fP2kffn2hmUiIiJvxWCWyAMF+wfLiv+tcHUxiIiIXI7dDIiIiIjIYzGYJSIiIiKPxWCWyAMhNVe28dnUDctERETein1miTxUTEKMq4tARETkcmyZJSIiIiKPxZZZIiIiIheIT0yWOf+ekb8PR8jVu7ES4u8rgf5+Ep+YJDEJSSJayqQBIQGm633EtnVi5e9F85GEWF+ZfmaLJCZpNr1WcGCAFM4ZLPVK55U+jUpJoL/7tIe6NJjdvHmzfPrpp7Jnzx65cuWKLFmyRDp37pzm32zcuFGGDx8uR44cUdOcvffee9K3b99MKzMRERG5f3AYEXU/w0FfRrdN6++j45LkTmyiC/eWr1y/ak93tfty+HKUrDl2Tcb/dUxealJaRravLOLtwWx0dLRUr15d+vfvL88888xDtw8PD5cOHTrIwIEDZd68ebJu3Tp58cUXpXDhwtKmTZtMKTMRERE9PJB0VquitW2jYhPlXnyyq3eDV9A0kW83h6tldwhoXRrMtmvXTt1sNWPGDCldurR8/vnn6n6lSpVky5Yt8sUXXzCYJSIiyuQAMy4hUW7c8ZVRe9ZKNANJrzPzn3B5o3VFl3c58Kg+s9u2bZOWLVuarEMQ+9prr1n9m7i4OHXTRUVFqf8TEhLUzdn018iM1yLvqbukxCRpWqKpYTnBx/3K6CruXG9knbfWW1KyJjvDb8mW/27IoYt35FpUvMTEJ6nH9KAxwN9HEhKTJSY+2eI6a9tGxSVmUoDpizeSCa9D7iZZE5n773/Sr1Ephz+3PecCjwpmIyIipGDBgibrcB8B6v379yUkJCTV30yYMEHGjRuXav3q1aslNDRUMsuaNWsy7bXIO+pueJ7h6v8Naza4uihuyV3rjTy73vDlfeKOj+y8JnLxno/EJSGQFAnwFUHjVGKyyP/HohbXG6+LThSJSvTBlq5+W0Tp9s++41LwzlFxtJiYmKwZzKbHyJEj1YAxHQJfDBxr3bq1hIWFOf318csCJ+dWrVpJQECA01+PHId155lYb57J2fWGFtDt/92UxfsvyYVbMRKbkKQuwVtr7bS07n5CsosH7BC5nyY1K0p7J7TM6lfSs1wwW6hQIbl69arJOtxHUGqpVRaCgoLUzRxOlpn5RZfZr0eOw7rzTKy3rFVvejD6z+lrcuD8Hbl+LzblcrwNfUBvRcdLxN14l7wfoqzM10ek72NlJcAJfWbtOX97VDDbsGFDWblypck6/JLHeiJvgilsS01J+SV8dthZyRaYzdVFIsrQgKRgf1+JifaVr05ulvuJyamC0at347EpEbmRAU1Ku3zwl8uD2Xv37snp06dNUm/t379f8uTJIyVKlFBdBC5duiQ//vijehwpub755ht56623VDqv9evXy6JFi2TFihUufBdErnEj5oari0BeDC2lW0/dkN/2XjC6bJ92SqWo+w9LneQrEhubie+CiNIDfb+ZZ/b/7d69W5544gnDfb1va58+fWTu3LlqIoXz588bHkdaLgSur7/+ukyZMkWKFSsms2bNYlouIiIHBae/7jkvR69Eyf2EpFSzEemtpbGJmly6E8uWUiIHCwvyk2K5QzJpBrD7kjd3ds4AllGPP/64aMi8awUCWkt/s2/fPieXjIjIe2ZDQiDLy/iUFeUO8Ze82QLccgYwVwaICQkJqttm+/aNs8TYAo/qM0tE5O1sbUF1/VSZ5O1yBftJtiD/TJsBDNvi9SoVzinP1i4mjcrlEz+MUKIsj8EsEZGbBanHIu6KpiVLoJ/pl3VsYrLcjklkCyq5TYCZMgNYjGTLFiLZgxlIkmswmCUiyoTL/T9sDZcdZ27K5Tv3UwUGvMxPxnIH+0lIoJ9DLns7u6XyweXqZlnicjV5JgazRB7I18dX6hSpY1gm97nsHxOfaBJwXL4Ty8v9WViov4/kzhaY7kvkSEOWLZAtmkQZwWCWyAOFBITIrgG7XF0M8fYUVPfjk+XOfV7291QFsweInx/CS9tbQH19faVgWLC0qVJI+j7mHjk2ibwdg1mIixP5/HORn38WOXNGJFs2kSZNRN5/X6RWLdueY+9ekY8/FvnnH5HISJEiRUSeflrknXcsb79nj8gHH4hs2SISHS1SpoxIr14ib7whEhjo0LdHROkb9c8UVFnncjxbQImyLgaziYkiHTqIrFtnGtwuXSqyapUIJmRo0SLt5/j7b5GOHdF56MG6s2dFvvhC/FetEv933zXdfvXqlO3jjaZXPHZMZNQokY0bRTDLmZ+fo94hkdez1meVo/7dZ0BSygxg0RKaLTTVDGAMRokoLQxmp017EMhWrSoybpwI8th+9FFKUNu3rwhmKQsKsvz3yJP78ssPAlm0xD7+uMi334osWSI+x45JpV9+EenaNeXx+/dF+vV7EMi+955IzZoiY8aIHD6cEujOmCEyZEhmvHvyUDEJMVJ5asrMK0eHHJXQgFDxdnp3gEW7z8rOk5gW9R8JCmCfVVdctk/PgKQHA4maciAREdmFwSwCR93MmSINGog884zIrl0pLa4XL4osXy7SpYvlvz96VOTcuZTlUqVEJkxIWa5RQwWzUGL9etEwRSNO0H/+KXL5cso2mLnsww9TltEtoWHDB2ViMEtpwGQj5yJTjru0Jh7J6l0Crt6NVTlWU3cH8JWrcfddWs6sEpzmzR5k8dJ9QlKy+Pj6SKVCbCklItfy7mD21q2Uy/uAQLNu3QePNWqUEswC+sFaC2bRP1aHvrYWlv1jYyXx0KGU50QfWePX0NWpk1IGtPCihfb2bZHcuTP8FomyUj9WdgnImNAAXymSM8hi39KgAH8JDvCT4nlCpUstBqdE5Dm8O5hFv1Zd3rym/VQLFHiwHB5u/TnKl8f1tJTuBgiM0ccW3Qy++MJ0uwsXUr9mwYIPlv39RfLkEbl69cF2DGbJCzBodV4LqjvPpU5E5CjeHcwii4DOPIOA8X3j7czlzy/Su7fIDz+IJCeLPPmk5e3QzcBRr0mURfKxRt1PlHvxya4umtsqmhP5S/1M8paGBvpJgbAQqV4slzxWPp80KJOXLahE5NW8O5g17haAwV7GjDMNGG9nyfTpKQPEZs9OyY4AVaqkBLd6N4ZcuRz7mkQe1Noam8BpWI3lDPKVbMH+JgOleJmfiCh9vDuYxYAt3c2bKYEoLvdDRMSDx0qXTvt5QkJSshd8+qnIyZMp3QXwN+XKGTbRKldO/Zp6lwLAa6MMlspG5EGBq7e3tlqaESo2SZOCOZhon4jIGbw7mEXQWalSSuspgklkMNAzCmzb9mA7TKBgi7CwlIFcgEwGmIBBRKJKlJAQPTht3Fjk669TlrduffC3eG29VRcpwthfltKA1rzK+SsbljMLA9fUKag4IxQRkWt5dzALAweKDBuWsjxgQMqsXJjNC/leoVixB/1gMbBr06YHg8L0ABWtsgh+W7VKGUi2Y4fIJ58YXuJE165SQ7+DyRKQhgvpufAamFChdu2U2caMy0SUBuSVPTL4SKamwIqO9Z5BWcaj/pmCiojIvTGYHTxYZNmylIkTjhwxTcGFfrBz51qfMEGHiRAwAAw3M0mvvSaXGzd+EMyiS8KcOQ9mABs/3vQPWrdmMEsuHZx1LOKuXL8bJ3fuJ2b5Pqsc9U9E5PkYzKKPLNJpff65yE8/pbS4YvAVugNgVq5atR7+HPXri7RvL7J/v8iNGyl/j+4Gr7wiyW3bpkxPax6woosBZhtD3tmYGJEyZUR69RJ54w1OZUuZOr3r8Yi7ZhMOZL1sAOyzSkSUdTGYBbS8jhqVckvLxo2W16OfLQJiS/Rpbs2hawFahInSOZ1t3Zkpk3zsGrDrodPZGre8bj51I0u1uoYF+Umx3CGG7gAVCuaQYgmXZFj3thIcZJb+joiIshwGs0QeCFPYHr1+1LCcVn/X8JvRcicLpMXS+7E+rEtAQkKCrFx5kf1aiYi8BINZoizAOHg9efWuR2cYyB3iL3mzBbAfKxER2YTBLJEHQreBoMRqEpxYXVp+tlWuRBlNuOFhgWv+HEFSqTAzBRARUfowmCXysNbXBTsvyNmbMVJIUjJheEIgi6A1X/ZATsNKREQOx2CWyE156qAttrYSEVFmYjBL5IYB7JT1J2XPuTtuP2gLg7KQSYCBKxERuYr3BrPR0ZbzuWJdcLDpdtb4+qZMgpDWtgkJ4hcbmzKxQkDAg/XILWthFLqC6UlDQ9O3LV4nOY3BP8iBm55t8R6SkhyzLcqrT8EaF/dgGt+Mbou6QJ0AJqSwlhbN1m31usN70evuYc+LY0c/rrAdtk8rJZy/f0r3gX9OyYKdF+XsbffN9xoW5CMl8uZIGZRVKrf0qVEw9aCs+zEp/2N/Bf5/WiwcYzjWrLFnW+SF1icxwWcCnw1r9YbjRa83a9um53Pv6HOEtW297Ryh1xv2Uc6cnnGOsPS5d8I5wu5tsQ+wL6zB503/bNizLeoMdWSt3lAX+mfZ0rbGXHmOSM+2We0ckWBUb9i37niOSGtfmNO8TGRkJGpTi0yp1tS39u1N/yA01PJ2uDVrZrptvnxWt02qXdt025IlrT9v5cqm2+K+tW3xPMbq1LG+LcpnDOW3ti3etzHsF2vbmh9Gzz6b9rb37j3Ytk+ftLe9du3BtoMHp71tePiDbUeMSHvbw4cfbDtmTJrbJmzd+mDbSZPSft4NGx5s+803FrdJ9PHVNpesrr366XKt7oertZJvL3e/24ilWuMB32kvPv2u9l2dTlqcr3/Ke9ft3Jn2fsA+1WFfp7Ut6kqHOkxrWxwDOhwbaWyb1KvXg21xzKX1vDhmjaW1rZPOEeqza4znCI85R6jPgwPPEYbb8uUPtp0zJ+1tFy16sC2W09oWz6XDa6S1LcqoQ9nT2tbDzhHq2NLxHOF25wjEaSpei4zUHsZ7W2aJMlm8r7+803qo/FH1cUn28xe5odaKO8gV7CcF/JKk0s4N8uyhddLo/EHx0zw3vRcREXkPH0S04kWioqIkZ86cEnn5soSFhTn98gASuP/999/Spl07CTB+PW+7hOiB3QwMddepkwTox4SdlxDjY2Jlzs4LMnvHRbl6L42/y2TZA3ylQsHs0qZKQenbpFxKdwF7Lgu68SVEQ7116CAB2bOnuW2WvYTogecIQ721aSMB7GbgMd0MTOqN3Qw85hyRYFxvbtrNIOr2bclZpIhERkZajteMeG/LLHaa8Y5Lazt7ntNcQoIk4aA2PgDB+MB5GHu2NX8dR21r/MF05LY4kegnE0duiw+nfvJL77Z63Rn3rbbhefVBXGOXH5b/rqdxcsxEeUIDpHS+bNKmSiHp+1hpyxMQ4H3aerzbsy1OwM7YFifrtD5zxseKtW2tcYdtve0codcb9pEenLr7OSKj2yJIMx5L4ahtEaTpga0jt7X0uTeuN+PyufM5IqPbgqefIxKs1Js7nSPs2BfeG8wSOZAewH6FLATn70iyi6935AoJkGaP5GeGASIiyvIYzBJlADIRvPP7Aflj/2WXB7Bl84VKt7olrLe8EhERZUEMZons5C7dCBLkqsT5HJX5vYZJ84rF2PpKREReicEskQe1wtYtmUtebfGIVC8RIjk/yaHWNSo3joEsERF5LQazRGlA0Lrtv5vy+dpTcuBiVKa/fu6QAGlqoe9rdLwdyaSJiIiyMAazRFa6EkxZe0qm7/CTpO17MvW1C+QIkhcbl2bfVyIiIhswmCUyD2LXnJRvNp7+/64Ezr98j1d4tGiYdKxeVPo0KmVzAJsvNJ/Ty0ZEROTuGMwSGfWHXbLvspo/LzP7v6YndVa2wGxy/c3rTisbERGRp2AwS+LtQWyv77fLjvDbmfJ6ZfNlk7FPVWHuVyIiIgdhMEteJ7NTa/n5iHSqXkQmPludfWCJiIgcjMEseV1/2KmbTktSGlNJOwJC1laVC0jvRqWlQZm8Dm+FvZ9wX9rNa6eW/+r5l4QE2DGlIBERURbCYJa8wp8HLstrC/c5PYjNSD9YeyRrybLp3CbDMhERkbdiMEtZvjvBiN/3y9WoeKd2IxjyeFkZ1qoC+8ESERFlMgazlGVbYocv2i8JSc7LTcDBXERERB4azP70008yY8YMCQ8Pl23btknJkiXlyy+/lNKlS0unTp0cX0oiO1pjn5u+VfZeuOOU50fI+kytIjLhGQ7mIiIicgd2fxtPnz5dhg8fLu3bt5c7d+5IUlKSWp8rVy4V0BK5Koid/PcJKTtqpVMCWXQlePWJsnJ6fHv5vGtNBrJERESe2jL79ddfy8yZM6Vz584yceJEw/o6derIiBEjHF0+IpcO7ioQlCyfdK8jTSsUYlcCIiKirBDMomtBzZo1U60PCgqS6OhoR5WLyKVdCuqXyi2z+9SWtatXSWM37RMbGhDq6iIQERF5XjCLfrH79+9X/WSNrVq1SipVquTIshFZDWK/XndKvlx3yuHP/UyNB5MbJCQkiLvCdLbRo/jjkYiIyO5gFv1lhwwZIrGxsaJpmuzcuVPmz58vEyZMkFmzZjmnlEROzFKANtdXnmBqLSIiIq8IZl988UUJCQmR9957T2JiYuR///ufFClSRKZMmSLdu3d3TinJ6zmjSwGDWCIiIi9NzdWzZ091QzB77949KVCggONLRuSkLgVZIYiNTYyVLou6qOXfu/4uwf7Bri4SERGR5wwAS0xMlPLly0toaKi6walTpyQgIEBKlSrljHKSF1p58IrqUhCb6Lg0Ba96eBCrS0pOkpWnVhqWiYiIvJXdyTL79u0rW7duTbV+x44d6jEiR/hw+VEZ/MtehwWy7asWlP/Gt5fhbSp6fCBLREREGWiZ3bdvnzz22GOp1jdo0ECGDh1q79MRpdJ/zk5Zf+K6Q56rfIFssuLVppzkgIiIKIuyO5j18fGRu3fvplofGRlpmA2MKD3iE5Ol6SfrJeJuXIafC22vX3evIU/WKOqQshEREZF7sru5qmnTpioNl3HgimWsa9y4sd0FmDp1qupnGxwcLPXr11epvtKCKXMrVKigMioUL15cXn/9dZUmjDy/W8Ej7/3lkEC2VvEwNe0sA1kiIqKsz+6W2U8++UQFtAgomzRpotb9888/EhUVJevXr7fruRYuXKjy1s6YMUMFsghU27RpIydOnLCYIeGXX36Rd955R2bPni2NGjWSkydPqn66aC2ePHmyvW+F3CRbQcvPN0r4zZgMP1eAr8gXXdkaS0RE5E3sbpmtXLmyHDx4ULp27SrXrl1TXQ569+4tx48fl6pVq9r1XAhABwwYIP369VPPi6AW2REQrFqCgWfor4vctmjNbd26tfTo0eOhrbnkvtkKyo1ameFAFl0KXmtRTo5/xNZYIiIib5OuPLOYJGH8+PEZeuH4+HjZs2ePjBw50rDO19dXWrZsKdu2bbP4N2iN/fnnn1XwWq9ePTlz5oysXLlSevXqZfV14uLi1E2HFmTAVKWZMV2p/hruPDWqK0xcdUK+//dchp+nRrEwWTCgvspQkJyUKI7MUuXOdRfoEyjxo+IN992xjK7izvVG1rHePBPrzTMleEC92VM2Hw1z0trpzp07KqBEy2xysmnqJLTS2uLy5ctStGhR1drasGFDw/q33npLNm3apFJ9WfLVV1/JiBEj1FS6yHc7cOBAmT59utXXGTt2rIwbN85ilwU9Ry5lrsXhPrIpAhcF0p8iy1c0eb5cstTO77hpbYmIiMg96LPMIsFAWFiYY1tm//zzTzX7F2b+wpOjv6oOy7YGs+mxceNG1SI8bdo01cf29OnTMmzYMPnwww9l9OjRFv8GLb/ol2vcMouBY+ii8LCd46hfFmvWrJFWrVqpSSW8vX9s9++2y/6I1Nkw7NGuSgH5omt1p+eLZd15JtabZ2K9eSbWm2dK8IB606+k28LuYPaNN96Q/v37q6AyIy2b+fLlEz8/P7l69arJetwvVKiQxb9BwIouBS+++KK6/+ijj0p0dLS89NJL8u6776puCuaCgoLUzRwqLzMrMLNfzx37xw79Za9kZAoExK7f9Kgp7asVkczkjnWH6Wx7LUnpXvPT0z9xOlsPqTd6ONabZ2K9eaYAN643e8pl9wCwS5cuyauvvprhS/SBgYFSu3ZtWbdunWEduizgvnG3A/MmZ/OAFQExpKO3BGWSj1ekzOaVnMF0W6c+bp/pgay7whS2vx39Td04nS0REXkzu1tmkTpr9+7dUqZMmQy/OC7/9+nTR+rUqaMGdCE1F1pakd0A0GUB/WqRwxY6duyoMiDUrFnT0M0ArbVYrwe15F7G/XlE5vx7NkOtsV91Y7otIiIiclAw26FDB3nzzTfl6NGj6jK/eTPwU089ZfNzdevWTa5fvy7vv/++RERESI0aNWTVqlVSsGBB9fj58+dNWmLfe+891S8X/6OFOH/+/CqQ/fjjj+19G+QB09KWyhMs60Y0d3rfWCIiIvKiYBZ5YeGDDz5I9RgCTXuntB06dKi6WRvwZczf31/GjBmjbuTenvxqsxy+nP6BXs0r5JPZ/eo7tExERESU9dgdzJqn4iIy12HKJjly5V66//6FxiVl9JP2TcBBRERE3ildkyYQOSOQ9fcV+ap75mcrICIiIi8LZjFICxMboE8rZvIyhkwH5J2ezEAgWzpviKx94wn2jyUiIiLnBrP79u2T9u3bqzRZCGrz5MkjN27cUKm6ChQowGDWS/WbvUMOpzOQZf9Y+4UGhMq9kfcMy0RERN7K7jyzr7/+usogcPv2bQkJCZHt27fLuXPnVM7Yzz77zDmlJLcPZDecvJHu/rEMZO2HwZbZArOpm/EsfERERN7G7mB2//79ahYwpMxCbte4uDg1PeykSZNk1KhRzikluXXWgvQEsugfO+1/NTnQi4iIiDI3mEVeWT33K7oVoN8s5MyZUy5cuJCx0pDH9ZFNT/qtUnlD5MRHnM0rI+IS46TvH33VDctERETeyu4+s5h9a9euXVK+fHlp1qyZmvAAfWZ/+uknqVqVrWzeov+c9PWRrVI4u6wY1swpZfImicmJ8sOBH9Ty1PZTJUiCXF0kIiIiz2iZHT9+vBQuXFgtY+at3Llzy6BBg9RMXt99950zykhuZsyyw7L+hP1dCxjIEhERkctbZuvUqWNYRjcDTD9L3uOFuTtl3XH7p6itWji7LGcgS0RERK5umSXvld5A9olH8jGQJSIiIte1zNaqVUvWrVunuhSgz2xaqYD27t3ryPKRm/hw+eF0B7Jz+jP1FhEREbkwmO3UqZMEBaUMMOncubOTikLuauXBy/L9lnN2/13VIjkYyBIREZHrg9kxY8ao/5OSkuSJJ56QatWqSa5cuZxbMnILScmavLpgn91/VypPsCx/talTykRERESUrgFgmCShdevWcuzYMQazXuLZ6f9KYrL9EyKsG9HcWUWi/5/C9tqIa4ZlIiIib2X3ADDkkj1z5oxzSkNul0t234VIu//um//VEj9fTrHqTOi3nj9bfnXjdLZEROTN7A5mP/roIxkxYoQsX75crly5IlFRUSY3yhpe/GGn3blk0SI74/la0rZqSh5iIiIiIrfLM9u+fXv1/1NPPWXSIqRpmrqPfrXk2ZbvvyRrj123+1fR0Q/aSSAiWnI6TGE7/O/hanlym8kS5M8ZwIiIyDvZHcxu2LDBOSUhtxnwNWzh/nR1LWAgm7nT2U7bPU0tT2o1idPZEhGR17I7mG3WjMnvs7LnZvwrSZp9f/NC49LSvhq7FhAREZEHBLO6mJgYOX/+vMTHx5usR9ou8tzuBXvP2zfgq0XF/DL6ycpOKxMRERGRQ4PZ69evS79+/eSvv/6y+Dj7zHpP94LmFfLL933rOa1MRERERA9jdyfH1157Te7cuSM7duyQkJAQWbVqlfzwww9Svnx5WbZsmb1PR27iuen2dS+oWSxMZvdjIEtEREQe1jK7fv16Wbp0qdSpU0d8fX2lZMmS0qpVKwkLC5MJEyZIhw4dnFNScpoPlx+WvXbkk/XzEfltcGOnlomIiIjIKS2z0dHRUqBAAbWcO3du1e0AHn30Udm7d6+9T0cutvLgZfl+yzm7/ubL7jU5KQIRERF5ZstshQoV5MSJE1KqVCmpXr26fPvtt2p5xowZUrgwR7R7XD/ZBfb1k61dIpd0rF7EaWUi24QEhEj4sHDDMhERkbeyO5gdNmyYmvkLxowZI23btpV58+ZJYGCgzJ071xllJCem4UpI1uzqXrBoYCOnlols4+vjK6VylXJ1MYiIiDwnmH322WflxRdflJ49expm/qpdu7acO3dOjh8/LiVKlJB8+fI5s6zk4jRc7F5AREREHttn9vbt22pwF4LW999/X86cOaPWh4aGSq1atRjIelj3gtcXHbDrb1pWKsDuBW4kPile3lz9prphmYiIyFvZHMyuW7dOBbAvvPCC/PzzzyoVV/PmzeWXX36RuLg455aSHOrrdSft6l5Qq3hOmdWnrlPLRPZJSEqQz7Z9pm5YJiIi8lZ2ZTNAGq6xY8eqoHbNmjVSpEgRGTBggBr4NWTIENmzZ4/zSkoOa5WduuE/m7f39xH5ddBjTi0TERERUaal5tKhVRYttBERESq/7IIFC6R+/frpLghljlfn77GrVfYL9pMlIiIiN2Z3NgNj4eHhKoMBbpGRkdKyZUvHlYycklN2xaGrNm/PNFxERESU5VpmY2NjVYssWmbRb/bHH39U/WgR2GJqW3Lf7gXD7Rj0hQODabiIiIgoy7TM7ty5U2bPni0LFy5UAe3TTz+tgtcWLVoYUnWRew/6ik1Mtnn7V1uUZ/cCIiIiyjrBbIMGDdSMXx9++KHKNYupbClrDvoK9veVV1qUd2qZiIiIiDI1mN29e7fKJ0tZPxXX5K412Crr5jCF7eFBhw3LRERE3srmYJaBrHe0yj75aGFpX62wU8tEjpnOtkqBKq4uBhERkeem5qKs1yqLnLJTetR0epmIiIiI3CI1F2WtVtmhzTnoy1NgCtvx/4xXy6OajJJAv0BXF4mIiMglGMxmYfa0ygb4+nDQlwfBFLbjNo1Ty282epPBLBEReS12M8ii7G2VHfJEObbKEhERUdZsma1Zs6bNuWT37t2b0TJRJk9by1ZZIiIiytLBbOfOnQ3LmDBh2rRpUrlyZWnYsKFat337djly5IgMHjzYeSUlp01by1ZZIiIiytLB7JgxYwzLL774orz66qtq8gTzbS5cuOD4EpJTp61lqywRERF5VZ/ZX3/9VXr37p1q/fPPPy+///67o8pFmTRtLVtliYiIyKuC2ZCQEPn3339Trce64OBgR5WL0tkqO30Tp60lIiIi72F3aq7XXntNBg0apAZ61atXT63bsWOHzJ49W0aPHu2MMpKNtp+5KXGJnLbWGwT7B8vOF3calomIiLyV3cHsO++8I2XKlJEpU6bIzz//rNZVqlRJ5syZI127dnVGGclGP28/Z/O2nLbWs/n5+kndonVdXQwiIiLPnDQBQSsDV/frYrD2aIRN23LaWiIiIvLqSRPu3Lkjs2bNklGjRsmtW7fUOnQ7uHTpkqPLR3bN9mXbtpy2NmtMZ/vpv5+qG5aJiIi8ld0tswcPHpSWLVtKzpw55ezZsypVV548eWTx4sVy/vx5+fHHH51TUnLIbF9MxZV1prN9a+1banlw3cGczpaIiLyW3S2zw4cPl759+8qpU6dMshe0b99eNm/e7Ojykc2tsrYN/GpZuQBbZYmIiMh7g9ldu3bJyy+/nGp90aJFJSLCtj6b5Lp0XM/XL+XU8hARERG5dTAbFBQkUVFRqdafPHlS8ufP76hykY2+WX/K5nRcyCvboGxep5eJiIiIyG2D2aeeeko++OADSUhIUPd9fHxUX9m3335bunTp4owyUhqtstM2nLZ5+4HNyrKLAREREXl3MPv555/LvXv3pECBAnL//n1p1qyZlCtXTnLkyCEff/yx3QWYOnWqlCpVSvW/rV+/vuzcmZIIPq1MCkOGDJHChQurVuJHHnlEVq5cKd7aVzYuyfZWWQ78IiIiIvH2bAbIYrBmzRrZsmWLymyAwLZWrVoqw4G9Fi5cqAaUzZgxQwWyX375pbRp00ZOnDihgmVz8fHx0qpVK/XYb7/9pvrpnjt3TnLlyiXe2Co7c0u4zdtzti8iIiLKitI1aQI0btxY3TJi8uTJMmDAAOnXr5+6j6B2xYoVampczDRmDuuR13br1q0SEBCg1qFV1xvtDL8l0XFJNm1bs3hOzvaVxWAK2w19NhiWiYiIvFW6gtl169ap27Vr1yQ5OTlVwGkLtLLu2bNHRo4caVjn6+urWni3bdtm8W+WLVsmDRs2VN0Mli5dqgac/e9//1P9df38/MSbrD5yxeZtR7Su6NSykGums3281OOuLgYREZHnBbPjxo1TA8Dq1Kmj+q1iAFh63LhxQ5KSkqRgwYIm63H/+PHjFv/mzJkzsn79eunZs6fqJ3v69GkZPHiwGow2ZswYi38TFxenbjo9EwP+Rh/E5kz6azjytdDF4Nc9F23aNiTAV2qXCMuU95rVOKPuyPlYb56J9eaZWG+eKcED6s2esvlommbbCKL/hwB20qRJ0qtXL8mIy5cvqz6v6DKA1lbdW2+9JZs2bZIdO3ak+hsM9oqNjZXw8HBDSyy6Knz66ady5YrllsqxY8eqANzcL7/8IqGhoeKJTkX6yDdHbWuJblooSbqUtquKyQMkaomy+sZqtdw6X2vx90l3jyEiIiK3ExMTo66+R0ZGSlhYWJrb2v0NiO4BjRo1kozKly+fCkivXr1qsh73CxUqZDWQRl9Z4y4FlSpVUpM1oFyBgamn9EQ3BgwyM26ZLV68uLRu3fqhO8dRvywwYA4D1/R+vhn10YpjInLBpm1fal9f6pfO45DX9TbOqDtHiY6Plmc/e1Ytf9LjE8kWmM3VRXIb7lxvZB3rzTOx3jxTggfUm6U5DRwWzL744ouqVXP06NGSEQg8a9eurfredu7cWa1D/1vcHzp0qMW/eeyxx9RrYzv0r9Una0CQaymQBaTvws0cKi8zK9BRr4cuBov3X7Zp27Bgf2lYjtPXZlRmHyu2CNAC3Lp87oD7xTOx3jwT680zBbhxvdlTLruDWVzm/+6772Tt2rVSrVq1VC+Gy/62Qotpnz59VP/bevXqqdRc0dHRhuwGvXv3Vl0RJkyYoO4PGjRIvvnmGxk2bJi88sorcurUKRk/fry8+uqr4k1ZDO7G2pbFoEutogxkiYiIKEuzO5hFbtkaNWqo5cOHD5s8Zu9gsG7dusn169fl/fffV10F8LyrVq0yDArDzGJ6Cyyge8Dff/8tr7/+ugqkEegisEU2A29hTxaD1lWYjouIiIiyNruD2Q0bUnJbOgq6FFjrVrBx48ZU6zBYbPv27eKN0MVg/q4LNncxqMe+skRERJTF2T2dLbnON+tPSWyCaV5fa9jFgIiIiLyBTS2zzzzzjMydO1eN/sdyWhYvXuyospFZq+y3m8/YvD27GBAREZE3sCmYzZkzp6E/LJYp820/c1Ni4m0b+JU3WyC7GGRxQf5BsrzHcsMyERGRt7IpmJ0zZ47FZco8P28/Z/O2H3aqyi4GWZy/r790eKSDq4tBRETkcuwz6yFdDDafvG7TtjWL55T21djFgIiIiLxDuubA/O2332TRokUqdRZm3jK2d+9eR5WNjHLLRtvYxWBE64pOLw+5XkJSgsw7NE8t93y0pwT4uWfSayIiIrdrmf3qq6/UpAbIBbtv3z412UHevHnlzJkz0q5dO+eU0stFRMXatF1ooJ80KJvX6eUh14tPipd+S/upG5aJiIi8ld3B7LRp09QMYF9//bWaQvatt95S8/tiFq7IyEjnlNLL/XvKti4G7asWYl9ZIiIi8ip2B7PoWtCoUSO1HBISInfv3lXLvXr1kvnz5zu+hF4O/WVXHLJt1q/HyuVzenmIiIiIPDqYLVSokNy6dUstlyhRwjAbV3h4uGia5vgSejmk5Lpv40QJhXKGOL08RERERB4dzDZv3lyWLVumltF39vXXX5dWrVpJt27d5Omnn3ZGGb2arSm5sgdx+loiIiLyPnZnM0B/2eTklJbCIUOGqMFfW7dulaeeekpefvllZ5TRa9mTkqtJ+bzsL0tERERex+5g1tfXV9103bt3VzdybUqu5+uXcnp5iIiIiDwymD148KDNT1itWrWMlIeMMCUXWYMpbBc9u8iwTERE5K1sCmZr1KghPj4+Dx3ghW2SkmxrSaSHY0ouSms62+eqPOfqYhAREXlGMItMBZT5/WXXHL1q07ZMyUVERETeyqZgtmTJks4vCaXqLxsZm2jTtkzJ5X0SkxNlybElavnpSk+rlloiIiJvlK5vwBMnTqgZwI4dO6buV6pUSV555RWpUKGCo8vntWztL5srJIApubxQXGKcdP2tq1q+N/Ke+AcymCUiIu9kd57Z33//XapWrSp79uyR6tWrq9vevXvVOjxGmdtftmWlAuwvS0RERF7L7uact956S0aOHCkffPCByfoxY8aox7p06eLI8nkl9pclIiIiclLL7JUrV6R3796p1j///PPqMco49pclIiIiclIw+/jjj8s///yTav2WLVukSZMm9j4dWbD2aIRN2+UKZX9ZIiIi8m52dzPAtLVvv/226jPboEEDtW779u3y66+/yrhx42TZsmUm25L9XQyW7L9k07b9GpVmf1kiIiLyanYHs4MHD1b/T5s2Td0sPQacQCH9XQxuRSc8dLvsQf4ytHm5TCkTERERUZYJZpOTk51TErIrJddzdYqxVdaLBfoFypxOcwzLRERE3sqhySljYmIkNDTUkU/pdWxNyVUsFwd+ebMAvwDpW6Ovq4tBRETkeQPAWrRoIZcupe7TuWPHDqlRo4ajyuWV7EnJlScbW+OIiIiI7A5mg4ODpVq1arJw4UJDt4OxY8eqTAbt27d3Rhm9BlNykT3T2a44uULdsExEROSt7O5msGLFCpk6dar0799fli5dKmfPnpVz587J8uXLpXXr1s4ppZe4dtfGKWyZksvrYTrbJ+c/qZY5nS0REXmzdH0DDhkyRC5evCiffPKJ+Pv7y8aNG6VRo0aOL52XyZc9yKbt+jYsxcFfREREROnpZnD79m01Ze306dPl22+/la5du6oWWfM0XWS/neE3bdqubim2yhIRERGlq2W2atWqUrp0adm3b5/6f8CAAar/LHLMogsCbpS+wV8/bD1n07Y3ouOcXh4iIiKiLNkyO3DgQNm8ebMKZHXdunWTAwcOSHx8vKPL51WDv+7cf/hkCVAgR7DTy0NERESUJVtmR48ebXF9sWLFZM2aNY4ok1eydbKEXCEc/EVERERkd8vspEmT5P79+4b7//77r8TFPbjcfffuXZPpbMk+t+7Z1nWgZaUCHPxFREREZG8wO3LkSBWw6tq1a2cyeQJm/8KAMEqfi7djbNrusXL5nF4Wcn+Ywvabdt+oG6ezJSIib2ZzNwNN09K8Txkb/LX0wGWbtuVkCaRPZzuk3hBXF4OIiMjzBoCRcwZ/3Yp++OCvvNkC2V+WiIiIyAinDfKgmb861SjC/rKkJCUnyT/n/1HLTUo0ET9fP1cXiYiIyP2D2VmzZkn27NnVcmJiosydO1fy5Uvpw2ncn5acM/NXi0oFnV4W8gyxibHyxA9PGKazzRaYzdVFIiIicu9gtkSJEjJz5kzD/UKFCslPP/2UahtKB1u7H7ObMhEREVH6gtmzZ8/auinZaf3xqzZtx5m/iIiIiExxAJgbZDJYsv9BirO0cOYvIiIiIlMMZl2MmQyIiIiI0o/BrIutPRph03bMZEBERESUGoNZD+li0KpyIaeXh4iIiMjTMM+sC7GLAWVkBrBJLScZlomIiLxVuoLZ//77T+bMmaP+nzJlihQoUED++usvlZqrSpUqji9lFsXJEii9Av0C5c3H3nR1MYiIiDyvm8GmTZvk0UcflR07dsjixYvl3r17av2BAwdkzJgxzihjlsXJEoiIiIgyOZh955135KOPPpI1a9ZIYGCgYX3z5s1l+/btGSyOl+FkCZSB6Wx3XdqlblgmIiLyVnZ3Mzh06JD88ssvqdajq8GNGzccVS6vwMkSKCPT2dabVU8tczpbIiLyZna3zObKlUuuXLmSav2+ffukaNGijipXlsfJEoiIiIhcEMx2795d3n77bYmIiBAfHx9JTk6Wf//9V0aMGCG9e/d2QJG8AzMZEBEREbkgmB0/frxUrFhRihcvrgZ/Va5cWZo2bSqNGjWS9957zwFF8g7MZEBERETkgj6zGPQ1c+ZMGT16tBw+fFgFtDVr1pTy5cs7oDjeg5kMiIiIiFwQzG7ZskUaN26scsriRunETAZEREREmd/NACm4SpcuLaNGjZKjR49mvAQiMnXqVClVqpQEBwdL/fr1ZefOnTb93YIFC1S/3c6dO4unYSYDIiIiIhcEs5cvX5Y33nhDTZ5QtWpVqVGjhnz66ady8eLFdBVg4cKFMnz4cDXhwt69e6V69erSpk0buXbtWpp/d/bsWTXorEmTJuJpmMmAMgpT2I5pNkbdOJ0tERF5M7uD2Xz58snQoUNVBgNMZ/vcc8/JDz/8oFpW0Wprr8mTJ8uAAQOkX79+ajDZjBkzJDQ0VGbPnm31b5KSkqRnz54ybtw4KVOmjHgaZjIgR0xnO/bxseqGZSIiIm9ldzBrDN0NMCPYxIkT1RS3aK21R3x8vOzZs0datmz5oEC+vur+tm3brP7dBx98oCZpeOGFF8QTMZMBERERkYsGgOnQMjtv3jz57bffJDY2Vjp16iQTJkyw6zkwYxhaWQsWNB2xj/vHjx+3OgDt+++/l/3799v0GnFxceqmi4qKUv8nJCSom7Ppr2H8WnlDbdvtzSvky5Qyku115y6StWQ5duOYWq6Ur5L4+mTod2mW4s71Rtax3jwT680zJXhAvdlTNruD2ZEjR6qBV+g726pVK5kyZYoKZNE1wNnu3r0rvXr1UqnB0N3BFgiw0R3B3OrVqzOlzLo1a9YYlhOTRXzE7/8TFVhqedXU2qtHtsvKlHiFXMi47txFbFKsdD/UXS0veHSBBPuxb7Un1Bs9HOvNM7HePNMaN663mJgY5wWzmzdvljfffFO6du1qc0BpDf7ez89Prl41HdmP+4UKFUq1PfroYuBXx44dDeswAxn4+/vLiRMnpGzZsqmCbwwwM26ZxYQPrVu3lrCwMMmMXxY4WBD4BwSkDNTZEX5LtB270/grHxXoFqzSQOqzz6zLWKo7dxEdHy1yKGUZAyazBWZzdZHchjvXG1nHevNMrDfPlOAB9aZfSXdKMIvuBY6CCRhq164t69atM6TXQnCK+xhkZg4zjx069P/f4P8Ps46hxRYtxAhSzQUFBambOVReZlag8ettOHHDpr+5GZPotgeZN8nsY8UWAVqAW5fPHXC/eCbWm2divXmmADeuN3vKZVMwu2zZMmnXrp16Yiyn5amnnhJ7oNW0T58+UqdOHalXr558+eWXEh0drbIbQO/evaVo0aKquwDy0CIdmLFcuXKp/83Xuyum5SIiIiJyHJuCWbSaRkREqAwCaU1QgAkMMKDLHt26dZPr16/L+++/r14DeWtXrVplGBR2/vx5leEgq2BaLiIiIqJMDmb1fqnmy46CLgWWuhXAxo0b0/zbuXPniidhWi4iIiIix7G7yfPHH380SXVlnDMWj5Fjug60qpx6ABwRERERZTCYRV/WyMjIVOsxCEvv50rW1S6ZWx7W4IrHsR2RNZjCdkTDEerG6WyJiMib2Z3NQNM01TfW3MWLFyVnzpyOKleWtefcbUlOSTBrFR7Hdg3L5s2sYpGHwRS2n7b+1NXFICIi8pxgtmbNmiqIxa1FixYqr6sOg77Cw8Olbdu2zipnlmFrn1lbtyMiIiLyZjYHs3oWA0wjiyTt2bNnN8kXW6pUKenSpYtzSumFfWaZloseNp3t+cjzarlEzhKczpaIiLyWzcHsmDFj1P8IWpFOCzlfyX63o1MPnjNXOGcw03JRmu4n3JfSU0qr5Xsj73EGMCIi8lp295nFBAeU/gkTPlxx7KHbje5QmWm5iIiIiJwRzKJ/7BdffCGLFi1SExogJZexW7du2fuUXgMTJlyJfHhf2NzZAjOlPERERESezu6OduPGjZPJkyerrgZI0YXpaJ955hk1S9fYsWOdU8osgoO/iIiIiFwczM6bN09mzpwpb7zxhspo0KNHD5k1a5aajnb79u0OLl7WwsFfRERERC4OZiMiIuTRRx9Vy8hooE+g8OSTT8qKFSscXLyshRMmEBEREbk4mC1WrJhcuXJFLZctW1ZWr16tlnft2iVBQUEOLp73TphARERERE4YAPb000/LunXrpH79+vLKK6/I888/L99//70aDPb666/b+3RehX1myVH8ff1lcJ3BhmUiIiJvZfe34MSJEw3LGARWokQJ2bZtm5QvX146duzo6PJlKewzS44S5B8kUztMdXUxiIiIXC7DTToNGzZUN3o4TphARERE5IJgdtmyZTY/4VNPPZWR8mRZnDCBHEnTNLkRc0Mt5wvNJz4+PGaIiMg72RTMdu7c2aYnwxcqJlWg1Hafu80JE8hhYhJipMBnBdQyp7MlIiJvZlMwm5yc7PySZHHX7sbZuB0HfxERERE5LTUXpU+BHLalLePgLyIiIiInDgD74IMP0nwcM4FRanVK5pZcoQFyJybB4uPo8ViIg7+IiIiInBvMLlmyxOR+QkKChIeHq6ltMYkCg1nL1h67ZjWQBcylMKYjB38REREROTWY3bdvX6p1UVFR0rdvXzWhAlme1WvCyuNpboNW21aVC2VamYiIiIiyAof0mQ0LC5Nx48bJ6NGjHfF0Wc5/UT4SEZX2ADC02u4Mv5VpZSIiIiLKChw2D2ZkZKS6UWpR1nsXmGAmA7IVprDtU72PYZmIiMhb2f0t+NVXX6VK3n7lyhX56aefpF27do4sW5YRFmDbdsxkQPZMZzu381xXF4OIiMjzgtkvvvjC5L6vr6/kz59f+vTpIyNHjnRk2bKMsmGa5AoJkDv3rTfR5g4NYCYDIiIiImcHs8hcQOmhZeBRIkl1RQSzgEFoQCinsyUiIq/FSRMyaQDYnfuJaW7DAWBkDwSy2SdkVzc9qCUiIvJGdrfMxsbGytdffy0bNmyQa9eupZrqdu/evY4sX5bAAWBEREREbhLMvvDCC7J69Wp59tlnpV69ery8aQMOACMiIiJyk2B2+fLlsnLlSnnsscecUyIvHADGqWyJiIiIMqnPbNGiRSVHjhzpfDnvdOgW+sxyKlsiIiIilwezn3/+ubz99tty7tw5hxcmK0pK1mTx2bR3M6eyJSIiIsqkbgZ16tRRg8DKlCkjoaGhEhBg2iH01i2OyDe2+9xtuRPvY1Mmg4Zl82ZauYiIiIi8Mpjt0aOHXLp0ScaPHy8FCxbkALCHuHY3zsbtmMmAbOfn6yfPVn7WsExEROSt7A5mt27dKtu2bZPq1as7p0RZTIEcQTZux0wGZLtg/2D59blfXV0MIiIiz+szW7FiRbl//75zSpMF1SmZW3IFaipjgSVYX5iZDIiIiIgyJ5idOHGivPHGG7Jx40a5efOmREVFmdzIFDIUPFMq2ep0tcxkQERERJSJ3Qzatm2r/m/RokWqueLRfzYpKSkDxSEiW0THR6upbOHeyHuSLTCbq4tERETkGcEsprElx6XmQnvsuD+PqtRcbJ0lIiIicnIw26xZM3v/xKs9LDUXuhlciYxlai4iIiKizAhmN2/enObjTZs2TU85siym5iIiIiJyo2D28ccfT7XOONcs+8yaYmouIiIiIjfKZnD79m2T27Vr12TVqlVSt25dWb16tXNK6cGYmouIiIjIjVpmc+bMmWpdq1atJDAwUIYPHy579uxxVNmyBAzqqpU3WdZfsT5LE1NzEREREWVSMGsNprY9ceKEo54uy/j7yFVZf8V6A/hLTUtL26qFM7VM5PkwhW378u0Ny0RERN7K7mD24MGDqfLLXrlyRU2mUKNGDUeWLUuk5fpo5fE0t1l24Iq81bYSW2bJ7ulsV/xvhauLQURE5HnBLAJWDPhCEGusQYMGMnv2bEeWzeMh3VZEFLIZWA9UmZaLiIiIKBOD2fDwcJP7vr6+kj9/fgkO5mj89KbbYlouIiIiokwKZkuWLJnOl/I+tqbbYlouSs90tgU+K6CWr424xulsiYjIa9mcmmv9+vVSuXJliYqKSvVYZGSkVKlSRf755x9Hl8+jId1WoTDkmTXtkqFjWi7KiJiEGHUjIiLyZjYHs19++aUMGDBAwsLCLKbrevnll2Xy5MmOLp9Hw6Cu99pXVMvmvWb1+0zLRURERJQJweyBAwekbdu2Vh9v3bo1c8xa0KZKQWleOFmMJklTcJ9puYiIiIgyKZi9evWqBAQEWH3c399frl+/nsHiZN08s8lmPQ1w/7vN4bLq8BVXFY2IiIjIe4LZokWLyuHDh9PMP1u4MFsZ7c0zO+7Po2o7IiIiInJiMNu+fXsZPXq0xMamTiN1//59GTNmjDz55JPpKIL35pnVjPLMEhEREZETU3O99957snjxYnnkkUdk6NChUqFCBbX++PHjMnXqVElKSpJ33303HUXIuphnlpzF18dXmpVsZlgmIiLyVjZ/CxYsWFC2bt0qVatWlZEjR8rTTz+tbqNGjVLrtmzZorZJDwTDpUqVUhMv1K9fX3bu3Gl125kzZ0qTJk0kd+7c6tayZcs0t3cl5pklZwkJCJGNfTeqG5aJiIi8la+9EyasXLlSbty4ITt27JDt27erZawrXbp0ugqwcOFCGT58uOqmsHfvXqlevbq0adNGrl27ZnH7jRs3So8ePWTDhg2ybds2KV68uMqkcOnSJXE3zDNLRERE5Fzpuj6JFtG6detKvXr11HJGIDct8tf269dPTcowY8YMCQ0NldmzZ1vcft68eTJ48GCpUaOGVKxYUWbNmiXJycmybt06cec8s+aYZ5aIiIjIBdPZOlJ8fLzKTYtuCzpfX1/VdQCtrraIiYmRhIQEyZPHcutmXFycuun0GczwN7g5W/NH8kj/R5JlfniA3E9INqwvlDNI3m1XUVpUyJcp5SD76fXijvWD6WzLTy2vlk8NOcXpbD2k3sg61ptnYr15pgQPqDd7yubSYBZdFDBwzLyvLe5jYJkt3n77bSlSpIgKgC2ZMGGCjBs3LtX61atXqxbgzPBoHpHNVxLldIKvPBKWLK2KalIuZ7QkndsjK89lShEoA9asWSPuJjYpVm7cv6GW//77bwn2Y79rT6g3ejjWm2divXmmNW5cb2is9IhgNqMmTpwoCxYsUP1oMXjMErT6ok+uccus3s/W0tS8jrby4GUZt/SQ3IlP6UpwMspXoiRI3qteUc0ORuLWvwrxQW/VqlWaE4a4qmVWDqUso485W2Y9o97IOtabZ2K9eaYED6g3/Uq62wez+fLlEz8/PzW7mDHcL1SoUJp/+9lnn6lgdu3atVKtWjWr2wUFBambOVSesysQs3u99uvhVMO/rkbFySsLDsj052txOlsPkBnHir0CtAC3Lp874H7xTKw3z8R680zuXG/2lMulCSoDAwOldu3aJoO39MFcDRs2tPp3kyZNkg8//FBWrVolderUEXeEWb0wu1dKIGs6wEsPbjn7FxEREVHGuDzbOroAIHfsDz/8IMeOHZNBgwZJdHS0ym4AvXv3Nhkg9sknn6iZyJDtALlpIyIi1O3evXviTjCrF2b3soazfxERERFlnMv7zHbr1k2uX78u77//vgpKkXILLa76oLDz58+rDAe66dOnqywIzz77rMnzIE/t2LFjxV1w9i8iIiIiLwhmAdPj4mYJBncZO3v2rHgCzv5FzoQpbOsUSeliw+lsiYjIm7lFMJsVYVYvzO4VERlrcf4v9KItxNm/KJ0whe2uAbtcXQwiIiKXY5OOk2BWL8zulcI0nOXsX0RERESOwWDWiZB26+vu1SXMLLsEWmSZlouIiIgo49jNwMkwMcKFo0nyyUF/CQ30k+/71FVdC9giSxkRkxAjlaemtPwfHXJUQgMyZzY7IiIid8NgNhPEJ6f8nydboDQsm9fVxaEsQNM0ORd5zrBMRETkrdjNIBPEJKa0wiYna7Ltv5ucKIGIiIjIQRjMOtnfR67KT6dTdvPlyFjpMXO7NP5kvZrqloiIiIgyhsGsEyFgfWXBAYlJNF2PdF2Dft7LgJaIiIgogxjMOgm6Eoz78+j/J+UyHeyldzLA4+xyQERERJR+DGadZGf4LbkSaX2qWoSweBzbEREREVH6MJuBk1y7G+vQ7YiM+fj4SOX8lQ3LRERE3orBrJMUyBHs0O2IjCGv7JHBR1xdDCIiIpdjNwMnwcQIhXMGm/WWfQDr8Ti2IyIiIqL0YTDrJJjha0zHymZDvlLoAS4e50xgREREROnHYNaJ2lYtLF93ry4BZnu5UM5gmf58LfU4UXqns60yrYq6YZmIiMhbsc+sk7WpUlDK5dDkWKSP9KhXXJ6qXlR1LWCLLGUEprA9ev2oYZmIiMhbMZjNBIn/H2s0KJNXGpbN6+riEBEREWUZ7GaQCRKSU1phQwL8XF0UIiIioiyFwayTYYavuwkpy2dvRnPGLyIiIiIHYjDrRKsOX5HHP98sN+NSWmbHrzwujT9Zr9YTERERUcYxmHUSBKyDft4rEVFxJusjImPVega0RERERBnHYNYJ0JVg3J9HzbLLptDX4XF2OaD0whS2JXOWVDdOZ0tERN6M2QycYGf4LbkSGWv1cYSweBzbMbsBpXc627OvnXV1MYiIiFyOLbNOcO1urEO3IyIiIiLLGMw6QYEcwQ7djoiIiIgsYzDrBJjhq3DOYLHWkxHr8Ti2I0qP+wn3pe7MuuqGZSIiIm/FYNYJMFXtmI6VLT6mB7h4nFPaUnola8my+/JudcMyERGRt2Iw6yRtqxaW6c/XknzZA03WF8oZrNbjcSIiIiLKGGYzcCIErPmzBUiXb3dIWLC/fNurjupawBZZIiIiIsdgMOtkCUkpl4BzhwYyDRcRERGRg7GbgZPFJqYEs8EB3NVEREREjsYIy8li45PU/8EBfq4uChEREVGWw24GTnY/ISWYDWHLLDlYvtB8ri4CERGRyzGYzaRuBkFsmSUHyhaYTa6/ed3VxSAiInI5Nhc6UVKyJsevRKnlmPgkdZ+IiIiIHIfBrJOsOnxFGn+yXn7YfkHd33X2trqP9URERETkGAxmnQAB66Cf98qVyFiT9RGRsWo9A1rKKExh+/jcx9WN09kSEZE3Y59ZB0NXgnF/HhVLHQqwDtMl4PFWlQtx8gRKN0xhu+ncJsMyERGRt2LLrIPtDL+VqkXWPKDF49iOiIiIiDKGLbMOdu1urEO3I0qvfVf2ycIjC2Xzuc1yLvKcXI++LjmDc0qDYg3krUZvSZOSTdR2G89ulCd+eCLN5yqZs6Scfe2sxccm/TtJ3l77tuH+9A7TZWCdgQ5+N0RERJYxmHWwAjmCHbodUXp9u+dbdTN2I+aGLD+5XFaeWim/PverPFPpGZueK8AvwOL607dOy9iNYx1SXiIiovRgMOtg9UrnkcI5g9VgL0v9ZtFLtlDOYLUdkbMVyl5IXqj5gjQu0Vhu378t4zaNkxM3T6h+tsP/Hq6C2ZqFaso//f5J9beTt02WJceXqOXOFTpbfP6X/nxJ7ifel2D/YIlN5NUGIiLKfAxmHQyDusZ0rKyyFiBwNQ5o9eFeeJyDv8jZnq/2vExuM1lCA0IN6yrnryw1vq2hltH14Fr0NSmQrYAKdo3FJcbJP+dTAlxfH18ZVHdQqueftXeWbDi7QaoWqCrVClaTXw794vT3REREZI4DwJygbdXCMv35WqoF1hjuYz0eJ8ooBKnGgao5BKjmj5fPWz7Vc1iy6Mgi1SUB2pZrK2VylzF5/MrdK/LmmjdVoDur4ywJ8LXcDYGIiMjZ2DLrJAhYkX5r2+lrsvqfHdK6SX1pWK4AW2TJYdPZRo+Ktvvvfj/6u2G5SYkmkj0wu8Xtpu2eZlgeUndIqseH/jVU7sTekdfqvyb1i9WX6bun210WIiIiR2Aw60QIXOuXziM3j2nqfway5Ep7Lu+RV/56RS0H+QXJF22+sJoFYfvF7WoZLbJomTW25NgSWXxssZTKVUo+av5RJpSciIjIOnYzIPICW85vkeY/NpfIuEjx9/WX+V3mS+0itS1uO3XXVMPyoDqDVFcCY0NWprTUfvvkt6qFmIiIyJXYMkvkgZA5oMuiLmr5966/q2wC1qz+b7U8vfBpiUmIUS2yC59dKJ0qdrK4LboOzD88Xy2H+IdI/5r9U21z5V7KdMxtfm5j8TkGrRikbrffvi25gnOl6/0RERHZisEskQdKSk5SuWL1ZWvQJaD7790lPilesgVkk6Xdl0qLMi2sbj93/1wV9EL3qt0lTwhTyBERkXtjMEuURf165Ffp8XsPSdKSxEd8ZEyzMRLkH6S6HOjqFqmr1oGmaSYDuSwN/AJLfW2RlmvX5V1q+bnKz0mj4o1Uyy4REZGzMZglyqJWnFqhAlnQRJO31r6VapvwYeFqIBesPbNWTt48qZbrF61vtU/taw1eS7Vuf8R+QzDbvHRzTmdLRESZhgPAiMimdFxERETuiC2zRFnU3M5z1c1WS7otybTXIiIichS2zBIRERGRx/K6llkMcoGoqKhMeb2EhASJiYlRrxcQwCk/PYk71110fLRIbMoyypcUaD2jgbdx53oj61hvnon15pkSPKDe9DhNj9vS4qPZslUWcvHiRSlevLiri0FERERED3HhwgUpVqxYmtt4XTCbnJwsly9flhw5coiPj0+m/LJA8IzKCAsLc/rrkeOw7jwT680zsd48E+vNM0V5QL0hPL17964UKVJEfH3T7hXrdd0MsEMeFuE7Aw4Wdz1gKG2sO8/EevNMrDfPxHrzTGFuXm85c+a0aTsOACMiIiIij8VgloiIiIg8FoNZJwsKCpIxY8ao/8mzsO48E+vNM7HePBPrzTMFZbF687oBYERERESUdbBlloiIiIg8FoNZIiIiIvJYDGaJiIiIyGMxmCUiIiIij8Vg1smmTp0qpUqVkuDgYKlfv77s3LnT1UXyahMmTJC6deuqGeAKFCggnTt3lhMnTphsExsbK0OGDJG8efNK9uzZpUuXLnL16lWTbc6fPy8dOnSQ0NBQ9TxvvvmmJCYmZvK78U4TJ05Us/e99tprhnWsM/d16dIlef7551XdhISEyKOPPiq7d+82PI4xyO+//74ULlxYPd6yZUs5deqUyXPcunVLevbsqZK758qVS1544QW5d++eC96Nd0hKSpLRo0dL6dKlVZ2ULVtWPvzwQ1VXOtab623evFk6duyoZsjCOfGPP/4wedxRdXTw4EFp0qSJimMwa9ikSZPE7SCbATnHggULtMDAQG327NnakSNHtAEDBmi5cuXSrl696uqiea02bdpoc+bM0Q4fPqzt379fa9++vVaiRAnt3r17hm0GDhyoFS9eXFu3bp22e/durUGDBlqjRo0MjycmJmpVq1bVWrZsqe3bt09buXKlli9fPm3kyJEuelfeY+fOnVqpUqW0atWqacOGDTOsZ525p1u3bmklS5bU+vbtq+3YsUM7c+aM9vfff2unT582bDNx4kQtZ86c2h9//KEdOHBAe+qpp7TSpUtr9+/fN2zTtm1brXr16tr27du1f/75RytXrpzWo0cPF72rrO/jjz/W8ubNqy1fvlwLDw/Xfv31Vy179uzalClTDNuw3lwP57F3331XW7x4MX5laEuWLDF53BF1FBkZqRUsWFDr2bOn+t6cP3++FhISon377beaO2Ew60T16tXThgwZYriflJSkFSlSRJswYYJLy0UPXLt2TZ0ENm3apO7fuXNHCwgIUCdv3bFjx9Q227ZtM5xAfH19tYiICMM206dP18LCwrS4uDgXvAvvcPfuXa18+fLamjVrtGbNmhmCWdaZ+3r77be1xo0bW308OTlZK1SokPbpp58a1qE+g4KC1JcmHD16VNXlrl27DNv89ddfmo+Pj3bp0iUnvwPv1KFDB61///4m65555hkV0ADrzf2IWTDrqDqaNm2aljt3bpPzJD7XFSpU0NwJuxk4SXx8vOzZs0c16+t8fX3V/W3btrm0bPRAZGSk+j9Pnjzqf9RZQkKCSb1VrFhRSpQoYag3/I9LpQULFjRs06ZNG4mKipIjR45k+nvwFuhGgG4CxnUDrDP3tWzZMqlTp44899xzqmtHzZo1ZebMmYbHw8PDJSIiwqTuMBc7umQZ1x0uf+J5dNge59MdO3Zk8jvyDo0aNZJ169bJyZMn1f0DBw7Ili1bpF27duo+6839hTuojrBN06ZNJTAw0OTcie55t2/fFnfh7+oCZFU3btxQ/Y6MvzwB948fP+6yctEDycnJqt/lY489JlWrVlXr8OHHhxYfcPN6w2P6NpbqVX+MHG/BggWyd+9e2bVrV6rHWGfu68yZMzJ9+nQZPny4jBo1StXfq6++quqrT58+hn1vqW6M6w6BsDF/f3/1A5R15xzvvPOO+qGHH4V+fn7qu+zjjz9WfSuB9eb+IhxUR/gffafNn0N/LHfu3OIOGMySV7f0HT58WLU4kPu6cOGCDBs2TNasWaMGIJBn/WBEq8/48ePVfbTM4jM3Y8YMFcySe1q0aJHMmzdPfvnlF6lSpYrs379f/fDHQCPWG7kjdjNwknz58qlftOYjqnG/UKFCLisXpRg6dKgsX75cNmzYIMWKFTOsR92gi8idO3es1hv+t1Sv+mPkWOhGcO3aNalVq5ZqNcBt06ZN8tVXX6lltBKwztwTRlFXrlzZZF2lSpVUZgnjfZ/WeRL/o/6NIQsFRmGz7pwDmT7QOtu9e3fVPadXr17y+uuvq2wwwHpzf4UcVEeecu5kMOskuIxWu3Zt1e/IuJUC9xs2bOjSsnkz9JNHILtkyRJZv359qssnqLOAgACTekPfIHz56vWG/w8dOmRyEkCrIVKbmH9xU8a1aNFC7W+0Duk3tPbhkqe+zDpzT+jCY576Dv0wS5YsqZbx+cMXonHd4fI2+usZ1x1+qOBHjQ6fXZxP0f+PHC8mJkb1mzSGxhnsc2C9ub/SDqojbIMUYBiXYHzurFChgtt0MVBcPQItq6fmwsjBuXPnqlGDL730kkrNZTyimjLXoEGDVKqSjRs3aleuXDHcYmJiTNI8IV3X+vXrVZqnhg0bqpt5mqfWrVur9F6rVq3S8ufPzzRPmcg4mwGwztw3lZq/v79K9XTq1Clt3rx5WmhoqPbzzz+bpA/CeXHp0qXawYMHtU6dOllMH1SzZk2V3mvLli0qqwVTPDlPnz59tKJFixpScyH1E1LZvfXWW4ZtWG/ukeEFqQb37dunshJMnjxZLZ87d85hdYQMCEjN1atXL5WaC3ENPsNMzeVlvv76a/Uli3yzSNWFXG7kOvjAW7oh96wOH/TBgwerdCT40D799NMq4DV29uxZrV27dirfHk7yb7zxhpaQkOCCd+SdzINZ1pn7+vPPP9UPCfywr1ixovbdd9+ZPI4UQqNHj1ZfmNimRYsW2okTJ0y2uXnzpvqCRa5TpFPr16+f+iIn54iKilKfL3x3BQcHa2XKlFH5TI3TM7HeXG/Dhg0Wv8/69Onj0DpCjlqk2MNz4EcOgmR344N/XN06TERERESUHuwzS0REREQei8EsEREREXksBrNERERE5LEYzBIRERGRx2IwS0REREQei8EsEREREXksBrNERERE5LEYzBKRVzt79qz4+PioqXHdxfHjx6VBgwYSHBwsNWrUcHVxiIjcGoNZInKpvn37qmBy4sSJJuv/+OMPtd4bjRkzRrJlyyYnTpwwmVvdXEREhLzyyitSpkwZCQoKkuLFi0vHjh3T/BtvPcY6d+7s6mIQkZMwmCUil0ML5CeffCK3b9+WrCI+Pj7df/vff/9J48aNpWTJkpI3b16rLcq1a9eW9evXy6effiqHDh2SVatWyRNPPCFDhgzJQMmJiDwLg1kicrmWLVtKoUKFZMKECVa3GTt2bKpL7l9++aWUKlUqVQvc+PHjpWDBgpIrVy754IMPJDExUd58803JkyePFCtWTObMmWPx0n6jRo1UYF21alXZtGmTyeOHDx+Wdu3aSfbs2dVz9+rVS27cuGF4/PHHH5ehQ4fKa6+9Jvny5ZM2bdpYfB/JycmqTCgHWlPxnhCE6tAavWfPHrUNlvG+LRk8eLB6fOfOndKlSxd55JFHpEqVKjJ8+HDZvn27Ybvz589Lp06dVLnDwsKka9eucvXq1VT7dfbs2VKiRAm1HZ47KSlJJk2apOqlQIEC8vHHH5u8Pl57+vTpap+EhISo1uHffvvNZBsE2M2bN1ePIyh/6aWX5N69e6nq67PPPpPChQurbRCIJyQkGLaJi4uTESNGSNGiRVVrdf369WXjxo2Gx+fOnavq+e+//5ZKlSqp8rdt21auXLlieH8//PCDLF26VJUZN/w9fmygvvC6qHP8cEjr+CMi98Vglohczs/PTwWgX3/9tVy8eDFDz4WWysuXL8vmzZtl8uTJ6pL9k08+Kblz55YdO3bIwIED5eWXX071Ogh233jjDdm3b580bNhQXa6/efOmeuzOnTsqKKtZs6bs3r1bBZ8ICBEYGkPQFBgYKP/++6/MmDHDYvmmTJkin3/+uQrgDh48qILep556Sk6dOqUeRxCGoBRlwTICOXO3bt1SZUDghwDPHII7PXBGIIvtEZyvWbNGzpw5I926dUvVEvzXX3+p55w/f758//330qFDB7WP8HdoNX/vvffU/jM2evRoFUgfOHBAevbsKd27d5djx46px6Kjo9V7w37ftWuX/Prrr7J27VoVQBrbsGGDen38j/2H4BQ3Hbbftm2bLFiwQO2v5557TgWr+v6CmJgYtT9/+uknVe8I4PX9hv9RT3qAixt+tHz11VeybNkyWbRokerOMW/ePJMfRkTkQTQiIhfq06eP1qlTJ7XcoEEDrX///mp5yZIlmvEpasyYMVr16tVN/vaLL77QSpYsafJcuJ+UlGRYV6FCBa1JkyaG+4mJiVq2bNm0+fPnq/vh4eHqdSZOnGjYJiEhQStWrJj2ySefqPsffvih1rp1a5PXvnDhgvq7EydOqPvNmjXTatas+dD3W6RIEe3jjz82WVe3bl1t8ODBhvt4n3i/1uzYsUO99uLFi9N8rdWrV2t+fn7a+fPnDeuOHDmi/nbnzp3qPl4nNDRUi4qKMmzTpk0brVSpUqn244QJEwz38RwDBw40eb369etrgwYNUsvfffedljt3bu3evXuGx1esWKH5+vpqERERJvWFOtE999xzWrdu3dTyuXPnVPkvXbpk8jotWrTQRo4cqZbnzJmjynL69GnD41OnTtUKFixo8RjTvfLKK1rz5s215OTkNPchEbk/tswSkdtACyBa5/TWvfRAq6av74NTG7oEPProoyatwLicfe3aNZO/Q2uszt/fX+rUqWMoB1oe0XKIS9j6rWLFiuoxtCrq0Ic1LVFRUarV+LHHHjNZj/v2vOeUWPLh8JwYFIabrnLlyqrl1vj10CKZI0cOk32G7cz3Y1r7TL+vPy/+r169uknLMd4nWovREmpcX6gTHS7766+Dbgro7oAuFMb7Hq3Fxvs9NDRUypYta/E5rEEXB2SwqFChgrz66quyevXqNLcnIvfl7+oCEBHpmjZtqi5Njxw5UgUbxhBYmQdxxn0rdQEBASb30UfS0joEVbZCP090O0CwbQ6Bk87SJX9nKF++vHoP6OfrCM7YZxl5bf11sN8R6KIPsXHACwhq03qOhwX8tWrVkvDwcNW9At0f0BUBfbfN+/0SkftjyywRuRWk6Przzz9VP0lj+fPnV6mojIMUR+aGNR40hQFjCKAwoEgPfI4cOaJaMMuVK2dysyeAxQCsIkWKqD61xnAfLaG2wkA2BP1Tp05VfVPNoY8voPwXLlxQN93Ro0fV4/a8ni37TL+v7zP8jxZt4/LhfeJHCVpDbYE+ymiZRSur+X7HwDRboR8znsdSfaD/8MyZM2XhwoXy+++/q/7FRORZGMwSkVtBlwAMJsIAHWPIFnD9+nU1wh6XmBHIoVXNUfB8S5YsUa2dGFiFNGH9+/dXj+E+gpwePXqowUx4fYye79evn8UgKS0YaIYWXgRPuNz+zjvvqKB82LBhdpcXr12vXj0VhGFAFC7tY7/pl//R0qjvz71796rMB71795ZmzZqpbhQZhUFdyIJw8uRJNdAOz68P8MJrIktAnz59VCYIdNNATlxkgUCXBVugewGeB2VevHixaknFayDrwIoVK2wuJ36EYPAY9jcyUKBFH4MDMdgN9Y3y470gQNYHzxGR52AwS0RuB2mpzC9po6Vv2rRpKohDX0wENZZG+mekRRg3PPeWLVvUSHek2AK9NRXBY+vWrVWAiBRcCHyM+5XaAv0zkT4L2QrwPMgggNdC1wF7IBUWAlTklcVzIZ1Yq1at1IQJSJmlX25HSipkFEAXDgS3+DsE0o4wbtw4lWWgWrVq8uOPP6rgUG/xRT9WBPz4EVC3bl159tlnpUWLFvLNN9/Y9RpIo4ZgFu8RLbpI5YUfFEgjZqsBAwaov0UAjxZ+1CX6COOHEdahfMjbu3LlSrvrk4hczwejwFxdCCIi8iwIlNGSzZm1iMjV+BOUiIiIiDwWg1kiIiIi8lhMzUVERHZjDzUichdsmSUiIiIij8VgloiIiIg8FoNZIiIiIvJYDGaJiIiIyGMxmCUiIiIij8VgloiIiIg8FoNZIiIiIvJYDGaJiIiIyGMxmCUiIiIi8VT/BxhrLyUQaIc3AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 800x450 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def pca_col(df, name, column_name='embeddings', variance_threshold=0.90):\n",
    "    embeddings = np.array(df[column_name].tolist())\n",
    "    \n",
    "    scaler = StandardScaler()\n",
    "    embeddings_standardized = scaler.fit_transform(embeddings)\n",
    "    \n",
    "    pca = PCA()\n",
    "    pca.fit(embeddings_standardized)\n",
    "    explained_variance = pca.explained_variance_ratio_\n",
    "\n",
    "    # Components to retain (in our case 90% of variance)\n",
    "    n_components = np.argmax(np.cumsum(explained_variance) >= variance_threshold) + 1\n",
    "        \n",
    "    plt.figure(figsize=(8, 4.5))\n",
    "    plt.axhline(y=variance_threshold, color='r', linestyle='--')\n",
    "    x_value = np.argmax(np.cumsum(explained_variance) >= variance_threshold) + 1\n",
    "    plt.axvline(x=x_value, color='g', linestyle='--')\n",
    "    plt.text(x_value + 10, 0.025, f'{x_value}', color='green', size='large', weight='bold')\n",
    "    plt.text(- 50, variance_threshold + 0.025, f'{variance_threshold:.2f}', color='red', size='large', weight='bold')\n",
    "\n",
    "    plt.plot(np.cumsum(explained_variance), marker='o')\n",
    "    plt.title(f'Cumulative Explained Variance by PCA Components for {name}')\n",
    "    plt.xlabel('Number of Components')\n",
    "    plt.ylabel('Cumulative Explained Variance')\n",
    "    plt.grid(True)\n",
    "    plt.show()\n",
    "    \n",
    "    pca = PCA(n_components=n_components)\n",
    "    reduced_embeddings = pca.fit_transform(embeddings_standardized)\n",
    "    \n",
    "    # Integrate reduced embeddings into the DataFrame\n",
    "    df[f'reduced_{column_name}'] = reduced_embeddings.tolist()\n",
    "    return df\n",
    "\n",
    "# Apply PCA and add column for each dataset\n",
    "X_df_BP = pca_col(X_df_BP, 'BP')\n",
    "#x_df_CC = pca_col(x_df_CC, 'CC')\n",
    "#x_df_MF = pca_col(x_df_MF, 'MF')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Save the datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure the directory exists\n",
    "datasets_path = Path('../data/datasets')\n",
    "datasets_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# Save the DataFrame to a CSV file\n",
    "X_df_BP.to_csv(datasets_path / 'X_df_BP.csv', index=False)\n",
    "y_df_BP.to_csv(datasets_path / 'y_df_BP.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MODELS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "from sklearn.model_selection import train_test_split\n",
    "from torch.utils.data import DataLoader, TensorDataset\n",
    "from sklearn.metrics import classification_report\n",
    "from torchsummary import summary\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X shape: (83064, 274), y shape: (83064, 1487)\n",
      "y_df_CC: 0    (GO:0005575, GO:0110165, GO:0000785, GO:003299...\n",
      "1    (GO:0005575, GO:0110165, GO:0071944, GO:000588...\n",
      "2    (GO:0005575, GO:0110165, GO:0005622, GO:000582...\n",
      "3     (GO:0005575, GO:0110165, GO:0005576, GO:0005615)\n",
      "4    (GO:0005575, GO:0005737, GO:0042175, GO:003299...\n",
      "Name: GO_term, dtype: object\n",
      "y.head: [[0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]]\n",
      "type(y_df_CC): <class 'pandas.core.series.Series'>, type(y): <class 'numpy.ndarray'>\n"
     ]
    }
   ],
   "source": [
    "# Load data\n",
    "\n",
    "X = X_df_BP['reduced_embeddings']  # Assuming embeddings are already reduced\n",
    "X = pd.DataFrame(X.tolist())  # Convert list of lists to DataFrame\n",
    "\n",
    "# Preprocess target (MultiLabelBinarizer for multi-label classification)\n",
    "mlb = MultiLabelBinarizer()\n",
    "y = mlb.fit_transform(y_df_BP)\n",
    "\n",
    "print(f\"X shape: {X.shape}, y shape: {y.shape}\")\n",
    "print(f\"y_df_CC: {y_df_CC[:5]}\\ny.head: {y[:5]}\")\n",
    "print(f\"type(y_df_CC): {type(y_df_CC)}, type(y): {type(y)}\")\n",
    "\n",
    "# Split data\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert NumPy array to PyTorch tensor\n",
    "X_train_tensor = torch.tensor(X_train.to_numpy(), dtype=torch.float32)\n",
    "X_test_tensor = torch.tensor(X_test.to_numpy(), dtype=torch.float32)\n",
    "y_train_tensor = torch.tensor(y_train, dtype=torch.float32).to(device)\n",
    "y_test_tensor = torch.tensor(y_test, dtype=torch.float32).to(device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# X_prova = X_train_tensor[:5000]\n",
    "# y_prova = y_train_tensor[:5000]\n",
    "\n",
    "# X_prova_test = X_test_tensor[:5000]\n",
    "# y_prova_test = y_test_tensor[:5000]\n",
    "\n",
    "# X_train_tensor = X_prova\n",
    "# y_train_tensor = y_prova.to(device)\n",
    "# X_test_tensor = X_prova_test\n",
    "# y_test_tensor = y_prova_test.to(device)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NN(nn.Module):\n",
    "    def __init__(self, input_size, hidden_sizes, dropouts, output_size):\n",
    "        super(NN, self).__init__()\n",
    "        \n",
    "        layers = []\n",
    "        in_size = input_size\n",
    "        \n",
    "        for i, hidden_size in enumerate(hidden_sizes):\n",
    "            layers.append(nn.Linear(in_size, hidden_size))\n",
    "            layers.append(nn.ReLU())\n",
    "            layers.append(nn.Dropout(dropouts[i]))\n",
    "            in_size = hidden_size\n",
    "        \n",
    "        layers.append(nn.Linear(in_size, output_size))\n",
    "        layers.append(nn.Sigmoid())  # For multilabel classification\n",
    "        \n",
    "        self.model = nn.Sequential(*layers)\n",
    "    \n",
    "    def forward(self, x):\n",
    "        return self.model(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Creating/Extracting the models DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>hidden_sizes</th>\n",
       "      <th>dropouts</th>\n",
       "      <th>lr</th>\n",
       "      <th>batch_size</th>\n",
       "      <th>epochs</th>\n",
       "      <th>macro_P</th>\n",
       "      <th>macro_R</th>\n",
       "      <th>macro_F1</th>\n",
       "      <th>weighted_P</th>\n",
       "      <th>weighted_R</th>\n",
       "      <th>weighted_F1</th>\n",
       "      <th>samples_P</th>\n",
       "      <th>samples_R</th>\n",
       "      <th>samples_F1</th>\n",
       "      <th>training_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NN1</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.1, 0.1]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1024</td>\n",
       "      <td>75</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.259</td>\n",
       "      <td>0.319</td>\n",
       "      <td>0.546</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.452</td>\n",
       "      <td>0.576</td>\n",
       "      <td>0.465</td>\n",
       "      <td>0.458</td>\n",
       "      <td>8.726524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NN2</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>2048</td>\n",
       "      <td>75</td>\n",
       "      <td>0.564</td>\n",
       "      <td>0.154</td>\n",
       "      <td>0.215</td>\n",
       "      <td>0.624</td>\n",
       "      <td>0.329</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.674</td>\n",
       "      <td>0.386</td>\n",
       "      <td>0.432</td>\n",
       "      <td>7.741851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NN3</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.4, 0.4]</td>\n",
       "      <td>0.005</td>\n",
       "      <td>1024</td>\n",
       "      <td>100</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.042</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.192</td>\n",
       "      <td>0.228</td>\n",
       "      <td>0.751</td>\n",
       "      <td>0.245</td>\n",
       "      <td>0.324</td>\n",
       "      <td>11.591921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NN4</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.1, 0.1]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>2048</td>\n",
       "      <td>75</td>\n",
       "      <td>0.574</td>\n",
       "      <td>0.183</td>\n",
       "      <td>0.255</td>\n",
       "      <td>0.622</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.677</td>\n",
       "      <td>0.399</td>\n",
       "      <td>0.439</td>\n",
       "      <td>3.417191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NN5</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>1024</td>\n",
       "      <td>75</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.228</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0.646</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.440</td>\n",
       "      <td>3.830458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name  hidden_sizes    dropouts     lr  batch_size  epochs  macro_P  macro_R  \\\n",
       "0  NN1  [4096, 2048]  [0.1, 0.1]  0.001        1024      75    0.464    0.259   \n",
       "1  NN2  [4096, 2048]  [0.2, 0.2]  0.003        2048      75    0.564    0.154   \n",
       "2  NN3  [4096, 2048]  [0.4, 0.4]  0.005        1024     100    0.203    0.029   \n",
       "3  NN4  [2048, 1024]  [0.1, 0.1]  0.001        2048      75    0.574    0.183   \n",
       "4  NN5  [2048, 1024]  [0.2, 0.2]  0.003        1024      75    0.538    0.169   \n",
       "\n",
       "   macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  samples_R  \\\n",
       "0     0.319       0.546       0.407        0.452      0.576      0.465   \n",
       "1     0.215       0.624       0.329        0.393      0.674      0.386   \n",
       "2     0.042       0.500       0.192        0.228      0.751      0.245   \n",
       "3     0.255       0.622       0.341        0.411      0.677      0.399   \n",
       "4     0.228       0.600       0.350        0.403      0.646      0.404   \n",
       "\n",
       "   samples_F1  training_time  \n",
       "0       0.458       8.726524  \n",
       "1       0.432       7.741851  \n",
       "2       0.324      11.591921  \n",
       "3       0.439       3.417191  \n",
       "4       0.440       3.830458  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if the file exists and create it if it doesn't\n",
    "datasets_path = Path('../data/datasets')\n",
    "nn_df_path = datasets_path / 'NN_BP.csv'\n",
    "\n",
    "if nn_df_path.exists():\n",
    "    \n",
    "    NN_BP = pd.read_csv(nn_df_path)\n",
    "    for c_list, c_int in zip(['hidden_sizes', 'dropouts'], ['batch_size', 'epochs']):\n",
    "        NN_BP[c_list] = NN_BP[c_list].apply(ast.literal_eval)\n",
    "        NN_BP[c_int] = NN_BP[c_int].astype(int)\n",
    "    \n",
    "    NN_BP['lr'] = NN_BP['lr'].astype(float)\n",
    "\n",
    "else:\n",
    "    if not datasets_path.exists():\n",
    "        datasets_path.mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    NN_BP = {\n",
    "        'name': [],\n",
    "        'hidden_sizes': [],\n",
    "        'dropouts': [],\n",
    "        'lr': [],\n",
    "        'batch_size': [],\n",
    "        'epochs': [],\n",
    "        'macro_P': [],\n",
    "        'macro_R': [],\n",
    "        'macro_F1': [],\n",
    "        'weighted_P': [],\n",
    "        'weighted_R': [],\n",
    "        'weighted_F1': [],\n",
    "        'samples_P': [],\n",
    "        'samples_R': [],\n",
    "        'samples_F1': [],\n",
    "        'training_time': [],\n",
    "    }\n",
    "    NN_BP = pd.DataFrame(NN_BP)\n",
    "\n",
    "\n",
    "NN_BP.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Filling NN_BP with model parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insert_parameters(df, names, hidden_sizes_list, dropouts_list, lrs, batch_sizes, epochs_list):\n",
    "    for name, hidden_sizes, dropouts, lr, batch_size, epochs in zip(names, hidden_sizes_list, dropouts_list, lrs, batch_sizes, epochs_list):\n",
    "        exists = name in df['name'].values\n",
    "        \n",
    "        if not exists:\n",
    "            new_row = {\n",
    "                'name': name,\n",
    "                'hidden_sizes': [hidden_sizes],\n",
    "                'dropouts': [dropouts],\n",
    "                'lr': lr,\n",
    "                'batch_size': int(batch_size),\n",
    "                'epochs': int(epochs),\n",
    "            }\n",
    "            new_row_df = pd.DataFrame(new_row)\n",
    "            df = pd.concat([df, new_row_df], ignore_index=True)\n",
    "        else:\n",
    "            print(f\"{name} already exists in the DataFrame.\")\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NN1 already exists in the DataFrame.\n",
      "NN2 already exists in the DataFrame.\n",
      "NN3 already exists in the DataFrame.\n",
      "NN4 already exists in the DataFrame.\n",
      "NN5 already exists in the DataFrame.\n",
      "NN6 already exists in the DataFrame.\n",
      "NN7 already exists in the DataFrame.\n",
      "NN8 already exists in the DataFrame.\n",
      "NN9 already exists in the DataFrame.\n",
      "NN10 already exists in the DataFrame.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>hidden_sizes</th>\n",
       "      <th>dropouts</th>\n",
       "      <th>lr</th>\n",
       "      <th>batch_size</th>\n",
       "      <th>epochs</th>\n",
       "      <th>macro_P</th>\n",
       "      <th>macro_R</th>\n",
       "      <th>macro_F1</th>\n",
       "      <th>weighted_P</th>\n",
       "      <th>weighted_R</th>\n",
       "      <th>weighted_F1</th>\n",
       "      <th>samples_P</th>\n",
       "      <th>samples_R</th>\n",
       "      <th>samples_F1</th>\n",
       "      <th>training_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NN1</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.1, 0.1]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1024</td>\n",
       "      <td>75</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.259</td>\n",
       "      <td>0.319</td>\n",
       "      <td>0.546</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.452</td>\n",
       "      <td>0.576</td>\n",
       "      <td>0.465</td>\n",
       "      <td>0.458</td>\n",
       "      <td>8.726524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NN2</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>2048</td>\n",
       "      <td>75</td>\n",
       "      <td>0.564</td>\n",
       "      <td>0.154</td>\n",
       "      <td>0.215</td>\n",
       "      <td>0.624</td>\n",
       "      <td>0.329</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.674</td>\n",
       "      <td>0.386</td>\n",
       "      <td>0.432</td>\n",
       "      <td>7.741851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NN3</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.4, 0.4]</td>\n",
       "      <td>0.005</td>\n",
       "      <td>1024</td>\n",
       "      <td>100</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.042</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.192</td>\n",
       "      <td>0.228</td>\n",
       "      <td>0.751</td>\n",
       "      <td>0.245</td>\n",
       "      <td>0.324</td>\n",
       "      <td>11.591921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NN4</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.1, 0.1]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>2048</td>\n",
       "      <td>75</td>\n",
       "      <td>0.574</td>\n",
       "      <td>0.183</td>\n",
       "      <td>0.255</td>\n",
       "      <td>0.622</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.677</td>\n",
       "      <td>0.399</td>\n",
       "      <td>0.439</td>\n",
       "      <td>3.417191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NN5</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>1024</td>\n",
       "      <td>75</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.228</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0.646</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.440</td>\n",
       "      <td>3.830458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name  hidden_sizes    dropouts     lr  batch_size  epochs  macro_P  macro_R  \\\n",
       "0  NN1  [4096, 2048]  [0.1, 0.1]  0.001        1024      75    0.464    0.259   \n",
       "1  NN2  [4096, 2048]  [0.2, 0.2]  0.003        2048      75    0.564    0.154   \n",
       "2  NN3  [4096, 2048]  [0.4, 0.4]  0.005        1024     100    0.203    0.029   \n",
       "3  NN4  [2048, 1024]  [0.1, 0.1]  0.001        2048      75    0.574    0.183   \n",
       "4  NN5  [2048, 1024]  [0.2, 0.2]  0.003        1024      75    0.538    0.169   \n",
       "\n",
       "   macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  samples_R  \\\n",
       "0     0.319       0.546       0.407        0.452      0.576      0.465   \n",
       "1     0.215       0.624       0.329        0.393      0.674      0.386   \n",
       "2     0.042       0.500       0.192        0.228      0.751      0.245   \n",
       "3     0.255       0.622       0.341        0.411      0.677      0.399   \n",
       "4     0.228       0.600       0.350        0.403      0.646      0.404   \n",
       "\n",
       "   samples_F1  training_time  \n",
       "0       0.458       8.726524  \n",
       "1       0.432       7.741851  \n",
       "2       0.324      11.591921  \n",
       "3       0.439       3.417191  \n",
       "4       0.440       3.830458  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "names = ['NN1', 'NN2', 'NN3', 'NN4', 'NN5', 'NN6', 'NN7', 'NN8', 'NN9', 'NN10']\n",
    "\n",
    "\n",
    "hidden_sizes = [[4096, 2048],   #NN1\n",
    "                [4096, 2048],   #NN2\n",
    "                [4096, 2048],   #NN3\n",
    "                [2048, 1024],   #NN4\n",
    "                [2048, 1024],   #NN5\n",
    "                [2048, 1024],   #NN6\n",
    "                [4096, 2048, 1024],   #NN7\n",
    "                [4096, 2048, 2048],   #NN8\n",
    "                [2048, 2048, 2048],   #NN9\n",
    "                [2048, 4096, 2048],   #NN10\n",
    "                [4096, 4096, 2048],   #NN11        \n",
    "                [4096, 4096, 2048]]   #NN12    \n",
    "\n",
    "\n",
    "dropouts = [[0.1, 0.1],   #NN1 \n",
    "            [0.2, 0.2],   #NN2\n",
    "            [0.4, 0.4],   #NN3\n",
    "            [0.1, 0.1],   #NN4\n",
    "            [0.2, 0.2],   #NN5\n",
    "            [0.4, 0.4],   #NN6\n",
    "            [0.1, 0.2, 0.3],  #NN7\n",
    "            [0.2, 0.2, 0.2],  #NN8\n",
    "            [0.2, 0.2, 0.2],  #NN9\n",
    "            [0.3, 0.3, 0.3],  #NN10\n",
    "            [0.1, 0.2, 0.3],  #NN11        \n",
    "            [0.1, 0.2, 0.3]]  #NN12 \n",
    "\n",
    "\n",
    "lrs = [0.001, #NN1 \n",
    "       0.003, #NN2\n",
    "       0.005, #NN3\n",
    "       0.001, #NN4\n",
    "       0.003, #NN5\n",
    "       0.005, #NN6\n",
    "       0.005, #NN7\n",
    "       0.003, #NN8\n",
    "       0.003, #NN9\n",
    "       0.005, #NN10\n",
    "       0.003, #NN11        \n",
    "       0.001] #NN12 \n",
    "\n",
    "\n",
    "batch_sizes = [1024, #NN1\n",
    "               2048, #NN2\n",
    "               1024, #NN3\n",
    "               2048, #NN4\n",
    "               1024, #NN5\n",
    "               1024, #NN6\n",
    "               2048, #NN7\n",
    "               1024, #NN8\n",
    "               1024, #NN9\n",
    "               2048, #NN10\n",
    "               2048, #N11        \n",
    "               1024] #N12 \n",
    "\n",
    "\n",
    "epochs = [75,  #NN1 \n",
    "          75,  #NN2\n",
    "          100, #NN3\n",
    "          75,  #NN4\n",
    "          75,  #NN5\n",
    "          100, #NN6\n",
    "          75,  #NN7\n",
    "          75,  #NN8\n",
    "          75,  #NN9\n",
    "          75,  #NN10\n",
    "          75,  #NN11\n",
    "          75]  #NN12\n",
    "\n",
    "\n",
    "\n",
    "NN_BP = insert_parameters(df = NN_BP, \n",
    "               names = names, \n",
    "               hidden_sizes_list=hidden_sizes, \n",
    "               dropouts_list=dropouts, \n",
    "               lrs=lrs, \n",
    "               batch_sizes=batch_sizes, \n",
    "               epochs_list=epochs)\n",
    "\n",
    "NN_BP.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Training the models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_model(df, name, input_size, output_size):\n",
    "\n",
    "    hidd_s = df[df['name'] == name]['hidden_sizes'].tolist()\n",
    "    dropouts = df[df['name'] == name]['dropouts'].tolist()\n",
    "    # Initialize the model\n",
    "    model = NN(\n",
    "        input_size = input_size,\n",
    "        hidden_sizes = hidd_s,\n",
    "        dropouts = dropouts, \n",
    "        output_size = output_size\n",
    "        ).to(device)\n",
    "    \n",
    "    # Loss function and optimizer\n",
    "    criterion = nn.BCELoss()  # Binary Cross-Entropy Loss for multilabel\n",
    "    optimizer = optim.Adam(model.parameters(), lr=df[df['name'] == name]['lr'])\n",
    "\n",
    "    return model, optimizer, criterion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_model(model, name, X_train, y_train, optimizer, criterion, batch_size, epochs):\n",
    "    # Determine the device (model should already be on this device)\n",
    "    device = next(model.parameters()).device\n",
    "\n",
    "    print(f\"\\nTraining {name} model:\")\n",
    "\n",
    "    # Move training data to the device\n",
    "    X_train = X_train.to(device)\n",
    "    y_train = y_train.to(device)\n",
    "\n",
    "    model.train()\n",
    "    start_time = time.time()  # Record the start time\n",
    "\n",
    "    for epoch in range(epochs):\n",
    "        for i in range(0, len(X_train), batch_size):\n",
    "            # Get the current batch\n",
    "            X_batch = X_train[i:i+batch_size]\n",
    "            y_batch = y_train[i:i+batch_size]\n",
    "\n",
    "            # Forward pass\n",
    "            outputs = model(X_batch)\n",
    "            loss = criterion(outputs, y_batch)\n",
    "\n",
    "            # Backward pass and optimization\n",
    "            optimizer.zero_grad()\n",
    "            loss.backward()\n",
    "            optimizer.step()\n",
    "        \n",
    "        \n",
    "        t = time.time() - start_time\n",
    "\n",
    "        if epoch == 0:\n",
    "            print(f\"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f} - time: {(t/60):.2f}min\")\n",
    "        elif epoch % 10 == 0:\n",
    "            print(f\"Epoch [{epoch}/{epochs}], Loss: {loss.item():.4f} - time: {(t/60):.2f}min\")\n",
    "\n",
    "    end_time = time.time()  # Record the end time\n",
    "\n",
    "    training_time = end_time - start_time  # Calculate the elapsed time\n",
    "    \n",
    "    return training_time / 60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_evaluation(df, name, model, X, y, target_names, training_time):\n",
    "\n",
    "    model.eval()\n",
    "    with torch.no_grad():\n",
    "        y_pred = model(X)\n",
    "        probabilities = torch.sigmoid(y_pred)  # Assuming multilabel classification with sigmoid activation\n",
    "        prob = probabilities[probabilities > 0.20].float()\n",
    "        \n",
    "        y_pred = (y_pred.cpu().numpy() > 0.5).astype(int)\n",
    "\n",
    "    exists = any(df[df['name'] == name])\n",
    "    \n",
    "    if exists:\n",
    "        \n",
    "        report = classification_report(y.cpu().numpy(), y_pred, target_names=target_names, output_dict=True)\n",
    "        # Convert the classification report to a DataFrame\n",
    "        report = pd.DataFrame(report).transpose().drop(columns=['support'])\n",
    "        \n",
    "        report.loc['macro avg'] = report.loc['macro avg'].apply(lambda x: round(x, 3))\n",
    "        report.loc['weighted avg'] = report.loc['weighted avg'].apply(lambda x: round(x, 3))\n",
    "        report.loc['samples avg'] = report.loc['samples avg'].apply(lambda x: round(x, 3))\n",
    "\n",
    "\n",
    "\n",
    "        df.loc[df['name'] == name, 'macro_P'] = report.loc['macro avg']['precision']\n",
    "        df.loc[df['name'] == name, 'macro_R'] = report.loc['macro avg']['recall']\n",
    "        df.loc[df['name'] == name, 'macro_F1'] = report.loc['macro avg']['f1-score']\n",
    "\n",
    "        df.loc[df['name'] == name, 'weighted_P'] = report.loc['weighted avg']['precision']\n",
    "        df.loc[df['name'] == name, 'weighted_R'] = report.loc['weighted avg']['recall']\n",
    "        df.loc[df['name'] == name, 'weighted_F1'] = report.loc['weighted avg']['f1-score']\n",
    "        \n",
    "        df.loc[df['name'] == name, 'samples_P'] = report.loc['samples avg']['precision']\n",
    "        df.loc[df['name'] == name, 'samples_R'] = report.loc['samples avg']['recall']\n",
    "        df.loc[df['name'] == name, 'samples_F1'] = report.loc['samples avg']['f1-score']\n",
    "\n",
    "        df.loc[df['name'] == name, 'training_time'] = training_time\n",
    "\n",
    "        return df, report, prob\n",
    "    else:\n",
    "        print(f\"{name} already exists in the DataFrame.\")\n",
    "        return df, None, prob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>hidden_sizes</th>\n",
       "      <th>dropouts</th>\n",
       "      <th>lr</th>\n",
       "      <th>batch_size</th>\n",
       "      <th>epochs</th>\n",
       "      <th>macro_P</th>\n",
       "      <th>macro_R</th>\n",
       "      <th>macro_F1</th>\n",
       "      <th>weighted_P</th>\n",
       "      <th>weighted_R</th>\n",
       "      <th>weighted_F1</th>\n",
       "      <th>samples_P</th>\n",
       "      <th>samples_R</th>\n",
       "      <th>samples_F1</th>\n",
       "      <th>training_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NN1</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.1, 0.1]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1024</td>\n",
       "      <td>75</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.259</td>\n",
       "      <td>0.319</td>\n",
       "      <td>0.546</td>\n",
       "      <td>0.407</td>\n",
       "      <td>0.452</td>\n",
       "      <td>0.576</td>\n",
       "      <td>0.465</td>\n",
       "      <td>0.458</td>\n",
       "      <td>8.726524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NN2</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>2048</td>\n",
       "      <td>75</td>\n",
       "      <td>0.564</td>\n",
       "      <td>0.154</td>\n",
       "      <td>0.215</td>\n",
       "      <td>0.624</td>\n",
       "      <td>0.329</td>\n",
       "      <td>0.393</td>\n",
       "      <td>0.674</td>\n",
       "      <td>0.386</td>\n",
       "      <td>0.432</td>\n",
       "      <td>7.741851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NN3</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.4, 0.4]</td>\n",
       "      <td>0.005</td>\n",
       "      <td>1024</td>\n",
       "      <td>100</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.042</td>\n",
       "      <td>0.500</td>\n",
       "      <td>0.192</td>\n",
       "      <td>0.228</td>\n",
       "      <td>0.751</td>\n",
       "      <td>0.245</td>\n",
       "      <td>0.324</td>\n",
       "      <td>11.591921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NN4</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.1, 0.1]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>2048</td>\n",
       "      <td>75</td>\n",
       "      <td>0.574</td>\n",
       "      <td>0.183</td>\n",
       "      <td>0.255</td>\n",
       "      <td>0.622</td>\n",
       "      <td>0.341</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.677</td>\n",
       "      <td>0.399</td>\n",
       "      <td>0.439</td>\n",
       "      <td>3.417191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NN5</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>1024</td>\n",
       "      <td>75</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.228</td>\n",
       "      <td>0.600</td>\n",
       "      <td>0.350</td>\n",
       "      <td>0.403</td>\n",
       "      <td>0.646</td>\n",
       "      <td>0.404</td>\n",
       "      <td>0.440</td>\n",
       "      <td>3.830458</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name  hidden_sizes    dropouts     lr  batch_size  epochs  macro_P  macro_R  \\\n",
       "0  NN1  [4096, 2048]  [0.1, 0.1]  0.001        1024      75    0.464    0.259   \n",
       "1  NN2  [4096, 2048]  [0.2, 0.2]  0.003        2048      75    0.564    0.154   \n",
       "2  NN3  [4096, 2048]  [0.4, 0.4]  0.005        1024     100    0.203    0.029   \n",
       "3  NN4  [2048, 1024]  [0.1, 0.1]  0.001        2048      75    0.574    0.183   \n",
       "4  NN5  [2048, 1024]  [0.2, 0.2]  0.003        1024      75    0.538    0.169   \n",
       "\n",
       "   macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  samples_R  \\\n",
       "0     0.319       0.546       0.407        0.452      0.576      0.465   \n",
       "1     0.215       0.624       0.329        0.393      0.674      0.386   \n",
       "2     0.042       0.500       0.192        0.228      0.751      0.245   \n",
       "3     0.255       0.622       0.341        0.411      0.677      0.399   \n",
       "4     0.228       0.600       0.350        0.403      0.646      0.404   \n",
       "\n",
       "   samples_F1  training_time  \n",
       "0       0.458       8.726524  \n",
       "1       0.432       7.741851  \n",
       "2       0.324      11.591921  \n",
       "3       0.439       3.417191  \n",
       "4       0.440       3.830458  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_names = [str(cls) for cls in mlb.classes_]\n",
    "\n",
    "report_path = datasets_path / 'BP_report.csv'\n",
    "probs_path = datasets_path / 'BP_report.csv'\n",
    "\n",
    "\n",
    "if report_path.exists() and probs_path.exists():\n",
    "    report = pd.read_csv(datasets_path / 'BP_reoprt.csv')\n",
    "    probs = pd.read_csv(datasets_path / 'BP_probs.csv')\n",
    "else:\n",
    "    report = {}\n",
    "    probs = {}\n",
    "\n",
    "for name in NN_BP['name']:\n",
    "\n",
    "    # Define model parameters\n",
    "    input_size = X_train_tensor.shape[1] \n",
    "    hidden_sizes = NN_BP[NN_BP['name'] == name]['hidden_sizes'].values[0]\n",
    "    dropouts = NN_BP[NN_BP['name'] == name]['dropouts'].values[0]\n",
    "    output_size = y_train_tensor.shape[1]\n",
    "    lr = float(NN_BP[NN_BP['name'] == name]['lr'].values[0])\n",
    "    batch_size = int(NN_BP[NN_BP['name'] == name]['batch_size'].values[0])\n",
    "    epochs = int(NN_BP[NN_BP['name'] == name]['epochs'].values[0])\n",
    "\n",
    "    # Initialize the model\n",
    "    model = NN(\n",
    "        input_size = input_size,\n",
    "        hidden_sizes = hidden_sizes,\n",
    "        dropouts = dropouts, \n",
    "        output_size = output_size\n",
    "        ).to(device)\n",
    "\n",
    "    # Loss function and optimizer\n",
    "    criterion = nn.BCELoss()  # Binary Cross-Entropy Loss for multilabel\n",
    "    optimizer = optim.Adam(model.parameters(), lr=lr)\n",
    "\n",
    "\n",
    "    # model, optimizer, criterion = create_model(NN_BP, name, X_train_tensor.shape[1] , y_train_tensor.shape[1])\n",
    "    if NN_BP[NN_BP['name'] == name]['training_time'].values[0] == 0:\n",
    "        # Train the model\n",
    "        training_time = train_model(model, name, X_train_tensor, y_train_tensor, optimizer, criterion, batch_size, epochs)\n",
    "\n",
    "        # Ensure X_test_tensor is on the same device as the model\n",
    "        device = next(model.parameters()).device\n",
    "        X_test_tensor = X_test_tensor.to(device)\n",
    "\n",
    "        # Evaluation\n",
    "        \n",
    "        NN_BP, report[name], probs[name] = model_evaluation(NN_BP, name, model, X_test_tensor, y_test_tensor, target_names, training_time)\n",
    "\n",
    "    # Save the DataFrame to a CSV file\n",
    "    NN_BP.to_csv(datasets_path / 'NN_BP.csv', index=False)\n",
    "    report = pd.DataFrame(report)\n",
    "    probs = pd.DataFrame(probs)\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "NN_BP.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Extracting results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(  name        hidden_sizes         dropouts     lr  batch_size  epochs  \\\n",
       " 7  NN8  [4096, 2048, 2048]  [0.2, 0.2, 0.2]  0.003        1024      75   \n",
       " 3  NN4        [2048, 1024]       [0.1, 0.1]  0.001        2048      75   \n",
       " 8  NN9  [2048, 2048, 2048]  [0.2, 0.2, 0.2]  0.003        1024      75   \n",
       " 1  NN2        [4096, 2048]       [0.2, 0.2]  0.003        2048      75   \n",
       " 4  NN5        [2048, 1024]       [0.2, 0.2]  0.003        1024      75   \n",
       " \n",
       "    training_time  \n",
       " 7      11.569981  \n",
       " 3       3.417191  \n",
       " 8       8.878847  \n",
       " 1       7.741851  \n",
       " 4       3.830458  ,\n",
       "    macro_P  macro_R  macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  \\\n",
       " 7    0.606    0.146     0.209       0.649       0.320        0.387      0.687   \n",
       " 3    0.574    0.183     0.255       0.622       0.341        0.411      0.677   \n",
       " 8    0.564    0.201     0.278       0.617       0.370        0.432      0.653   \n",
       " 1    0.564    0.154     0.215       0.624       0.329        0.393      0.674   \n",
       " 4    0.538    0.169     0.228       0.600       0.350        0.403      0.646   \n",
       " \n",
       "    samples_R  samples_F1  \n",
       " 7      0.374       0.424  \n",
       " 3      0.399       0.439  \n",
       " 8      0.422       0.452  \n",
       " 1      0.386       0.432  \n",
       " 4      0.404       0.440  )"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metrics = ['macro_P', 'macro_R', 'macro_F1', \n",
    "           'weighted_P', 'weighted_R', 'weighted_F1', \n",
    "           'samples_P', 'samples_R', 'samples_F1']\n",
    "\n",
    "\n",
    "\n",
    "BP_results = NN_BP.sort_values(by=metrics, ascending=False)[['macro_P', 'macro_R', 'macro_F1',\n",
    "                                                             'weighted_P', 'weighted_R', 'weighted_F1', \n",
    "                                                             'samples_P', 'samples_R', 'samples_F1']][:5]\n",
    "\n",
    "BP_params = NN_BP.sort_values(by=metrics, ascending=False)[['name', 'hidden_sizes', 'dropouts', 'lr', \n",
    "                                                'batch_size', 'epochs', 'training_time']][:5]\n",
    "\n",
    "BP_params.head(), BP_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(   macro_P  macro_R  macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  \\\n",
       " 7    0.606    0.146     0.209       0.649       0.320        0.387      0.687   \n",
       " 3    0.574    0.183     0.255       0.622       0.341        0.411      0.677   \n",
       " 8    0.564    0.201     0.278       0.617       0.370        0.432      0.653   \n",
       " 1    0.564    0.154     0.215       0.624       0.329        0.393      0.674   \n",
       " 4    0.538    0.169     0.228       0.600       0.350        0.403      0.646   \n",
       " \n",
       "    samples_R  samples_F1 name  \n",
       " 7      0.374       0.424  NN1  \n",
       " 3      0.399       0.439  NN2  \n",
       " 8      0.422       0.452  NN3  \n",
       " 1      0.386       0.432  NN4  \n",
       " 4      0.404       0.440  NN5  ,\n",
       "   name        hidden_sizes         dropouts     lr  batch_size  epochs  \\\n",
       " 7  NN1  [4096, 2048, 2048]  [0.2, 0.2, 0.2]  0.003        1024      75   \n",
       " 3  NN2        [2048, 1024]       [0.1, 0.1]  0.001        2048      75   \n",
       " 8  NN3  [2048, 2048, 2048]  [0.2, 0.2, 0.2]  0.003        1024      75   \n",
       " 1  NN4        [4096, 2048]       [0.2, 0.2]  0.003        2048      75   \n",
       " 4  NN5        [2048, 1024]       [0.2, 0.2]  0.003        1024      75   \n",
       " \n",
       "    training_time  \n",
       " 7      11.569981  \n",
       " 3       3.417191  \n",
       " 8       8.878847  \n",
       " 1       7.741851  \n",
       " 4       3.830458  )"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_names = [f'NN{i}' for i in range(1, BP_results.shape[0] + 1)]\n",
    "\n",
    "# Add the row means to the DataFrame\n",
    "BP_results['name'] = row_names\n",
    "BP_params['name'] = row_names\n",
    "\n",
    "# Save the DataFrame to a CSV file\n",
    "BP_results.to_csv(datasets_path / 'BP_results.csv', index=False)\n",
    "BP_params.to_csv(datasets_path / 'BP_params.csv', index=False)\n",
    "\n",
    "# Print the updated DataFrame\n",
    "BP_results.head(), BP_params.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### LaTex tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{lccccc}\n",
      "    \\toprule\n",
      "    \\textit{Metric} & \\textit{NN1} & \\textit{NN2} & \\textit{NN3} & \\textit{NN4} & \\textit{NN5} \\\\\n",
      "    \\midrule\n",
      "    Macro P & 0.61 & 0.57 & 0.56 & 0.56 & 0.54 \\\\\n",
      "    Macro R & 0.15 & 0.18 & 0.20 & 0.15 & 0.17 \\\\\n",
      "    Macro F1 & 0.21 & 0.26 & 0.28 & 0.21 & 0.23 \\\\\n",
      "    Weighted P & 0.65 & 0.62 & 0.62 & 0.62 & 0.60 \\\\\n",
      "    Weighted R & 0.32 & 0.34 & 0.37 & 0.33 & 0.35 \\\\\n",
      "    Weighted F1 & 0.39 & 0.41 & 0.43 & 0.39 & 0.40 \\\\\n",
      "    Samples P & 0.69 & 0.68 & 0.65 & 0.67 & 0.65 \\\\\n",
      "    Samples R & 0.37 & 0.40 & 0.42 & 0.39 & 0.40 \\\\\n",
      "    Samples F1 & 0.42 & 0.44 & 0.45 & 0.43 & 0.44 \\\\\n",
      "    \\bottomrule\n",
      "\\end{tabular}\n",
      "\\begin{table}[H]\n",
      "    \\centering\n",
      "    \\begin{tabular}{lcccccc}\n",
      "        \\toprule\n",
      "        \\textit{Model} & \\textit{Hidden Layer Sizes} & \\textit{Dropout} & \\textit{Learning Rate} & \\textit{Batch Size} & \\textit{Epochs} & \\textit{Regularization} \\\\\n",
      "        \\midrule\n",
      "        NN1 & [4096, 2048, 2048] & [0.2, 0.2, 0.2] & 0.003 & 1024 & 75 & None \\\\\n",
      "        NN2 & [2048, 1024] & [0.1, 0.1] & 0.001 & 2048 & 75 & None \\\\\n",
      "        NN3 & [2048, 2048, 2048] & [0.2, 0.2, 0.2] & 0.003 & 1024 & 75 & None \\\\\n",
      "        NN4 & [4096, 2048] & [0.2, 0.2] & 0.003 & 2048 & 75 & None \\\\\n",
      "        NN5 & [2048, 1024] & [0.2, 0.2] & 0.003 & 1024 & 75 & None \\\\\n",
      "        \\bottomrule\n",
      "    \\end{tabular}\n",
      "    \\caption{Model Parameters.}\n",
      "\\end{table}\n"
     ]
    }
   ],
   "source": [
    "def generate_latex_table(df):\n",
    "    metrics = ['macro_P', 'macro_R', 'macro_F1', 'weighted_P', 'weighted_R', 'weighted_F1', 'samples_P', 'samples_R', 'samples_F1']\n",
    "    metric_names = ['Macro P', 'Macro R', 'Macro F1', 'Weighted P', 'Weighted R', 'Weighted F1', 'Samples P', 'Samples R', 'Samples F1']\n",
    "    \n",
    "    # Initialize the LaTeX table\n",
    "    latex_table = \"\\\\begin{tabular}{l\" + \"c\" * len(df['name']) + \"}\\n\"\n",
    "    latex_table += \"    \\\\toprule\\n\"\n",
    "    latex_table += \"    \\\\textit{Metric} & \" + \" & \".join([f\"\\\\textit{{{name}}}\" for name in df['name']]) + \" \\\\\\\\\\n\"\n",
    "    latex_table += \"    \\\\midrule\\n\"\n",
    "    \n",
    "    for metric, metric_name in zip(metrics, metric_names):\n",
    "        values = df[metric]\n",
    "        formatted_values = [f\"{value:.2f}\" for value in values]\n",
    "        latex_table += f\"    {metric_name} & \" + \" & \".join(formatted_values) + \" \\\\\\\\\\n\"\n",
    "    \n",
    "    latex_table += \"    \\\\bottomrule\\n\"\n",
    "    latex_table += \"\\\\end{tabular}\"\n",
    "    \n",
    "    return latex_table\n",
    "\n",
    "\n",
    "def generate_parameters_latex_table(df):\n",
    "    # Initialize the LaTeX table\n",
    "    latex_table = \"\\\\begin{table}[H]\\n\"\n",
    "    latex_table += \"    \\\\centering\\n\"\n",
    "    latex_table += \"    \\\\begin{tabular}{lcccccc}\\n\"\n",
    "    latex_table += \"        \\\\toprule\\n\"\n",
    "    latex_table += \"        \\\\textit{Model} & \\\\textit{Hidden Layer Sizes} & \\\\textit{Dropout} & \\\\textit{Learning Rate} & \\\\textit{Batch Size} & \\\\textit{Epochs} & \\\\textit{Regularization} \\\\\\\\\\n\"\n",
    "    latex_table += \"        \\\\midrule\\n\"\n",
    "    \n",
    "    for index, row in df.iterrows():\n",
    "        model_name = row['name']\n",
    "        hidden_sizes = row['hidden_sizes']\n",
    "        dropouts = row['dropouts']\n",
    "        lr = row['lr']\n",
    "        batch_size = row['batch_size']\n",
    "        epochs = row['epochs']\n",
    "        regularization = 'None'  # Assuming regularization is not provided in the DataFrame\n",
    "        \n",
    "        latex_table += f\"        {model_name} & {hidden_sizes} & {dropouts} & {lr} & {batch_size} & {epochs} & {regularization} \\\\\\\\\\n\"\n",
    "    \n",
    "    latex_table += \"        \\\\bottomrule\\n\"\n",
    "    latex_table += \"    \\\\end{tabular}\\n\"\n",
    "    latex_table += \"    \\\\caption{Model Parameters.}\\n\"\n",
    "    latex_table += \"\\\\end{table}\"\n",
    "    \n",
    "    return latex_table\n",
    "\n",
    "# Generate the LaTeX table\n",
    "latex_table = generate_latex_table(BP_results)\n",
    "print(latex_table)\n",
    "\n",
    "latex_table = generate_parameters_latex_table(BP_params)\n",
    "print(latex_table)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
