{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Align\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import substitution_matrices\n",
    "from Bio.Data import IUPACData\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "import cafaeval\n",
    "from cafaeval.evaluation import cafa_eval\n",
    "from cafaeval.parser import obo_parser, gt_parser\n",
    "\n",
    "from pathlib import Path\n",
    "import os\n",
    "\n",
    "import h5py\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating the path to the main folder we are going to use to create our datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to train data\n",
    "training_data_path = Path('../data/train')\n",
    "\n",
    "# Path to test data\n",
    "test_data_path = Path('../data/test')\n",
    "\n",
    "# Path to baseline data\n",
    "baseline_data_path = Path('../data/baseline')\n",
    "\n",
    "# Path tp output folder\n",
    "output_folder = Path('../output')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training Set\n",
    "\n",
    "In this section we are **creating the Training Set**, fusing togheter the data from:\n",
    "- _train_set.tsv_\n",
    "- _train_ids.txt_\n",
    "- _train.fasta_\n",
    "- _train_embeddings.h5_\n",
    "- _train_protein2ipr.dat_\n",
    "- _go-basic.obo_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_set.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sub_ontology</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0110165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0043226</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID        sub_ontology     GO_term\n",
       "0  P91124  cellular_component  GO:0005575\n",
       "1  P91124  cellular_component  GO:0110165\n",
       "2  P91124  cellular_component  GO:0005737\n",
       "3  P91124  cellular_component  GO:0005622\n",
       "4  P91124  cellular_component  GO:0043226"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_set = pd.read_csv(training_data_path / 'train_set.tsv', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "train_set.rename(columns={'Protein_ID': 'ID', 'aspect' : 'sub_ontology'}, inplace=True)\n",
    "\n",
    "# Display the first few rows of the train set\n",
    "train_set.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_ids.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['P91124', 'Q55DL5', 'O81027', 'Q04418', 'Q8IXT2']\n"
     ]
    }
   ],
   "source": [
    "# Extracting train_ids.txt\n",
    "with open(training_data_path / 'train_ids.txt', 'r') as file:\n",
    "    train_ids = file.read().splitlines()\n",
    "\n",
    "# Display the first few IDs to verify\n",
    "print(train_ids[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set (123969).\n",
      "Proceeding with the analysis.\n"
     ]
    }
   ],
   "source": [
    "if len(train_ids) == len(train_set['ID'].unique()):\n",
    "    print(f\"The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set ({len(train_ids)}).\\n\"\n",
    "          \"Proceeding with the analysis.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train.fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: P91124\n",
      "Name: P91124\n",
      "Description: P91124\n",
      "Number of features: 0\n",
      "Seq('MAAQFYERNTSGMNADRFMTRLTDESTVNTMQRHYWTARQFIRTKLGKKEDEHL...NLL')\n"
     ]
    }
   ],
   "source": [
    "train_fasta_list = list(SeqIO.parse(training_data_path / 'train.fasta', 'fasta'))\n",
    "\n",
    "# Print the first sequence to verify\n",
    "print(train_fasta_list[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's transform it into a DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>num_features</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>P91124</td>\n",
       "      <td>P91124</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O81027</td>\n",
       "      <td>O81027</td>\n",
       "      <td>O81027</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q04418</td>\n",
       "      <td>Q04418</td>\n",
       "      <td>Q04418</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID    name description  num_features  \\\n",
       "0  P91124  P91124      P91124             0   \n",
       "1  Q55DL5  Q55DL5      Q55DL5             0   \n",
       "2  O81027  O81027      O81027             0   \n",
       "3  Q04418  Q04418      Q04418             0   \n",
       "4  Q8IXT2  Q8IXT2      Q8IXT2             0   \n",
       "\n",
       "                                            sequence  \n",
       "0  (M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...  \n",
       "1  (M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...  \n",
       "2  (M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...  \n",
       "3  (M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...  \n",
       "4  (M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract relevant information from SeqRecord\n",
    "train_fasta_dict = [{\n",
    "    'ID': record.id,\n",
    "    'name': record.name,\n",
    "    'description': record.description,\n",
    "    'num_features': len(record.features),\n",
    "    'sequence': record.seq,\n",
    "} for record in train_fasta_list]\n",
    "\n",
    "# Create a DataFrame from the extracted data\n",
    "train_fasta = pd.DataFrame(train_fasta_dict)\n",
    "\n",
    "# Display the DataFrame\n",
    "train_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking if `ID`, `name` and `description` have the same information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 differences between the ID and name columns.\n",
      "We have a total of 0 differences between the ID and description columns.\n"
     ]
    }
   ],
   "source": [
    "# Checking for differences between the ID and name columns\n",
    "diff_id_name = sum(train_fasta['ID'] != train_fasta['name'])\n",
    "\n",
    "# Checking for differences between the ID and description columns\n",
    "diff_id_description = sum(train_fasta['ID'] != train_fasta['description'])\n",
    "\n",
    "print(f\"We have a total of {diff_id_name} differences between the ID and name columns.\\nWe have a total of {diff_id_description} differences between the ID and description columns.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking if `num_features` has value different from 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 sequences with features.\n"
     ]
    }
   ],
   "source": [
    "num_features_values = sum(train_fasta['num_features'] != 0)\n",
    "\n",
    "print(f\"We have a total of {num_features_values} sequences with features.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Nice!**\n",
    "\n",
    "We found that `ID`, `name` and `description` columns are the same. We can remove two of them (`name`, `descritpion`). \n",
    "\n",
    "In the same way we saw that `num_features` is not very informative since it has only 0, let's remove it as well\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>(M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>(M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O81027</td>\n",
       "      <td>(M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q04418</td>\n",
       "      <td>(M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>(M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID                                           sequence\n",
       "0  P91124  (M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...\n",
       "1  Q55DL5  (M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...\n",
       "2  O81027  (M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...\n",
       "3  Q04418  (M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...\n",
       "4  Q8IXT2  (M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ..."
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_fasta.drop(columns=['name', 'description', 'num_features'], inplace=True)\n",
    "\n",
    "\n",
    "train_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_embeddings.h5`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>[0.0682, -0.04648, 0.001752, -0.00858, 0.00376...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW4</td>\n",
       "      <td>[0.002447, 0.007053, 0.06445, 0.00727, -0.0335...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>[0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings\n",
       "0  A0A009IHW8  [0.0682, -0.04648, 0.001752, -0.00858, 0.00376...\n",
       "1  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...\n",
       "2  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...\n",
       "3  A0A023FBW4  [0.002447, 0.007053, 0.06445, 0.00727, -0.0335...\n",
       "4  A0A023FBW7  [0.04932, 0.02069, 0.10864, 0.01634, -0.05106,..."
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"train_embeddings.h5\"\n",
    "\n",
    "data_list = []\n",
    "\n",
    "with h5py.File(training_data_path / filename, \"r\") as f:\n",
    "    for dataset_name in f.keys():\n",
    "        dataset = f[dataset_name][:]\n",
    "        data_list.append([dataset_name, dataset])\n",
    "\n",
    "train_embeddings = pd.DataFrame(data_list, columns=[\"ID\", \"embeddings\"])\n",
    "\n",
    "train_embeddings.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(123969, 1024)\n"
     ]
    }
   ],
   "source": [
    "train_embeddings_np = np.array(train_embeddings['embeddings'].tolist())\n",
    "print(train_embeddings_np.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_protein2ipr.dat`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR000157</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) domain</td>\n",
       "      <td>PS50104</td>\n",
       "      <td>133</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR000157</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) domain</td>\n",
       "      <td>SM00255</td>\n",
       "      <td>134</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR035897</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) dom...</td>\n",
       "      <td>G3DSA:3.40.50.10140</td>\n",
       "      <td>80</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR035897</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) dom...</td>\n",
       "      <td>SSF52200</td>\n",
       "      <td>128</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>IPR006910</td>\n",
       "      <td>Rad21/Rec8-like protein, N-terminal</td>\n",
       "      <td>PF04825</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        ipr                                             domain  \\\n",
       "0  A0A009IHW8  IPR000157  Toll/interleukin-1 receptor homology (TIR) domain   \n",
       "1  A0A009IHW8  IPR000157  Toll/interleukin-1 receptor homology (TIR) domain   \n",
       "2  A0A009IHW8  IPR035897  Toll/interleukin-1 receptor homology (TIR) dom...   \n",
       "3  A0A009IHW8  IPR035897  Toll/interleukin-1 receptor homology (TIR) dom...   \n",
       "4  A0A021WW32  IPR006910                Rad21/Rec8-like protein, N-terminal   \n",
       "\n",
       "              familyID  start  end  \n",
       "0              PS50104    133  266  \n",
       "1              SM00255    134  258  \n",
       "2  G3DSA:3.40.50.10140     80  266  \n",
       "3             SSF52200    128  249  \n",
       "4              PF04825      1  103  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_protein2ipr = pd.read_csv(training_data_path / 'train_protein2ipr.dat', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "train_protein2ipr.columns = ['ID', 'ipr', 'domain', 'familyID', 'start', 'end']\n",
    "\n",
    "# Display the first few rows of the train set\n",
    "train_protein2ipr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train protein2ipr ((1103545, 6)):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>(IPR000157, IPR000157, IPR035897, IPR035897)</td>\n",
       "      <td>(Toll/interleukin-1 receptor homology (TIR) do...</td>\n",
       "      <td>(PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)</td>\n",
       "      <td>(133, 134, 80, 128)</td>\n",
       "      <td>(266, 258, 266, 249)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FDY8</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                                ipr  \\\n",
       "0  A0A009IHW8       (IPR000157, IPR000157, IPR035897, IPR035897)   \n",
       "1  A0A021WW32                  (IPR006910, IPR039781, IPR049589)   \n",
       "2  A0A021WZA4  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "3  A0A023FBW7                                       (IPR045797,)   \n",
       "4  A0A023FDY8                                       (IPR045797,)   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Toll/interleukin-1 receptor homology (TIR) do...   \n",
       "1  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "2  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "3                                 (Evasins Class A,)   \n",
       "4                                 (Evasins Class A,)   \n",
       "\n",
       "                                            familyID  \\\n",
       "0  (PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)   \n",
       "1                      (PF04825, PTHR12585, cd21792)   \n",
       "2  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "3                                         (PF19429,)   \n",
       "4                                         (PF19429,)   \n",
       "\n",
       "                       start                             end  \n",
       "0        (133, 134, 80, 128)            (266, 258, 266, 249)  \n",
       "1                (1, 1, 327)                 (103, 607, 404)  \n",
       "2  (6, 31, 33, 329, 64, 286)  (496, 476, 175, 480, 258, 482)  \n",
       "3                      (38,)                          (110,)  \n",
       "4                      (38,)                          (110,)  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by 'ID' and aggregate other columns into lists\n",
    "train_protein2ipr_grouped = train_protein2ipr.groupby('ID').agg(lambda x: tuple(x)).reset_index()\n",
    "\n",
    "print(f\"Train protein2ipr ({train_protein2ipr.shape}):\")\n",
    "train_protein2ipr_grouped.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Still have to understand what *tizio, caio, sempronio* are."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `go-basic.obo`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All IDs with namespaces:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0048308, GO:0048311]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0007005]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0019952</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0050876</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                              name           namespace  \\\n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "1  GO:0000002  mitochondrial genome maintenance  biological_process   \n",
       "2  GO:0000003                      reproduction  biological_process   \n",
       "3  GO:0019952                      reproduction  biological_process   \n",
       "4  GO:0050876                      reproduction  biological_process   \n",
       "\n",
       "                       is_a part_of  \n",
       "0  [GO:0048308, GO:0048311]      []  \n",
       "1              [GO:0007005]      []  \n",
       "2              [GO:0008150]      []  \n",
       "3              [GO:0008150]      []  \n",
       "4              [GO:0008150]      []  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "file_path = \"../data/train/go-basic.obo\"  # Replace with your file path\n",
    "\n",
    "# Step 1: Initialize storage for GO terms\n",
    "go_terms = []\n",
    "\n",
    "# Step 2: Parse the .obo file\n",
    "with open(file_path, 'r') as file:\n",
    "    current_term = {}\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "        \n",
    "        # Start of a new term\n",
    "        if line == \"[Term]\":\n",
    "            if current_term:  # Save the previous term\n",
    "                go_terms.append(current_term)\n",
    "            current_term = {}  # Start a new term\n",
    "            \n",
    "        elif line.startswith(\"id:\"):\n",
    "            current_term['ID'] = line.split(\"id: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"alt_id:\"):\n",
    "            alt_id = line.split(\"alt_id: \")[1]\n",
    "            current_term.setdefault('alt_ids', []).append(alt_id)\n",
    "            \n",
    "        elif line.startswith(\"name:\"):\n",
    "            current_term['name'] = line.split(\"name: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"namespace:\"):\n",
    "            current_term['namespace'] = line.split(\"namespace: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"is_a\"):\n",
    "            match = re.search(r\"GO:\\d+\", line)  # Search for GO ID\n",
    "            if match:  # Check if a match was found\n",
    "                is_a_id = match.group()\n",
    "                current_term.setdefault('is_a', []).append(is_a_id)\n",
    "                \n",
    "        elif line.startswith(\"relationship: part_of\"):\n",
    "            match = re.search(r\"GO:\\d+\", line)  # Search for GO ID\n",
    "            if match:  # Check if a match was found\n",
    "                part_of_id = match.group()\n",
    "                current_term.setdefault('part_of', []).append(part_of_id)\n",
    "\n",
    "            \n",
    "    # Add the last term\n",
    "    if current_term:\n",
    "        go_terms.append(current_term)\n",
    "\n",
    "# Step 3: Create a unified list of all IDs (primary and alt_ids)\n",
    "expanded_terms = []\n",
    "for term in go_terms:\n",
    "    primary_id = term['ID']\n",
    "    alt_ids = term.get('alt_ids', [])\n",
    "    all_ids = [primary_id] + alt_ids\n",
    "    \n",
    "    for term_id in all_ids:\n",
    "        expanded_terms.append({\n",
    "            'ID': term_id,\n",
    "            'name': term.get('name'),\n",
    "            'namespace': term.get('namespace'),\n",
    "            'is_a': term.get('is_a', []),\n",
    "            'part_of': term.get('part_of', [])\n",
    "        })\n",
    "\n",
    "# Step 4: Convert to a DataFrame\n",
    "df = pd.DataFrame(expanded_terms)\n",
    "\n",
    "# Step 5: Filter by namespace and explode relationships\n",
    "df_is_a = df.explode('is_a').dropna(subset=['is_a'])\n",
    "df_part_of = df.explode('part_of').dropna(subset=['part_of'])\n",
    "\n",
    "\n",
    "print(\"All IDs with namespaces:\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'is_a' relationships:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0048308</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0048311</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0007005</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0019952</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                              name           namespace  \\\n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "1  GO:0000002  mitochondrial genome maintenance  biological_process   \n",
       "2  GO:0000003                      reproduction  biological_process   \n",
       "3  GO:0019952                      reproduction  biological_process   \n",
       "\n",
       "         is_a part_of  \n",
       "0  GO:0048308      []  \n",
       "0  GO:0048311      []  \n",
       "1  GO:0007005      []  \n",
       "2  GO:0008150      []  \n",
       "3  GO:0008150      []  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"'is_a' relationships:\")\n",
    "df_is_a.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'part_of' relationships:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GO:0000015</td>\n",
       "      <td>phosphopyruvate hydratase complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>[GO:1902494]</td>\n",
       "      <td>GO:0005829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GO:0000022</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GO:0000022</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0007052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GO:1905121</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GO:1905121</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0007052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID                               name           namespace  \\\n",
       "16  GO:0000015  phosphopyruvate hydratase complex  cellular_component   \n",
       "22  GO:0000022         mitotic spindle elongation  biological_process   \n",
       "22  GO:0000022         mitotic spindle elongation  biological_process   \n",
       "23  GO:1905121         mitotic spindle elongation  biological_process   \n",
       "23  GO:1905121         mitotic spindle elongation  biological_process   \n",
       "\n",
       "                        is_a     part_of  \n",
       "16              [GO:1902494]  GO:0005829  \n",
       "22  [GO:0051231, GO:1903047]  GO:0000070  \n",
       "22  [GO:0051231, GO:1903047]  GO:0007052  \n",
       "23  [GO:0051231, GO:1903047]  GO:0000070  \n",
       "23  [GO:0051231, GO:1903047]  GO:0007052  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"'part_of' relationships:\")\n",
    "df_part_of.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging all previous extractions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us collect the dataframes and check lengths. Noting that:\n",
    "- _train_set.tsv_ contains the proteins, their GO annotations and their corresponding aspects which will be used to subdivde it into three separate datasets\n",
    "- _train.fasta_ and _train_embeddings.h5_ can be grouped together since they refers to the whole input\n",
    "- _train_protein2ipr.dat_\n",
    "- _go-basic.obo_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: CC (84638, 2), MF (55698, 2), BP (83064, 2)\n",
      "Cellular Component (84638):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0000785, GO:003299...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0071944, GO:000588...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005622, GO:000582...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005576, GO:0005615)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>(GO:0005575, GO:0005737, GO:0042175, GO:003299...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                            GO_term\n",
       "0  A0A021WW32  (GO:0005575, GO:0110165, GO:0000785, GO:003299...\n",
       "1  A0A021WZA4  (GO:0005575, GO:0110165, GO:0071944, GO:000588...\n",
       "2  A0A023GPJ3  (GO:0005575, GO:0110165, GO:0005622, GO:000582...\n",
       "3  A0A023GUT0   (GO:0005575, GO:0110165, GO:0005576, GO:0005615)\n",
       "4  A0A023IM54  (GO:0005575, GO:0005737, GO:0042175, GO:003299..."
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def group_and_combine(df, sub_ontology_value):\n",
    "    return df[df['sub_ontology'] == sub_ontology_value].groupby('ID')['GO_term'].apply(tuple).reset_index()\n",
    "\n",
    "# Create three dataframes for each sub_ontology value\n",
    "df_CC = group_and_combine(train_set, 'cellular_component')\n",
    "df_MF = group_and_combine(train_set, 'molecular_function')\n",
    "df_BP = group_and_combine(train_set, 'biological_process')\n",
    "\n",
    "print(f\"Shapes: CC {df_CC.shape}, MF {df_MF.shape}, BP {df_BP.shape}\")\n",
    "\n",
    "print(f\"Cellular Component ({df_CC.shape[0]}):\")\n",
    "df_CC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows missing from train_protein2ipr_grouped: 6522\n",
      "Combined DataFrame shape: (123969, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>[0.0682, -0.04648, 0.001752, -0.00858, 0.00376...</td>\n",
       "      <td>(M, S, L, E, Q, K, K, G, A, D, I, I, S, K, I, ...</td>\n",
       "      <td>(IPR000157, IPR000157, IPR035897, IPR035897)</td>\n",
       "      <td>(Toll/interleukin-1 receptor homology (TIR) do...</td>\n",
       "      <td>(PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)</td>\n",
       "      <td>(133, 134, 80, 128)</td>\n",
       "      <td>(266, 258, 266, 249)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW4</td>\n",
       "      <td>[0.002447, 0.007053, 0.06445, 0.00727, -0.0335...</td>\n",
       "      <td>(M, T, S, H, G, A, V, K, I, A, I, F, A, V, I, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>[0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...</td>\n",
       "      <td>(M, K, V, L, L, Y, I, A, A, S, C, L, M, L, L, ...</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A009IHW8  [0.0682, -0.04648, 0.001752, -0.00858, 0.00376...   \n",
       "1  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...   \n",
       "2  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...   \n",
       "3  A0A023FBW4  [0.002447, 0.007053, 0.06445, 0.00727, -0.0335...   \n",
       "4  A0A023FBW7  [0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, S, L, E, Q, K, K, G, A, D, I, I, S, K, I, ...   \n",
       "1  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "2  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "3  (M, T, S, H, G, A, V, K, I, A, I, F, A, V, I, ...   \n",
       "4  (M, K, V, L, L, Y, I, A, A, S, C, L, M, L, L, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0       (IPR000157, IPR000157, IPR035897, IPR035897)   \n",
       "1                  (IPR006910, IPR039781, IPR049589)   \n",
       "2  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "3                                                NaN   \n",
       "4                                       (IPR045797,)   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Toll/interleukin-1 receptor homology (TIR) do...   \n",
       "1  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "2  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "3                                                NaN   \n",
       "4                                 (Evasins Class A,)   \n",
       "\n",
       "                                            familyID  \\\n",
       "0  (PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)   \n",
       "1                      (PF04825, PTHR12585, cd21792)   \n",
       "2  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "3                                                NaN   \n",
       "4                                         (PF19429,)   \n",
       "\n",
       "                       start                             end  \n",
       "0        (133, 134, 80, 128)            (266, 258, 266, 249)  \n",
       "1                (1, 1, 327)                 (103, 607, 404)  \n",
       "2  (6, 31, 33, 329, 64, 286)  (496, 476, 175, 480, 258, 482)  \n",
       "3                        NaN                             NaN  \n",
       "4                      (38,)                          (110,)  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## TRY NOT COMBINE EMBEDDINGS\n",
    "combined_train = pd.merge(train_embeddings, train_fasta, on='ID')\n",
    "combined_train = pd.merge(combined_train, train_protein2ipr_grouped, on='ID', how='left')\n",
    "\n",
    "# combined_train = pd.merge(train_fasta, train_protein2ipr_grouped, on='ID', how='left')\n",
    "\n",
    "missing_rows = combined_train[combined_train['ipr'].isna()].shape[0]\n",
    "print(f\"Number of rows missing from train_protein2ipr_grouped: {missing_rows}\")\n",
    "\n",
    "# combined_train = combined_train.drop('domain', axis=1)\n",
    "\n",
    "print(f\"Combined DataFrame shape: {combined_train.shape}\")\n",
    "combined_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: CC (84638, 8), MF (55698, 8), BP (83064, 8)\n",
      "Full df CC (84638, 8):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0000785, GO:003299...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0071944, GO:000588...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>(M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...</td>\n",
       "      <td>(IPR013087, IPR013087, IPR013087, IPR013087, I...</td>\n",
       "      <td>(Zinc finger C2H2-type, Zinc finger C2H2-type,...</td>\n",
       "      <td>(PF00096, PF00096, PF00096, PF00096, PS00028, ...</td>\n",
       "      <td>(644, 956, 983, 1013, 197, 646, 674, 957, 985,...</td>\n",
       "      <td>(666, 977, 1005, 1034, 218, 666, 694, 977, 100...</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005622, GO:000582...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>(M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...</td>\n",
       "      <td>(IPR000065, IPR000065, IPR009079, IPR009079)</td>\n",
       "      <td>(Leptin, Leptin, Four-helical cytokine-like, c...</td>\n",
       "      <td>(PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...</td>\n",
       "      <td>(28, 7, 24, 29)</td>\n",
       "      <td>(172, 172, 172, 172)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005576, GO:0005615)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>(M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...</td>\n",
       "      <td>(IPR003006, IPR003597, IPR003597, IPR007110, I...</td>\n",
       "      <td>(Immunoglobulin/major histocompatibility compl...</td>\n",
       "      <td>(PS00290, PF07654, SM00407, PS50835, PR01669, ...</td>\n",
       "      <td>(382, 312, 312, 294, 55, 135, 199, 247, 271, 3...</td>\n",
       "      <td>(388, 391, 394, 401, 63, 146, 216, 261, 287, 3...</td>\n",
       "      <td>(GO:0005575, GO:0005737, GO:0042175, GO:003299...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                           sequence  \\\n",
       "0  A0A021WW32  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "1  A0A021WZA4  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "2  A0A023GPJ3  (M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...   \n",
       "3  A0A023GUT0  (M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...   \n",
       "4  A0A023IM54  (M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0                  (IPR006910, IPR039781, IPR049589)   \n",
       "1  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "2  (IPR013087, IPR013087, IPR013087, IPR013087, I...   \n",
       "3       (IPR000065, IPR000065, IPR009079, IPR009079)   \n",
       "4  (IPR003006, IPR003597, IPR003597, IPR007110, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "1  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "2  (Zinc finger C2H2-type, Zinc finger C2H2-type,...   \n",
       "3  (Leptin, Leptin, Four-helical cytokine-like, c...   \n",
       "4  (Immunoglobulin/major histocompatibility compl...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                      (PF04825, PTHR12585, cd21792)   \n",
       "1  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "2  (PF00096, PF00096, PF00096, PF00096, PS00028, ...   \n",
       "3  (PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...   \n",
       "4  (PS00290, PF07654, SM00407, PS50835, PR01669, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                        (1, 1, 327)   \n",
       "1                          (6, 31, 33, 329, 64, 286)   \n",
       "2  (644, 956, 983, 1013, 197, 646, 674, 957, 985,...   \n",
       "3                                    (28, 7, 24, 29)   \n",
       "4  (382, 312, 312, 294, 55, 135, 199, 247, 271, 3...   \n",
       "\n",
       "                                                 end  \\\n",
       "0                                    (103, 607, 404)   \n",
       "1                     (496, 476, 175, 480, 258, 482)   \n",
       "2  (666, 977, 1005, 1034, 218, 666, 694, 977, 100...   \n",
       "3                               (172, 172, 172, 172)   \n",
       "4  (388, 391, 394, 401, 63, 146, 216, 261, 287, 3...   \n",
       "\n",
       "                                             GO_term  \n",
       "0  (GO:0005575, GO:0110165, GO:0000785, GO:003299...  \n",
       "1  (GO:0005575, GO:0110165, GO:0071944, GO:000588...  \n",
       "2  (GO:0005575, GO:0110165, GO:0005622, GO:000582...  \n",
       "3   (GO:0005575, GO:0110165, GO:0005576, GO:0005615)  \n",
       "4  (GO:0005575, GO:0005737, GO:0042175, GO:003299...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge combined_train with df_ab\n",
    "df_CC_full= pd.merge(combined_train, df_CC, on='ID', how='right')\n",
    "df_MF_full= pd.merge(combined_train, df_MF, on='ID', how='right')\n",
    "df_BP_full= pd.merge(combined_train, df_BP, on='ID', how='right')\n",
    "\n",
    "print(f\"Shapes: CC {df_CC_full.shape}, MF {df_MF_full.shape}, BP {df_BP_full.shape}\")\n",
    "\n",
    "print(f\"Full df CC {df_CC_full.shape}:\")\n",
    "df_CC_full.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: x_df_CC (84638, 7), y_df_CC (84638,), x_df_MF (55698, 7), y_df_MF (55698,), x_df_BP (83064, 7), y_df_BP (83064,)\n",
      "CC Input:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>(M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...</td>\n",
       "      <td>(IPR013087, IPR013087, IPR013087, IPR013087, I...</td>\n",
       "      <td>(Zinc finger C2H2-type, Zinc finger C2H2-type,...</td>\n",
       "      <td>(PF00096, PF00096, PF00096, PF00096, PS00028, ...</td>\n",
       "      <td>(644, 956, 983, 1013, 197, 646, 674, 957, 985,...</td>\n",
       "      <td>(666, 977, 1005, 1034, 218, 666, 694, 977, 100...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>(M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...</td>\n",
       "      <td>(IPR000065, IPR000065, IPR009079, IPR009079)</td>\n",
       "      <td>(Leptin, Leptin, Four-helical cytokine-like, c...</td>\n",
       "      <td>(PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...</td>\n",
       "      <td>(28, 7, 24, 29)</td>\n",
       "      <td>(172, 172, 172, 172)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>(M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...</td>\n",
       "      <td>(IPR003006, IPR003597, IPR003597, IPR007110, I...</td>\n",
       "      <td>(Immunoglobulin/major histocompatibility compl...</td>\n",
       "      <td>(PS00290, PF07654, SM00407, PS50835, PR01669, ...</td>\n",
       "      <td>(382, 312, 312, 294, 55, 135, 199, 247, 271, 3...</td>\n",
       "      <td>(388, 391, 394, 401, 63, 146, 216, 261, 287, 3...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                           sequence  \\\n",
       "0  A0A021WW32  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "1  A0A021WZA4  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "2  A0A023GPJ3  (M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...   \n",
       "3  A0A023GUT0  (M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...   \n",
       "4  A0A023IM54  (M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0                  (IPR006910, IPR039781, IPR049589)   \n",
       "1  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "2  (IPR013087, IPR013087, IPR013087, IPR013087, I...   \n",
       "3       (IPR000065, IPR000065, IPR009079, IPR009079)   \n",
       "4  (IPR003006, IPR003597, IPR003597, IPR007110, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "1  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "2  (Zinc finger C2H2-type, Zinc finger C2H2-type,...   \n",
       "3  (Leptin, Leptin, Four-helical cytokine-like, c...   \n",
       "4  (Immunoglobulin/major histocompatibility compl...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                      (PF04825, PTHR12585, cd21792)   \n",
       "1  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "2  (PF00096, PF00096, PF00096, PF00096, PS00028, ...   \n",
       "3  (PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...   \n",
       "4  (PS00290, PF07654, SM00407, PS50835, PR01669, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                        (1, 1, 327)   \n",
       "1                          (6, 31, 33, 329, 64, 286)   \n",
       "2  (644, 956, 983, 1013, 197, 646, 674, 957, 985,...   \n",
       "3                                    (28, 7, 24, 29)   \n",
       "4  (382, 312, 312, 294, 55, 135, 199, 247, 271, 3...   \n",
       "\n",
       "                                                 end  \n",
       "0                                    (103, 607, 404)  \n",
       "1                     (496, 476, 175, 480, 258, 482)  \n",
       "2  (666, 977, 1005, 1034, 218, 666, 694, 977, 100...  \n",
       "3                               (172, 172, 172, 172)  \n",
       "4  (388, 391, 394, 401, 63, 146, 216, 261, 287, 3...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_df_CC = df_CC_full.iloc[:, :-1]\n",
    "y_df_CC = df_CC_full.iloc[:, -1]\n",
    "x_df_MF = df_MF_full.iloc[:, :-1]\n",
    "y_df_MF = df_MF_full.iloc[:, -1]\n",
    "x_df_BP = df_BP_full.iloc[:, :-1]\n",
    "y_df_BP = df_BP_full.iloc[:, -1]\n",
    "\n",
    "print(f\"Shapes: x_df_CC {x_df_CC.shape}, y_df_CC {y_df_CC.shape}, x_df_MF {x_df_MF.shape}, y_df_MF {y_df_MF.shape}, x_df_BP {x_df_BP.shape}, y_df_BP {y_df_BP.shape}\")\n",
    "\n",
    "print(\"CC Input:\")\n",
    "x_df_CC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CC Target:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    (GO:0005575, GO:0110165, GO:0000785, GO:003299...\n",
       "1    (GO:0005575, GO:0110165, GO:0071944, GO:000588...\n",
       "2    (GO:0005575, GO:0110165, GO:0005622, GO:000582...\n",
       "3     (GO:0005575, GO:0110165, GO:0005576, GO:0005615)\n",
       "4    (GO:0005575, GO:0005737, GO:0042175, GO:003299...\n",
       "Name: GO_term, dtype: object"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"CC Target:\")\n",
    "y_df_CC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x_emb_CC shape: (84638, 1024)\n",
      "x_emb_MF shape: (55698, 1024)\n",
      "x_emb_BP shape: (83064, 1024)\n"
     ]
    }
   ],
   "source": [
    "# Filter the IDs based on sub_ontology\n",
    "cc_ids = train_set[train_set['sub_ontology'] == 'cellular_component']['ID'].unique()\n",
    "mf_ids = train_set[train_set['sub_ontology'] == 'molecular_function']['ID'].unique()\n",
    "bp_ids = train_set[train_set['sub_ontology'] == 'biological_process']['ID'].unique()\n",
    "\n",
    "# Create a dictionary to map IDs to their index in train_embeddings\n",
    "id_to_index = {id_: idx for idx, id_ in enumerate(train_embeddings['ID'])}\n",
    "\n",
    "# Get the indices for each sub_ontology\n",
    "cc_indices = [id_to_index[id_] for id_ in cc_ids if id_ in id_to_index]\n",
    "mf_indices = [id_to_index[id_] for id_ in mf_ids if id_ in id_to_index]\n",
    "bp_indices = [id_to_index[id_] for id_ in bp_ids if id_ in id_to_index]\n",
    "\n",
    "# Create the numpy matrices\n",
    "x_emb_CC = train_embeddings_np[cc_indices]\n",
    "x_emb_MF = train_embeddings_np[mf_indices]\n",
    "x_emb_BP = train_embeddings_np[bp_indices]\n",
    "\n",
    "print(f\"x_emb_CC shape: {x_emb_CC.shape}\")\n",
    "print(f\"x_emb_MF shape: {x_emb_MF.shape}\")\n",
    "print(f\"x_emb_BP shape: {x_emb_BP.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let us export the dataframe so that we don't have to run this code each time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrames have been exported successfully.\n"
     ]
    }
   ],
   "source": [
    "import filecmp\n",
    "\n",
    "# Function to export DataFrame to CSV with check\n",
    "def export_dataframe(df, filename, check=False):\n",
    "    output_path = output_folder / filename\n",
    "    temp_path = output_folder / f'temp_{filename}'\n",
    "    \n",
    "    # Save DataFrame to a temporary file\n",
    "    df.to_csv(temp_path, index=False)\n",
    "    \n",
    "    # Check if the file already exists\n",
    "    if output_path.exists():\n",
    "        # Compare the temporary file with the existing file\n",
    "        if filecmp.cmp(temp_path, output_path, shallow=False):\n",
    "            print(f\"{filename} is already up-to-date.\")\n",
    "            temp_path.unlink()  # Remove the temporary file\n",
    "        else:\n",
    "            if check:\n",
    "                # Ask the user if they want to overwrite the file\n",
    "                response = input(f\"{filename} already exists and is different. Do you want to overwrite it? (yes/no): \")\n",
    "            if not check or response.lower() == 'yes':\n",
    "                os.remove(output_path)\n",
    "                temp_path.rename(output_path)\n",
    "                print(f\"{filename} has been overwritten.\")\n",
    "            else:\n",
    "                temp_path.unlink()  # Remove the temporary file\n",
    "                print(f\"{filename} has not been overwritten.\")\n",
    "    else:\n",
    "        temp_path.rename(output_path)\n",
    "        print(f\"{filename} has been created.\")\n",
    "\n",
    "# Export DataFrames to CSV files with check\n",
    "#export_dataframe(x_df_CC, 'x_df_CC.csv')\n",
    "#export_dataframe(y_df_CC, 'y_df_CC.csv')\n",
    "#export_dataframe(x_df_MF, 'x_df_MF.csv')\n",
    "#export_dataframe(y_df_MF, 'y_df_MF.csv')\n",
    "#export_dataframe(x_df_BP, 'x_df_BP.csv')\n",
    "#export_dataframe(y_df_BP, 'y_df_BP.csv')\n",
    "\n",
    "print(\"DataFrames have been exported successfully.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Numpy arrays have been exported successfully.\n"
     ]
    }
   ],
   "source": [
    "# Function to export numpy array to .npy file\n",
    "def export_numpy_array(np_array, filename):\n",
    "    output_path = output_folder / filename\n",
    "    np.save(output_path, np_array)\n",
    "    print(f\"{filename} has been created.\")\n",
    "\n",
    "### Export numpy arrays to .npy files\n",
    "#export_numpy_array(x_emb_CC, 'x_emb_CC.npy')\n",
    "#export_numpy_array(x_emb_MF, 'x_emb_MF.npy')\n",
    "#export_numpy_array(x_emb_BP, 'x_emb_BP.npy')\n",
    "\n",
    "print(\"Numpy arrays have been exported successfully.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test set\n",
    "\n",
    "In this section we are **creating the Test set**, fusing togheter the data from:\n",
    "- _test_ids.txt_\n",
    "- _test.fasta_\n",
    "- _test_embeddings.h5_\n",
    "- _test_protein2ipr.dat_\n",
    "- _blast_test_results.tsv_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_ids.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['O43747', 'Q969H0', 'Q9JMA2', 'P18065', 'A0A8I6AN32']\n"
     ]
    }
   ],
   "source": [
    "# Extracting test_ids.txt\n",
    "with open(test_data_path / 'test_ids.txt', 'r') as file:\n",
    "    test_ids = file.read().splitlines()\n",
    "\n",
    "# Display the first few IDs to verify\n",
    "print(test_ids[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test.fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: O43747\n",
      "Name: O43747\n",
      "Description: O43747\n",
      "Number of features: 0\n",
      "Seq('MPAPIRLRELIRTIRTARTQAEEREMIQKECAAIRSSFREEDNTYRCRNVAKLL...SWQ')\n"
     ]
    }
   ],
   "source": [
    "test_fasta_list = list(SeqIO.parse(test_data_path / 'test.fasta', 'fasta'))\n",
    "\n",
    "# Print the first sequence to verify\n",
    "print(test_fasta_list[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>num_features</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O43747</td>\n",
       "      <td>O43747</td>\n",
       "      <td>O43747</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q969H0</td>\n",
       "      <td>Q969H0</td>\n",
       "      <td>Q969H0</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P18065</td>\n",
       "      <td>P18065</td>\n",
       "      <td>P18065</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        name description  num_features  \\\n",
       "0      O43747      O43747      O43747             0   \n",
       "1      Q969H0      Q969H0      Q969H0             0   \n",
       "2      Q9JMA2      Q9JMA2      Q9JMA2             0   \n",
       "3      P18065      P18065      P18065             0   \n",
       "4  A0A8I6AN32  A0A8I6AN32  A0A8I6AN32             0   \n",
       "\n",
       "                                            sequence  \n",
       "0  (M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...  \n",
       "1  (M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...  \n",
       "2  (M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...  \n",
       "3  (M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...  \n",
       "4  (M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract relevant information from SeqRecord\n",
    "test_fasta_dict = [{\n",
    "    'ID': record.id,\n",
    "    'name': record.name,\n",
    "    'description': record.description,\n",
    "    'num_features': len(record.features),\n",
    "    'sequence': record.seq,\n",
    "} for record in test_fasta_list]\n",
    "\n",
    "# Create a DataFrame from the extracted data\n",
    "test_fasta = pd.DataFrame(test_fasta_dict)\n",
    "\n",
    "# Display the DataFrame\n",
    "test_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check if ID, name and description are the same thing, as well as seeing if num_features has relevant information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 differences between the ID and name columns.\n",
      "We have a total of 0 differences between the ID and description columns.\n"
     ]
    }
   ],
   "source": [
    "# Checking for differences between the ID and name columns\n",
    "diff_id_name = sum(test_fasta['ID'] != test_fasta['name'])\n",
    "\n",
    "# Checking for differences between the ID and description columns\n",
    "diff_id_description = sum(test_fasta['ID'] != test_fasta['description'])\n",
    "\n",
    "print(f\"We have a total of {diff_id_name} differences between the ID and name columns.\\nWe have a total of {diff_id_description} differences between the ID and description columns.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 sequences with features.\n"
     ]
    }
   ],
   "source": [
    "num_features_values = sum(test_fasta['num_features'] != 0)\n",
    "\n",
    "print(f\"We have a total of {num_features_values} sequences with features.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's drop the useless columns, as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O43747</td>\n",
       "      <td>(M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q969H0</td>\n",
       "      <td>(M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>(M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P18065</td>\n",
       "      <td>(M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>(M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                           sequence\n",
       "0      O43747  (M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...\n",
       "1      Q969H0  (M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...\n",
       "2      Q9JMA2  (M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...\n",
       "3      P18065  (M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...\n",
       "4  A0A8I6AN32  (M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ..."
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_fasta.drop(columns=['name', 'description', 'num_features'], inplace=True)\n",
    "\n",
    "\n",
    "test_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's also check if the IDs are the same as the ones in the txt file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set (1000).\n",
      "Proceeding with the analysis.\n"
     ]
    }
   ],
   "source": [
    "len_ID = len(test_fasta['ID'].unique()) # assigned because gave problem on else statement print\n",
    "\n",
    "if len(test_ids) == len_ID:\n",
    "    print(f\"The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set ({len(test_ids)}).\\n\"\n",
    "          \"Proceeding with the analysis.\")\n",
    "else:\n",
    "    print(f'The numbers are not the same: test_ids are {len(test_ids)}, while the length of the fasta file is {len_ID})')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_embeddings.h5`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>[0.00979, -0.03973, 0.03653, -0.006447, -0.040...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>[0.02786, -0.01154, 0.008865, -0.01765, 0.0073...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4P506</td>\n",
       "      <td>[0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K1A2</td>\n",
       "      <td>[0.00882, 0.0835, -0.001374, -0.0003645, -0.06...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2K1V4</td>\n",
       "      <td>[0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings\n",
       "0  A0A0B4JCV4  [0.00979, -0.03973, 0.03653, -0.006447, -0.040...\n",
       "1  A0A0B4KHT0  [0.02786, -0.01154, 0.008865, -0.01765, 0.0073...\n",
       "2  A0A0B4P506  [0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...\n",
       "3  A0A0G2K1A2  [0.00882, 0.0835, -0.001374, -0.0003645, -0.06...\n",
       "4  A0A0G2K1V4  [0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0...."
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_list = []\n",
    "\n",
    "with h5py.File(test_data_path / \"test_embeddings.h5\", \"r\") as f:\n",
    "    for dataset_name in f.keys():\n",
    "        dataset = f[dataset_name][:]\n",
    "        data_list.append([dataset_name, dataset])\n",
    "\n",
    "test_embeddings = pd.DataFrame(data_list, columns=[\"ID\", \"embeddings\"])\n",
    "\n",
    "test_embeddings.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1000, 1024)\n"
     ]
    }
   ],
   "source": [
    "test_embeddings_np = np.array(test_embeddings['embeddings'].tolist())\n",
    "print(test_embeddings_np.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_protein2ipr.dat`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>IPR039915</td>\n",
       "      <td>TACC family</td>\n",
       "      <td>PTHR13924</td>\n",
       "      <td>38</td>\n",
       "      <td>1206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PF00643</td>\n",
       "      <td>177</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PF00643</td>\n",
       "      <td>236</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PS50119</td>\n",
       "      <td>173</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PS50119</td>\n",
       "      <td>235</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        ipr                  domain   familyID  start   end\n",
       "0  A0A0B4JCV4  IPR039915             TACC family  PTHR13924     38  1206\n",
       "1  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PF00643    177   219\n",
       "2  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PF00643    236   274\n",
       "3  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PS50119    173   220\n",
       "4  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PS50119    235   282"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_protein2ipr = pd.read_csv(test_data_path / 'test_protein2ipr.dat', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "test_protein2ipr.columns = ['ID', 'ipr', 'domain', 'familyID', 'start', 'end']\n",
    "\n",
    "# Remove 'domain' that is useless\n",
    "test_protein2ipr.drop('domain', axis=1)\n",
    "\n",
    "test_protein2ipr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test protein2ipr ((11263, 6)):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>(IPR039915,)</td>\n",
       "      <td>(TACC family,)</td>\n",
       "      <td>(PTHR13924,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(1206,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>(IPR000315, IPR000315, IPR000315, IPR000315, I...</td>\n",
       "      <td>(B-box-type zinc finger, B-box-type zinc finge...</td>\n",
       "      <td>(PF00643, PF00643, PS50119, PS50119, SM00336, ...</td>\n",
       "      <td>(177, 236, 173, 235, 173, 235, 976, 826, 988, ...</td>\n",
       "      <td>(219, 274, 220, 282, 220, 276, 1048, 839, 1004...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4P506</td>\n",
       "      <td>(IPR003417, IPR003417, IPR036552, IPR036552)</td>\n",
       "      <td>(Core-binding factor, beta subunit, Core-bindi...</td>\n",
       "      <td>(PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)</td>\n",
       "      <td>(1, 1, 1, 4)</td>\n",
       "      <td>(164, 168, 142, 140)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K1A2</td>\n",
       "      <td>(IPR010255, IPR019791, IPR019791, IPR019791, I...</td>\n",
       "      <td>(Haem peroxidase superfamily, Haem peroxidase,...</td>\n",
       "      <td>(SSF48113, PF03098, PR00457, PR00457, PR00457,...</td>\n",
       "      <td>(142, 148, 172, 226, 374, 392, 417, 470, 598, ...</td>\n",
       "      <td>(718, 692, 183, 241, 392, 412, 443, 480, 618, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2K1V4</td>\n",
       "      <td>(IPR000048, IPR000048, IPR001609, IPR001609, I...</td>\n",
       "      <td>(IQ motif, EF-hand binding site, IQ motif, EF-...</td>\n",
       "      <td>(PS50096, SM00015, PF00063, PR00193, PR00193, ...</td>\n",
       "      <td>(789, 788, 89, 116, 172, 228, 459, 513, 86, 80...</td>\n",
       "      <td>(818, 810, 774, 135, 197, 255, 487, 541, 786, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                                ipr  \\\n",
       "0  A0A0B4JCV4                                       (IPR039915,)   \n",
       "1  A0A0B4KHT0  (IPR000315, IPR000315, IPR000315, IPR000315, I...   \n",
       "2  A0A0B4P506       (IPR003417, IPR003417, IPR036552, IPR036552)   \n",
       "3  A0A0G2K1A2  (IPR010255, IPR019791, IPR019791, IPR019791, I...   \n",
       "4  A0A0G2K1V4  (IPR000048, IPR000048, IPR001609, IPR001609, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0                                     (TACC family,)   \n",
       "1  (B-box-type zinc finger, B-box-type zinc finge...   \n",
       "2  (Core-binding factor, beta subunit, Core-bindi...   \n",
       "3  (Haem peroxidase superfamily, Haem peroxidase,...   \n",
       "4  (IQ motif, EF-hand binding site, IQ motif, EF-...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                                       (PTHR13924,)   \n",
       "1  (PF00643, PF00643, PS50119, PS50119, SM00336, ...   \n",
       "2  (PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)   \n",
       "3  (SSF48113, PF03098, PR00457, PR00457, PR00457,...   \n",
       "4  (PS50096, SM00015, PF00063, PR00193, PR00193, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                              (38,)   \n",
       "1  (177, 236, 173, 235, 173, 235, 976, 826, 988, ...   \n",
       "2                                       (1, 1, 1, 4)   \n",
       "3  (142, 148, 172, 226, 374, 392, 417, 470, 598, ...   \n",
       "4  (789, 788, 89, 116, 172, 228, 459, 513, 86, 80...   \n",
       "\n",
       "                                                 end  \n",
       "0                                            (1206,)  \n",
       "1  (219, 274, 220, 282, 220, 276, 1048, 839, 1004...  \n",
       "2                               (164, 168, 142, 140)  \n",
       "3  (718, 692, 183, 241, 392, 412, 443, 480, 618, ...  \n",
       "4  (818, 810, 774, 135, 197, 255, 487, 541, 786, ...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by 'ID' and aggregate other columns into lists\n",
    "test_protein2ipr_grouped = test_protein2ipr.groupby('ID').agg(lambda x: tuple(x)).reset_index()\n",
    "\n",
    "print(f\"Test protein2ipr ({test_protein2ipr.shape}):\")\n",
    "test_protein2ipr_grouped.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `blast_test_results.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows missing from train_protein2ipr_grouped: 19\n",
      "Combined DataFrame shape: (1000, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O43747</td>\n",
       "      <td>(M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...</td>\n",
       "      <td>(IPR002553, IPR008152, IPR008152, IPR008153, I...</td>\n",
       "      <td>(Clathrin/coatomer adaptor, adaptin-like, N-te...</td>\n",
       "      <td>(PF01602, PF02883, SM00809, PS50180, G3DSA:1.2...</td>\n",
       "      <td>(23, 706, 699, 702, 1, 695, 22, 3)</td>\n",
       "      <td>(573, 817, 817, 817, 588, 821, 599, 820)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q969H0</td>\n",
       "      <td>(M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...</td>\n",
       "      <td>(IPR001680, IPR001680, IPR001680, IPR001680, I...</td>\n",
       "      <td>(WD40 repeat, WD40 repeat, WD40 repeat, WD40 r...</td>\n",
       "      <td>(PF00400, PF00400, PF00400, PF00400, PF00400, ...</td>\n",
       "      <td>(373, 411, 451, 491, 532, 571, 611, 391, 417, ...</td>\n",
       "      <td>(407, 447, 487, 527, 567, 607, 650, 416, 456, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>(M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...</td>\n",
       "      <td>(IPR002616, IPR002616, IPR004803, IPR004803, I...</td>\n",
       "      <td>(tRNA-guanine(15) transglycosylase-like, tRNA-...</td>\n",
       "      <td>(PF01702, TIGR00449, MF_00168, TIGR00430, G3DS...</td>\n",
       "      <td>(27, 22, 15, 21, 16, 22)</td>\n",
       "      <td>(380, 380, 379, 380, 382, 380)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P18065</td>\n",
       "      <td>(M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...</td>\n",
       "      <td>(IPR000716, IPR000716, IPR000716, IPR000716, I...</td>\n",
       "      <td>(Thyroglobulin type-1, Thyroglobulin type-1, T...</td>\n",
       "      <td>(PF00086, PS00484, PS51162, SM00211, cd00191, ...</td>\n",
       "      <td>(227, 257, 224, 258, 226, 42, 38, 40, 39, 87, ...</td>\n",
       "      <td>(306, 291, 306, 310, 306, 111, 134, 133, 136, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>(M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...</td>\n",
       "      <td>(IPR001876, IPR001876, IPR001876, IPR001876, I...</td>\n",
       "      <td>(Zinc finger, RanBP2-type, Zinc finger, RanBP2...</td>\n",
       "      <td>(PF00641, PS01358, PS50199, SM00547, SSF90209)</td>\n",
       "      <td>(177, 181, 177, 179, 170)</td>\n",
       "      <td>(208, 202, 208, 205, 209)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                           sequence  \\\n",
       "0      O43747  (M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...   \n",
       "1      Q969H0  (M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...   \n",
       "2      Q9JMA2  (M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...   \n",
       "3      P18065  (M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...   \n",
       "4  A0A8I6AN32  (M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0  (IPR002553, IPR008152, IPR008152, IPR008153, I...   \n",
       "1  (IPR001680, IPR001680, IPR001680, IPR001680, I...   \n",
       "2  (IPR002616, IPR002616, IPR004803, IPR004803, I...   \n",
       "3  (IPR000716, IPR000716, IPR000716, IPR000716, I...   \n",
       "4  (IPR001876, IPR001876, IPR001876, IPR001876, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Clathrin/coatomer adaptor, adaptin-like, N-te...   \n",
       "1  (WD40 repeat, WD40 repeat, WD40 repeat, WD40 r...   \n",
       "2  (tRNA-guanine(15) transglycosylase-like, tRNA-...   \n",
       "3  (Thyroglobulin type-1, Thyroglobulin type-1, T...   \n",
       "4  (Zinc finger, RanBP2-type, Zinc finger, RanBP2...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0  (PF01602, PF02883, SM00809, PS50180, G3DSA:1.2...   \n",
       "1  (PF00400, PF00400, PF00400, PF00400, PF00400, ...   \n",
       "2  (PF01702, TIGR00449, MF_00168, TIGR00430, G3DS...   \n",
       "3  (PF00086, PS00484, PS51162, SM00211, cd00191, ...   \n",
       "4     (PF00641, PS01358, PS50199, SM00547, SSF90209)   \n",
       "\n",
       "                                               start  \\\n",
       "0                 (23, 706, 699, 702, 1, 695, 22, 3)   \n",
       "1  (373, 411, 451, 491, 532, 571, 611, 391, 417, ...   \n",
       "2                           (27, 22, 15, 21, 16, 22)   \n",
       "3  (227, 257, 224, 258, 226, 42, 38, 40, 39, 87, ...   \n",
       "4                          (177, 181, 177, 179, 170)   \n",
       "\n",
       "                                                 end  \n",
       "0           (573, 817, 817, 817, 588, 821, 599, 820)  \n",
       "1  (407, 447, 487, 527, 567, 607, 650, 416, 456, ...  \n",
       "2                     (380, 380, 379, 380, 382, 380)  \n",
       "3  (306, 291, 306, 310, 306, 111, 134, 133, 136, ...  \n",
       "4                          (208, 202, 208, 205, 209)  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combined_test = pd.merge(test_embeddings, test_fasta, on='ID')\n",
    "# combined_test = pd.merge(combined_test, test_protein2ipr_grouped, on='ID', how='left')\n",
    "\n",
    "combined_test = pd.merge(test_fasta, test_protein2ipr_grouped, on='ID', how='left')\n",
    "\n",
    "missing_rows = combined_test[combined_test['ipr'].isna()].shape[0]\n",
    "print(f\"Number of rows missing from train_protein2ipr_grouped: {missing_rows}\")\n",
    "\n",
    "print(f\"Combined DataFrame shape: {combined_test.shape}\")\n",
    "combined_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combined_train and combined_test have the same columns: True\n"
     ]
    }
   ],
   "source": [
    "# Check if combined_train and combined_test have the same columns\n",
    "print(\"Combined_train and combined_test have the same columns:\" , set(combined_train.columns) == set(combined_test.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "### EXPORT\n",
    "#export_dataframe(combined_test, 'x_df_test.csv')\n",
    "#export_numpy_array(test_embeddings, 'test_emb.npy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PCA for embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n# Standardize\\nscaler = StandardScaler()\\nembeddings_standardized = scaler.fit_transform(train_embeddings_np)\\n\\n# Apply PCA\\n\",\\npca = PCA()\\npca.fit(embeddings_standardized)\\n\\n# Plot cumulative explained variance\\nexplained_variance = pca.explained_variance_ratio_\\nplt.figure(figsize=(10, 6))\\nplt.plot(np.cumsum(explained_variance), marker=\\'o\\')\\nplt.title(\\'Cumulative Explained Variance by PCA Components\\')\\nplt.xlabel(\\'Number of Components\\')\\nplt.ylabel(\\'Cumulative Explained Variance\\')\\nplt.grid(True)\\nplt.show()\\n\\n# Choose the number of components to retain (in our case 90% of variance)\\nn_components = np.argmax(np.cumsum(explained_variance) >= 0.90) + 1\\nprint(f\"Number of components to retain 90% variance: {n_components}\")\\n\\n# Apply PCA with chosen number of components\\npca = PCA(n_components=n_components)\\nreduced_embeddings_np = pca.fit_transform(embeddings_standardized)\\n\\n##Integrate reduced embeddings into combined_df\\n#combined_train[\\'reduced_embeddings\\'] = reduced_embeddings.tolist()\\n#print(combined_train[[\\'embeddings\\', \\'reduced_embeddings\\']].head())\\n'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "# Standardize\n",
    "scaler = StandardScaler()\n",
    "embeddings_standardized = scaler.fit_transform(train_embeddings_np)\n",
    "\n",
    "# Apply PCA\\n\",\n",
    "pca = PCA()\n",
    "pca.fit(embeddings_standardized)\n",
    "\n",
    "# Plot cumulative explained variance\n",
    "explained_variance = pca.explained_variance_ratio_\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.plot(np.cumsum(explained_variance), marker='o')\n",
    "plt.title('Cumulative Explained Variance by PCA Components')\n",
    "plt.xlabel('Number of Components')\n",
    "plt.ylabel('Cumulative Explained Variance')\n",
    "plt.grid(True)\n",
    "plt.show()\n",
    "\n",
    "# Choose the number of components to retain (in our case 90% of variance)\n",
    "n_components = np.argmax(np.cumsum(explained_variance) >= 0.90) + 1\n",
    "print(f\"Number of components to retain 90% variance: {n_components}\")\n",
    "\n",
    "# Apply PCA with chosen number of components\n",
    "pca = PCA(n_components=n_components)\n",
    "reduced_embeddings_np = pca.fit_transform(embeddings_standardized)\n",
    "\n",
    "##Integrate reduced embeddings into combined_df\n",
    "#combined_train['reduced_embeddings'] = reduced_embeddings.tolist()\n",
    "#print(combined_train[['embeddings', 'reduced_embeddings']].head())\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll see that considering each df separately doesn't change much."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'embeddings'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3791\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3790\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 3791\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[0;32m   3792\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[1;32mindex.pyx:152\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mindex.pyx:181\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'embeddings'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[46], line 37\u001b[0m\n\u001b[0;32m     33\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m df\n\u001b[0;32m     35\u001b[0m \u001b[38;5;66;03m# Apply PCA and add column for each dataset\u001b[39;00m\n\u001b[0;32m     36\u001b[0m \u001b[38;5;66;03m#x_df_BP = pca_col(x_df_BP, 'BP')\u001b[39;00m\n\u001b[1;32m---> 37\u001b[0m x_df_CC \u001b[38;5;241m=\u001b[39m pca_col(x_df_CC, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mCC\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "Cell \u001b[1;32mIn[46], line 2\u001b[0m, in \u001b[0;36mpca_col\u001b[1;34m(df, name, column_name, variance_threshold)\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpca_col\u001b[39m(df, name, column_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124membeddings\u001b[39m\u001b[38;5;124m'\u001b[39m, variance_threshold\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.90\u001b[39m):\n\u001b[1;32m----> 2\u001b[0m     embeddings \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray(df[column_name]\u001b[38;5;241m.\u001b[39mtolist())\n\u001b[0;32m      4\u001b[0m     scaler \u001b[38;5;241m=\u001b[39m StandardScaler()\n\u001b[0;32m      5\u001b[0m     embeddings_standardized \u001b[38;5;241m=\u001b[39m scaler\u001b[38;5;241m.\u001b[39mfit_transform(embeddings)\n",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\Lib\\site-packages\\pandas\\core\\frame.py:3893\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3891\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[0;32m   3892\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[1;32m-> 3893\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mget_loc(key)\n\u001b[0;32m   3894\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[0;32m   3895\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[1;32mc:\\ProgramData\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3798\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3793\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[0;32m   3794\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[0;32m   3795\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[0;32m   3796\u001b[0m     ):\n\u001b[0;32m   3797\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[1;32m-> 3798\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[0;32m   3799\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[0;32m   3800\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[0;32m   3801\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[0;32m   3802\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[0;32m   3803\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'embeddings'"
     ]
    }
   ],
   "source": [
    "def pca_col(df, name, column_name='embeddings', variance_threshold=0.90):\n",
    "    embeddings = np.array(df[column_name].tolist())\n",
    "    \n",
    "    scaler = StandardScaler()\n",
    "    embeddings_standardized = scaler.fit_transform(embeddings)\n",
    "    \n",
    "    pca = PCA()\n",
    "    pca.fit(embeddings_standardized)\n",
    "    explained_variance = pca.explained_variance_ratio_\n",
    "\n",
    "    # Components to retain (in our case 90% of variance)\n",
    "    n_components = np.argmax(np.cumsum(explained_variance) >= variance_threshold) + 1\n",
    "        \n",
    "    plt.figure(figsize=(8, 4.5))\n",
    "    plt.axhline(y=variance_threshold, color='r', linestyle='--')\n",
    "    x_value = np.argmax(np.cumsum(explained_variance) >= variance_threshold) + 1\n",
    "    plt.axvline(x=x_value, color='g', linestyle='--')\n",
    "    plt.text(x_value + 10, 0.025, f'{x_value}', color='green', size='large', weight='bold')\n",
    "    plt.text(- 50, variance_threshold + 0.025, f'{variance_threshold:.2f}', color='red', size='large', weight='bold')\n",
    "\n",
    "    plt.plot(np.cumsum(explained_variance), marker='o')\n",
    "    plt.title(f'Cumulative Explained Variance by PCA Components for {name}')\n",
    "    plt.xlabel('Number of Components')\n",
    "    plt.ylabel('Cumulative Explained Variance')\n",
    "    plt.grid(True)\n",
    "    plt.show()\n",
    "    \n",
    "    pca = PCA(n_components=n_components)\n",
    "    reduced_embeddings = pca.fit_transform(embeddings_standardized)\n",
    "    \n",
    "    # Integrate reduced embeddings into the DataFrame\n",
    "    df[f'reduced_{column_name}'] = reduced_embeddings.tolist()\n",
    "    return df\n",
    "\n",
    "# Apply PCA and add column for each dataset\n",
    "#x_df_BP = pca_col(x_df_BP, 'BP')\n",
    "x_df_CC = pca_col(x_df_CC, 'CC')\n",
    "#x_df_MF = pca_col(x_df_MF, 'MF')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export_numpy_array(x_remb_CC, 'x_remb_CC.npy')\n",
    "#export_numpy_array(x_remb_MF, 'x_remb_MF.npy')\n",
    "#export_numpy_array(x_remb_BP, 'x_remb_BP.npy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "from sklearn.metrics import classification_report\n",
    "\n",
    "device = torch.device(\"cpu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data\n",
    "X = x_df_CC['reduced_embeddings']  # Assuming embeddings are already reduced\n",
    "X = pd.DataFrame(X.tolist())  # Convert list of lists to DataFrame\n",
    "\n",
    "\n",
    "# Preprocess target (MultiLabelBinarizer for multi-label classification)\n",
    "mlb = MultiLabelBinarizer()\n",
    "y = mlb.fit_transform(y_df_BP)\n",
    "\n",
    "# Split data\n",
    "X_train_full, X_test_full, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [1/50], Loss: 6.5290\n",
      "Epoch [2/50], Loss: 3.4102\n",
      "Epoch [3/50], Loss: 3.3666\n",
      "Epoch [4/50], Loss: 3.3397\n",
      "Epoch [5/50], Loss: 3.3185\n",
      "Epoch [6/50], Loss: 3.3038\n",
      "Epoch [7/50], Loss: 3.2919\n",
      "Epoch [8/50], Loss: 3.2807\n",
      "Epoch [9/50], Loss: 3.2744\n",
      "Epoch [10/50], Loss: 3.2660\n",
      "Epoch [11/50], Loss: 3.2584\n",
      "Epoch [12/50], Loss: 3.2526\n",
      "Epoch [13/50], Loss: 3.2452\n",
      "Epoch [14/50], Loss: 3.2391\n",
      "Epoch [15/50], Loss: 3.2315\n",
      "Epoch [16/50], Loss: 3.2239\n",
      "Epoch [17/50], Loss: 3.2161\n",
      "Epoch [18/50], Loss: 3.2070\n",
      "Epoch [19/50], Loss: 3.1974\n",
      "Epoch [20/50], Loss: 3.1873\n",
      "Epoch [21/50], Loss: 3.1742\n",
      "Epoch [22/50], Loss: 3.1631\n",
      "Epoch [23/50], Loss: 3.1518\n",
      "Epoch [24/50], Loss: 3.1411\n",
      "Epoch [25/50], Loss: 3.1278\n",
      "Epoch [26/50], Loss: 3.1160\n",
      "Epoch [27/50], Loss: 3.1030\n",
      "Epoch [28/50], Loss: 3.0884\n",
      "Epoch [29/50], Loss: 3.0768\n",
      "Epoch [30/50], Loss: 3.0668\n",
      "Epoch [31/50], Loss: 3.0547\n",
      "Epoch [32/50], Loss: 3.0441\n",
      "Epoch [33/50], Loss: 3.0311\n",
      "Epoch [34/50], Loss: 3.0207\n",
      "Epoch [35/50], Loss: 3.0067\n",
      "Epoch [36/50], Loss: 2.9898\n",
      "Epoch [37/50], Loss: 2.9840\n",
      "Epoch [38/50], Loss: 2.9744\n",
      "Epoch [39/50], Loss: 2.9614\n",
      "Epoch [40/50], Loss: 2.9494\n",
      "Epoch [41/50], Loss: 2.9371\n",
      "Epoch [42/50], Loss: 2.9262\n",
      "Epoch [43/50], Loss: 2.9173\n",
      "Epoch [44/50], Loss: 2.9072\n",
      "Epoch [45/50], Loss: 2.8995\n",
      "Epoch [46/50], Loss: 2.8880\n",
      "Epoch [47/50], Loss: 2.8787\n",
      "Epoch [48/50], Loss: 2.8653\n",
      "Epoch [49/50], Loss: 2.8560\n",
      "Epoch [50/50], Loss: 2.8470\n",
      "\n",
      "Neural Network Results:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\eliat\\AppData\\Roaming\\Python\\Python311\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "  GO:0000118       0.00      0.00      0.00        44\n",
      "  GO:0000123       0.00      0.00      0.00        64\n",
      "  GO:0000124       0.00      0.00      0.00        26\n",
      "  GO:0000131       0.00      0.00      0.00        19\n",
      "  GO:0000137       0.00      0.00      0.00        14\n",
      "  GO:0000138       0.00      0.00      0.00        18\n",
      "  GO:0000139       0.00      0.00      0.00       116\n",
      "  GO:0000145       0.00      0.00      0.00        10\n",
      "  GO:0000151       0.00      0.00      0.00       140\n",
      "  GO:0000152       0.00      0.00      0.00        33\n",
      "  GO:0000176       0.00      0.00      0.00        11\n",
      "  GO:0000178       0.00      0.00      0.00        19\n",
      "  GO:0000228       0.00      0.00      0.00       155\n",
      "  GO:0000235       0.00      0.00      0.00         8\n",
      "  GO:0000307       0.00      0.00      0.00        20\n",
      "  GO:0000313       0.00      0.00      0.00        95\n",
      "  GO:0000314       0.00      0.00      0.00        48\n",
      "  GO:0000315       0.00      0.00      0.00        44\n",
      "  GO:0000322       0.00      0.00      0.00       153\n",
      "  GO:0000323       0.00      0.00      0.00       372\n",
      "  GO:0000324       0.00      0.00      0.00       146\n",
      "  GO:0000325       0.00      0.00      0.00       151\n",
      "  GO:0000329       0.00      0.00      0.00        77\n",
      "  GO:0000407       0.00      0.00      0.00        11\n",
      "  GO:0000421       0.00      0.00      0.00        11\n",
      "  GO:0000428       0.00      0.00      0.00        92\n",
      "  GO:0000502       0.00      0.00      0.00        80\n",
      "  GO:0000775       0.00      0.00      0.00       165\n",
      "  GO:0000776       0.00      0.00      0.00       111\n",
      "  GO:0000779       0.00      0.00      0.00       119\n",
      "  GO:0000781       0.00      0.00      0.00        57\n",
      "  GO:0000785       0.00      0.00      0.00       441\n",
      "  GO:0000786       0.00      0.00      0.00        16\n",
      "  GO:0000791       0.00      0.00      0.00        40\n",
      "  GO:0000792       0.00      0.00      0.00        59\n",
      "  GO:0000793       0.00      0.00      0.00       187\n",
      "  GO:0000794       0.00      0.00      0.00        48\n",
      "  GO:0000795       0.00      0.00      0.00        14\n",
      "  GO:0000803       0.00      0.00      0.00        17\n",
      "  GO:0000922       0.00      0.00      0.00        47\n",
      "  GO:0000930       0.00      0.00      0.00         9\n",
      "  GO:0000932       0.00      0.00      0.00        44\n",
      "  GO:0000935       0.00      0.00      0.00        15\n",
      "  GO:0000940       0.00      0.00      0.00        22\n",
      "  GO:0000974       0.00      0.00      0.00         9\n",
      "  GO:0001411       0.00      0.00      0.00        12\n",
      "  GO:0001533       0.00      0.00      0.00        14\n",
      "  GO:0001650       0.00      0.00      0.00        72\n",
      "  GO:0001669       0.00      0.00      0.00        40\n",
      "  GO:0001673       0.00      0.00      0.00        19\n",
      "  GO:0001674       0.00      0.00      0.00        10\n",
      "  GO:0001725       0.00      0.00      0.00        18\n",
      "  GO:0001726       0.00      0.00      0.00        30\n",
      "  GO:0001750       0.00      0.00      0.00        27\n",
      "  GO:0001772       0.00      0.00      0.00         9\n",
      "  GO:0001891       0.00      0.00      0.00        19\n",
      "  GO:0001917       0.00      0.00      0.00        17\n",
      "  GO:0005575       1.00      1.00      1.00     16928\n",
      "  GO:0005576       0.00      0.00      0.00      1337\n",
      "  GO:0005581       0.00      0.00      0.00        19\n",
      "  GO:0005604       0.00      0.00      0.00        42\n",
      "  GO:0005615       0.00      0.00      0.00       992\n",
      "  GO:0005618       0.00      0.00      0.00       289\n",
      "  GO:0005622       0.80      1.00      0.88     13465\n",
      "  GO:0005628       0.00      0.00      0.00        22\n",
      "  GO:0005634       0.30      0.03      0.06      5432\n",
      "  GO:0005635       0.00      0.00      0.00       277\n",
      "  GO:0005637       0.00      0.00      0.00        10\n",
      "  GO:0005640       0.00      0.00      0.00        13\n",
      "  GO:0005643       0.00      0.00      0.00        57\n",
      "  GO:0005654       0.08      0.00      0.00      2061\n",
      "  GO:0005657       0.00      0.00      0.00        46\n",
      "  GO:0005665       0.00      0.00      0.00        18\n",
      "  GO:0005667       0.00      0.00      0.00       215\n",
      "  GO:0005669       0.00      0.00      0.00        19\n",
      "  GO:0005680       0.00      0.00      0.00        18\n",
      "  GO:0005681       0.00      0.00      0.00        94\n",
      "  GO:0005684       0.00      0.00      0.00        29\n",
      "  GO:0005685       0.00      0.00      0.00        12\n",
      "  GO:0005686       0.00      0.00      0.00        10\n",
      "  GO:0005694       0.00      0.00      0.00       811\n",
      "  GO:0005700       0.00      0.00      0.00        52\n",
      "  GO:0005703       0.00      0.00      0.00        15\n",
      "  GO:0005705       0.00      0.00      0.00        17\n",
      "  GO:0005721       0.00      0.00      0.00        21\n",
      "  GO:0005730       0.00      0.00      0.00       636\n",
      "  GO:0005732       0.00      0.00      0.00        11\n",
      "  GO:0005737       0.60      0.96      0.74     10168\n",
      "  GO:0005739       0.00      0.00      0.00      1734\n",
      "  GO:0005740       0.00      0.00      0.00       439\n",
      "  GO:0005741       0.00      0.00      0.00       102\n",
      "  GO:0005742       0.00      0.00      0.00         9\n",
      "  GO:0005743       0.00      0.00      0.00       298\n",
      "  GO:0005746       0.00      0.00      0.00        88\n",
      "  GO:0005747       0.00      0.00      0.00        52\n",
      "  GO:0005750       0.00      0.00      0.00        14\n",
      "  GO:0005751       0.00      0.00      0.00         5\n",
      "  GO:0005753       0.00      0.00      0.00        33\n",
      "  GO:0005758       0.00      0.00      0.00        31\n",
      "  GO:0005759       0.00      0.00      0.00       241\n",
      "  GO:0005761       0.00      0.00      0.00        92\n",
      "  GO:0005762       0.00      0.00      0.00        41\n",
      "  GO:0005763       0.00      0.00      0.00        48\n",
      "  GO:0005764       0.00      0.00      0.00       225\n",
      "  GO:0005765       0.00      0.00      0.00       112\n",
      "  GO:0005766       0.00      0.00      0.00        31\n",
      "  GO:0005767       0.00      0.00      0.00         9\n",
      "  GO:0005768       0.00      0.00      0.00       393\n",
      "  GO:0005769       0.00      0.00      0.00       104\n",
      "  GO:0005770       0.00      0.00      0.00        90\n",
      "  GO:0005771       0.00      0.00      0.00        21\n",
      "  GO:0005773       0.00      0.00      0.00       616\n",
      "  GO:0005774       0.00      0.00      0.00       250\n",
      "  GO:0005775       0.00      0.00      0.00        41\n",
      "  GO:0005776       0.00      0.00      0.00        35\n",
      "  GO:0005777       0.00      0.00      0.00       199\n",
      "  GO:0005778       0.00      0.00      0.00        34\n",
      "  GO:0005782       0.00      0.00      0.00        20\n",
      "  GO:0005783       0.00      0.00      0.00      1115\n",
      "  GO:0005788       0.00      0.00      0.00        80\n",
      "  GO:0005789       0.00      0.00      0.00       315\n",
      "  GO:0005790       0.00      0.00      0.00        12\n",
      "  GO:0005791       0.00      0.00      0.00        19\n",
      "  GO:0005793       0.00      0.00      0.00        24\n",
      "  GO:0005794       0.00      0.00      0.00       776\n",
      "  GO:0005795       0.00      0.00      0.00        65\n",
      "  GO:0005796       0.00      0.00      0.00        17\n",
      "  GO:0005797       0.00      0.00      0.00        16\n",
      "  GO:0005798       0.00      0.00      0.00        44\n",
      "  GO:0005801       0.00      0.00      0.00        26\n",
      "  GO:0005802       0.00      0.00      0.00       126\n",
      "  GO:0005811       0.00      0.00      0.00        56\n",
      "  GO:0005813       0.00      0.00      0.00       256\n",
      "  GO:0005814       0.00      0.00      0.00        37\n",
      "  GO:0005815       0.00      0.00      0.00       399\n",
      "  GO:0005816       0.00      0.00      0.00        45\n",
      "  GO:0005818       0.00      0.00      0.00        11\n",
      "  GO:0005819       0.00      0.00      0.00       207\n",
      "  GO:0005826       0.00      0.00      0.00        12\n",
      "  GO:0005829       0.00      0.00      0.00      3581\n",
      "  GO:0005832       0.00      0.00      0.00        11\n",
      "  GO:0005834       0.00      0.00      0.00         9\n",
      "  GO:0005838       0.00      0.00      0.00        26\n",
      "  GO:0005839       0.00      0.00      0.00        29\n",
      "  GO:0005840       0.00      0.00      0.00       298\n",
      "  GO:0005844       0.00      0.00      0.00        54\n",
      "  GO:0005847       0.00      0.00      0.00        12\n",
      "  GO:0005849       0.00      0.00      0.00        14\n",
      "  GO:0005852       0.00      0.00      0.00        14\n",
      "  GO:0005856       0.00      0.00      0.00      1061\n",
      "  GO:0005865       0.00      0.00      0.00        13\n",
      "  GO:0005871       0.00      0.00      0.00        12\n",
      "  GO:0005874       0.00      0.00      0.00       154\n",
      "  GO:0005875       0.00      0.00      0.00        55\n",
      "  GO:0005876       0.00      0.00      0.00        45\n",
      "  GO:0005879       0.00      0.00      0.00        15\n",
      "  GO:0005881       0.00      0.00      0.00        57\n",
      "  GO:0005882       0.00      0.00      0.00        23\n",
      "  GO:0005884       0.00      0.00      0.00        36\n",
      "  GO:0005885       0.00      0.00      0.00        10\n",
      "  GO:0005886       0.29      0.00      0.00      3178\n",
      "  GO:0005891       0.00      0.00      0.00        10\n",
      "  GO:0005892       0.00      0.00      0.00        15\n",
      "  GO:0005901       0.00      0.00      0.00        38\n",
      "  GO:0005902       0.00      0.00      0.00        27\n",
      "  GO:0005903       0.00      0.00      0.00        40\n",
      "  GO:0005905       0.00      0.00      0.00        21\n",
      "  GO:0005911       0.00      0.00      0.00       390\n",
      "  GO:0005912       0.00      0.00      0.00        68\n",
      "  GO:0005918       0.00      0.00      0.00        14\n",
      "  GO:0005921       0.00      0.00      0.00        15\n",
      "  GO:0005923       0.00      0.00      0.00        23\n",
      "  GO:0005925       0.00      0.00      0.00       116\n",
      "  GO:0005927       0.00      0.00      0.00        12\n",
      "  GO:0005929       0.00      0.00      0.00       521\n",
      "  GO:0005930       0.00      0.00      0.00       111\n",
      "  GO:0005933       0.00      0.00      0.00        67\n",
      "  GO:0005934       0.00      0.00      0.00        22\n",
      "  GO:0005935       0.00      0.00      0.00        51\n",
      "  GO:0005937       0.00      0.00      0.00        42\n",
      "  GO:0005938       0.00      0.00      0.00       258\n",
      "  GO:0005942       0.00      0.00      0.00        11\n",
      "  GO:0008021       0.00      0.00      0.00        79\n",
      "  GO:0008023       0.00      0.00      0.00        27\n",
      "  GO:0008076       0.00      0.00      0.00        31\n",
      "  GO:0008180       0.00      0.00      0.00        24\n",
      "  GO:0008287       0.00      0.00      0.00        32\n",
      "  GO:0008305       0.00      0.00      0.00        13\n",
      "  GO:0008328       0.00      0.00      0.00        23\n",
      "  GO:0009274       0.00      0.00      0.00       149\n",
      "  GO:0009277       0.00      0.00      0.00        49\n",
      "  GO:0009279       0.00      0.00      0.00        22\n",
      "  GO:0009295       0.00      0.00      0.00        51\n",
      "  GO:0009505       0.00      0.00      0.00        78\n",
      "  GO:0009506       0.00      0.00      0.00       188\n",
      "  GO:0009507       0.00      0.00      0.00       451\n",
      "  GO:0009524       0.00      0.00      0.00        13\n",
      "  GO:0009526       0.00      0.00      0.00       167\n",
      "  GO:0009528       0.00      0.00      0.00        11\n",
      "  GO:0009532       0.00      0.00      0.00       155\n",
      "  GO:0009534       0.00      0.00      0.00        94\n",
      "  GO:0009535       0.00      0.00      0.00        75\n",
      "  GO:0009536       0.00      0.00      0.00       652\n",
      "  GO:0009570       0.00      0.00      0.00       153\n",
      "  GO:0009579       0.00      0.00      0.00       127\n",
      "  GO:0009705       0.00      0.00      0.00        29\n",
      "  GO:0009706       0.00      0.00      0.00        11\n",
      "  GO:0009897       0.00      0.00      0.00       107\n",
      "  GO:0009898       0.00      0.00      0.00        73\n",
      "  GO:0009925       0.00      0.00      0.00       131\n",
      "  GO:0009941       0.00      0.00      0.00       110\n",
      "  GO:0009986       0.00      0.00      0.00       290\n",
      "  GO:0010008       0.00      0.00      0.00       106\n",
      "  GO:0010287       0.00      0.00      0.00        12\n",
      "  GO:0010369       0.00      0.00      0.00        13\n",
      "  GO:0010494       0.00      0.00      0.00        65\n",
      "  GO:0012505       0.10      0.00      0.00      2489\n",
      "  GO:0012506       0.00      0.00      0.00       291\n",
      "  GO:0012507       0.00      0.00      0.00        15\n",
      "  GO:0014069       0.00      0.00      0.00        96\n",
      "  GO:0014704       0.00      0.00      0.00        15\n",
      "  GO:0015030       0.00      0.00      0.00        18\n",
      "  GO:0015629       0.00      0.00      0.00       213\n",
      "  GO:0015630       0.00      0.00      0.00       695\n",
      "  GO:0015934       0.00      0.00      0.00       146\n",
      "  GO:0015935       0.00      0.00      0.00       114\n",
      "  GO:0016010       0.00      0.00      0.00         7\n",
      "  GO:0016020       0.30      0.01      0.02      4773\n",
      "  GO:0016028       0.00      0.00      0.00        12\n",
      "  GO:0016234       0.00      0.00      0.00        30\n",
      "  GO:0016235       0.00      0.00      0.00        12\n",
      "  GO:0016323       0.00      0.00      0.00       108\n",
      "  GO:0016324       0.00      0.00      0.00       173\n",
      "  GO:0016327       0.00      0.00      0.00        10\n",
      "  GO:0016328       0.00      0.00      0.00        29\n",
      "  GO:0016363       0.00      0.00      0.00        27\n",
      "  GO:0016459       0.00      0.00      0.00        17\n",
      "  GO:0016469       0.00      0.00      0.00        53\n",
      "  GO:0016471       0.00      0.00      0.00        15\n",
      "  GO:0016514       0.00      0.00      0.00        14\n",
      "  GO:0016528       0.00      0.00      0.00        28\n",
      "  GO:0016529       0.00      0.00      0.00        23\n",
      "  GO:0016581       0.00      0.00      0.00        12\n",
      "  GO:0016591       0.00      0.00      0.00        58\n",
      "  GO:0016592       0.00      0.00      0.00        32\n",
      "  GO:0016604       0.00      0.00      0.00       291\n",
      "  GO:0016605       0.00      0.00      0.00        20\n",
      "  GO:0016607       0.00      0.00      0.00       126\n",
      "  GO:0017053       0.00      0.00      0.00        42\n",
      "  GO:0018995       0.00      0.00      0.00       167\n",
      "  GO:0019005       0.00      0.00      0.00        24\n",
      "  GO:0019028       0.00      0.00      0.00        16\n",
      "  GO:0019866       0.00      0.00      0.00       319\n",
      "  GO:0019867       0.00      0.00      0.00       222\n",
      "  GO:0019897       0.00      0.00      0.00        45\n",
      "  GO:0019898       0.00      0.00      0.00        75\n",
      "  GO:0020002       0.00      0.00      0.00        32\n",
      "  GO:0020003       0.00      0.00      0.00        20\n",
      "  GO:0020007       0.00      0.00      0.00        14\n",
      "  GO:0020011       0.00      0.00      0.00        14\n",
      "  GO:0020015       0.00      0.00      0.00        34\n",
      "  GO:0020016       0.00      0.00      0.00        20\n",
      "  GO:0020023       0.00      0.00      0.00        46\n",
      "  GO:0020036       0.00      0.00      0.00        26\n",
      "  GO:0022624       0.00      0.00      0.00        29\n",
      "  GO:0022625       0.00      0.00      0.00       100\n",
      "  GO:0022626       0.00      0.00      0.00       189\n",
      "  GO:0022627       0.00      0.00      0.00        65\n",
      "  GO:0030014       0.00      0.00      0.00        12\n",
      "  GO:0030016       0.00      0.00      0.00       137\n",
      "  GO:0030017       0.00      0.00      0.00       117\n",
      "  GO:0030018       0.00      0.00      0.00        54\n",
      "  GO:0030027       0.00      0.00      0.00        45\n",
      "  GO:0030054       0.00      0.00      0.00      1149\n",
      "  GO:0030055       0.00      0.00      0.00       130\n",
      "  GO:0030117       0.00      0.00      0.00        38\n",
      "  GO:0030118       0.00      0.00      0.00        23\n",
      "  GO:0030119       0.00      0.00      0.00        15\n",
      "  GO:0030120       0.00      0.00      0.00        28\n",
      "  GO:0030125       0.00      0.00      0.00        14\n",
      "  GO:0030131       0.00      0.00      0.00        13\n",
      "  GO:0030133       0.00      0.00      0.00       131\n",
      "  GO:0030134       0.00      0.00      0.00        23\n",
      "  GO:0030135       0.00      0.00      0.00       102\n",
      "  GO:0030136       0.00      0.00      0.00        68\n",
      "  GO:0030137       0.00      0.00      0.00        13\n",
      "  GO:0030139       0.00      0.00      0.00       173\n",
      "  GO:0030140       0.00      0.00      0.00        14\n",
      "  GO:0030141       0.00      0.00      0.00       223\n",
      "  GO:0030175       0.00      0.00      0.00        46\n",
      "  GO:0030286       0.00      0.00      0.00        15\n",
      "  GO:0030288       0.00      0.00      0.00        43\n",
      "  GO:0030312       0.00      0.00      0.00       621\n",
      "  GO:0030313       0.00      0.00      0.00        69\n",
      "  GO:0030315       0.00      0.00      0.00        25\n",
      "  GO:0030424       0.00      0.00      0.00       307\n",
      "  GO:0030425       0.00      0.00      0.00       247\n",
      "  GO:0030426       0.00      0.00      0.00        67\n",
      "  GO:0030427       0.00      0.00      0.00       169\n",
      "  GO:0030428       0.00      0.00      0.00        26\n",
      "  GO:0030430       0.00      0.00      0.00        87\n",
      "  GO:0030445       0.00      0.00      0.00        10\n",
      "  GO:0030446       0.00      0.00      0.00         9\n",
      "  GO:0030479       0.00      0.00      0.00        19\n",
      "  GO:0030496       0.00      0.00      0.00        60\n",
      "  GO:0030532       0.00      0.00      0.00        42\n",
      "  GO:0030658       0.00      0.00      0.00        54\n",
      "  GO:0030659       0.00      0.00      0.00       282\n",
      "  GO:0030660       0.00      0.00      0.00        19\n",
      "  GO:0030662       0.00      0.00      0.00        48\n",
      "  GO:0030665       0.00      0.00      0.00        28\n",
      "  GO:0030666       0.00      0.00      0.00        41\n",
      "  GO:0030667       0.00      0.00      0.00        73\n",
      "  GO:0030669       0.00      0.00      0.00        18\n",
      "  GO:0030670       0.00      0.00      0.00         9\n",
      "  GO:0030672       0.00      0.00      0.00        32\n",
      "  GO:0030684       0.00      0.00      0.00        46\n",
      "  GO:0030686       0.00      0.00      0.00        14\n",
      "  GO:0030687       0.00      0.00      0.00         6\n",
      "  GO:0030863       0.00      0.00      0.00        77\n",
      "  GO:0030864       0.00      0.00      0.00        44\n",
      "  GO:0030867       0.00      0.00      0.00         9\n",
      "  GO:0030880       0.00      0.00      0.00        94\n",
      "  GO:0030894       0.00      0.00      0.00        26\n",
      "  GO:0030964       0.00      0.00      0.00        63\n",
      "  GO:0030981       0.00      0.00      0.00        18\n",
      "  GO:0030990       0.00      0.00      0.00        18\n",
      "  GO:0030992       0.00      0.00      0.00        11\n",
      "  GO:0031010       0.00      0.00      0.00         7\n",
      "  GO:0031011       0.00      0.00      0.00        14\n",
      "  GO:0031012       0.00      0.00      0.00       307\n",
      "  GO:0031045       0.00      0.00      0.00        12\n",
      "  GO:0031082       0.00      0.00      0.00         7\n",
      "  GO:0031090       0.10      0.00      0.00      1502\n",
      "  GO:0031091       0.00      0.00      0.00        17\n",
      "  GO:0031093       0.00      0.00      0.00        13\n",
      "  GO:0031097       0.00      0.00      0.00         5\n",
      "  GO:0031143       0.00      0.00      0.00        14\n",
      "  GO:0031160       0.00      0.00      0.00        12\n",
      "  GO:0031201       0.00      0.00      0.00        22\n",
      "  GO:0031234       0.00      0.00      0.00        22\n",
      "  GO:0031248       0.00      0.00      0.00        68\n",
      "  GO:0031252       0.00      0.00      0.00       120\n",
      "  GO:0031253       0.00      0.00      0.00        93\n",
      "  GO:0031256       0.00      0.00      0.00        41\n",
      "  GO:0031261       0.00      0.00      0.00        19\n",
      "  GO:0031312       0.00      0.00      0.00        16\n",
      "  GO:0031410       0.04      0.00      0.00      1000\n",
      "  GO:0031430       0.00      0.00      0.00        15\n",
      "  GO:0031461       0.00      0.00      0.00        82\n",
      "  GO:0031463       0.00      0.00      0.00        14\n",
      "  GO:0031514       0.00      0.00      0.00        82\n",
      "  GO:0031519       0.00      0.00      0.00        26\n",
      "  GO:0031526       0.00      0.00      0.00        21\n",
      "  GO:0031594       0.00      0.00      0.00        80\n",
      "  GO:0031672       0.00      0.00      0.00        26\n",
      "  GO:0031674       0.00      0.00      0.00        63\n",
      "  GO:0031901       0.00      0.00      0.00        20\n",
      "  GO:0031902       0.00      0.00      0.00        25\n",
      "  GO:0031903       0.00      0.00      0.00        34\n",
      "  GO:0031907       0.00      0.00      0.00        20\n",
      "  GO:0031941       0.00      0.00      0.00        10\n",
      "  GO:0031965       0.00      0.00      0.00       108\n",
      "  GO:0031966       0.00      0.00      0.00       418\n",
      "  GO:0031967       0.00      0.00      0.00       878\n",
      "  GO:0031968       0.00      0.00      0.00       195\n",
      "  GO:0031969       0.00      0.00      0.00        18\n",
      "  GO:0031970       0.00      0.00      0.00        34\n",
      "  GO:0031974       0.14      0.00      0.01      3082\n",
      "  GO:0031975       0.00      0.00      0.00       947\n",
      "  GO:0031976       0.00      0.00      0.00        95\n",
      "  GO:0031977       0.00      0.00      0.00        15\n",
      "  GO:0031981       0.09      0.00      0.00      2657\n",
      "  GO:0031982       0.06      0.00      0.00      1356\n",
      "  GO:0031983       0.00      0.00      0.00        64\n",
      "  GO:0031984       0.00      0.00      0.00       490\n",
      "  GO:0031985       0.00      0.00      0.00        52\n",
      "  GO:0032010       0.00      0.00      0.00         8\n",
      "  GO:0032040       0.00      0.00      0.00        37\n",
      "  GO:0032153       0.00      0.00      0.00       122\n",
      "  GO:0032154       0.00      0.00      0.00        41\n",
      "  GO:0032156       0.00      0.00      0.00        16\n",
      "  GO:0032279       0.00      0.00      0.00       104\n",
      "  GO:0032391       0.00      0.00      0.00        12\n",
      "  GO:0032420       0.00      0.00      0.00        13\n",
      "  GO:0032421       0.00      0.00      0.00        14\n",
      "  GO:0032432       0.00      0.00      0.00        27\n",
      "  GO:0032541       0.00      0.00      0.00        15\n",
      "  GO:0032587       0.00      0.00      0.00        12\n",
      "  GO:0032588       0.00      0.00      0.00        29\n",
      "  GO:0032589       0.00      0.00      0.00        23\n",
      "  GO:0032590       0.00      0.00      0.00        14\n",
      "  GO:0032806       0.00      0.00      0.00        12\n",
      "  GO:0032809       0.00      0.00      0.00        13\n",
      "  GO:0032838       0.00      0.00      0.00       240\n",
      "  GO:0032991       0.15      0.00      0.00      3276\n",
      "  GO:0032993       0.00      0.00      0.00       499\n",
      "  GO:0033176       0.00      0.00      0.00        18\n",
      "  GO:0033178       0.00      0.00      0.00        16\n",
      "  GO:0033643       0.00      0.00      0.00       165\n",
      "  GO:0033644       0.00      0.00      0.00        43\n",
      "  GO:0033646       0.00      0.00      0.00       112\n",
      "  GO:0033647       0.00      0.00      0.00        47\n",
      "  GO:0033648       0.00      0.00      0.00        47\n",
      "  GO:0033655       0.00      0.00      0.00        73\n",
      "  GO:0034357       0.00      0.00      0.00        95\n",
      "  GO:0034399       0.00      0.00      0.00        74\n",
      "  GO:0034451       0.00      0.00      0.00        39\n",
      "  GO:0034518       0.00      0.00      0.00        13\n",
      "  GO:0034702       0.00      0.00      0.00       112\n",
      "  GO:0034703       0.00      0.00      0.00        72\n",
      "  GO:0034704       0.00      0.00      0.00        19\n",
      "  GO:0034705       0.00      0.00      0.00        37\n",
      "  GO:0034708       0.00      0.00      0.00        47\n",
      "  GO:0034774       0.00      0.00      0.00        64\n",
      "  GO:0035097       0.00      0.00      0.00        37\n",
      "  GO:0035267       0.00      0.00      0.00         8\n",
      "  GO:0035371       0.00      0.00      0.00        16\n",
      "  GO:0035577       0.00      0.00      0.00        13\n",
      "  GO:0035578       0.00      0.00      0.00        17\n",
      "  GO:0035579       0.00      0.00      0.00        19\n",
      "  GO:0035580       0.00      0.00      0.00        13\n",
      "  GO:0035770       0.00      0.00      0.00       146\n",
      "  GO:0035838       0.00      0.00      0.00        19\n",
      "  GO:0035861       0.00      0.00      0.00        34\n",
      "  GO:0035869       0.00      0.00      0.00        37\n",
      "  GO:0036064       0.00      0.00      0.00        86\n",
      "  GO:0036126       0.00      0.00      0.00        47\n",
      "  GO:0036379       0.00      0.00      0.00        27\n",
      "  GO:0036452       0.00      0.00      0.00        11\n",
      "  GO:0036464       0.00      0.00      0.00       138\n",
      "  GO:0036477       0.00      0.00      0.00       387\n",
      "  GO:0042025       0.00      0.00      0.00        35\n",
      "  GO:0042170       0.00      0.00      0.00        93\n",
      "  GO:0042175       0.00      0.00      0.00       333\n",
      "  GO:0042383       0.00      0.00      0.00        48\n",
      "  GO:0042470       0.00      0.00      0.00         9\n",
      "  GO:0042555       0.00      0.00      0.00        10\n",
      "  GO:0042575       0.00      0.00      0.00        19\n",
      "  GO:0042579       0.00      0.00      0.00       199\n",
      "  GO:0042581       0.00      0.00      0.00        33\n",
      "  GO:0042582       0.00      0.00      0.00        31\n",
      "  GO:0042597       0.00      0.00      0.00        50\n",
      "  GO:0042641       0.00      0.00      0.00        22\n",
      "  GO:0042644       0.00      0.00      0.00        12\n",
      "  GO:0042645       0.00      0.00      0.00        27\n",
      "  GO:0042646       0.00      0.00      0.00        15\n",
      "  GO:0042651       0.00      0.00      0.00        87\n",
      "  GO:0042734       0.00      0.00      0.00        77\n",
      "  GO:0042763       0.00      0.00      0.00        24\n",
      "  GO:0042764       0.00      0.00      0.00        23\n",
      "  GO:0042788       0.00      0.00      0.00        33\n",
      "  GO:0042995       0.00      0.00      0.00      1236\n",
      "  GO:0043005       0.00      0.00      0.00       557\n",
      "  GO:0043025       0.00      0.00      0.00       263\n",
      "  GO:0043073       0.00      0.00      0.00        30\n",
      "  GO:0043186       0.00      0.00      0.00        20\n",
      "  GO:0043189       0.00      0.00      0.00         8\n",
      "  GO:0043190       0.00      0.00      0.00        15\n",
      "  GO:0043194       0.00      0.00      0.00        13\n",
      "  GO:0043195       0.00      0.00      0.00        48\n",
      "  GO:0043197       0.00      0.00      0.00        65\n",
      "  GO:0043198       0.00      0.00      0.00        11\n",
      "  GO:0043202       0.00      0.00      0.00        16\n",
      "  GO:0043204       0.00      0.00      0.00        48\n",
      "  GO:0043209       0.00      0.00      0.00        48\n",
      "  GO:0043226       0.68      0.94      0.79     11588\n",
      "  GO:0043227       0.62      0.88      0.73     10584\n",
      "  GO:0043228       0.22      0.00      0.01      3062\n",
      "  GO:0043229       0.66      0.90      0.76     11139\n",
      "  GO:0043230       0.00      0.00      0.00       436\n",
      "  GO:0043231       0.59      0.76      0.66     10012\n",
      "  GO:0043232       0.23      0.00      0.01      3058\n",
      "  GO:0043233       0.14      0.00      0.01      3082\n",
      "  GO:0043235       0.00      0.00      0.00       118\n",
      "  GO:0043292       0.00      0.00      0.00       160\n",
      "  GO:0043296       0.00      0.00      0.00        52\n",
      "  GO:0043332       0.00      0.00      0.00        39\n",
      "  GO:0043596       0.00      0.00      0.00        30\n",
      "  GO:0043601       0.00      0.00      0.00        18\n",
      "  GO:0043656       0.00      0.00      0.00       113\n",
      "  GO:0043657       0.00      0.00      0.00       166\n",
      "  GO:0043679       0.00      0.00      0.00        78\n",
      "  GO:0044232       0.00      0.00      0.00        27\n",
      "  GO:0044291       0.00      0.00      0.00        19\n",
      "  GO:0044295       0.00      0.00      0.00        12\n",
      "  GO:0044297       0.00      0.00      0.00       291\n",
      "  GO:0044298       0.00      0.00      0.00        14\n",
      "  GO:0044304       0.00      0.00      0.00        36\n",
      "  GO:0044306       0.00      0.00      0.00        78\n",
      "  GO:0044309       0.00      0.00      0.00        67\n",
      "  GO:0044391       0.00      0.00      0.00       260\n",
      "  GO:0044423       0.00      0.00      0.00        34\n",
      "  GO:0044732       0.00      0.00      0.00        28\n",
      "  GO:0044853       0.00      0.00      0.00        45\n",
      "  GO:0045111       0.00      0.00      0.00        41\n",
      "  GO:0045121       0.00      0.00      0.00       104\n",
      "  GO:0045169       0.00      0.00      0.00        13\n",
      "  GO:0045171       0.00      0.00      0.00        46\n",
      "  GO:0045172       0.00      0.00      0.00         9\n",
      "  GO:0045177       0.00      0.00      0.00       258\n",
      "  GO:0045178       0.00      0.00      0.00       145\n",
      "  GO:0045179       0.00      0.00      0.00        21\n",
      "  GO:0045202       0.00      0.00      0.00       601\n",
      "  GO:0045211       0.00      0.00      0.00        87\n",
      "  GO:0045239       0.00      0.00      0.00        13\n",
      "  GO:0045259       0.00      0.00      0.00        35\n",
      "  GO:0045271       0.00      0.00      0.00        58\n",
      "  GO:0045273       0.00      0.00      0.00        13\n",
      "  GO:0045275       0.00      0.00      0.00        14\n",
      "  GO:0045277       0.00      0.00      0.00         7\n",
      "  GO:0045293       0.00      0.00      0.00        13\n",
      "  GO:0045334       0.00      0.00      0.00        22\n",
      "  GO:0045335       0.00      0.00      0.00       102\n",
      "  GO:0045495       0.00      0.00      0.00        27\n",
      "  GO:0046540       0.00      0.00      0.00        12\n",
      "  GO:0048046       0.00      0.00      0.00        48\n",
      "  GO:0048471       0.00      0.00      0.00       246\n",
      "  GO:0048475       0.00      0.00      0.00        38\n",
      "  GO:0048770       0.00      0.00      0.00         9\n",
      "  GO:0048786       0.00      0.00      0.00        40\n",
      "  GO:0048787       0.00      0.00      0.00        13\n",
      "  GO:0051233       0.00      0.00      0.00        30\n",
      "  GO:0051285       0.00      0.00      0.00        14\n",
      "  GO:0051286       0.00      0.00      0.00       113\n",
      "  GO:0055028       0.00      0.00      0.00        16\n",
      "  GO:0055029       0.00      0.00      0.00        79\n",
      "  GO:0055035       0.00      0.00      0.00        77\n",
      "  GO:0055037       0.00      0.00      0.00        48\n",
      "  GO:0055038       0.00      0.00      0.00        12\n",
      "  GO:0055044       0.00      0.00      0.00       188\n",
      "  GO:0055120       0.00      0.00      0.00        30\n",
      "  GO:0060076       0.00      0.00      0.00        24\n",
      "  GO:0060170       0.00      0.00      0.00        21\n",
      "  GO:0060187       0.00      0.00      0.00       117\n",
      "  GO:0060205       0.00      0.00      0.00        64\n",
      "  GO:0060293       0.00      0.00      0.00        22\n",
      "  GO:0061174       0.00      0.00      0.00         4\n",
      "  GO:0061645       0.00      0.00      0.00        21\n",
      "  GO:0061695       0.00      0.00      0.00       188\n",
      "  GO:0062023       0.00      0.00      0.00       156\n",
      "  GO:0062039       0.00      0.00      0.00        74\n",
      "  GO:0062040       0.00      0.00      0.00        74\n",
      "  GO:0065010       0.00      0.00      0.00       435\n",
      "  GO:0070013       0.14      0.00      0.01      3082\n",
      "  GO:0070062       0.00      0.00      0.00       393\n",
      "  GO:0070069       0.00      0.00      0.00        26\n",
      "  GO:0070160       0.00      0.00      0.00        46\n",
      "  GO:0070161       0.00      0.00      0.00       515\n",
      "  GO:0070382       0.00      0.00      0.00        86\n",
      "  GO:0070461       0.00      0.00      0.00        41\n",
      "  GO:0070469       0.00      0.00      0.00        97\n",
      "  GO:0070603       0.00      0.00      0.00        68\n",
      "  GO:0070820       0.00      0.00      0.00        34\n",
      "  GO:0070821       0.00      0.00      0.00        13\n",
      "  GO:0070822       0.00      0.00      0.00        12\n",
      "  GO:0070847       0.00      0.00      0.00        13\n",
      "  GO:0070938       0.00      0.00      0.00        19\n",
      "  GO:0070971       0.00      0.00      0.00        11\n",
      "  GO:0071011       0.00      0.00      0.00        55\n",
      "  GO:0071013       0.00      0.00      0.00        49\n",
      "  GO:0071014       0.00      0.00      0.00         8\n",
      "  GO:0071162       0.00      0.00      0.00        12\n",
      "  GO:0071782       0.00      0.00      0.00        16\n",
      "  GO:0071944       0.38      0.01      0.01      3847\n",
      "  GO:0072562       0.00      0.00      0.00        30\n",
      "  GO:0072686       0.00      0.00      0.00        90\n",
      "  GO:0072687       0.00      0.00      0.00        14\n",
      "  GO:0080008       0.00      0.00      0.00        14\n",
      "  GO:0090406       0.00      0.00      0.00        20\n",
      "  GO:0090545       0.00      0.00      0.00        12\n",
      "  GO:0090571       0.00      0.00      0.00        13\n",
      "  GO:0090575       0.00      0.00      0.00       108\n",
      "  GO:0090665       0.00      0.00      0.00         7\n",
      "  GO:0090734       0.00      0.00      0.00        40\n",
      "  GO:0097014       0.00      0.00      0.00       228\n",
      "  GO:0097038       0.00      0.00      0.00        22\n",
      "  GO:0097060       0.00      0.00      0.00       150\n",
      "  GO:0097225       0.00      0.00      0.00        12\n",
      "  GO:0097346       0.00      0.00      0.00        19\n",
      "  GO:0097386       0.00      0.00      0.00         6\n",
      "  GO:0097447       0.00      0.00      0.00       247\n",
      "  GO:0097517       0.00      0.00      0.00        18\n",
      "  GO:0097525       0.00      0.00      0.00        37\n",
      "  GO:0097526       0.00      0.00      0.00        12\n",
      "  GO:0097542       0.00      0.00      0.00        16\n",
      "  GO:0097546       0.00      0.00      0.00        17\n",
      "  GO:0097708       0.04      0.00      0.00      1001\n",
      "  GO:0097729       0.00      0.00      0.00        65\n",
      "  GO:0097730       0.00      0.00      0.00        68\n",
      "  GO:0097731       0.00      0.00      0.00        41\n",
      "  GO:0097733       0.00      0.00      0.00        39\n",
      "  GO:0098533       0.00      0.00      0.00        23\n",
      "  GO:0098552       0.00      0.00      0.00       207\n",
      "  GO:0098554       0.00      0.00      0.00        11\n",
      "  GO:0098562       0.00      0.00      0.00        92\n",
      "  GO:0098576       0.00      0.00      0.00         8\n",
      "  GO:0098588       0.00      0.00      0.00       821\n",
      "  GO:0098590       0.00      0.00      0.00       571\n",
      "  GO:0098636       0.00      0.00      0.00        16\n",
      "  GO:0098644       0.00      0.00      0.00        13\n",
      "  GO:0098685       0.00      0.00      0.00        23\n",
      "  GO:0098686       0.00      0.00      0.00         5\n",
      "  GO:0098687       0.00      0.00      0.00       241\n",
      "  GO:0098688       0.00      0.00      0.00         4\n",
      "  GO:0098791       0.00      0.00      0.00       169\n",
      "  GO:0098793       0.00      0.00      0.00       244\n",
      "  GO:0098794       0.00      0.00      0.00       238\n",
      "  GO:0098796       0.00      0.00      0.00       612\n",
      "  GO:0098797       0.00      0.00      0.00       246\n",
      "  GO:0098798       0.00      0.00      0.00       269\n",
      "  GO:0098799       0.00      0.00      0.00        17\n",
      "  GO:0098800       0.00      0.00      0.00       134\n",
      "  GO:0098802       0.00      0.00      0.00        78\n",
      "  GO:0098803       0.00      0.00      0.00        92\n",
      "  GO:0098827       0.00      0.00      0.00       331\n",
      "  GO:0098839       0.00      0.00      0.00        30\n",
      "  GO:0098852       0.00      0.00      0.00       189\n",
      "  GO:0098857       0.00      0.00      0.00       105\n",
      "  GO:0098858       0.00      0.00      0.00        84\n",
      "  GO:0098862       0.00      0.00      0.00        52\n",
      "  GO:0098878       0.00      0.00      0.00        25\n",
      "  GO:0098978       0.00      0.00      0.00       183\n",
      "  GO:0098982       0.00      0.00      0.00        32\n",
      "  GO:0098984       0.00      0.00      0.00       108\n",
      "  GO:0099023       0.00      0.00      0.00        43\n",
      "  GO:0099080       0.00      0.00      0.00       641\n",
      "  GO:0099081       0.00      0.00      0.00       390\n",
      "  GO:0099086       0.00      0.00      0.00        18\n",
      "  GO:0099091       0.00      0.00      0.00        13\n",
      "  GO:0099092       0.00      0.00      0.00        10\n",
      "  GO:0099501       0.00      0.00      0.00        32\n",
      "  GO:0099503       0.00      0.00      0.00       335\n",
      "  GO:0099512       0.00      0.00      0.00       382\n",
      "  GO:0099513       0.00      0.00      0.00       214\n",
      "  GO:0099522       0.00      0.00      0.00        11\n",
      "  GO:0099568       0.00      0.00      0.00       296\n",
      "  GO:0099572       0.00      0.00      0.00       116\n",
      "  GO:0099634       0.00      0.00      0.00        38\n",
      "  GO:0099738       0.00      0.00      0.00        56\n",
      "  GO:0101002       0.00      0.00      0.00        35\n",
      "  GO:0101003       0.00      0.00      0.00        12\n",
      "  GO:0101031       0.00      0.00      0.00        17\n",
      "  GO:0110085       0.00      0.00      0.00        11\n",
      "  GO:0110165       0.99      1.00      1.00     16764\n",
      "  GO:0120025       0.00      0.00      0.00      1222\n",
      "  GO:0120111       0.00      0.00      0.00        12\n",
      "  GO:0120114       0.00      0.00      0.00        49\n",
      "  GO:0120119       0.00      0.00      0.00        11\n",
      "  GO:0140220       0.00      0.00      0.00        14\n",
      "  GO:0140445       0.00      0.00      0.00        14\n",
      "  GO:0140513       0.00      0.00      0.00       774\n",
      "  GO:0140534       0.00      0.00      0.00        61\n",
      "  GO:0140535       0.00      0.00      0.00       594\n",
      "  GO:0150034       0.00      0.00      0.00       142\n",
      "  GO:1902493       0.00      0.00      0.00        70\n",
      "  GO:1902494       0.00      0.00      0.00      1119\n",
      "  GO:1902495       0.00      0.00      0.00       215\n",
      "  GO:1902554       0.00      0.00      0.00        62\n",
      "  GO:1902555       0.00      0.00      0.00        15\n",
      "  GO:1902562       0.00      0.00      0.00        18\n",
      "  GO:1902911       0.00      0.00      0.00        70\n",
      "  GO:1903293       0.00      0.00      0.00        36\n",
      "  GO:1903561       0.00      0.00      0.00       410\n",
      "  GO:1904724       0.00      0.00      0.00        11\n",
      "  GO:1904813       0.00      0.00      0.00        23\n",
      "  GO:1904949       0.00      0.00      0.00       125\n",
      "  GO:1905348       0.00      0.00      0.00        19\n",
      "  GO:1905354       0.00      0.00      0.00        20\n",
      "  GO:1905360       0.00      0.00      0.00        12\n",
      "  GO:1905368       0.00      0.00      0.00       117\n",
      "  GO:1905369       0.00      0.00      0.00        87\n",
      "  GO:1990023       0.00      0.00      0.00        16\n",
      "  GO:1990204       0.00      0.00      0.00       135\n",
      "  GO:1990234       0.00      0.00      0.00       488\n",
      "  GO:1990351       0.00      0.00      0.00       231\n",
      "  GO:1990752       0.00      0.00      0.00        20\n",
      "  GO:1990904       0.00      0.00      0.00       529\n",
      "\n",
      "   micro avg       0.75      0.43      0.54    222146\n",
      "   macro avg       0.01      0.01      0.01    222146\n",
      "weighted avg       0.40      0.43      0.39    222146\n",
      " samples avg       0.75      0.51      0.56    222146\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Define the Neural Network\n",
    "class MultilabelNN(nn.Module):\n",
    "    def __init__(self, input_size, hidden_sizes, output_size):\n",
    "        super(MultilabelNN, self).__init__()\n",
    "        self.model = nn.Sequential(\n",
    "            nn.Linear(input_size, hidden_sizes[0]),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(0.3),  # Dropout for regularization\n",
    "            nn.Linear(hidden_sizes[0], hidden_sizes[1]),\n",
    "            nn.ReLU(),\n",
    "            nn.Dropout(0.3),\n",
    "            nn.Linear(hidden_sizes[1], output_size),\n",
    "            nn.Sigmoid()  # For multilabel classification\n",
    "        )\n",
    "    \n",
    "    def forward(self, x):\n",
    "        return self.model(x)\n",
    "\n",
    "# Define model parameters\n",
    "input_size = X_train_tensor.shape[1]  # 300 features\n",
    "hidden_sizes = [4096, 2048, 1024]            # Hidden layer sizes\n",
    "output_size = y_train.shape[1]    # Number of labels\n",
    "\n",
    "# Initialize the model\n",
    "model = MultilabelNN(input_size, hidden_sizes, output_size).to(device)\n",
    "\n",
    "# Loss function and optimizer\n",
    "\n",
    "criterion = nn.BCELoss()  # Binary Cross-Entropy Loss for multilabel\n",
    "optimizer = optim.Adam(model.parameters(), lr=0.005)\n",
    "\n",
    "# Training loop\n",
    "epochs = 100\n",
    "batch_size = 512\n",
    "\n",
    "def train_model(model, X_train, y_train, optimizer, criterion, batch_size, epochs):\n",
    "    # Determine the device (model should already be on this device)\n",
    "    device = next(model.parameters()).device\n",
    "\n",
    "    # Move training data to the device\n",
    "    X_train = X_train.to(device)\n",
    "    y_train = y_train.to(device)\n",
    "\n",
    "    model.train()\n",
    "    for epoch in range(epochs):\n",
    "        for i in range(0, len(X_train), batch_size):\n",
    "            # Get the current batch\n",
    "            X_batch = X_train[i:i+batch_size]\n",
    "            y_batch = y_train[i:i+batch_size]\n",
    "\n",
    "            # Forward pass\n",
    "            outputs = model(X_batch)\n",
    "            loss = criterion(outputs, y_batch)\n",
    "\n",
    "            # Backward pass and optimization\n",
    "            optimizer.zero_grad()\n",
    "            loss.backward()\n",
    "            optimizer.step()\n",
    "\n",
    "        print(f\"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f}\")\n",
    "\n",
    "\n",
    "# Train the model\n",
    "train_model(model, X_train_tensor, y_train_tensor, optimizer, criterion, batch_size, epochs)\n",
    "\n",
    "# Ensure X_test_tensor is on the same device as the model\n",
    "device = next(model.parameters()).device\n",
    "X_test_tensor = X_test_tensor.to(device)\n",
    "\n",
    "# Evaluation\n",
    "model.eval()\n",
    "with torch.no_grad():\n",
    "    y_pred = model(X_test_tensor)\n",
    "\n",
    "    # Move predictions back to the CPU for further processing (e.g., NumPy conversion)\n",
    "    y_pred_binary = (y_pred.cpu().numpy() > 0.5).astype(int)\n",
    "\n",
    "\n",
    "# Classification report\n",
    "target_names = [str(cls) for cls in mlb.classes_]\n",
    "print(\"\\nNeural Network Results:\")\n",
    "print(classification_report(y_test, y_pred_binary, target_names=target_names))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
