{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\Bio\\pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio import Align\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import substitution_matrices\n",
    "from Bio.Data import IUPACData\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "'''import cafaeval\n",
    "from cafaeval.evaluation import cafa_eval\n",
    "from cafaeval.parser import obo_parser, gt_parser'''\n",
    "\n",
    "from pathlib import Path\n",
    "import os\n",
    "import ast\n",
    "import h5py\n",
    "\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating the path to the main folder we are going to use to create our datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to train data\n",
    "training_data_path = Path('../data/train')\n",
    "\n",
    "# Path to test data\n",
    "test_data_path = Path('../data/test')\n",
    "\n",
    "# Path to baseline data\n",
    "baseline_data_path = Path('../data/baseline')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training Set\n",
    "\n",
    "In this section we are **creating the Training Set**, fusing togheter the data from:\n",
    "- _train_set.tsv_\n",
    "- _train_ids.txt_\n",
    "- _train.fasta_\n",
    "- _train_embeddings.h5_\n",
    "- _train_protein2ipr.dat_\n",
    "- _go-basic.obo_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_set.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sub_ontology</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0110165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0005622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P91124</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0043226</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID        sub_ontology     GO_term\n",
       "0  P91124  cellular_component  GO:0005575\n",
       "1  P91124  cellular_component  GO:0110165\n",
       "2  P91124  cellular_component  GO:0005737\n",
       "3  P91124  cellular_component  GO:0005622\n",
       "4  P91124  cellular_component  GO:0043226"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_set = pd.read_csv(training_data_path / 'train_set.tsv', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "train_set.rename(columns={'Protein_ID': 'ID', 'aspect' : 'sub_ontology'}, inplace=True)\n",
    "\n",
    "# Display the first few rows of the train set\n",
    "train_set.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_ids.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['P91124', 'Q55DL5', 'O81027', 'Q04418', 'Q8IXT2']\n"
     ]
    }
   ],
   "source": [
    "# Extracting train_ids.txt\n",
    "with open(training_data_path / 'train_ids.txt', 'r') as file:\n",
    "    train_ids = file.read().splitlines()\n",
    "\n",
    "# Display the first few IDs to verify\n",
    "print(train_ids[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set (123969).\n",
      "Proceeding with the analysis.\n"
     ]
    }
   ],
   "source": [
    "if len(train_ids) == len(train_set['ID'].unique()):\n",
    "    print(f\"The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set ({len(train_ids)}).\\n\"\n",
    "          \"Proceeding with the analysis.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train.fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: P91124\n",
      "Name: P91124\n",
      "Description: P91124\n",
      "Number of features: 0\n",
      "Seq('MAAQFYERNTSGMNADRFMTRLTDESTVNTMQRHYWTARQFIRTKLGKKEDEHL...NLL')\n"
     ]
    }
   ],
   "source": [
    "train_fasta_list = list(SeqIO.parse(training_data_path / 'train.fasta', 'fasta'))\n",
    "\n",
    "# Print the first sequence to verify\n",
    "print(train_fasta_list[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's transform it into a DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>num_features</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>P91124</td>\n",
       "      <td>P91124</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O81027</td>\n",
       "      <td>O81027</td>\n",
       "      <td>O81027</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q04418</td>\n",
       "      <td>Q04418</td>\n",
       "      <td>Q04418</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID    name description  num_features  \\\n",
       "0  P91124  P91124      P91124             0   \n",
       "1  Q55DL5  Q55DL5      Q55DL5             0   \n",
       "2  O81027  O81027      O81027             0   \n",
       "3  Q04418  Q04418      Q04418             0   \n",
       "4  Q8IXT2  Q8IXT2      Q8IXT2             0   \n",
       "\n",
       "                                            sequence  \n",
       "0  (M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...  \n",
       "1  (M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...  \n",
       "2  (M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...  \n",
       "3  (M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...  \n",
       "4  (M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract relevant information from SeqRecord\n",
    "train_fasta_dict = [{\n",
    "    'ID': record.id,\n",
    "    'name': record.name,\n",
    "    'description': record.description,\n",
    "    'num_features': len(record.features),\n",
    "    'sequence': record.seq,\n",
    "} for record in train_fasta_list]\n",
    "\n",
    "# Create a DataFrame from the extracted data\n",
    "train_fasta = pd.DataFrame(train_fasta_dict)\n",
    "\n",
    "# Display the DataFrame\n",
    "train_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking if `ID`, `name` and `description` have the same information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 differences between the ID and name columns.\n",
      "We have a total of 0 differences between the ID and description columns.\n"
     ]
    }
   ],
   "source": [
    "# Checking for differences between the ID and name columns\n",
    "diff_id_name = sum(train_fasta['ID'] != train_fasta['name'])\n",
    "\n",
    "# Checking for differences between the ID and description columns\n",
    "diff_id_description = sum(train_fasta['ID'] != train_fasta['description'])\n",
    "\n",
    "print(f\"We have a total of {diff_id_name} differences between the ID and name columns.\\nWe have a total of {diff_id_description} differences between the ID and description columns.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking if `num_features` has value different from 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 sequences with features.\n"
     ]
    }
   ],
   "source": [
    "num_features_values = sum(train_fasta['num_features'] != 0)\n",
    "\n",
    "print(f\"We have a total of {num_features_values} sequences with features.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Nice!**\n",
    "\n",
    "We found that `ID`, `name` and `description` columns are the same. We can remove two of them (`name`, `descritpion`). \n",
    "\n",
    "In the same way we saw that `num_features` is not very informative since it has only 0, let's remove it as well\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P91124</td>\n",
       "      <td>(M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q55DL5</td>\n",
       "      <td>(M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O81027</td>\n",
       "      <td>(M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q04418</td>\n",
       "      <td>(M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q8IXT2</td>\n",
       "      <td>(M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ID                                           sequence\n",
       "0  P91124  (M, A, A, Q, F, Y, E, R, N, T, S, G, M, N, A, ...\n",
       "1  Q55DL5  (M, S, E, I, L, I, D, S, E, L, L, S, C, P, I, ...\n",
       "2  O81027  (M, Q, W, N, G, V, R, R, A, H, S, I, W, C, K, ...\n",
       "3  Q04418  (M, D, L, L, G, D, I, V, E, K, D, T, S, D, S, ...\n",
       "4  Q8IXT2  (M, E, P, S, D, M, P, A, G, Y, H, C, P, L, D, ..."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_fasta.drop(columns=['name', 'description', 'num_features'], inplace=True)\n",
    "\n",
    "\n",
    "train_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_embeddings.h5`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>[0.0682, -0.04648, 0.001752, -0.00858, 0.00376...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW4</td>\n",
       "      <td>[0.002447, 0.007053, 0.06445, 0.00727, -0.0335...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>[0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings\n",
       "0  A0A009IHW8  [0.0682, -0.04648, 0.001752, -0.00858, 0.00376...\n",
       "1  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...\n",
       "2  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...\n",
       "3  A0A023FBW4  [0.002447, 0.007053, 0.06445, 0.00727, -0.0335...\n",
       "4  A0A023FBW7  [0.04932, 0.02069, 0.10864, 0.01634, -0.05106,..."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filename = \"train_embeddings.h5\"\n",
    "\n",
    "data_list = []\n",
    "\n",
    "with h5py.File(training_data_path / filename, \"r\") as f:\n",
    "    for dataset_name in f.keys():\n",
    "        dataset = f[dataset_name][:]\n",
    "        data_list.append([dataset_name, dataset])\n",
    "\n",
    "train_embeddings = pd.DataFrame(data_list, columns=[\"ID\", \"embeddings\"])\n",
    "\n",
    "train_embeddings.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `train_protein2ipr.dat`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR000157</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) domain</td>\n",
       "      <td>PS50104</td>\n",
       "      <td>133</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR000157</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) domain</td>\n",
       "      <td>SM00255</td>\n",
       "      <td>134</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR035897</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) dom...</td>\n",
       "      <td>G3DSA:3.40.50.10140</td>\n",
       "      <td>80</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>IPR035897</td>\n",
       "      <td>Toll/interleukin-1 receptor homology (TIR) dom...</td>\n",
       "      <td>SSF52200</td>\n",
       "      <td>128</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>IPR006910</td>\n",
       "      <td>Rad21/Rec8-like protein, N-terminal</td>\n",
       "      <td>PF04825</td>\n",
       "      <td>1</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        ipr                                             domain  \\\n",
       "0  A0A009IHW8  IPR000157  Toll/interleukin-1 receptor homology (TIR) domain   \n",
       "1  A0A009IHW8  IPR000157  Toll/interleukin-1 receptor homology (TIR) domain   \n",
       "2  A0A009IHW8  IPR035897  Toll/interleukin-1 receptor homology (TIR) dom...   \n",
       "3  A0A009IHW8  IPR035897  Toll/interleukin-1 receptor homology (TIR) dom...   \n",
       "4  A0A021WW32  IPR006910                Rad21/Rec8-like protein, N-terminal   \n",
       "\n",
       "              familyID  start  end  \n",
       "0              PS50104    133  266  \n",
       "1              SM00255    134  258  \n",
       "2  G3DSA:3.40.50.10140     80  266  \n",
       "3             SSF52200    128  249  \n",
       "4              PF04825      1  103  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_protein2ipr = pd.read_csv(training_data_path / 'train_protein2ipr.dat', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "train_protein2ipr.columns = ['ID', 'ipr', 'domain', 'familyID', 'start', 'end']\n",
    "\n",
    "# Display the first few rows of the train set\n",
    "train_protein2ipr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train protein2ipr ((1103545, 6)):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>(IPR000157, IPR000157, IPR035897, IPR035897)</td>\n",
       "      <td>(Toll/interleukin-1 receptor homology (TIR) do...</td>\n",
       "      <td>(PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)</td>\n",
       "      <td>(133, 134, 80, 128)</td>\n",
       "      <td>(266, 258, 266, 249)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FDY8</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                                ipr  \\\n",
       "0  A0A009IHW8       (IPR000157, IPR000157, IPR035897, IPR035897)   \n",
       "1  A0A021WW32                  (IPR006910, IPR039781, IPR049589)   \n",
       "2  A0A021WZA4  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "3  A0A023FBW7                                       (IPR045797,)   \n",
       "4  A0A023FDY8                                       (IPR045797,)   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Toll/interleukin-1 receptor homology (TIR) do...   \n",
       "1  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "2  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "3                                 (Evasins Class A,)   \n",
       "4                                 (Evasins Class A,)   \n",
       "\n",
       "                                            familyID  \\\n",
       "0  (PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)   \n",
       "1                      (PF04825, PTHR12585, cd21792)   \n",
       "2  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "3                                         (PF19429,)   \n",
       "4                                         (PF19429,)   \n",
       "\n",
       "                       start                             end  \n",
       "0        (133, 134, 80, 128)            (266, 258, 266, 249)  \n",
       "1                (1, 1, 327)                 (103, 607, 404)  \n",
       "2  (6, 31, 33, 329, 64, 286)  (496, 476, 175, 480, 258, 482)  \n",
       "3                      (38,)                          (110,)  \n",
       "4                      (38,)                          (110,)  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by 'ID' and aggregate other columns into lists\n",
    "train_protein2ipr_grouped = train_protein2ipr.groupby('ID').agg(lambda x: tuple(x)).reset_index()\n",
    "\n",
    "print(f\"Train protein2ipr ({train_protein2ipr.shape}):\")\n",
    "train_protein2ipr_grouped.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Still have to understand what *tizio, caio, sempronio* are."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `go-basic.obo`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All IDs with namespaces:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0048308, GO:0048311]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0007005]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0019952</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0050876</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0008150]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                              name           namespace  \\\n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "1  GO:0000002  mitochondrial genome maintenance  biological_process   \n",
       "2  GO:0000003                      reproduction  biological_process   \n",
       "3  GO:0019952                      reproduction  biological_process   \n",
       "4  GO:0050876                      reproduction  biological_process   \n",
       "\n",
       "                       is_a part_of  \n",
       "0  [GO:0048308, GO:0048311]      []  \n",
       "1              [GO:0007005]      []  \n",
       "2              [GO:0008150]      []  \n",
       "3              [GO:0008150]      []  \n",
       "4              [GO:0008150]      []  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "file_path = \"../data/train/go-basic.obo\"  # Replace with your file path\n",
    "\n",
    "# Step 1: Initialize storage for GO terms\n",
    "go_terms = []\n",
    "\n",
    "# Step 2: Parse the .obo file\n",
    "with open(file_path, 'r') as file:\n",
    "    current_term = {}\n",
    "    for line in file:\n",
    "        line = line.strip()\n",
    "        \n",
    "        # Start of a new term\n",
    "        if line == \"[Term]\":\n",
    "            if current_term:  # Save the previous term\n",
    "                go_terms.append(current_term)\n",
    "            current_term = {}  # Start a new term\n",
    "            \n",
    "        elif line.startswith(\"id:\"):\n",
    "            current_term['ID'] = line.split(\"id: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"alt_id:\"):\n",
    "            alt_id = line.split(\"alt_id: \")[1]\n",
    "            current_term.setdefault('alt_ids', []).append(alt_id)\n",
    "            \n",
    "        elif line.startswith(\"name:\"):\n",
    "            current_term['name'] = line.split(\"name: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"namespace:\"):\n",
    "            current_term['namespace'] = line.split(\"namespace: \")[1]\n",
    "            \n",
    "        elif line.startswith(\"is_a\"):\n",
    "            match = re.search(r\"GO:\\d+\", line)  # Search for GO ID\n",
    "            if match:  # Check if a match was found\n",
    "                is_a_id = match.group()\n",
    "                current_term.setdefault('is_a', []).append(is_a_id)\n",
    "                \n",
    "        elif line.startswith(\"relationship: part_of\"):\n",
    "            match = re.search(r\"GO:\\d+\", line)  # Search for GO ID\n",
    "            if match:  # Check if a match was found\n",
    "                part_of_id = match.group()\n",
    "                current_term.setdefault('part_of', []).append(part_of_id)\n",
    "\n",
    "            \n",
    "    # Add the last term\n",
    "    if current_term:\n",
    "        go_terms.append(current_term)\n",
    "\n",
    "# Step 3: Create a unified list of all IDs (primary and alt_ids)\n",
    "expanded_terms = []\n",
    "for term in go_terms:\n",
    "    primary_id = term['ID']\n",
    "    alt_ids = term.get('alt_ids', [])\n",
    "    all_ids = [primary_id] + alt_ids\n",
    "    \n",
    "    for term_id in all_ids:\n",
    "        expanded_terms.append({\n",
    "            'ID': term_id,\n",
    "            'name': term.get('name'),\n",
    "            'namespace': term.get('namespace'),\n",
    "            'is_a': term.get('is_a', []),\n",
    "            'part_of': term.get('part_of', [])\n",
    "        })\n",
    "\n",
    "# Step 4: Convert to a DataFrame\n",
    "df = pd.DataFrame(expanded_terms)\n",
    "\n",
    "# Step 5: Filter by namespace and explode relationships\n",
    "df_is_a = df.explode('is_a').dropna(subset=['is_a'])\n",
    "df_part_of = df.explode('part_of').dropna(subset=['part_of'])\n",
    "\n",
    "\n",
    "print(\"All IDs with namespaces:\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'is_a' relationships:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0048308</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0000001</td>\n",
       "      <td>mitochondrion inheritance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0048311</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0000002</td>\n",
       "      <td>mitochondrial genome maintenance</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0007005</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0000003</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0019952</td>\n",
       "      <td>reproduction</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                              name           namespace  \\\n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "0  GO:0000001         mitochondrion inheritance  biological_process   \n",
       "1  GO:0000002  mitochondrial genome maintenance  biological_process   \n",
       "2  GO:0000003                      reproduction  biological_process   \n",
       "3  GO:0019952                      reproduction  biological_process   \n",
       "\n",
       "         is_a part_of  \n",
       "0  GO:0048308      []  \n",
       "0  GO:0048311      []  \n",
       "1  GO:0007005      []  \n",
       "2  GO:0008150      []  \n",
       "3  GO:0008150      []  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"'is_a' relationships:\")\n",
    "df_is_a.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'part_of' relationships:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>is_a</th>\n",
       "      <th>part_of</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GO:0000015</td>\n",
       "      <td>phosphopyruvate hydratase complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>[GO:1902494]</td>\n",
       "      <td>GO:0005829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GO:0000022</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GO:0000022</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0007052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GO:1905121</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0000070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GO:1905121</td>\n",
       "      <td>mitotic spindle elongation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>[GO:0051231, GO:1903047]</td>\n",
       "      <td>GO:0007052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID                               name           namespace  \\\n",
       "16  GO:0000015  phosphopyruvate hydratase complex  cellular_component   \n",
       "22  GO:0000022         mitotic spindle elongation  biological_process   \n",
       "22  GO:0000022         mitotic spindle elongation  biological_process   \n",
       "23  GO:1905121         mitotic spindle elongation  biological_process   \n",
       "23  GO:1905121         mitotic spindle elongation  biological_process   \n",
       "\n",
       "                        is_a     part_of  \n",
       "16              [GO:1902494]  GO:0005829  \n",
       "22  [GO:0051231, GO:1903047]  GO:0000070  \n",
       "22  [GO:0051231, GO:1903047]  GO:0007052  \n",
       "23  [GO:0051231, GO:1903047]  GO:0000070  \n",
       "23  [GO:0051231, GO:1903047]  GO:0007052  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"'part_of' relationships:\")\n",
    "df_part_of.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging all previous extractions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us collect the dataframes and check lengths. Noting that:\n",
    "- _train_set.tsv_ contains the proteins, their GO annotations and their corresponding aspects which will be used to subdivde it into three separate datasets\n",
    "- _train.fasta_ and _train_embeddings.h5_ can be grouped together since they refers to the whole input\n",
    "- _train_protein2ipr.dat_\n",
    "- _go-basic.obo_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: CC (84638, 2), MF (55698, 2), BP (83064, 2)\n",
      "Cellular Component (84638):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0000785, GO:003299...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0071944, GO:000588...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005622, GO:000582...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005576, GO:0005615)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>(GO:0005575, GO:0005737, GO:0042175, GO:003299...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                            GO_term\n",
       "0  A0A021WW32  (GO:0005575, GO:0110165, GO:0000785, GO:003299...\n",
       "1  A0A021WZA4  (GO:0005575, GO:0110165, GO:0071944, GO:000588...\n",
       "2  A0A023GPJ3  (GO:0005575, GO:0110165, GO:0005622, GO:000582...\n",
       "3  A0A023GUT0   (GO:0005575, GO:0110165, GO:0005576, GO:0005615)\n",
       "4  A0A023IM54  (GO:0005575, GO:0005737, GO:0042175, GO:003299..."
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def group_and_combine(df, sub_ontology_value):\n",
    "    return df[df['sub_ontology'] == sub_ontology_value].groupby('ID')['GO_term'].apply(tuple).reset_index()\n",
    "\n",
    "# Create three dataframes for each sub_ontology value\n",
    "df_CC = group_and_combine(train_set, 'cellular_component')\n",
    "df_MF = group_and_combine(train_set, 'molecular_function')\n",
    "df_BP = group_and_combine(train_set, 'biological_process')\n",
    "\n",
    "print(f\"Shapes: CC {df_CC.shape}, MF {df_MF.shape}, BP {df_BP.shape}\")\n",
    "\n",
    "print(f\"Cellular Component ({df_CC.shape[0]}):\")\n",
    "df_CC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows missing from train_protein2ipr_grouped: 6522\n",
      "Combined DataFrame shape: (123969, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A009IHW8</td>\n",
       "      <td>[0.0682, -0.04648, 0.001752, -0.00858, 0.00376...</td>\n",
       "      <td>(M, S, L, E, Q, K, K, G, A, D, I, I, S, K, I, ...</td>\n",
       "      <td>(IPR000157, IPR000157, IPR035897, IPR035897)</td>\n",
       "      <td>(Toll/interleukin-1 receptor homology (TIR) do...</td>\n",
       "      <td>(PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)</td>\n",
       "      <td>(133, 134, 80, 128)</td>\n",
       "      <td>(266, 258, 266, 249)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023FBW4</td>\n",
       "      <td>[0.002447, 0.007053, 0.06445, 0.00727, -0.0335...</td>\n",
       "      <td>(M, T, S, H, G, A, V, K, I, A, I, F, A, V, I, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023FBW7</td>\n",
       "      <td>[0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...</td>\n",
       "      <td>(M, K, V, L, L, Y, I, A, A, S, C, L, M, L, L, ...</td>\n",
       "      <td>(IPR045797,)</td>\n",
       "      <td>(Evasins Class A,)</td>\n",
       "      <td>(PF19429,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(110,)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A009IHW8  [0.0682, -0.04648, 0.001752, -0.00858, 0.00376...   \n",
       "1  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...   \n",
       "2  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...   \n",
       "3  A0A023FBW4  [0.002447, 0.007053, 0.06445, 0.00727, -0.0335...   \n",
       "4  A0A023FBW7  [0.04932, 0.02069, 0.10864, 0.01634, -0.05106,...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, S, L, E, Q, K, K, G, A, D, I, I, S, K, I, ...   \n",
       "1  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "2  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "3  (M, T, S, H, G, A, V, K, I, A, I, F, A, V, I, ...   \n",
       "4  (M, K, V, L, L, Y, I, A, A, S, C, L, M, L, L, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0       (IPR000157, IPR000157, IPR035897, IPR035897)   \n",
       "1                  (IPR006910, IPR039781, IPR049589)   \n",
       "2  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "3                                                NaN   \n",
       "4                                       (IPR045797,)   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Toll/interleukin-1 receptor homology (TIR) do...   \n",
       "1  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "2  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "3                                                NaN   \n",
       "4                                 (Evasins Class A,)   \n",
       "\n",
       "                                            familyID  \\\n",
       "0  (PS50104, SM00255, G3DSA:3.40.50.10140, SSF52200)   \n",
       "1                      (PF04825, PTHR12585, cd21792)   \n",
       "2  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "3                                                NaN   \n",
       "4                                         (PF19429,)   \n",
       "\n",
       "                       start                             end  \n",
       "0        (133, 134, 80, 128)            (266, 258, 266, 249)  \n",
       "1                (1, 1, 327)                 (103, 607, 404)  \n",
       "2  (6, 31, 33, 329, 64, 286)  (496, 476, 175, 480, 258, 482)  \n",
       "3                        NaN                             NaN  \n",
       "4                      (38,)                          (110,)  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_train = pd.merge(train_embeddings, train_fasta, on='ID')\n",
    "combined_train = pd.merge(combined_train, train_protein2ipr_grouped, on='ID', how='left')\n",
    "\n",
    "missing_rows = combined_train[combined_train['ipr'].isna()].shape[0]\n",
    "print(f\"Number of rows missing from train_protein2ipr_grouped: {missing_rows}\")\n",
    "\n",
    "# combined_train = combined_train.drop('domain', axis=1)\n",
    "\n",
    "print(f\"Combined DataFrame shape: {combined_train.shape}\")\n",
    "combined_train.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: CC (84638, 9), MF (55698, 9), BP (83064, 9)\n",
      "Full df CC (84638, 9):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GO_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0000785, GO:003299...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0071944, GO:000588...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>[0.01512, 0.01102, 0.0217, -0.02512, 0.0396, 0...</td>\n",
       "      <td>(M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...</td>\n",
       "      <td>(IPR013087, IPR013087, IPR013087, IPR013087, I...</td>\n",
       "      <td>(Zinc finger C2H2-type, Zinc finger C2H2-type,...</td>\n",
       "      <td>(PF00096, PF00096, PF00096, PF00096, PS00028, ...</td>\n",
       "      <td>(644, 956, 983, 1013, 197, 646, 674, 957, 985,...</td>\n",
       "      <td>(666, 977, 1005, 1034, 218, 666, 694, 977, 100...</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005622, GO:000582...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>[-0.00414, -0.01288, 0.0716, 0.01605, -0.03983...</td>\n",
       "      <td>(M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...</td>\n",
       "      <td>(IPR000065, IPR000065, IPR009079, IPR009079)</td>\n",
       "      <td>(Leptin, Leptin, Four-helical cytokine-like, c...</td>\n",
       "      <td>(PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...</td>\n",
       "      <td>(28, 7, 24, 29)</td>\n",
       "      <td>(172, 172, 172, 172)</td>\n",
       "      <td>(GO:0005575, GO:0110165, GO:0005576, GO:0005615)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>[-0.01651, 0.02525, 0.04333, 0.01558, -0.01678...</td>\n",
       "      <td>(M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...</td>\n",
       "      <td>(IPR003006, IPR003597, IPR003597, IPR007110, I...</td>\n",
       "      <td>(Immunoglobulin/major histocompatibility compl...</td>\n",
       "      <td>(PS00290, PF07654, SM00407, PS50835, PR01669, ...</td>\n",
       "      <td>(382, 312, 312, 294, 55, 135, 199, 247, 271, 3...</td>\n",
       "      <td>(388, 391, 394, 401, 63, 146, 216, 261, 287, 3...</td>\n",
       "      <td>(GO:0005575, GO:0005737, GO:0042175, GO:003299...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...   \n",
       "1  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...   \n",
       "2  A0A023GPJ3  [0.01512, 0.01102, 0.0217, -0.02512, 0.0396, 0...   \n",
       "3  A0A023GUT0  [-0.00414, -0.01288, 0.0716, 0.01605, -0.03983...   \n",
       "4  A0A023IM54  [-0.01651, 0.02525, 0.04333, 0.01558, -0.01678...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "1  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "2  (M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...   \n",
       "3  (M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...   \n",
       "4  (M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0                  (IPR006910, IPR039781, IPR049589)   \n",
       "1  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "2  (IPR013087, IPR013087, IPR013087, IPR013087, I...   \n",
       "3       (IPR000065, IPR000065, IPR009079, IPR009079)   \n",
       "4  (IPR003006, IPR003597, IPR003597, IPR007110, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "1  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "2  (Zinc finger C2H2-type, Zinc finger C2H2-type,...   \n",
       "3  (Leptin, Leptin, Four-helical cytokine-like, c...   \n",
       "4  (Immunoglobulin/major histocompatibility compl...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                      (PF04825, PTHR12585, cd21792)   \n",
       "1  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "2  (PF00096, PF00096, PF00096, PF00096, PS00028, ...   \n",
       "3  (PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...   \n",
       "4  (PS00290, PF07654, SM00407, PS50835, PR01669, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                        (1, 1, 327)   \n",
       "1                          (6, 31, 33, 329, 64, 286)   \n",
       "2  (644, 956, 983, 1013, 197, 646, 674, 957, 985,...   \n",
       "3                                    (28, 7, 24, 29)   \n",
       "4  (382, 312, 312, 294, 55, 135, 199, 247, 271, 3...   \n",
       "\n",
       "                                                 end  \\\n",
       "0                                    (103, 607, 404)   \n",
       "1                     (496, 476, 175, 480, 258, 482)   \n",
       "2  (666, 977, 1005, 1034, 218, 666, 694, 977, 100...   \n",
       "3                               (172, 172, 172, 172)   \n",
       "4  (388, 391, 394, 401, 63, 146, 216, 261, 287, 3...   \n",
       "\n",
       "                                             GO_term  \n",
       "0  (GO:0005575, GO:0110165, GO:0000785, GO:003299...  \n",
       "1  (GO:0005575, GO:0110165, GO:0071944, GO:000588...  \n",
       "2  (GO:0005575, GO:0110165, GO:0005622, GO:000582...  \n",
       "3   (GO:0005575, GO:0110165, GO:0005576, GO:0005615)  \n",
       "4  (GO:0005575, GO:0005737, GO:0042175, GO:003299...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge combined_train with df_ab\n",
    "df_CC_full= pd.merge(combined_train, df_CC, on='ID', how='right')\n",
    "df_MF_full= pd.merge(combined_train, df_MF, on='ID', how='right')\n",
    "df_BP_full= pd.merge(combined_train, df_BP, on='ID', how='right')\n",
    "\n",
    "print(f\"Shapes: CC {df_CC_full.shape}, MF {df_MF_full.shape}, BP {df_BP_full.shape}\")\n",
    "\n",
    "print(f\"Full df CC {df_CC_full.shape}:\")\n",
    "df_CC_full.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shapes: x_df_CC (84638, 8), y_df_CC (84638,), x_df_MF (55698, 8), y_df_MF (55698,), x_df_BP (83064, 8), y_df_BP (83064,)\n",
      "CC Input:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A021WW32</td>\n",
       "      <td>[-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...</td>\n",
       "      <td>(M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...</td>\n",
       "      <td>(IPR006910, IPR039781, IPR049589)</td>\n",
       "      <td>(Rad21/Rec8-like protein, N-terminal, Rad21/Re...</td>\n",
       "      <td>(PF04825, PTHR12585, cd21792)</td>\n",
       "      <td>(1, 1, 327)</td>\n",
       "      <td>(103, 607, 404)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A021WZA4</td>\n",
       "      <td>[0.007904, 0.0877, -0.001715, 0.03766, 0.01788...</td>\n",
       "      <td>(M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...</td>\n",
       "      <td>(IPR004481, IPR004481, IPR004837, IPR004837, I...</td>\n",
       "      <td>(Sodium/potassium/calcium exchanger, Sodium/po...</td>\n",
       "      <td>(PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...</td>\n",
       "      <td>(6, 31, 33, 329, 64, 286)</td>\n",
       "      <td>(496, 476, 175, 480, 258, 482)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A023GPJ3</td>\n",
       "      <td>[0.01512, 0.01102, 0.0217, -0.02512, 0.0396, 0...</td>\n",
       "      <td>(M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...</td>\n",
       "      <td>(IPR013087, IPR013087, IPR013087, IPR013087, I...</td>\n",
       "      <td>(Zinc finger C2H2-type, Zinc finger C2H2-type,...</td>\n",
       "      <td>(PF00096, PF00096, PF00096, PF00096, PS00028, ...</td>\n",
       "      <td>(644, 956, 983, 1013, 197, 646, 674, 957, 985,...</td>\n",
       "      <td>(666, 977, 1005, 1034, 218, 666, 694, 977, 100...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A023GUT0</td>\n",
       "      <td>[-0.00414, -0.01288, 0.0716, 0.01605, -0.03983...</td>\n",
       "      <td>(M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...</td>\n",
       "      <td>(IPR000065, IPR000065, IPR009079, IPR009079)</td>\n",
       "      <td>(Leptin, Leptin, Four-helical cytokine-like, c...</td>\n",
       "      <td>(PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...</td>\n",
       "      <td>(28, 7, 24, 29)</td>\n",
       "      <td>(172, 172, 172, 172)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A023IM54</td>\n",
       "      <td>[-0.01651, 0.02525, 0.04333, 0.01558, -0.01678...</td>\n",
       "      <td>(M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...</td>\n",
       "      <td>(IPR003006, IPR003597, IPR003597, IPR007110, I...</td>\n",
       "      <td>(Immunoglobulin/major histocompatibility compl...</td>\n",
       "      <td>(PS00290, PF07654, SM00407, PS50835, PR01669, ...</td>\n",
       "      <td>(382, 312, 312, 294, 55, 135, 199, 247, 271, 3...</td>\n",
       "      <td>(388, 391, 394, 401, 63, 146, 216, 261, 287, 3...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A021WW32  [-0.01643, -0.001583, 0.00389, 0.0734, 0.01243...   \n",
       "1  A0A021WZA4  [0.007904, 0.0877, -0.001715, 0.03766, 0.01788...   \n",
       "2  A0A023GPJ3  [0.01512, 0.01102, 0.0217, -0.02512, 0.0396, 0...   \n",
       "3  A0A023GUT0  [-0.00414, -0.01288, 0.0716, 0.01605, -0.03983...   \n",
       "4  A0A023IM54  [-0.01651, 0.02525, 0.04333, 0.01558, -0.01678...   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, F, Y, E, H, I, I, L, A, K, K, G, P, L, A, ...   \n",
       "1  (M, K, Y, I, N, C, T, Q, P, A, I, D, D, F, P, ...   \n",
       "2  (M, D, R, D, A, E, E, G, R, P, L, S, L, V, N, ...   \n",
       "3  (M, R, C, P, G, V, S, L, W, G, L, L, C, L, G, ...   \n",
       "4  (M, K, L, L, P, L, L, V, V, V, A, L, G, L, A, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0                  (IPR006910, IPR039781, IPR049589)   \n",
       "1  (IPR004481, IPR004481, IPR004837, IPR004837, I...   \n",
       "2  (IPR013087, IPR013087, IPR013087, IPR013087, I...   \n",
       "3       (IPR000065, IPR000065, IPR009079, IPR009079)   \n",
       "4  (IPR003006, IPR003597, IPR003597, IPR007110, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0  (Rad21/Rec8-like protein, N-terminal, Rad21/Re...   \n",
       "1  (Sodium/potassium/calcium exchanger, Sodium/po...   \n",
       "2  (Zinc finger C2H2-type, Zinc finger C2H2-type,...   \n",
       "3  (Leptin, Leptin, Four-helical cytokine-like, c...   \n",
       "4  (Immunoglobulin/major histocompatibility compl...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                      (PF04825, PTHR12585, cd21792)   \n",
       "1  (PTHR10846, TIGR00367, PF01699, PF01699, G3DSA...   \n",
       "2  (PF00096, PF00096, PF00096, PF00096, PS00028, ...   \n",
       "3  (PF02024, PTHR11724, G3DSA:1.20.1250.10, SSF47...   \n",
       "4  (PS00290, PF07654, SM00407, PS50835, PR01669, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                        (1, 1, 327)   \n",
       "1                          (6, 31, 33, 329, 64, 286)   \n",
       "2  (644, 956, 983, 1013, 197, 646, 674, 957, 985,...   \n",
       "3                                    (28, 7, 24, 29)   \n",
       "4  (382, 312, 312, 294, 55, 135, 199, 247, 271, 3...   \n",
       "\n",
       "                                                 end  \n",
       "0                                    (103, 607, 404)  \n",
       "1                     (496, 476, 175, 480, 258, 482)  \n",
       "2  (666, 977, 1005, 1034, 218, 666, 694, 977, 100...  \n",
       "3                               (172, 172, 172, 172)  \n",
       "4  (388, 391, 394, 401, 63, 146, 216, 261, 287, 3...  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_df_CC = df_CC_full.iloc[:, :-1]\n",
    "y_df_CC = df_CC_full.iloc[:, -1]\n",
    "\n",
    "x_df_MF = df_MF_full.iloc[:, :-1]\n",
    "y_df_MF = df_MF_full.iloc[:, -1]\n",
    "\n",
    "X_df_BP = df_BP_full.iloc[:, :-1]\n",
    "y_df_BP = df_BP_full.iloc[:, -1]\n",
    "\n",
    "print(f\"Shapes: x_df_CC {x_df_CC.shape}, y_df_CC {y_df_CC.shape}, x_df_MF {x_df_MF.shape}, y_df_MF {y_df_MF.shape}, x_df_BP {X_df_BP.shape}, y_df_BP {y_df_BP.shape}\")\n",
    "\n",
    "print(\"CC Input:\")\n",
    "x_df_CC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CC Target:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    (GO:0005575, GO:0110165, GO:0000785, GO:003299...\n",
       "1    (GO:0005575, GO:0110165, GO:0071944, GO:000588...\n",
       "2    (GO:0005575, GO:0110165, GO:0005622, GO:000582...\n",
       "3     (GO:0005575, GO:0110165, GO:0005576, GO:0005615)\n",
       "4    (GO:0005575, GO:0005737, GO:0042175, GO:003299...\n",
       "Name: GO_term, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"CC Target:\")\n",
    "y_df_CC.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test set\n",
    "\n",
    "In this section we are **creating the Test set**, fusing togheter the data from:\n",
    "- _test_ids.txt_\n",
    "- _test.fasta_\n",
    "- _test_embeddings.h5_\n",
    "- _test_protein2ipr.dat_\n",
    "- _blast_test_results.tsv_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_ids.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['O43747', 'Q969H0', 'Q9JMA2', 'P18065', 'A0A8I6AN32']\n"
     ]
    }
   ],
   "source": [
    "# Extracting test_ids.txt\n",
    "with open(test_data_path / 'test_ids.txt', 'r') as file:\n",
    "    test_ids = file.read().splitlines()\n",
    "\n",
    "# Display the first few IDs to verify\n",
    "print(test_ids[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test.fasta`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: O43747\n",
      "Name: O43747\n",
      "Description: O43747\n",
      "Number of features: 0\n",
      "Seq('MPAPIRLRELIRTIRTARTQAEEREMIQKECAAIRSSFREEDNTYRCRNVAKLL...SWQ')\n"
     ]
    }
   ],
   "source": [
    "test_fasta_list = list(SeqIO.parse(test_data_path / 'test.fasta', 'fasta'))\n",
    "\n",
    "# Print the first sequence to verify\n",
    "print(test_fasta_list[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>num_features</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O43747</td>\n",
       "      <td>O43747</td>\n",
       "      <td>O43747</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q969H0</td>\n",
       "      <td>Q969H0</td>\n",
       "      <td>Q969H0</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P18065</td>\n",
       "      <td>P18065</td>\n",
       "      <td>P18065</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>0</td>\n",
       "      <td>(M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        name description  num_features  \\\n",
       "0      O43747      O43747      O43747             0   \n",
       "1      Q969H0      Q969H0      Q969H0             0   \n",
       "2      Q9JMA2      Q9JMA2      Q9JMA2             0   \n",
       "3      P18065      P18065      P18065             0   \n",
       "4  A0A8I6AN32  A0A8I6AN32  A0A8I6AN32             0   \n",
       "\n",
       "                                            sequence  \n",
       "0  (M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...  \n",
       "1  (M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...  \n",
       "2  (M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...  \n",
       "3  (M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...  \n",
       "4  (M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract relevant information from SeqRecord\n",
    "test_fasta_dict = [{\n",
    "    'ID': record.id,\n",
    "    'name': record.name,\n",
    "    'description': record.description,\n",
    "    'num_features': len(record.features),\n",
    "    'sequence': record.seq,\n",
    "} for record in test_fasta_list]\n",
    "\n",
    "# Create a DataFrame from the extracted data\n",
    "test_fasta = pd.DataFrame(test_fasta_dict)\n",
    "\n",
    "# Display the DataFrame\n",
    "test_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's check if ID, name and description are the same thing, as well as seeing if num_features has relevant information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 differences between the ID and name columns.\n",
      "We have a total of 0 differences between the ID and description columns.\n"
     ]
    }
   ],
   "source": [
    "# Checking for differences between the ID and name columns\n",
    "diff_id_name = sum(test_fasta['ID'] != test_fasta['name'])\n",
    "\n",
    "# Checking for differences between the ID and description columns\n",
    "diff_id_description = sum(test_fasta['ID'] != test_fasta['description'])\n",
    "\n",
    "print(f\"We have a total of {diff_id_name} differences between the ID and name columns.\\nWe have a total of {diff_id_description} differences between the ID and description columns.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We have a total of 0 sequences with features.\n"
     ]
    }
   ],
   "source": [
    "num_features_values = sum(test_fasta['num_features'] != 0)\n",
    "\n",
    "print(f\"We have a total of {num_features_values} sequences with features.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's drop the useless columns, as before"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>O43747</td>\n",
       "      <td>(M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q969H0</td>\n",
       "      <td>(M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q9JMA2</td>\n",
       "      <td>(M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P18065</td>\n",
       "      <td>(M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A8I6AN32</td>\n",
       "      <td>(M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                           sequence\n",
       "0      O43747  (M, P, A, P, I, R, L, R, E, L, I, R, T, I, R, ...\n",
       "1      Q969H0  (M, N, Q, E, L, L, S, V, G, S, K, R, R, R, T, ...\n",
       "2      Q9JMA2  (M, A, A, V, G, S, P, G, S, L, E, S, A, P, R, ...\n",
       "3      P18065  (M, L, P, R, V, G, C, P, A, L, P, L, P, P, P, ...\n",
       "4  A0A8I6AN32  (M, A, S, N, D, Y, T, Q, Q, A, T, Q, S, Y, G, ..."
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_fasta.drop(columns=['name', 'description', 'num_features'], inplace=True)\n",
    "\n",
    "\n",
    "test_fasta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's also check if the IDs are the same as the ones in the txt file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set (1000).\n",
      "Proceeding with the analysis.\n"
     ]
    }
   ],
   "source": [
    "len_ID = len(test_fasta['ID'].unique()) # assigned because gave problem on else statement print\n",
    "\n",
    "if len(test_ids) == len_ID:\n",
    "    print(f\"The number of IDs in train_ids.txt is equal to the number of unique IDs in the train set ({len(test_ids)}).\\n\"\n",
    "          \"Proceeding with the analysis.\")\n",
    "else:\n",
    "    print(f'The numbers are not the same: test_ids are {len(test_ids)}, while the length of the fasta file is {len_ID})')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_embeddings.h5`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>[0.00979, -0.03973, 0.03653, -0.006447, -0.040...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>[0.02786, -0.01154, 0.008865, -0.01765, 0.0073...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4P506</td>\n",
       "      <td>[0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K1A2</td>\n",
       "      <td>[0.00882, 0.0835, -0.001374, -0.0003645, -0.06...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2K1V4</td>\n",
       "      <td>[0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings\n",
       "0  A0A0B4JCV4  [0.00979, -0.03973, 0.03653, -0.006447, -0.040...\n",
       "1  A0A0B4KHT0  [0.02786, -0.01154, 0.008865, -0.01765, 0.0073...\n",
       "2  A0A0B4P506  [0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...\n",
       "3  A0A0G2K1A2  [0.00882, 0.0835, -0.001374, -0.0003645, -0.06...\n",
       "4  A0A0G2K1V4  [0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0...."
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_list = []\n",
    "\n",
    "with h5py.File(test_data_path / \"test_embeddings.h5\", \"r\") as f:\n",
    "    for dataset_name in f.keys():\n",
    "        dataset = f[dataset_name][:]\n",
    "        data_list.append([dataset_name, dataset])\n",
    "\n",
    "test_embeddings = pd.DataFrame(data_list, columns=[\"ID\", \"embeddings\"])\n",
    "\n",
    "test_embeddings.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `test_protein2ipr.dat`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>IPR039915</td>\n",
       "      <td>TACC family</td>\n",
       "      <td>PTHR13924</td>\n",
       "      <td>38</td>\n",
       "      <td>1206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PF00643</td>\n",
       "      <td>177</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PF00643</td>\n",
       "      <td>236</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PS50119</td>\n",
       "      <td>173</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>IPR000315</td>\n",
       "      <td>B-box-type zinc finger</td>\n",
       "      <td>PS50119</td>\n",
       "      <td>235</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID        ipr                  domain   familyID  start   end\n",
       "0  A0A0B4JCV4  IPR039915             TACC family  PTHR13924     38  1206\n",
       "1  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PF00643    177   219\n",
       "2  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PF00643    236   274\n",
       "3  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PS50119    173   220\n",
       "4  A0A0B4KHT0  IPR000315  B-box-type zinc finger    PS50119    235   282"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_protein2ipr = pd.read_csv(test_data_path / 'test_protein2ipr.dat', sep='\\t')\n",
    "\n",
    "# Rename Protein_ID and aspect columns\n",
    "test_protein2ipr.columns = ['ID', 'ipr', 'domain', 'familyID', 'start', 'end']\n",
    "\n",
    "# Remove 'domain' that is useless\n",
    "test_protein2ipr.drop('domain', axis=1)\n",
    "\n",
    "test_protein2ipr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test protein2ipr ((11263, 6)):\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>(IPR039915,)</td>\n",
       "      <td>(TACC family,)</td>\n",
       "      <td>(PTHR13924,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(1206,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>(IPR000315, IPR000315, IPR000315, IPR000315, I...</td>\n",
       "      <td>(B-box-type zinc finger, B-box-type zinc finge...</td>\n",
       "      <td>(PF00643, PF00643, PS50119, PS50119, SM00336, ...</td>\n",
       "      <td>(177, 236, 173, 235, 173, 235, 976, 826, 988, ...</td>\n",
       "      <td>(219, 274, 220, 282, 220, 276, 1048, 839, 1004...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4P506</td>\n",
       "      <td>(IPR003417, IPR003417, IPR036552, IPR036552)</td>\n",
       "      <td>(Core-binding factor, beta subunit, Core-bindi...</td>\n",
       "      <td>(PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)</td>\n",
       "      <td>(1, 1, 1, 4)</td>\n",
       "      <td>(164, 168, 142, 140)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K1A2</td>\n",
       "      <td>(IPR010255, IPR019791, IPR019791, IPR019791, I...</td>\n",
       "      <td>(Haem peroxidase superfamily, Haem peroxidase,...</td>\n",
       "      <td>(SSF48113, PF03098, PR00457, PR00457, PR00457,...</td>\n",
       "      <td>(142, 148, 172, 226, 374, 392, 417, 470, 598, ...</td>\n",
       "      <td>(718, 692, 183, 241, 392, 412, 443, 480, 618, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2K1V4</td>\n",
       "      <td>(IPR000048, IPR000048, IPR001609, IPR001609, I...</td>\n",
       "      <td>(IQ motif, EF-hand binding site, IQ motif, EF-...</td>\n",
       "      <td>(PS50096, SM00015, PF00063, PR00193, PR00193, ...</td>\n",
       "      <td>(789, 788, 89, 116, 172, 228, 459, 513, 86, 80...</td>\n",
       "      <td>(818, 810, 774, 135, 197, 255, 487, 541, 786, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                                ipr  \\\n",
       "0  A0A0B4JCV4                                       (IPR039915,)   \n",
       "1  A0A0B4KHT0  (IPR000315, IPR000315, IPR000315, IPR000315, I...   \n",
       "2  A0A0B4P506       (IPR003417, IPR003417, IPR036552, IPR036552)   \n",
       "3  A0A0G2K1A2  (IPR010255, IPR019791, IPR019791, IPR019791, I...   \n",
       "4  A0A0G2K1V4  (IPR000048, IPR000048, IPR001609, IPR001609, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0                                     (TACC family,)   \n",
       "1  (B-box-type zinc finger, B-box-type zinc finge...   \n",
       "2  (Core-binding factor, beta subunit, Core-bindi...   \n",
       "3  (Haem peroxidase superfamily, Haem peroxidase,...   \n",
       "4  (IQ motif, EF-hand binding site, IQ motif, EF-...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                                       (PTHR13924,)   \n",
       "1  (PF00643, PF00643, PS50119, PS50119, SM00336, ...   \n",
       "2  (PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)   \n",
       "3  (SSF48113, PF03098, PR00457, PR00457, PR00457,...   \n",
       "4  (PS50096, SM00015, PF00063, PR00193, PR00193, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                              (38,)   \n",
       "1  (177, 236, 173, 235, 173, 235, 976, 826, 988, ...   \n",
       "2                                       (1, 1, 1, 4)   \n",
       "3  (142, 148, 172, 226, 374, 392, 417, 470, 598, ...   \n",
       "4  (789, 788, 89, 116, 172, 228, 459, 513, 86, 80...   \n",
       "\n",
       "                                                 end  \n",
       "0                                            (1206,)  \n",
       "1  (219, 274, 220, 282, 220, 276, 1048, 839, 1004...  \n",
       "2                               (164, 168, 142, 140)  \n",
       "3  (718, 692, 183, 241, 392, 412, 443, 480, 618, ...  \n",
       "4  (818, 810, 774, 135, 197, 255, 487, 541, 786, ...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Group by 'ID' and aggregate other columns into lists\n",
    "test_protein2ipr_grouped = test_protein2ipr.groupby('ID').agg(lambda x: tuple(x)).reset_index()\n",
    "\n",
    "print(f\"Test protein2ipr ({test_protein2ipr.shape}):\")\n",
    "test_protein2ipr_grouped.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extracting `blast_test_results.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows missing from train_protein2ipr_grouped: 19\n",
      "Combined DataFrame shape: (1000, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>embeddings</th>\n",
       "      <th>sequence</th>\n",
       "      <th>ipr</th>\n",
       "      <th>domain</th>\n",
       "      <th>familyID</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0B4JCV4</td>\n",
       "      <td>[0.00979, -0.03973, 0.03653, -0.006447, -0.040...</td>\n",
       "      <td>(M, E, F, D, D, A, E, N, G, L, G, M, G, F, G, ...</td>\n",
       "      <td>(IPR039915,)</td>\n",
       "      <td>(TACC family,)</td>\n",
       "      <td>(PTHR13924,)</td>\n",
       "      <td>(38,)</td>\n",
       "      <td>(1206,)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0B4KHT0</td>\n",
       "      <td>[0.02786, -0.01154, 0.008865, -0.01765, 0.0073...</td>\n",
       "      <td>(M, D, M, D, L, E, Q, L, K, N, D, F, L, P, L, ...</td>\n",
       "      <td>(IPR000315, IPR000315, IPR000315, IPR000315, I...</td>\n",
       "      <td>(B-box-type zinc finger, B-box-type zinc finge...</td>\n",
       "      <td>(PF00643, PF00643, PS50119, PS50119, SM00336, ...</td>\n",
       "      <td>(177, 236, 173, 235, 173, 235, 976, 826, 988, ...</td>\n",
       "      <td>(219, 274, 220, 282, 220, 276, 1048, 839, 1004...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0B4P506</td>\n",
       "      <td>[0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...</td>\n",
       "      <td>(M, P, R, V, V, P, D, Q, R, S, K, F, E, N, E, ...</td>\n",
       "      <td>(IPR003417, IPR003417, IPR036552, IPR036552)</td>\n",
       "      <td>(Core-binding factor, beta subunit, Core-bindi...</td>\n",
       "      <td>(PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)</td>\n",
       "      <td>(1, 1, 1, 4)</td>\n",
       "      <td>(164, 168, 142, 140)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2K1A2</td>\n",
       "      <td>[0.00882, 0.0835, -0.001374, -0.0003645, -0.06...</td>\n",
       "      <td>(M, K, L, F, L, A, L, A, G, L, L, A, P, L, A, ...</td>\n",
       "      <td>(IPR010255, IPR019791, IPR019791, IPR019791, I...</td>\n",
       "      <td>(Haem peroxidase superfamily, Haem peroxidase,...</td>\n",
       "      <td>(SSF48113, PF03098, PR00457, PR00457, PR00457,...</td>\n",
       "      <td>(142, 148, 172, 226, 374, 392, 417, 470, 598, ...</td>\n",
       "      <td>(718, 692, 183, 241, 392, 412, 443, 480, 618, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2K1V4</td>\n",
       "      <td>[0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0....</td>\n",
       "      <td>(M, S, S, D, A, E, M, A, V, F, G, E, A, A, P, ...</td>\n",
       "      <td>(IPR000048, IPR000048, IPR001609, IPR001609, I...</td>\n",
       "      <td>(IQ motif, EF-hand binding site, IQ motif, EF-...</td>\n",
       "      <td>(PS50096, SM00015, PF00063, PR00193, PR00193, ...</td>\n",
       "      <td>(789, 788, 89, 116, 172, 228, 459, 513, 86, 80...</td>\n",
       "      <td>(818, 810, 774, 135, 197, 255, 487, 541, 786, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID                                         embeddings  \\\n",
       "0  A0A0B4JCV4  [0.00979, -0.03973, 0.03653, -0.006447, -0.040...   \n",
       "1  A0A0B4KHT0  [0.02786, -0.01154, 0.008865, -0.01765, 0.0073...   \n",
       "2  A0A0B4P506  [0.01643, 0.01802, 0.03702, -0.0591, 0.0356, 0...   \n",
       "3  A0A0G2K1A2  [0.00882, 0.0835, -0.001374, -0.0003645, -0.06...   \n",
       "4  A0A0G2K1V4  [0.0659, 0.0929, -0.001803, 0.0226, 0.0383, 0....   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  (M, E, F, D, D, A, E, N, G, L, G, M, G, F, G, ...   \n",
       "1  (M, D, M, D, L, E, Q, L, K, N, D, F, L, P, L, ...   \n",
       "2  (M, P, R, V, V, P, D, Q, R, S, K, F, E, N, E, ...   \n",
       "3  (M, K, L, F, L, A, L, A, G, L, L, A, P, L, A, ...   \n",
       "4  (M, S, S, D, A, E, M, A, V, F, G, E, A, A, P, ...   \n",
       "\n",
       "                                                 ipr  \\\n",
       "0                                       (IPR039915,)   \n",
       "1  (IPR000315, IPR000315, IPR000315, IPR000315, I...   \n",
       "2       (IPR003417, IPR003417, IPR036552, IPR036552)   \n",
       "3  (IPR010255, IPR019791, IPR019791, IPR019791, I...   \n",
       "4  (IPR000048, IPR000048, IPR001609, IPR001609, I...   \n",
       "\n",
       "                                              domain  \\\n",
       "0                                     (TACC family,)   \n",
       "1  (B-box-type zinc finger, B-box-type zinc finge...   \n",
       "2  (Core-binding factor, beta subunit, Core-bindi...   \n",
       "3  (Haem peroxidase superfamily, Haem peroxidase,...   \n",
       "4  (IQ motif, EF-hand binding site, IQ motif, EF-...   \n",
       "\n",
       "                                            familyID  \\\n",
       "0                                       (PTHR13924,)   \n",
       "1  (PF00643, PF00643, PS50119, PS50119, SM00336, ...   \n",
       "2  (PF02312, PTHR10276, G3DSA:2.40.250.10, SSF50723)   \n",
       "3  (SSF48113, PF03098, PR00457, PR00457, PR00457,...   \n",
       "4  (PS50096, SM00015, PF00063, PR00193, PR00193, ...   \n",
       "\n",
       "                                               start  \\\n",
       "0                                              (38,)   \n",
       "1  (177, 236, 173, 235, 173, 235, 976, 826, 988, ...   \n",
       "2                                       (1, 1, 1, 4)   \n",
       "3  (142, 148, 172, 226, 374, 392, 417, 470, 598, ...   \n",
       "4  (789, 788, 89, 116, 172, 228, 459, 513, 86, 80...   \n",
       "\n",
       "                                                 end  \n",
       "0                                            (1206,)  \n",
       "1  (219, 274, 220, 282, 220, 276, 1048, 839, 1004...  \n",
       "2                               (164, 168, 142, 140)  \n",
       "3  (718, 692, 183, 241, 392, 412, 443, 480, 618, ...  \n",
       "4  (818, 810, 774, 135, 197, 255, 487, 541, 786, ...  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_test = pd.merge(test_embeddings, test_fasta, on='ID')\n",
    "combined_test = pd.merge(combined_test, test_protein2ipr_grouped, on='ID', how='left')\n",
    "\n",
    "missing_rows = combined_test[combined_test['ipr'].isna()].shape[0]\n",
    "print(f\"Number of rows missing from train_protein2ipr_grouped: {missing_rows}\")\n",
    "\n",
    "print(f\"Combined DataFrame shape: {combined_test.shape}\")\n",
    "combined_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combined_train and combined_test have the same columns: True\n"
     ]
    }
   ],
   "source": [
    "# Check if combined_train and combined_test have the same columns\n",
    "print(\"Combined_train and combined_test have the same columns:\" , set(combined_train.columns) == set(combined_test.columns))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PCA for embeddings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll see that considering each df separately doesn't change much."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAArMAAAGuCAYAAACDR47DAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAB/o0lEQVR4nO3dd1hT1/8H8HcIYQuCyHABVasi7q1171mtVmvd61fFWVe1raK2ilpr7XBUq+Kuo9o6qIp7a92z1ipKVRAVBWWG5Pz+4JtoTMAEEpKQ9+t5eEzOPffez80J4ePJuedIhBACRERERERWyM7cARARERER5RaTWSIiIiKyWkxmiYiIiMhqMZklIiIiIqvFZJaIiIiIrBaTWSIiIiKyWkxmiYiIiMhqMZklIiIiIqvFZJaIiIiIrBaTWQIAXL58GQMGDEBQUBCcnJzg5uaG6tWrY+7cuUhISDB3eDmaNm0aJBJJrvaNjIzEtGnTdG4LDAxE//79cx9YLjVp0gQSiUTnT2BgoMnP3aRJk3zfN68iIiIgkUhw9+7dbOtUq1YNxYsXh0KhyLZOgwYN4O3tjYyMjDzHdOjQIUgkEhw6dCjPxzIn1XVs2bLFpOdRtaHqx97eHiVKlMCAAQPw4MEDrfp37tzBiBEj8O6778LZ2RkuLi6oWLEivvzyS531AeCDDz6ARCLBiBEjDI4vKSkJM2fORM2aNeHu7g5HR0cEBgZi4MCBOH/+vMHHI00pKSmYNm2a0X9fNm7ciIoVK8LZ2RkSiQQXL1406vFfp/pdef3H09MTderUwapVq7TqBwYGatR1c3NDnTp1sHr1apPFWFDZmzsAMr9ly5YhNDQU5cqVw4QJExAcHAy5XI6zZ89iyZIlOHnyJLZt22buME0iMjISCxcu1JnQbtu2De7u7vkfFIB33nkH69at0yp3dHQ0QzT6WbRokblDyNGgQYMwcuRI7NmzB+3atdPa/s8//+DEiRMYM2YMHBwc8ny+6tWr4+TJkwgODs7zsWzJypUrUb58eaSmpuLIkSMIDw/H4cOHceXKFbi6ugIAdu7ciY8++gje3t4YMWIEqlWrBolEgitXrmDFihXYtWsXLly4oHHc+Ph47Ny5EwCwbt06zJs3D05OTnrFdPv2bbRq1Qrx8fEYOnQopk+fDjc3N9y9exebNm1CjRo18Pz5c3h4eBj3xbAhKSkpmD59OgAY7T/Fjx8/Rp8+fdCmTRssWrQIjo6OePfdd41y7JzMmjULTZs2BQA8efIEq1evRv/+/ZGUlISRI0dq1G3QoAHmzZsHALh//z7mzZuHfv36ITk5GcOGDTN5rAWGIJt24sQJIZVKRZs2bURaWprW9vT0dPHHH3+YITL9hYWFidy+lYcPH57rfU2lcePGomLFimY7d+PGjc1y7rxYuXKlACCio6OzrZOQkCCcnJxE165ddW7/7LPPBABx+fLlPMWSkZEh5HJ5no5hSQ4ePCgAiM2bN5v0PKo2/OuvvzTKp0yZIgCItWvXCiGEuHPnjnB1dRXVqlUTz58/1zqOUqkUv/32m1b5N998IwCI9u3bCwBi3bp1esWVmZkpKlWqJNzd3cWVK1d01omMjBTJycl6HY90e/z4sQAgwsLCjHbMY8eOCQBi48aNRjtmTu2c3e+KQqEQgYGBol69ehrlAQEBon379hplz549E+7u7qJMmTJGi9kWcJiBjZs1axYkEgmWLl2qs9fPwcEBnTp1Uj+XSCQ6ezHf/Epe9ZXhgQMHMGTIEBQpUgTu7u7o27cvkpOTERcXh+7du6Nw4cLw9/fH+PHjIZfL1ftn9xXt3bt3IZFIEBERkeN1bdy4Ea1atYK/vz+cnZ1RoUIFTJo0CcnJyeo6/fv3x8KFC9XXpfpRfVX9+jU9fvwYDg4OmDJlita5/v77b0gkEvzwww/qsri4OHzyyScoUaIEHBwcEBQUhOnTpyMzMzPHuPUlhEC7du1QpEgRxMTEqMtTUlJQsWJFVKhQQX2tqmEYFy5cwAcffAB3d3d4eHigd+/eePz48VvPNX36dNSpUwdeXl5wd3dH9erVsXz5cgghNOq9OcxA1Vbz5s3D/PnzERQUBDc3N9SrVw+nTp3SOs/Zs2fRqVMneHl5wcnJCdWqVcOmTZu06p06dQoNGjSAk5MTihUrhsmTJ2u8d7Lj6emJLl26YMeOHXj69KnGNoVCgTVr1qBWrVqoVKkS/v33XwwYMABly5aFi4sLihcvjo4dO+LKlSsa+6nep2vWrMG4ceNQvHhxODo64t9//9X5Hj579iw++ugjBAYGwtnZGYGBgejZsyfu3buncVzV78/BgwcxbNgweHt7o0iRIvjggw/w8OFDrWtbv3496tWrBzc3N7i5uaFq1apYvny5Rp19+/ahefPmcHd3h4uLCxo0aID9+/e/9XVTSUtLw9ixY+Hn5wdnZ2c0btxYo/dzzZo1kEgkOHnypNa+M2bMgEwm0xn729StWxcA1K/R/PnzkZycjEWLFunsCZVIJPjggw+0ylesWAFfX1+sWrUKzs7OWLFihV7n//3333HlyhVMnjwZISEhOuu0bdsWLi4u6ufHjh1D8+bNUahQIbi4uKB+/frYtWuXxj55/YxU/X7NnTsXM2fORKlSpeDk5ISaNWvqbFdDYtL3fbdx40bUq1cPrq6ucHNzQ+vWrbV6xPv37w83Nzf8+++/aNeuHdzc3FCyZEmMGzcO6enp6mspWrQogKzPG9Vn8eufv//3f/+HkiVLwtHREUWLFkWDBg2wb9++7JoN/fv3x3vvvQcA6NGjByQSicbn0/bt21GvXj24uLigUKFCaNmypdZ7V/XZef78eXTr1g2enp4oXbp0tufMjp2dHdzc3CCTyd5at3DhwihXrpzWZwLljMmsDVMoFDhw4ABq1KiBkiVLmuQcgwcPhoeHB3799Vd8+eWXWL9+PYYMGYL27dujSpUq2LJlC/r164dvv/0WP/74o9HOe+vWLbRr1w7Lly/H7t27MWbMGGzatAkdO3ZU15kyZQq6desGADh58qT6x9/fX+t4RYsWRYcOHbBq1SoolUqNbStXroSDgwN69eoFICuRrV27Nvbs2YOpU6fizz//xKBBgxAeHo4hQ4bofQ2ZmZlaP6pzqxIoFxcXdO/eXf1HLjQ0FNHR0di0aZP6K1mVLl26oEyZMtiyZQumTZuG33//Ha1bt35rInj37l188skn2LRpE7Zu3YoPPvgAI0eOxFdffaXXdSxcuBBRUVFYsGAB1q1bh+TkZLRr1w6JiYnqOgcPHkSDBg3w/PlzLFmyBH/88QeqVq2KHj16aPzH5fr162jevDmeP3+OiIgILFmyBBcuXMDXX3+tVyyDBg1CRkYG1q5dq1G+Z88ePHz4EIMGDQIAPHz4EEWKFMHs2bOxe/duLFy4EPb29qhTpw5u3rypddzJkycjJiYGS5YswY4dO+Dj46Pz/Hfv3kW5cuWwYMEC7NmzB3PmzEFsbCxq1aqFJ0+eaNUfPHgwZDIZ1q9fj7lz5+LQoUPo3bu3Rp2pU6eiV69eKFasGCIiIrBt2zb069dP44/h2rVr0apVK7i7u2PVqlXYtGkTvLy80Lp1a70T2s8//xx37tzBL7/8gl9++QUPHz5EkyZNcOfOHQBZCYOfn5/6P4gqmZmZ+Pnnn9GlSxcUK1ZMr3O97t9//wUAdbKzd+9e+Pr6qpNcfZw4cQI3btxA3759UaRIEXTt2hUHDhxAdHT0W/fdu3cvAKBz5856nevw4cNo1qwZEhMTsXz5cmzYsAGFChVCx44dsXHjRq36ef2M/Omnn7B7924sWLAAa9euhZ2dHdq2bauRmOUmpre972bNmoWePXsiODgYmzZtwpo1a/DixQs0bNgQ169f16grl8vRqVMnNG/eHH/88QcGDhyI7777DnPmzAEA+Pv7Y/fu3QCyfkdVn8WqzoM+ffrg999/x9SpU7F371788ssvaNGihdZ/Sl83ZcoU9Xtx1qxZOHnypHoo1Pr16/H+++/D3d0dGzZswPLly/Hs2TM0adIEx44d0zrWBx98gDJlymDz5s1YsmRJtudUUSqV6s/sR48eYfbs2bh69arWa6iLXC7HvXv31O930pO5u4bJfOLi4gQA8dFHH+m9D7L5GiggIED069dP/Vz1leHIkSM16nXu3FkAEPPnz9cor1q1qqhevbr6uerrmoMHD2rUi46OFgDEypUr1WVvG2agVCqFXC4Xhw8fFgDEpUuX1NtyGmbw5jVt375dABB79+5Vl2VmZopixYppfHX9ySefCDc3N3Hv3j2N482bN08AENeuXcs2ViGyvuoHoPNn0KBBGnWPHTsm7O3txZgxY8SKFSsEAPHLL79o1FG9Pp9++qlG+bp16zS+vlWdO6dhBgqFQsjlcjFjxgxRpEgRoVQqs91X1VaVKlUSmZmZ6vIzZ84IAGLDhg3qsvLly4tq1appfT3foUMH4e/vLxQKhRBCiB49eghnZ2cRFxenrpOZmSnKly//1mEGQmS9F4KCgkTlypU1yrt27SpcXFxEYmKizv0yMzNFRkaGKFu2rMbrqHqfNmrUSGuf7N7Dbx735cuXwtXVVXz//ffqctXvT2hoqEb9uXPnCgAiNjZWCJH1lbtUKhW9evXK9hzJycnCy8tLdOzYUaNcoVCIKlWqiNq1a2e77+vXUb16dY32vnv3rpDJZGLw4MHqsrCwMOHg4CAePXqkLtu4caMAIA4fPpzjeVTXfOrUKSGXy8WLFy/Ezp07RdGiRUWhQoXUbe7k5CTq1q2b47HeNHDgQAFA3LhxQ+OapkyZ8tZ927RpIwDoHIalS926dYWPj4948eKFuiwzM1OEhISIEiVKqF/DvH5Gqn6/ihUrJlJTU9XlSUlJwsvLS7Ro0SLXMb3tfRcTEyPs7e21Yn/x4oXw8/MT3bt3V5f169dPABCbNm3SqNuuXTtRrlw59fOchhm4ubmJMWPGaJW/ja6v/RUKhShWrJioVKmS+nNFFbuPj4+oX7++ukz12Tl16lSDzvfmj52dnfjiiy+06gcEBIh27doJuVwu5HK5iI6OVr9eEyZMMPh6bRl7ZsmkOnTooPG8QoUKAID27dtrlRvza5U7d+7g448/hp+fH6RSKWQyGRo3bgwAuHHjRq6O2bZtW/j5+WHlypXqMlWP3sCBA9VlO3fuRNOmTVGsWDGNXtW2bdsCyOoleZvSpUvjr7/+0vp5c5hDgwYNMHPmTCxYsADDhg1D79691b2Lb1L1HKt0794d9vb2OHjwYI6xHDhwAC1atICHh4f6tZw6dSqePn2K+Pj4t15L+/btIZVK1c8rV64M4NXXxv/++y/+/vtvdXyvv2bt2rVDbGysujf04MGDaN68OXx9fdXHk0ql6NGjx1vjALJ6tAcMGIDLly/j3LlzAICnT59ix44d6Nq1q/qGv8zMTMyaNQvBwcFwcHCAvb09HBwccOvWLZ3vn65du+p1/pcvX+Kzzz5DmTJlYG9vD3t7e7i5uSE5OVnncV8f4gNov3ZRUVFQKBQYPnx4tuc8ceIEEhIS0K9fP61e/jZt2uCvv/7SGH6TnY8//lhj1pCAgADUr19f4/2jumFl2bJl6rKffvoJlSpVQqNGjd56DiBrWIFMJkOhQoXQoUMH+Pn54c8//9Roc0O8fPkSmzZtQv369VG+fHkAQOPGjVG6dGlERERofdOSF8nJyTh9+jS6desGNzc3dblUKkWfPn1w//59rZ79vH5GfvDBBxo3sql6XI8cOQKFQpGrmN72vtuzZw8yMzPRt29fjfeUk5MTGjdurDU8TCKRaHwrpjqmvp/5tWvXRkREBL7++mucOnVKr2FF2bl58yYePnyIPn36wM7uVQrk5uaGrl274tSpU0hJSdHYR9/fb5U5c+aoP7OjoqIwceJEzJ49GxMmTNCqGxkZCZlMBplMhqCgIGzatAkjR47U+9smysLZDGyYt7c3XFxc9PqqLbe8vLw0nqvuEtdVnpaWZpRzvnz5Eg0bNoSTkxO+/vprvPvuu3BxccF///2HDz74AKmpqbk6rr29Pfr06YMff/wRz58/R+HChREREQF/f3+0bt1aXe/Ro0fYsWNHtuOjdH2d/CbV2Dd99OrVC1OmTEF6errOD0sVPz8/respUqRIjl/VnTlzBq1atUKTJk2wbNky9Rjg33//HTNnztTrtSxSpIjGc9XYbNW+jx49AgCMHz8e48eP13kM1Wv29OlTrevQdW05GTBgAKZNm4aVK1eiRo0aWLduHTIyMjT+EzB27FgsXLgQn332GRo3bgxPT0/Y2dlh8ODBOq9Z19AUXT7++GPs378fU6ZMQa1ateDu7g6JRIJ27drpPO7bXjvVmOcSJUpke07V66saUqNLQkKC1rCUN2X3ul+6dEn93NfXFz169MDPP/+MSZMm4dq1azh69Ch+/vnnHI/9utWrV6NChQqwt7eHr6+v1mtbqlQpgz6zNm7ciJcvX6J79+54/vy5urx79+4IDw9HVFSUxu/vm0qVKgUAiI6OVifD2Xn27BmEEDrfD6ohFm/+vuX1MzK7dsnIyMDLly/x4sULg2PS93e2Vq1aWscEoJEkAoCLi4vWzBGOjo56f+Zv3LgRX3/9NX755RdMmTIFbm5u6NKlC+bOnWvQ7z7w6lqzez2USiWePXumMQZa399vlXfeeUfj87tFixZ49uwZvv32WwwaNEjjffTee+/hu+++g0QigYuLC0qXLm2U2VRsDZNZGyaVStG8eXP8+eefuH//fo5/EFUcHR3Vg/Zfl1NClBuqD743z6VPInjgwAE8fPgQhw4dUvfGAtD4Q5ZbAwYMwDfffINff/0VPXr0wPbt2zFmzBiNnkdvb29UrlwZM2fO1HmM3IwbzI5CoUCvXr3g6ekJR0dHDBo0CMePH9f5YRgXF4fixYurn2dmZuLp06daf7he9+uvv0Imk2Hnzp0af4x+//13o12Dt7c3gKxxp7pu3AGAcuXKAcj6IxsXF6e1XVdZdkqUKIFWrVph/fr1+Pbbb7Fy5UqUKVNGo+dw7dq16Nu3L2bNmqWx75MnT1C4cGGtY+ozz3FiYiJ27tyJsLAwTJo0SV2enp6e67mcVePq7t+/n+24d9Xr++OPP2Y7zlSfXs/sXvc33z+jR4/GmjVr8Mcff2D37t0oXLiw1rcCOalQoUKO/5Fr3bo1fvzxR5w6dUqvcbOqG+HGjBmDMWPG6NyeUzLbunVrLF26FL///rtGu+mi+k9PbGys1jbVDVSq9jCW7NrFwcEBbm5usLe3N3pMqvpbtmxBQEBALqI2jLe3NxYsWIAFCxYgJiYG27dvx6RJkxAfH68ea6sv1fs1u9fDzs4Onp6eGuW5ncf8dZUrV4YQApcvX9ZIZj08PPTuuKDscZiBjZs8eTKEEBgyZIjOieLlcjl27Nihfh4YGIjLly9r1Dlw4ABevnxp1LhUiwO8ea7t27e/dV/VB8+bszPo6h16s8fhbSpUqIA6depg5cqVWL9+PdLT0zFgwACNOh06dMDVq1dRunRp1KxZU+vHmMlsWFgYjh49inXr1mHjxo24dOlStr2zb85bu2nTJmRmZuY4p6Nq8vrXk/XU1FSsWbPGKPEDWYlq2bJlcenSJZ2vV82aNVGoUCEAQNOmTbF//351zxCQldDruoklJ4MGDcKzZ88wdepUXLx4EQMGDND4gyWRSLTeP7t27cp2Mn59SCQSCCG0jvvLL7/kuJBDTlq1agWpVIrFixdnW6dBgwYoXLgwrl+/nu3rq09P0IYNGzRmsLh37x5OnDih9f6pUaMG6tevjzlz5mDdunXo37//W3t9DfHpp5/C1dUVoaGhGjcRqggh1PNi37hxAydPnkTXrl1x8OBBrR/VDUk5/Wf8/fffR6VKlRAeHo6rV6/qrLNnzx6kpKTA1dUVderUwdatWzU+U5RKJdauXYsSJUoYfZ7TrVu3avRwvnjxAjt27EDDhg0hlUpNElPr1q1hb2+P27dvZ/ueMpS+n8WlSpXCiBEj0LJly1wtVlGuXDkUL14c69ev13g/Jycn47ffflPPcGBsqsUasrs5lPKGPbM2rl69eli8eDFCQ0NRo0YNDBs2DBUrVoRcLseFCxewdOlShISEqMc79enTB1OmTMHUqVPRuHFjXL9+HT/99JPRJwv38/NDixYtEB4eDk9PTwQEBGD//v3YunXrW/etX78+PD09MXToUISFhUEmk2HdunUaX4eqVKpUCUDWGKe2bdtCKpWicuXKOf5xHzhwID755BM8fPgQ9evXV/caqsyYMQNRUVGoX78+Ro0ahXLlyiEtLQ13795FZGQklixZ8tZe8NTUVJ3TVwGvpiqKiopCeHg4pkyZgubNmwMAwsPDMX78eDRp0gRdunTR2G/r1q2wt7dHy5Ytce3aNUyZMgVVqlRB9+7ds42jffv2mD9/Pj7++GP83//9H54+fYp58+YZffGGn3/+GW3btkXr1q3Rv39/FC9eHAkJCbhx4wbOnz+PzZs3AwC+/PJLbN++Hc2aNcPUqVPh4uKChQsX6jXm83WdOnWCt7c3vvnmG0ilUvTr109je4cOHRAREYHy5cujcuXKOHfuHL755hu9vr3Ijru7Oxo1aoRvvvkG3t7eCAwMxOHDh7F8+XKdvb36CAwMxOeff46vvvoKqamp6NmzJzw8PHD9+nU8efJEPbn/jz/+iH79+iEhIQHdunWDj48PHj9+jEuXLuHx48c5JsMq8fHx6NKlC4YMGYLExESEhYXByckJkydP1qo7evRo9XRIoaGhubq27AQFBam/Galatap60QQga7aLFStWQAiBLl26qHtlJ06ciNq1a2sd68WLF9i/fz/Wrl2L0aNH6zyfVCrFtm3b0KpVK9SrVw/Dhg1D06ZN4erqinv37mHLli3YsWMHnj17BiDrd7Bly5Zo2rQpxo8fDwcHByxatAhXr17Fhg0bjNLL92Z8LVu2xNixY6FUKjFnzhwkJSWpFyAwRUyBgYGYMWMGvvjiC9y5cwdt2rSBp6cnHj16hDNnzsDV1VXj/PooVKgQAgIC8Mcff6B58+bw8vKCt7c3PD090bRpU3z88ccoX748ChUqhL/++gu7d+/O9pucnNjZ2WHu3Lno1asXOnTogE8++QTp6en45ptv8Pz5c8yePdvgY77p1q1b6s/vxMRE7Nu3D8uXL0fNmjXRsGHDPB+fdDDbrWdkUS5evCj69esnSpUqJRwcHNSTkk+dOlXEx8er66Wnp4uJEyeKkiVLCmdnZ9G4cWNx8eLFbGczeHMCdNXdoY8fP9Yo79evn3B1ddUoi42NFd26dRNeXl7Cw8ND9O7dW5w9e1av2QxOnDgh6tWrJ1xcXETRokXF4MGDxfnz57X2TU9PF4MHDxZFixYVEolE4474N69JJTExUTg7OwsAYtmyZTpfz8ePH4tRo0aJoKAgIZPJhJeXl6hRo4b44osvxMuXL3Xuo5LTbAYAhFwuFw8fPhQ+Pj6iWbNmGnfkKpVK0bFjR1G4cGH1dahen3PnzomOHTsKNzc3UahQIdGzZ0+Nu85V535zNoMVK1aIcuXKCUdHR/HOO++I8PBwsXz5cq3ZA7KbzeCbb77RukbouGv50qVLonv37sLHx0fIZDLh5+cnmjVrJpYsWaJR7/jx46Ju3brC0dFR+Pn5iQkTJoilS5fqNZvB6z799FMBQLRr105r27Nnz8SgQYOEj4+PcHFxEe+99544evSo1jXmtKCArtkM7t+/L7p27So8PT1FoUKFRJs2bcTVq1f1/v3JboaE1atXi1q1agknJyfh5uYmqlWrpvE+F0KIw4cPi/bt2wsvLy8hk8lE8eLFRfv27d+6GILqnGvWrBGjRo0SRYsWFY6OjqJhw4bi7NmzOvdJT08Xjo6Ook2bNjke+3XZXXN2bt++LUJDQ0WZMmWEo6OjcHZ2FsHBwWLs2LEiOjpaZGRkCB8fH1G1atVsj5GZmSlKlCghKlWq9NbzPX/+XHz11VeievXqws3NTchkMlGqVCnRu3dvcfz4cY26R48eFc2aNROurq7C2dlZ1K1bV+zYsUOv69X3M1L1+zVnzhwxffp0UaJECeHg4CCqVasm9uzZoxV/XmLK7n33+++/i6ZNmwp3d3fh6OgoAgICRLdu3cS+ffuyjfvN63zdvn37RLVq1YSjo6MAIPr16yfS0tLE0KFDReXKlYW7u7twdnYW5cqVE2FhYW9dqCKn38/ff/9d1KlTRzg5OQlXV1fRvHlzrXbMri3edr7Xf1xdXUVwcLAICwvTmi1F16IJlDsSId6Y+ZyICpRp06Zh+vTpePz4sdHH6xHpsmPHDnTq1Am7du3SuXQw5d3du3cRFBSEb775JtsbJ4lsBYcZEBGRUVy/fh337t3DuHHjULVqVfV0dEREpsQbwIiIyChCQ0PRqVMneHp6mmR8KBGRLhxmQERERERWiz2zRERERGS1mMwSERERkdViMktEREREVsvmZjNQKpV4+PAhChUqxJsTiIiIiCyQEAIvXrxAsWLFYGeXc9+rzSWzDx8+zHYNcyIiIiKyHP/9999bV1+0uWRWtcb7f//9B3d393w5p1wux969e9GqVSvIZLJ8OSflnaW3W3JGMop9WwwA8HDcQ7g6uJo5Istg6e1GurHdrBPbzXpZetslJSWhZMmS6rwtJzaXzKqGFri7u+drMuvi4gJ3d3eLfMOQbpbebtIMKeCU9djd3Z3J7P9YeruRbmw368R2s17W0nb6DAnlDWBEREREZLVsrmeWqKCQ2knRrmw79WMiIiJbxGSWyEo52Tth18e7zB0GERGRWXGYARERERFZLSazRERERGS1mMwSWankjGS4znKF6yxXJGckmzscIiIis+CYWSIrliJPMXcIREREZsWeWSIiIiKyWuyZJSIiIjKDjEwlVh6/g73XHiExNR0OUikyMhVIkSsggQTOMjs42L8qg8haRODN8tzUtZdK8PSZHWZfPQQ7u7fvn6kUcHeWoVWwH/o3CIKDveX0h5o1mT1y5Ai++eYbnDt3DrGxsdi2bRs6d+6c4z6HDx/G2LFjce3aNRQrVgwTJ07E0KFD8ydgIiIisliq5HDP1Tg8epEGZ/u8J33G3P/1us9SMvE0RW7mV8wOyMgwaI+z954j/M+/8X+NgjC5XbCJ4jKMWZPZ5ORkVKlSBQMGDEDXrl3fWj86Ohrt2rXDkCFDsHbtWhw/fhyhoaEoWrSoXvsTERGRabwtkTRmr6KuMstIDm2DAPDzkWgAsIiE1qzJbNu2bdG2bVu96y9ZsgSlSpXCggULAAAVKlTA2bNnMW/ePCazREREbzB1gpmcrkBqqh2mXjiAxLRMc18u5bNlR6MxrlV5sw85sKoxsydPnkSrVq00ylq3bo3ly5dDLpdDJpNp7ZOeno709HT186SkJACAXC6HXJ4//4NTnSe/zkfGYentpshUoFGpRurHcollxpnfLL3dSDdbbreMTCVWnbqLvdceIT4pHeK1pFFmL4E8U4mUDKXeZUDW/skZinxKMO2ATCaytkgpgIjjtzGgfqDRj23IZ4FVJbNxcXHw9fXVKPP19UVmZiaePHkCf39/rX3Cw8Mxffp0rfK9e/fCxcXFZLHqEhUVla/nI+Ow5HYb6zUWAHAw6qCZI7E8ltxulD1LbzelAG4+l+BMPHD/pQQZSkBmB9jbAZlKIEMBSCTaZYDu8lQlkCEkACRmvS6i3Dp64W/4Pr9u9OOmpOg/9aRVJbNA1v82XyeE0FmuMnnyZIwdO1b9PCkpCSVLlkSrVq3g7u5uukBfI5fLERUVhZYtW+rsPSbLxHazTmw365Qf7ZaXHlAASMtU4nlqJoRJoiOyTg2rlUc7E/TMqr5J14dVJbN+fn6Ii4vTKIuPj4e9vT2KFCmicx9HR0c4Ojpqlctksnz/Q2eOc1Lesd2sE9vNOuXUbq+P/4xLSjVovGdSaiZe/i8pJSLjsJMA/RuUhswEY2YN+fy2qmS2Xr162LFjh0bZ3r17UbNmTf7RIpuTnJGMwO8DAQB3R9+Fq4OreQMi0kN2NySlyzPx5Hn2c14yGSWyPEMaWsZ8s2ZNZl++fIl///1X/Tw6OhoXL16El5cXSpUqhcmTJ+PBgwdYvXo1AGDo0KH46aefMHbsWAwZMgQnT57E8uXLsWHDBnNdApFZPUl5Yu4QyEYplAKnbj/F0X/jcSnmOR6/TENKRs53xyenK/A8xxuS7JD4wrA5L4ko/0kAzjOrcvbsWTRt2lT9XDW2tV+/foiIiEBsbCxiYmLU24OCghAZGYlPP/0UCxcuRLFixfDDDz9wWi4iojxSKAVO3HqCLef/w38JKUiTK7KdkikhOQOPXmRw7CiRkXk6SVHEzSEfVwB7CXtHJ64AlhdNmjRR38ClS0REhFZZ48aNcf78eRNGRURUMOi7GlJapsCD52lMTqnAcXeUooSns0WuAGbuBFEulyMyMhLt2jWx+qGaVjVmlojI1ql6UDefi8GNuBcQQqlzPXeuhkTm9mYiaeoVwMydHJL5MJklIrIAOX3Nzx5UMhZjJ5hZK4CloYiHK5wdZfD3cELtoCLoVz+QiSTlGyazREQm9npv6vXYJKRkZGokBkxS6XWFnaRwdbQ3ytfeTg6mTTBffVXd0Oq/qibrxWSWyErZSexQs1hN9WMyjzcT1VS5QmNs6tMUOZ6ncKL9gszXTYYibo65/trc1dEeFfw90K1GCdQv4w2pHVcDIzIEk1kiK+Usc8ZfQ/4ydxgFXnZTUEkggUIpeFe/FTOkB1SVjLo4SOHj7owqJQqjQVlv1H2nCJNPIjNjMktENi+7u/45BZVlM/TrePaAEhVMTGYB4OVLYPZsYPNm4N49wMUFqFsXmDwZaNhQv2P88w/w1VfA/v3AkydAkSJAixZAWBgQEKBd/9KlrPpHjgCJiYC/P9CpEzBlClC0qHGvj8jG6Rqzyrv+LYPqhqSsFcBS4OKie85LJqNElB0ms8nJQKNGwIULr8rS04E//wT27AHWrQM++ijnY5w7BzRpkpUUq8TFAWvXAjt3Anv3atY/cABo3x5IS3tVdu8e8OOPwK5dwIkTgK9vni+NCrYUeQqCF2atvnJ9+HW4yFzMHJF5qYYDHP4nDgeu2OGHf44iNVOB1AwlnqdyzGp+8HSSwtlBmuPd8dndkFSQ5rwkovzFZParr14lst27Az/9BFy+nNVLmpICfPIJ0KoV4OWV/TGGD3+VyP7wAzBwYFYS/MknwPPnkA4dCkydmrVdLgf6989KZO3sgDVrgDZtgK+/Br77DrhzBxg3LisRJsqBEAL3Eu+pH9sK1ZCAvdceITE1HQ5SXcMB7ICXqWaMsmDwdZNBKpXovHkpNVMJVwf2lBKR+dl2MisEsGLFq+dz52Z9xd+8eVZiGxEBJCUBGzcCw4bpPkZiInD6dNZjZ2dgxAhAIgH+7/+Azz4Dnj+H3fnzcI+Ozqqzdy/w339Zjxs1Aj7+OOvxrFnA4sVZSe6mTcDChYCHh0kum8ga6EpaHz5Pw/O0THOHZrVyWg1JrlBCYidBBT8mp0RkXWw7mY2OBh4/znpcqJDm2NZKlV49PnUq+2Q29bXeH0n2H/yF//0364Eq8X3zHE5OQNmywJUrWb23588DTZvqeSFE1kvXzVccx2qY4h4OcJZJuRoSEdkk205mHz169bhwYc1tr/eKvl7vTb6+QLFiwMOHWcMSfvrp1TCD58/V1RwTE413TiIrpGs+1uQ0BXtac6D6mp89qERE2bPtZPZ1b445fP15Dj2ukEiA8HCgX7+s56NGZf28QWmv46XO7TmJLNybva0KBedjfZ2LvQSFXWQa41CZpBIR5Y5tJ7OvzxjwWi8qgKyxsLrq6dK3b1av6pw5wMWLWVN7vfde1njbgwcBACk+PsY9J5GFeDNxZW+r9opQdhI7+Lo7oXVFft1PRGRstp3MvvMO4OMDxMdnzUZw796rcbNXrryqV6fO24/1/vtZPyppaUDFigAA4eCAp/97jLp1X9W5elWzvmpcrUwGVK+eiwsiWyKRSBBcNFj9OD+8eVPWs2TbHNvq6SSFy/8m63eU2cNJJkVJLxd0rc4eVSKi/GbbySyQNb519uysxxMnZs0icOlS1gIKAODuDvTokfW4f39g1aqsxwcPZs0tC2QlpdevZy2w4OkJ3LoFfP551jRbAJQDByJDNR62ZUugZMmsGQ2OHAE2bMiamuurr17NO9u9e9Z5iXLgInPBtdBrJju+LSeur9/1zymoiIgsG5PZL7/MWhzhwoWsKbE2bXq1zc4O+PnnnOeYBbJ6VFUJ75saN4Zy9mzg0KGs5zJZ1pRfqkUTVFNzqQQFAd9+m9urIcoV1c1ZW87/h/8SUnDvaUqBTlxd7CXwdHXgXf9ERAUAk1lXV+Dw4azxrps2aS9n26jR249RpgzQrl3WeNknTwBHRyA4GOjdO2vhhDc1a5Y13ddXX2WdOzExa0YELmdL+eT1mQXO3E1AXFKGuUMyCdWKVHZ2HLNKRFRQMZkFsuaY/frrrJ+cRERk/bwpJCRrGdrsyHX0cFWpAmzZYkiURBpS5CmotawWAOCvIX/luJztm8nro6SCM7OAp5MURdwc1MMByvkVQgn5A4z+qA2cHB3MHR4REZkYk1kiKyWEwPXH19WPX/f6DAN3nybjWUqm1SevqqQ1TSHgWyj7Xla5XI7IyPsc10pEZCOYzBIVABmZSqw/dRun7zzFhZjnVj3eVXXzFcexEhGRPpjMElmpjEwlCqV3RiHl+6g244i5w8kVXzcZiro7w9/DCbWDiqBf/UAmrUREZBAms0RW4vWhA/88eoGXGUp4YbC5w9KLu2PWPKycj5WIiIyNySyRhXp9uqyTt58i/qV1zDjg7ihFqSKu7G0lIqJ8wWSWyIKoEtjvD/yDc/eeW/xNW55OUhR1d+LYViIiMhvbTWaTkwGpVLtcKgWcnDTrZcfODnB2fntduRx26emaZSkpgMgmVZFIsua6zU3d1FRAqcw+ZlfX3NVNSwMUCuPUdXHJihsA0tOBzEzj1HV2zmoTAMjI0D0lmiF15XJIVdcik+l3XCenV+8ruTyrfnYcHQF7+6zhA0dvYeNf9xGdkGaxCaynkx2KujtnJa7lfdC/up924pqeCqQj6/Vy+N+0WEpl1nstO4bUtbfPet2ArN+JlBTtOqp2S09/1W7Z1VUx5PfeFJ8Ruura2meEqt2SkwEPD+v4jFB5/ffeBJ8RBtfNzMx6LbLj4PDqd8OQugrFq5UqVV5vNxeXV7/Luuq+zpyfEbmpWxA/I1Rtl5KS9TunYimfETm9Fm8SNiYxMVEAEIlZzar9066d5g4uLrrrAUI0bqxZ19s727oJZcqIjIyMV3UDArI/bnCw5nGDg7OvGxCgWbdmzezrentr1m3cOPu6Li6addu1y77um2+jbt1yrvvy5au6/frlXDc+/lXd0NCc60ZHv6o7fnzOda9efVU3LCzHuvITJ17VnTs35+MePPiq7k8/6ayTKbETx0pVFuE/7BR1Z0aJgM92Wt7P+D9E3U+Wi64fzxE/1+oi0u3ss65d5cyZnF+HsLBXda9ezbnu+PGv6kZH51w3NPRV3fj4HOsq+vR5Vffly5yP262b5ns4p7om+owQNWtq1uVnRBYr+IwQZ868qmuEzwj1z86dr+quXJlz3U2bXtXdtCnnuitXvqq7c2fOdX/66VXdgwdzrmtlnxGiX79XdW34M0JpoZ8RiYAAIBITE8Xb2G7PLFE+y7Czx6RWI/BHSGMopDLgAZDVjWl+rjI7FHdQosKZg+h2ZT/qx1yGVOTwv20iIiILIRFCCHMHkZ+SkpLg4eGBxIcP4e7url3BBF8PyOVy7N67F226dIFM9XWNrX2F+DorGWYgl8uxZ88etH7/fchU7wkDv0JUpKXjxJ0ETNt7C7ef5vCVWz6TAAjycsZHtUqgf8MyWcMFDPla0IK/QlS3W/v2kLm55VhXrSB+hZhdXQv9jFC3W+vWkHGYQRYrGGag0W4cZmB4XTN+Rqjbrk2brN85FQv5jEh69gwexYohMTFRd772GtvtmXV11XzhcqpnyDF1kcuhVP3CqLz+JnsbQ+q+/kY3Zt3XfzGNWdfR8dWHiTHrOji8+vDLbV25HIrX//DoeVxLvInLw8keIcU9UKVEYTQo64267xTRnhZLKtX//W5IXTs709SVSHTXVbXb6++V7OpmxxLq2tpnhKrdXF1fJaeAZX9G5LWuTPYqUTRmXXv7V4mtMevq+r1/vd1ej8+SPyPyWhewjLp5/YxQtd2b2yzlM8KA18J2k1kiI1IoBU7dfopVJ6Ox70Y8lGbOYD2d7VHax40zDBARUYHHZJYoDzIylZj02yX8cekhFGYeYvqOtws+qlWKySsREdkUJrNEBlL1ws7dcwOX7ieZLQ4PZykepO3DS/v9iP/iKNyd3MwWCxERkbkwmSXSk7l7YSUAKhV3R8cqxdGvfiDkylS4hbcFAC4LS0RENovJLNFbZGQq0XvFCZyOfpbv57YDUCOgMEY1fxf1y3hrJK1y61jdloiIyKSYzBLpoFAKHL/1BLMuSDD65L58PXcxdyd0rlY8+1kHiIiISI3JLNFrFEqB76P+wcLD//5vKIGOJY9NoLS3C3rk4uYtbxdvE0ZFRERk+ZjMEiFrKMHkrZex9fyDfJsXtrS3K6Z1qqg1fEBfrg6ueDzhsQkiIyIish5MZsmmZWQq0Wf5qXwbD+tbyBGD3gvi9FlERERGwmSWbFJ+JrESAF2qFsPsblWYwBIRERkZk1myGaplZqftvIrbj3NYg9sIJABqBRbGyGbasxAYS6o8FW3XZU3N9WevP+EsM2BZQSIiogKCySzZhB2XHmLspouQK0w7Ijav42ANoRRKHL53WP2YiIjIFjGZpQJLtVLX579fxr2nqSY7j50E6FKtGMI/4DACIiKi/MZklgqkHZceYsKWS0iTm67HMj97YYmIiEi3XCWza9aswZIlSxAdHY2TJ08iICAACxYsQFBQEN5//31jx0ikN4VS4MPFJ3D+v+cmO8cHvJmLiIjIYhj813jx4sUYO3Ys2rVrh+fPn0OhUAAAChcujAULFhg7PiK9KJQCC6L+QenPI02SyEoAjGpaGrdntcP8j6oxkSUiIrIQBv9F/vHHH7Fs2TJ88cUXkEpfrY5Us2ZNXLlyxajBEeljx6WHKD/lTyzYf8vox5ZCYHijIPw7qx3Gti7P4QREREQWxuBhBtHR0ahWrZpWuaOjI5KTk40SFJE+TDmkoGoJd3zaoiwS/j6NDi3LWmwS6yJzMXcIREREZmVwMhsUFISLFy8iICBAo/zPP/9EcHCw0QIjysmOSw8xasMFoy89WyfQE2sG14WDvR3kcjkibxr5BEbk6uCK5M/5H0giIrJtBiezEyZMwPDhw5GWlgYhBM6cOYMNGzYgPDwcv/zyiyliJFLLyFSiww9H8U/8S6Melzd1ERERWSeDk9kBAwYgMzMTEydOREpKCj7++GMUL14c33//PT766CNTxEgEhVJg5PrziLwaZ9TjjmpaGqNblrPYYQRERESUs1xNzTVkyBAMGTIET548gVKphI+Pj7HjIlLbcekhRv96AUojjiloF+KLHz+uYdVJbFpmGrpu6goA+K37b3CydzJzRERERPkvVzeAZWZmomzZsvD29laX37p1CzKZDIGBgcaMj2yYQinQfckJnIt5brRjdq1ecFbqUigViLwVqX5MRERkiwz+i96/f3+cOHFCq/z06dPo37+/MWIiwo5LD1H2i0ijJbLVS7rj9qx2+LY754glIiIqSAzumb1w4QIaNGigVV63bl2MGDHCKEGRbRsU8Rf2/x1vlGPJ7IDvuldFh6rFjXI8IiIisiwGJ7MSiQQvXrzQKk9MTFSvBkaUGwqlQPN5B3E3ITXPx5IAGN28DEY2f9eqx8USERFRzgz+vrVhw4YIDw/XSFwVCgXCw8Px3nvvGRzAokWLEBQUBCcnJ9SoUQNHjx7Nsf66detQpUoVuLi4wN/fHwMGDMDTp08NPi9ZlsjLsSjzeaRREtnqJd3x76x2GMNZCoiIiAo8g3tm586di0aNGqFcuXJo2LAhAODo0aNISkrCgQMHDDrWxo0bMWbMGCxatAgNGjTAzz//jLZt2+L69esoVaqUVv1jx46hb9+++O6779CxY0c8ePAAQ4cOxeDBg7Ft2zZDL4UsgDGn3LKTAD/04JACIiIiW2Jwz2xwcDAuX76M7t27Iz4+Hi9evEDfvn3x999/IyQkxKBjzZ8/H4MGDcLgwYNRoUIFLFiwACVLlsTixYt11j916hQCAwMxatQoBAUF4b333sMnn3yCs2fPGnoZZAEiL8ei3BeReU5kJciaL/bWzHZMZImIiGxMruaZLVasGGbNmpWnE2dkZODcuXOYNGmSRnmrVq10zpYAAPXr18cXX3yByMhItG3bFvHx8diyZQvat2+f7XnS09ORnp6ufp6UlAQAkMvlkMvleboGfanOk1/nswazd9/E8uP38nycqiXc8euQOpDaSaBUZMKYM1RZers5SByQ8XmG+rmlxpnfLL3dSDe2m3Viu1kvS287Q+KSCCEMnor++fPnOHPmDOLj46FUKjW29e3bV69jPHz4EMWLF8fx48dRv359dfmsWbOwatUq3Lx5U+d+W7ZswYABA5CWlobMzEx06tQJW7ZsgUwm01l/2rRpmD59ulb5+vXr4eLiolesZFzboiU4FGeHrD7V3JFAoE8ZJWoUNeJKCkRERGQRVKvMJiYmwt3dPce6BvfM7tixA7169UJycjIKFSoEieRVQiKRSPROZl/f53VCCK0ylevXr2PUqFGYOnUqWrdujdjYWEyYMAFDhw7F8uXLde4zefJkjB07Vv08KSkJJUuWRKtWrd764hiLXC5HVFQUWrZsmW3SbQsUSoExv17EobjHeTpO6+Ci+L5HVZPf3MV2s05sN+vEdrNObDfrZeltp/omXR8GJ7Pjxo3DwIEDMWvWrDz1bHp7e0MqlSIuTnO8ZHx8PHx9fXXuEx4ejgYNGmDChAkAgMqVK8PV1RUNGzbE119/DX9/f619HB0d4ejoqFUuk8nyvfHMcU5LEXk5FqM2nEdmHjpS7STATz2roV3lYsYLTA+W2m5pmWnos60PAGBNlzVczvYNltpulDO2m3Viu1kvS207Q2Iy+AawBw8eYNSoUXn+it7BwQE1atRAVFSURnlUVJTGsIPXpaSkwM5OM2SpVAogq0eXLNPMXdcRuj5viWyglxNuzWyX74msJVMoFdhyfQu2XN/C5WyJiMhmGZzMtm7d2mizB4wdOxa//PILVqxYgRs3buDTTz9FTEwMhg4dCiBriMDrwxY6duyIrVu3YvHixbhz5w6OHz+OUaNGoXbt2ihWjEmOJfpq5zUsOxqdp2M0L++NQxObc85YIiIi0mLwMIP27dtjwoQJuH79OipVqqTVDdypUye9j9WjRw88ffoUM2bMQGxsLEJCQhAZGYmAgAAAQGxsLGJiYtT1+/fvjxcvXuCnn37CuHHjULhwYTRr1gxz5swx9DIoH0zfcQ0rj9/N9f6cN5aIiIjexuBkdsiQIQCAGTNmaG2TSCQGL2kbGhqK0NBQndsiIiK0ykaOHImRI0cadA7KfwNXnsGBm7m/0at6SQ9sHtaAvbFERESUI4OT2Ten4iJ6U4cfjuDqwxe53n9AgwCEdTRsAQ4iIiKyTblaNIEoO+2/P4xrsS9zvf+QhoH4on1FI0ZEREREBVmuktnk5GQcPnwYMTExyMjI0Ng2atQoowRG1qdDHhJZezvgh4/yf9otIiIism4GJ7MXLlxAu3btkJKSguTkZHh5eeHJkydwcXGBj48Pk1kbNXDlaVzNZSLL8bG54yJzwcvJL9WPiYiIbJHBU3N9+umn6NixIxISEuDs7IxTp07h3r17qFGjBubNm2eKGMnChW2/igM3n+Rq3wENArB1+HtMZHNBIpHA1cEVrg6u2a6aR0REVNAZnMxevHgR48aNg1QqhVQqRXp6OkqWLIm5c+fi888/N0WMZMEGrjyDVSfu5WrfIQ0DeaMXERER5YnByaxMJlP3Avn6+qrngfXw8NCYE5YKvo4/HMnV9Fv2dsCij6vxRq88Ss9MR//f+6P/7/2Rnplu7nCIiIjMwuAxs9WqVcPZs2fx7rvvomnTppg6dSqePHmCNWvWoFKlSqaIkSzQwJWncSUX028FFnHG/nFNOazACDKVmVh1aRUAYGG7hXCEo5kjIiIiyn8G98zOmjUL/v7+AICvvvoKRYoUwbBhwxAfH4+lS5caPUCyPNN35G6MbEV/Nxya0IyJLBERERmNwT2zNWvWVD8uWrQoIiMjjRoQWbavdl7DyuOGj5EN9HLCrtGNTRARERER2TKDe2bJdn218xqWH7tr8H72dsD+8c2MHxARERHZPL16ZqtXr479+/fD09MT1apVy3EaoPPnzxstOLIcM3flLpEFgJ8+rs6hBURERGQSeiWz77//Phwds24u6dy5synjIQsUefkhlh29a/B+DlIJfuhZDW1C/I0fFBERERH0TGbDwsIAAAqFAk2aNEHlypXh6elp0sDIMiiUAqN+vWDwfn7uDjg+qQV7ZImIiMikDLoBTCqVonXr1rhx4waTWRvx4eLjyFQato8UYCKbD1xkLogfH69+TEREZIsMvgGsUqVKuHPnjiliIQvz1c6rOP9fosH7/cgxsvlCIpGgqGtRFHUtyuVsiYjIZhmczM6cORPjx4/Hzp07ERsbi6SkJI0fKhgiLz/E8mOGT8H1SaMgtKvMMbJERESUPwyeZ7ZNmzYAgE6dOmn0BgkhIJFIoFAojBcdmYVCKfDpxosG7/fTR9XQoWox4wdEOqVnpmPsnrEAgPmt58PRniuAERGR7TE4mT148KAp4iAL8uHi40hXCIP2YSKb/zKVmVh0dhEAYG7LuVzOloiIbJLByWzjxlzFqSAbvOqMweNkB70XxESWiIiIzMLgZFYlJSUFMTExyMjI0CivXLlynoMi89h58QH23Xhs0D7VS3pgSodgE0VERERElDODk9nHjx9jwIAB+PPPP3Vu55hZ66RQCow2cJysvQTYPKyBaQIiIiIi0oPBsxmMGTMGz549w6lTp+Ds7Izdu3dj1apVKFu2LLZv326KGCkffLjkOAwcJosfenIKLiIiIjIvg3tmDxw4gD/++AO1atWCnZ0dAgIC0LJlS7i7uyM8PBzt27c3RZxkQjsvPsD5GMPGyQ5pyCm4iIiIyPwM7plNTk6Gj48PAMDLywuPH2eNsaxUqRLOnz9v3OjI5BRKgU83XTJonwENAvFFe46TJSIiIvMzuGe2XLlyuHnzJgIDA1G1alX8/PPPCAwMxJIlS+Dvz546a/Ph4uOQK/UfX1C9pAfCOlY0YUSkL2eZM6JHR6sfExER2SKDk9kxY8YgNjYWABAWFobWrVtj3bp1cHBwQEREhLHjIxMydBouKW/4sih2EjsEFg40dxhERERmpXcy27lzZwwePBg9e/aEnV3W6IRq1arh7t27+Pvvv1GqVCl4e3ubLFAyrtxMw7Xgo2q84YuIiIgsit5jZlNTU9G5c2eUKFECn3/+OW7dugUAcHFxQfXq1ZnIWpHcjJOtUaowOlbhwgiWJEORgQl7J2DC3gnIUGS8fQciIqICSO9kds+ePbh79y6GDRuGTZs2oXz58mjUqBFWr16N1NRUU8ZIRvbj/n8MGidrLwE2Da1vwogoN+QKOeadnId5J+dBrpCbOxwiIiKzMGg2gxIlSmDKlCn4999/sW/fPgQEBCA0NBR+fn745JNPcPr0aVPFSUaiUAosPHjboH04nywRERFZKoOn5lJp2rQp1qxZg9jYWMydOxdbtmxBgwa8OcjSGdorO+g9zidLRERElsvg2Qxed+fOHURERCAiIgKJiYlo0aKFseIiEzC0V7Z6SQ9M6cD5ZImIiMhyGdwzm5qaitWrV6Np06YoW7Ys1qxZg8GDByM6Ohq7d+82RYxkJKM2nNO7V9YOnIaLiIiILJ/ePbMnTpzAypUrsWnTJmRkZKBz587Ys2cPe2OtROTlh9h15ZHe9Uc1L8txskRERGTx9E5m33vvPVSpUgUzZ85Er1694Onpacq4yIgUSoGxBkzFJbOTYGTzsiaMiIiIiMg49E5mz549i+rVq5syFjKRH/f/g7RMpd71hzctw15ZK+Asc8bVYVfVj4mIiGyR3sksE1nrZOhNX072duyVtRJ2EjtU9Klo7jCIiIjMKtdTc5F1MHQqrvndq7JXloiIiKxGnqbmIstmaK9sh0r+nFPWimQoMjDr6CwAwOcNP4eD1MHMEREREeU/JrMFmCG9svYS4Pue1UwcERmTXCHH9MPTAQAT6k9gMktERDaJwwwKKEN7ZUc041RcREREZH306pmtVq0aJBL9Ep3z58/nKSAyDkN6ZTkVFxEREVkrvZLZzp07qx+npaVh0aJFCA4ORr169QAAp06dwrVr1xAaGmqSIMkwhvbKciouIiIislZ6JbNhYWHqx4MHD8aoUaPw1VdfadX577//jBsd5Yohy9ayV5aIiIismcFjZjdv3oy+fftqlffu3Ru//fabUYKi3DN02Vr2yhIREZE1MziZdXZ2xrFjx7TKjx07BicnJ6MERbmjUApM+O2y3vXZK0tERETWzuCpucaMGYNhw4bh3LlzqFu3LoCsMbMrVqzA1KlTjR4g6e/UnadITlfoXZ+9stbNyd4JZwafUT8mIiKyRQYns5MmTcI777yD77//HuvXrwcAVKhQAREREejevbvRAyT9rT11T++6XLbW+kntpKhVvJa5wyAiIjKrXC2a0L17dyauFkahFNh3PU7v+ly2loiIiAqCXC2a8Pz5c/zyyy/4/PPPkZCQACBrftkHDx4YNTjSX9a8svrV5bK1BUOGIgPfHP8G3xz/BhmKDHOHQ0REZBYG98xevnwZLVq0gIeHB+7evYvBgwfDy8sL27Ztw71797B69WpTxEk5MGReWS5bW3DIFXJM3DcRABBaK5TL2RIRkU0yuGd27Nix6N+/P27duqUxe0Hbtm1x5MgRowZH+jFkta+WFX05vICIiIgKDIOT2b/++guffPKJVnnx4sURF6f/mE0yDoVSYNmxaL3r964TaLpgiIiIiPKZwcmsk5MTkpKStMpv3ryJokWLGiUo0t+Z6AS9p+NysrdD3dJFTBwRERERUf4xOJl9//33MWPGDMjlcgCARCJBTEwMJk2ahK5duxo9QMrZ0iP/6l13aOPSHGJAREREBYrByey8efPw+PFj+Pj4IDU1FY0bN0aZMmVQqFAhzJw50+AAFi1ahKCgIDg5OaFGjRo4evRojvXT09PxxRdfICAgAI6OjihdujRWrFhh8HkLgsjLD3Hw5hO96jpIudoXERERFTwGz2bg7u6OY8eO4cCBAzh//jyUSiWqV6+OFi1aGHzyjRs3YsyYMVi0aBEaNGiAn3/+GW3btsX169dRqlQpnft0794djx49wvLly1GmTBnEx8cjMzPT4HNbO4VS4Ms/rupdv1edUuyVJSIiogInV4smAECzZs3QrFmzPJ18/vz5GDRoEAYPHgwAWLBgAfbs2YPFixcjPDxcq/7u3btx+PBh3LlzB15eXgCAwMDAPMVgrc5EJyAhWa53/VYVOa9sQeNk74SD/Q6qHxMREdmiXCWz+/fvx/79+xEfHw+lUnOmfn2/8s/IyMC5c+cwadIkjfJWrVrhxIkTOvfZvn07atasiblz52LNmjVwdXVFp06d8NVXX8HZ2Tk3l2K19l6L1btuEVcH1A7yMmE0ZA5SOymaBDYxdxhERERmZXAyO336dMyYMQM1a9aEv78/JJLcfXX95MkTKBQK+Pr6apT7+vpmO8XXnTt3cOzYMTg5OWHbtm148uQJQkNDkZCQkG0SnZ6ejvT0dPVz1UwMcrlcfRObqanOY6zzKZQCm8/d17t+WIfyUCoyodRv0gP6H2O3G+UPtpt1YrtZJ7ab9bL0tjMkLoOT2SVLliAiIgJ9+vQxdFed3kyGhRDZJshKpRISiQTr1q2Dh4cHgKyhCt26dcPChQt19s6Gh4dj+vTpWuV79+6Fi4uLEa5Af1FRUUY5zq1ECV6mS/WqG1xYARFzHpExRjm1TTJWuxlbpsjE3id7AQCtvFvBXpLrUUMFkqW2G+WM7Wad2G7Wy1LbLiUlRe+6Bv/1y8jIQP369Q3dTYu3tzekUqlWL2x8fLxWb62Kv78/ihcvrk5kAaBChQoQQuD+/fsoW1b7bv3Jkydj7Nix6udJSUkoWbIkWrVqBXd39zxfhz7kcjmioqLQsmVLyGSyPB/v6103APynV93PP6iDOhxikCvGbjdjS85IRrd53QAAc3rOgauDq5kjsgyW3m6kG9vNOrHdrJelt52uNQ2yY3AyO3jwYKxfvx5TpkwxdFcNDg4OqFGjBqKiotClSxd1eVRUFN5//32d+zRo0ACbN2/Gy5cv4ebmBgD4559/YGdnhxIlSujcx9HREY6OjlrlMpks3xvPGOdUKAW2XnyoV113J3vUK+PDWQzyyBzvFX3IxKuYLDVGc+JrYp3YbtaJ7Wa9LLXtDInJ4GQ2LS0NS5cuxb59+1C5cmWtk82fP1/vY40dOxZ9+vRBzZo1Ua9ePSxduhQxMTEYOnQogKxe1QcPHmD16tUAgI8//hhfffUVBgwYgOnTp+PJkyeYMGECBg4caDM3gJ2JTsCLNP0Gv3atXpyJLBERERVoBiezly9fRtWqVQEAV69qznNq6M1gPXr0wNOnTzFjxgzExsYiJCQEkZGRCAgIAADExsYiJubVYE83NzdERUVh5MiRqFmzJooUKYLu3bvj66+/NvQyrJYhsxhwOi4iIiIq6AxOZg8ePGjUAEJDQxEaGqpzW0REhFZZ+fLlLXawsqkplAJbzus3i4G7kz2n4yIiIqICz+DlbMl8OMSAiIiISJNePbMffPABIiIi4O7ujg8++CDHulu3bjVKYKSNQwyIiIiINOmVzHp4eKjHw74+LRblHw4xoDc52jtiZ8+d6sdERES2SK9kduXKlTofU/7hEAN6k72dPdq/297cYRAREZkVx8xaibikNL3rcogBERER2YpcrX+5ZcsWbNq0CTExMcjIyNDYdv78eaMERpqO33qsVz0OMbAdcoUc666sAwD0qtQLMqnlTXpNRERkagb3zP7www8YMGAAfHx8cOHCBdSuXRtFihTBnTt30LZtW1PEaPMUSoFdV/S7+YtDDGxHhiIDA/4YgAF/DECGIuPtOxARERVABiezixYtwtKlS/HTTz/BwcEBEydORFRUFEaNGoXExERTxGjzfjpwC6lypV51OcSAiIiIbInByWxMTAzq168PAHB2dsaLFy8AAH369MGGDRuMGx1BoRT4+cgdveoWdpFxiAERERHZFIOTWT8/Pzx9+hQAEBAQgFOnTgEAoqOjIYQwbnSEU3eeIiVDv1kMBtQP4hADIiIisikGJ7PNmjXDjh07AACDBg3Cp59+ipYtW6JHjx7o0qWL0QO0dWtP3dOrnpO9HUY0K2PiaIiIiIgsi8GzGSxduhRKZdb4zaFDh8LLywvHjh1Dx44dMXToUKMHaMsUSoEj/+g3i0HT8kXZK0tEREQ2x+Bk1s7ODnZ2rzp0u3fvju7duxs1KMpyJjoByXoOMehdJ9C0wRARERFZIL2S2cuXL+t9wMqVK+c6GNKk70IJLg5S1C1dxMTRkKVxtHfEpm6b1I+JiIhskV7JbNWqVSGRSN56g5dEIoFCoV9PIr2dvgsltAvx4xADG2RvZ48PK35o7jCIiIjMSq9kNjo62tRx0BsUSoGo64/0qtugjLeJoyEiIiKyTHolswEBAaaOg95wJjoBiWmZetX183A2cTRkiTKVmdh2YxsAoEuFLrC3y9Xq1ERERFYtV3/9bt68iR9//BE3btyARCJB+fLlMXLkSJQrV87Y8dksfcfLFnbmQgm2Kj0zHd23ZN18+XLyS9g7MJklIiLbY/A8s1u2bEFISAjOnTuHKlWqoHLlyjh//jxCQkKwefNmU8Rok/QdL9uigg/HyxIREZHNMrgrZ+LEiZg8eTJmzJihUR4WFobPPvsMH37IG1LyiuNliYiIiPRjcM9sXFwc+vbtq1Xeu3dvxMXFGSUoW8fxskRERET6MTiZbdKkCY4ePapVfuzYMTRs2NAoQdm6+Bd6jpd14XhZIiIism0GDzPo1KkTPvvsM5w7dw5169YFAJw6dQqbN2/G9OnTsX37do26ZDhvN/0mwO9fL5DjZYmIiMimGZzMhoaGAgAWLVqERYsW6dwGcAGFvDgT/VSverUC2StLREREts3gZFapVJoiDvofhVJg2VH9Fql4kpxu4mjIkjlIHbDy/ZXqx0RERLbIqBNTpqSkwMXFxZiHtDk/HbiFlAz9erR9CjmZOBqyZDKpDP2r9jd3GERERGaVqxvA7t+/r1V++vRpVK1a1Rgx2SyFUmDl8bt61eXNX0RERES5SGbd3d1RuXJl/PrrrwCyhh1MmzYNjRo14g1feXQmOgHPU+V61R1QP4g3f9m4TGUmdv2zC7v+2YVMpX5TuRERERU0Bg8z2L59O5YsWYLBgwdj+/btuHv3LmJiYrBr1y60aNHCFDHaDH2n5HJxkGJEszImjoYsXXpmOjps6ACAy9kSEZHtytVfv6FDh+LevXuYM2cO7O3tcejQIdSvX9/Ysdkcfafk+r+G77BXloiIiAi5GGbw7NkzdO3aFYsXL8bPP/+M7t27o1WrVlrTdJHhOCUXERERkWEM7pkNCQlBUFAQLly4gKCgIAwZMgQbN25EaGgodu3ahV27dpkizgJPoRRYdeKeXnU5JRcRERFRFoN7ZocOHYojR44gKChIXdajRw9cunQJGRkZRg3Olhhy8xen5CIiIiLKYnDP7JQpU3SWlyhRAlFRUXkOyFbFJel381dhZ07JRURERKSid8/s3LlzkZqaqn5+5MgRpKe/+rr7xYsXGsvZkmESXuo3dKBFBR/e/EVERET0P3ons5MnT8aLFy/Uzzt06IAHDx6on6ekpODnn382bnQ25P6zFL3qNSjjbeJIyFo4SB3wU9uf8FPbn7icLRER2Sy9hxkIIXJ8TrmnUAr8cemhXnX9PJxNHA1ZC5lUhuG1h5s7DCIiIrMy+AYwMr4z0QlISH77zV9FXB04XpaIiIjoNVwyyALou/LX+1WLcbwsqSmUChyNOQoAaFiqIaR2UjNHRERElP8MSmZ/+eUXuLm5AQAyMzMREREBb++sMZyvj6clw+i78lfzCr4mjoSsSVpmGpquagogazlbVwdXM0dERESU//ROZkuVKoVly5apn/v5+WHNmjVadSgX9B1+zGHKRERERBr0Tmbv3r1rwjBs24G/H+lVjyt/EREREWniDWBmplAKbLv44O0VwZW/iIiIiN7EZNbMOJMBERERUe4xmTUzzmRARERElHtMZs2MMxkQERER5R7nmTU3zmRAuSSTyjC3xVz1YyIiIluUq2T29u3bWLlyJW7fvo3vv/8ePj4+2L17N0qWLImKFSsaO8YCjTMZUG45SB0wocEEc4dBRERkVgYPMzh8+DAqVaqE06dPY+vWrXj58iUA4PLlywgLCzN6gAUZZzIgIiIiyhuDk9lJkybh66+/RlRUFBwcHNTlTZs2xcmTJ40aXEHHmQwoLxRKBf568Bf+evAXFEqFucMhIiIyC4OHGVy5cgXr16/XKi9atCiePn1qlKBsBWcyoLxIy0xD7V9qA+BytkREZLsM7pktXLgwYmNjtcovXLiA4sWLGyUoW8GZDIiIiIjyxuBk9uOPP8Znn32GuLg4SCQSKJVKHD9+HOPHj0ffvn1NEWPBxZkMiIiIiPLE4GR25syZKFWqFIoXL46XL18iODgYjRo1Qv369fHll1+aIsYCizMZEBEREeWNwWNmZTIZ1q1bhxkzZuDChQtQKpWoVq0aypYta4r4CizOZEBERESUdwYns4cPH0bjxo1RunRplC5d2hQx2QTOZEBERESUdwYPM2jZsiVKlSqFSZMm4erVq3kOYNGiRQgKCoKTkxNq1KiBo0eP6rXf8ePHYW9vj6pVq+Y5BnPYdz1Or3qcyYCIiIgoewYnsw8fPsTEiRNx9OhRVK5cGZUrV8bcuXNx//59g0++ceNGjBkzBl988QUuXLiAhg0bom3btoiJiclxv8TERPTt2xfNmzc3+JyWwJAhBi2D/UwcDVkrmVSGsMZhCGscxuVsiYjIZhmczHp7e2PEiBE4fvw4bt++jR49emD16tUIDAxEs2bNDDrW/PnzMWjQIAwePBgVKlTAggULULJkSSxevDjH/T755BN8/PHHqFevnqHhWwQOMSBjcJA6YFqTaZjWZBocpA5v34GIiKgAMjiZfV1QUBAmTZqE2bNno1KlSjh8+LDe+2ZkZODcuXNo1aqVRnmrVq1w4sSJbPdbuXIlbt++bdVL53KxBCIiIiLjMPgGMJXjx49j3bp12LJlC9LS0tCpUyfMmjVL7/2fPHkChUIBX1/NBQF8fX0RF6d7POmtW7cwadIkHD16FPb2+oWenp6O9PRXU1slJSUBAORyOeTyt/eOGoPqPKp/i7joF3uzct75FiNpe7PdLI1SKHHjyQ0AQAXvCrCT5On/pgWGpbcb6cZ2s05sN+tl6W1nSFwGJ7Off/45NmzYgIcPH6JFixZYsGABOnfuDBcXF0MPBQCQSDR7HoUQWmUAoFAo8PHHH2P69Ol499139T5+eHg4pk+frlW+d+/eXMecW1FRUQCATCUggfR/ayHo6nkVkAB4dO0UIm/kX3ykm6rdLE2aIg0fXfkIAPBrpV/hJOUUbq+z1HajnLHdrBPbzXpZatulpKToXVcihDBofan69eujV69e6NGjB7y9vQ0OTiUjIwMuLi7YvHkzunTpoi4fPXo0Ll68qDVk4fnz5/D09IRUKlWXKZVKCCEglUqxd+9enWN2dfXMlixZEk+ePIG7u3uu4zeEXC5HVFQUWrZsCZlMhtPRCei94uxb91s7sCbqcMys2bzZbpYmOSMZnvM8AQDPxj+Dq4OrmSOyDJbebqQb2806sd2sl6W3XVJSEry9vZGYmPjWfM3gntmcxrMawsHBATVq1EBUVJRGMhsVFYX3339fq767uzuuXLmiUbZo0SIcOHAAW7ZsQVBQkM7zODo6wtHRUatcJpPle+Opznnw5hO96j9NybTIN5itMcd7RR8y8SomS43RnPiaWCe2m3Viu1kvS207Q2LSK5ndvn072rZtC5lMhu3bt+dYt1OnTnqffOzYsejTpw9q1qyJevXqYenSpYiJicHQoUMBAJMnT8aDBw+wevVq2NnZISQkRGN/Hx8fODk5aZVbMq78RURERGQ8eiWznTt3RlxcHHx8fNC5c+ds60kkEigUCr1P3qNHDzx9+hQzZsxAbGwsQkJCEBkZiYCAAABAbGzsW+ectTaclouIiIjIePRKZpVKpc7HxhAaGorQ0FCd2yIiInLcd9q0aZg2bZpR4zE1TstFREREZDwGz+WzevVqjRuqVDIyMrB69WqjBFWQ6Tt0gCt/EREREb2dwcnsgAEDkJiYqFX+4sULDBgwwChBFWQ1Ajzxtg5XO0lWPaKcyKQyjK83HuPrjedytkREZLMMns0gu3lg79+/Dw8PD6MEVZCdu/cMyrdMhqYUWfXqlS6SP0GRVXKQOuCbVt+YOwwiIiKz0juZrVatGiQSCSQSCZo3b66xApdCoUB0dDTatGljkiALEn3HzOpbj4iIiMiW6Z3MqmYxuHjxIlq3bg03Nzf1NgcHBwQGBqJr165GD7Cg0XfMLKflordRCiViErNm+yjlUYrL2RIRkU3SO5kNCwsDAAQGBqJHjx5wcmKylRvPkrVvnnuTv4cTp+Wit0qVpyLo+6zFQl5OfskVwIiIyCYZPGa2X79+pojDJiiUAl/tuvHWelPaB3NaLiIiIiI9GJzMKhQKfPfdd9i0aRNiYmKQkZGhsT0hIcFowRU0Z+89Q2zi28fCero65EM0RERERNbP4EF206dPx/z589G9e3ckJiZi7Nix+OCDD2BnZ2d1Cxjkt/gXbx9ikFWPN38RERER6cPgZHbdunVYtmwZxo8fD3t7e/Ts2RO//PILpk6dilOnTpkixgLDp5CjnvU4HpmIiIhIHwYns3FxcahUqRIAwM3NTb2AQocOHbBr1y7jRlfAVCtZmAsmEBERERmRwclsiRIlEBsbCwAoU6YM9u7dCwD466+/4OioX8+jrbrw33O9F0wgIiIiorcz+AawLl26YP/+/ahTpw5Gjx6Nnj17Yvny5YiJicGnn35qihgLDI6ZJWOyt7NHaM1Q9WMiIiJbZPBfwNmzZ6sfd+vWDSVKlMCJEydQpkwZdOrUyajBFTQcM0vG5GjviIXtF5o7DCIiIrPKc3dO3bp1UbduXWPEUuAlvMx4ax0umEBERESkP72S2e3bt+t9QPbO6qYUwJzdN99ajwsmkL6EEHiS8gQA4O3iDYmE7xsiIrI9eiWznTt31utgEokECoUiL/EUWLeTJIhLevuYWS6YQPpKkafAZ54PAC5nS0REtkuvZFapVJo6jgIvSa5fPd78RURERKQ/g6fmotxxl+lXjzd/EREREenP4BvAZsyYkeP2qVOn5jqYgqy0u4Cfu2OOQw148xcRERGRYQxOZrdt26bxXC6XIzo6Gvb29ihdujST2WzYSYAOlfzwy/F72dbpVMWfN38RERERGcDgZPbChQtaZUlJSejfvz+6dOlilKAKIqUAdl6Jy7HO9kuxmNimAhNaIiIiIj0ZZcysu7s7ZsyYgSlTphjjcAWSPrMZxCam4Ux0Qj5FRERERGT9jLYG5vPnz5GYmGiswxU4nM2AjM3ezh79qvRTPyYiIrJFBv8F/OGHHzSeCyEQGxuLNWvWoE2bNkYLrKDhbAZkbI72jojoHGHuMIiIiMzK4GT2u+++03huZ2eHokWLol+/fpg8ebLRAitoXurRM8vZDIiIiIgMY3AyGx0dbYo4CjSFUuD3e28fnsylbMkQQgikyFMAAC4yFy5nS0RENomLJuSDs/ee4XnG2xMNLmVLhkiRp8At3A1u4W7qpJaIiMjWGNwzm5aWhh9//BEHDx5EfHy81lK358+fN1pwBUX8i5xnMXhVjzd/ERERERnC4GR24MCBiIqKQrdu3VC7dm1+takHn0KOetbjzV9EREREhjA4md21axciIyPRoEEDU8RTINUM8ISLvUBKpu7EXwLAjzd/ERERERnM4DGzxYsXR6FChUwRS4G170Y8UjKz3y4AhHXkzV9EREREhjI4mf3222/x2Wef4d69e6aIp8BRKAW+jvw7xzqFXWRoGeyXTxERERERFRwGDzOoWbMm0tLS8M4778DFxQUymeZqAAkJXI71dWeiE/63jG32va7PU+Q4E52AeqWL5F9gRERERAWAwclsz5498eDBA8yaNQu+vr68Aewt9J2hgDMZkKGkdlJ0C+6mfkxERGSLDE5mT5w4gZMnT6JKlSqmiKfA0XeGAs5kQIZysnfC5g83mzsMIiIiszJ4zGz58uWRmppqilgKpNpBXvBzd0TWbV7aJOAytkRERES5ZXAyO3v2bIwbNw6HDh3C06dPkZSUpPFDmqR2EnzZrny22zmTAREREVHuGTzMoE2bNgCA5s2ba5QLISCRSKBQKIwTGRHlKDkjGW7hbgCAl5NfwtXB1cwRERER5T+Dk9mDBw+aIo4C621Tc0kATN9xHS2D/dg7S0RERGQgg5PZxo0bmyKOAuttU3MJALGJaZyai4iIiCgXDE5mjxw5kuP2Ro0a5TqYgohTcxERERGZjsHJbJMmTbTKXp9rlmNmNXFqLiIiIiLTMXg2g2fPnmn8xMfHY/fu3ahVqxb27t1rihitGqfmIiIiIjIdg3tmPTw8tMpatmwJR0dHfPrppzh37pxRAisopHYSdKjkh1+O3822DqfmIiIiIsodg5PZ7BQtWhQ3b9401uEKjN1XY7H8+L1st/9foyC0CfHPx4iooJDaSdGubDv1YyIiIltkcDJ7+fJljedCCMTGxmL27Nlc4vYNCqXA9B3X/zfAQHfP6/ZLsZjYpgJ7ZslgTvZO2PXxLnOHQUREZFYGJ7NVq1aFRCKBEJpjQOvWrYsVK1YYLbCC4Ex0AmITc56lgNNyEREREeWewclsdHS0xnM7OzsULVoUTk68G/9NnJaLiIiIyLQMTmYDAgJMEUeBxGm5yJSSM5LhM88HABA/Pp7L2RIRkU3Se2quAwcOIDg4GElJSVrbEhMTUbFiRRw9etSowVm72kFe8Pdwyma0LKflorxLkacgRZ5i7jCIiIjMRu9kdsGCBRgyZAjc3d21tnl4eOCTTz7B/PnzjRqctZPaSRDWMfh/zzTHGKsSXE7LRURERJR7eiezly5dQps2bbLd3qpVK84xq0ObEH8MahCg1TsrkXBaLiIiIqK80juZffToEWQyWbbb7e3t8fjxY6MEVZCo5pl9c/0vpQCWHonG7quxZomLiIiIqCDQO5ktXrw4rly5ku32y5cvw9+fvYyv02ee2ek7rkOh1L3ULRERERHlTO9ktl27dpg6dSrS0rSnkUpNTUVYWBg6dOhg1OCs3dvmmRV4Nc8sERERERlO76m5vvzyS2zduhXvvvsuRowYgXLlykEikeDGjRtYuHAhFAoFvvjiC1PGanU4zyyZkp3EDo0DGqsfExER2SK9/wL6+vrixIkTCAkJweTJk9GlSxd07twZn3/+OUJCQnD8+HH4+voaHMCiRYsQFBQEJycn1KhRI8fpvbZu3YqWLVuiaNGicHd3R7169bBnzx6Dz5lfOM8smZKzzBmH+h/Cof6H4CxzNnc4REREZmFQd05AQAAiIyPx5MkTnD59GqdOncKTJ08QGRmJwMBAg0++ceNGjBkzBl988QUuXLiAhg0bom3btoiJidFZ/8iRI2jZsiUiIyNx7tw5NG3aFB07dsSFCxcMPnd+4DyzRERERKaVq+8mPT09UatWLdSuXRuenp65Pvn8+fMxaNAgDB48GBUqVMCCBQtQsmRJLF68WGf9BQsWYOLEiahVqxbKli2LWbNmoWzZstixY0euYzAlzjNLREREZFoGL2drLBkZGTh37hwmTZqkUd6qVSucOHFCr2MolUq8ePECXl7Z92ymp6cjPT1d/Vy1gplcLodcLs9F5IZpXs4bCz4MwWdbryBN8arcz8MRX7Qtj+blvPMlDjKcql0stX2SM5JRdmFZAMCt4be4nO3/WHq7kW5sN+vEdrNelt52hsRltmT2yZMnUCgUWuNsfX19ERcXp9cxvv32WyQnJ6N79+7Z1gkPD8f06dO1yvfu3QsXFxfDgs4lpQAqFLbDhacSlHRVolMpgTIeyVDcO4fIe/kSAuVBVFSUuUPQKU2RhiepTwAAe/bsgZOUY69fZ6ntRjlju1kntpv1stS2S0nRf6l2syWzKhKJ5lfsQgitMl02bNiAadOm4Y8//oCPj0+29SZPnoyxY8eqnyclJaFkyZJo1aqVzqV5jW3PtUeYFfk3HiVl9Q7/l2yH3x444ssq5dG6ouE3zFH+kcvliIqKQsuWLXNcMMRckjOSgf9N/dy6dWv2zP6Ppbcb6cZ2s05sN+tl6W2n+iZdH2ZLZr29vSGVSrV6YePj4986K8LGjRsxaNAgbN68GS1atMixrqOjIxwdHbXKZTKZyRtv99VYjPz1ktbqX4+S0jHy10tY3Ls6l7O1AvnxXskNmXgVk6XGaE58TawT2806sd2sl6W2nSExmW1ySgcHB9SoUUOrezsqKgr169fPdr8NGzagf//+WL9+Pdq3b2/qMHNNc/UvTaoyrv5FRERElDdmHWYwduxY9OnTBzVr1kS9evWwdOlSxMTEYOjQoQCyhgg8ePAAq1evBpCVyPbt2xfff/896tatq+7VdXZ2hoeHh9muQxdDVv+qV7pI/gVGREREVICYNZnt0aMHnj59ihkzZiA2NhYhISGIjIxEQEAAACA2NlZjztmff/4ZmZmZGD58OIYPH64u79evHyIiIvI7/Bxx9S8iIiIi0zP7DWChoaEIDQ3Vue3NBPXQoUOmD8hIuPoXmZqdxA41i9VUPyYiIrJFZk9mCyrV6l9xiWk6x81KAPhx9S/KA2eZM/4a8pe5wyAiIjIrdueYyOurf7050RhX/yIiIiIyDiazJtQmxB+Le1eHr7vm1GB+Hk6clouIiIjICDjMwMTahPijTkBhVJt5AACwon8tNH63KHtkKc9S5CkIXpjV+399+HW4yPJnRTsiIiJLwmQ2H6TKFQAAiQRo8m5R2DGRJSMQQuBe4j31YyIiIlvEYQb5IClNDgBwkEpwOjqBCyUQERERGQmTWRPbfTUWvZafBQCkZwr0XHYK7805gN1XY80cGREREZH1YzJrQruvxmLY2vN4mpyhUR6XmIZha88zoSUiIiLKIyazJqJQCkzfcV3nHLOqsuk7rnPIAREREVEeMJk1kTPRCYhNzH6pWgEgNjENZ6IT8i8oIiIiogKGsxmYSPyL7BPZ3NQjepNEIkFw0WD1YyIiIlvEZNZEfAo5GbUe0ZtcZC64FnrN3GEQERGZFYcZmEjtIC/4ezhpLWWrIgHg7+GE2kFe+RkWERERUYHCZNZEpHYShHUM1rlNleCGdQzmSmBEREREecBk1oTahPhjce/qcHWQapT7eThhce/qaBPib6bIqCBIkaeg4qKKqLioIlLkKeYOh4iIyCw4ZtbE2oT448g/8Vh/5j5aVvDBwPfeQe0gL/bIUp4JIXD98XX1YyIiIlvEZDYfpGcqAQCVi7ujXukiZo6GiIiIqODgMIN8kCbPSmad3hhuQERERER5w2TWxBRKgdjnqQCyFkngil9ERERExsNk1oR2X43Fe3MO4Px/iQCAFcfv4b05B7D7aqyZIyMiIiIqGJjMmsjuq7EYtva81pK2cYlpGLb2PBNaIiIiIiNgMmsCCqXA9B3XoWtAgaps+o7rHHJAeSKRSBDgEYAAjwAuZ0tERDaLsxmYwJnoBK0e2dcJZI2fPROdwNkNKNdcZC64O+auucMgIiIyK/bMmkD8i+wT2dzUIyIiIiLdmMyagE8hJ6PWIyIiIiLdmMyaQO0gL/h7OCG7UYwSAP4eTqgd5JWfYVEBkypPRa1ltVBrWS2kylPNHQ4REZFZMJk1AamdBGEdgwFAK6FVPQ/rGMwlbSlPlEKJsw/P4uzDs1AKpbnDISIiMgsmsybSJsQfi3tXh6+7o0a5n4cTFveujjYh/maKjIiIiKjg4GwGJtQmxB+1A4ug+tdRAICIftXRsJwfe2SJiIiIjIQ9syaWocj6+tdOItCgjDcTWSIiIiIjYjJrYmlyBQBAxleaiIiIyOiYYplY6v+SWQe+0kRERERGxzGzJsZklkzJ28Xb3CEQERGZFZNZE+MwAzIVVwdXPJ7w2NxhEBERmRVTLBNLSc8EAGQogdPRCVAohZkjIiIiIio4mMya0O6rsRi/5TIAICFdgt4rzuK9OQew+2qsmSMjIiIiKhiYzJrI7quxGLb2PJ6nyDXK4xLTMGzteSa0lGep8lQ0iWiCJhFNuJwtERHZLI6ZNQGFUmD6juvQNaBAIGtJ2+k7rqNlMBdQoNxTCiUO3zusfkxERGSL2DNrAmeiExCbmJbtdgEgNjENZ6IT8i8oIiIiogKIyawJxL/IPpHNTT0iIiIi0o3DDEzAp5CTUesR5caxmGNYf2U9Tt0/hQcvHuB52nP4u/mjsm9lfNbgMzQo1QAAcOjuITRd1TTHYwV4BODumLs6t4UfDcfnBz5XP1/cfjGG1hxqtOsgIiLKCZNZE6gd5AV/DyfEJabpHDcrAeDn4YTaQV75HRrZkLWX1+Lncz9rlN1LvId7ifew458dWNphKYbUGKLXsQo5FtJZfuPxDUw/PD3PsRIREeUWhxmYgNROgrCOwQCyEtfXqZ6HdQzmzV9kUnYSO3QL7oa9vffi5eSXuP/pfXQL7qbePnn/ZCiUCjQJbAIRJrR+eob0VNcdXG2w1vGVQomB2wciXZEOV5lrvlwTERHRm5jMmkibEH8s7l0dfh6aQwn8PJywuHd1tAnxN1NkVJC4yFzgInPRuW12i9nY/OFmtCzdEq4OrijuXhyL2i1Sb3+a+hSPU3SvIHY/6T42X98MACjsVBiDqg/SqvPdye9w6v4p1CleB10qdDHC1RARERmOwwxMqE2IP1oG++Hkv/HYe/Q0WjWsg3plfNgjS0bh6uCK5M+Ts93u7uiuVZYiT1E/dpG5oIhzEZ37/nD6B2Qqs1av+7/q/wc3BzeN7f8m/IspB6fAUeqIFe+vwNzjc3NzCURERHnGZNbEpHYS1AnywtMbAnWCvJjIktkIITA+arz6+Sc1PoFMKtOq9zLjJZadXwYAkNnJMKrOKK3jDPxjIFIzUxHePBzBRYNNGzgREVEOOMyAyAZkKDLQZ1sfbLm+BQDQLKgZwpuH66y7/PxyPE97DgD4KOQjFHcvrrH9pzM/4WjMUdQsVhMT6k8wadxERERvw55ZIiuVlpmGrpu6AgB+6/4bnOx1T/WWlJ6ELhu74ED0AQBAp3Kd8GvXX+Fo76hVV6FU4PvT36ufj6s3TqvO5P2TAQDDaw3HhbgLALLG36rEJMbg7MOzqOZXDVI7aS6vjoiISD9MZomslEKpQOStSPVjXR4kPUC79e1w+dFlAFkJ6Pdtvs82ydz29zZEP48GALR4pwWq+FXRqpMszxqnO+CPATqPEX4sHOHHwvHss2co7FTYoGsiIiIyFJNZogLqavxVtF3XFveT7kMCCea0mIMJDXIeFjD/5Hz1Y129skRERJaGySxRATXvxDzcT7oPABAQmLhvIibum6hR52C/g2gS2AQAcOr+KZy8fxIAULFoRbQp00bncUWY9lIg/X/vj1WXVgHgCmBERJS/eAMYEQEAvj35rfoxe2WJiMhasGeWqICK6ByBiM4Retff/OHmfDsXERGRsbBnloiIiIisls31zAqRNd4vKSkp384pl8uRkpKCpKQkyGTak9STZbL0dkvOSAbSsh4nJSVB4aB7RgNbY+ntRrqx3awT2816WXrbqfI0Vd6WE4nQp1YBcv/+fZQsWdLcYRARERHRW/z3338oUaJEjnVsLplVKpV4+PAhChUqBIkkf5aWTUpKQsmSJfHff//B3d09X85Jecd2s05sN+vEdrNObDfrZeltJ4TAixcvUKxYMdjZ5Twq1uaGGdjZ2b01wzcVd3d3i3zDUM7YbtaJ7Wad2G7Wie1mvSy57Tw8PPSqxxvAiIiIiMhqMZklIiIiIqvFZDYfODo6IiwsDI6OjuYOhQzAdrNObDfrxHazTmw361WQ2s7mbgAjIiIiooKDPbNEREREZLWYzBIRERGR1WIyS0RERERWi8ksEREREVktJrMmtmjRIgQFBcHJyQk1atTA0aNHzR2SzQoPD0etWrVQqFAh+Pj4oHPnzrh586ZGHSEEpk2bhmLFisHZ2RlNmjTBtWvXNOqkp6dj5MiR8Pb2hqurKzp16oT79+/n56XYtPDwcEgkEowZM0ZdxnazXA8ePEDv3r1RpEgRuLi4oGrVqjh37px6O9vO8mRmZuLLL79EUFAQnJ2d8c4772DGjBlQKpXqOmw38zty5Ag6duyIYsWKQSKR4Pfff9fYbqw2evbsGfr06QMPDw94eHigT58+eP78uYmvzkCCTObXX38VMplMLFu2TFy/fl2MHj1auLq6inv37pk7NJvUunVrsXLlSnH16lVx8eJF0b59e1GqVCnx8uVLdZ3Zs2eLQoUKid9++01cuXJF9OjRQ/j7+4ukpCR1naFDh4rixYuLqKgocf78edG0aVNRpUoVkZmZaY7LsilnzpwRgYGBonLlymL06NHqcrabZUpISBABAQGif//+4vTp0yI6Olrs27dP/Pvvv+o6bDvL8/XXX4siRYqInTt3iujoaLF582bh5uYmFixYoK7DdjO/yMhI8cUXX4jffvtNABDbtm3T2G6sNmrTpo0ICQkRJ06cECdOnBAhISGiQ4cO+XWZemEya0K1a9cWQ4cO1SgrX768mDRpkpkiotfFx8cLAOLw4cNCCCGUSqXw8/MTs2fPVtdJS0sTHh4eYsmSJUIIIZ4/fy5kMpn49ddf1XUePHgg7OzsxO7du/P3AmzMixcvRNmyZUVUVJRo3LixOpllu1muzz77TLz33nvZbmfbWab27duLgQMHapR98MEHonfv3kIItpslejOZNVYbXb9+XQAQp06dUtc5efKkACD+/vtvE1+V/jjMwEQyMjJw7tw5tGrVSqO8VatWOHHihJmiotclJiYCALy8vAAA0dHRiIuL02gzR0dHNG7cWN1m586dg1wu16hTrFgxhISEsF1NbPjw4Wjfvj1atGihUc52s1zbt29HzZo18eGHH8LHxwfVqlXDsmXL1NvZdpbpvffew/79+/HPP/8AAC5duoRjx46hXbt2ANhu1sBYbXTy5El4eHigTp066jp169aFh4eHRbWjvbkDKKiePHkChUIBX19fjXJfX1/ExcWZKSpSEUJg7NixeO+99xASEgIA6nbR1Wb37t1T13FwcICnp6dWHbar6fz66684f/48/vrrL61tbDfLdefOHSxevBhjx47F559/jjNnzmDUqFFwdHRE37592XYW6rPPPkNiYiLKly8PqVQKhUKBmTNnomfPngD4O2cNjNVGcXFx8PHx0Tq+j4+PRbUjk1kTk0gkGs+FEFpllP9GjBiBy5cv49ixY1rbctNmbFfT+e+//zB69Gjs3bsXTk5O2dZju1kepVKJmjVrYtasWQCAatWq4dq1a1i8eDH69u2rrse2sywbN27E2rVrsX79elSsWBEXL17EmDFjUKxYMfTr109dj+1m+YzRRrrqW1o7cpiBiXh7e0MqlWr9zyU+Pl7rf0qUv0aOHInt27fj4MGDKFGihLrcz88PAHJsMz8/P2RkZODZs2fZ1iHjOnfuHOLj41GjRg3Y29vD3t4ehw8fxg8//AB7e3v16852szz+/v4IDg7WKKtQoQJiYmIA8HfOUk2YMAGTJk3CRx99hEqVKqFPnz749NNPER4eDoDtZg2M1UZ+fn549OiR1vEfP35sUe3IZNZEHBwcUKNGDURFRWmUR0VFoX79+maKyrYJITBixAhs3boVBw4cQFBQkMb2oKAg+Pn5abRZRkYGDh8+rG6zGjVqQCaTadSJjY3F1atX2a4m0rx5c1y5cgUXL15U/9SsWRO9evXCxYsX8c4777DdLFSDBg20pr/7559/EBAQAIC/c5YqJSUFdnaa6YFUKlVPzcV2s3zGaqN69eohMTERZ86cUdc5ffo0EhMTLasdzXHXma1QTc21fPlycf36dTFmzBjh6uoq7t69a+7QbNKwYcOEh4eHOHTokIiNjVX/pKSkqOvMnj1beHh4iK1bt4orV66Inj176pzKpESJEmLfvn3i/PnzolmzZpxuJp+9PpuBEGw3S3XmzBlhb28vZs6cKW7duiXWrVsnXFxcxNq1a9V12HaWp1+/fqJ48eLqqbm2bt0qvL29xcSJE9V12G7m9+LFC3HhwgVx4cIFAUDMnz9fXLhwQT39p7HaqE2bNqJy5cri5MmT4uTJk6JSpUqcmsvWLFy4UAQEBAgHBwdRvXp19TRQlP8A6PxZuXKluo5SqRRhYWHCz89PODo6ikaNGokrV65oHCc1NVWMGDFCeHl5CWdnZ9GhQwcRExOTz1dj295MZtlulmvHjh0iJCREODo6ivLly4ulS5dqbGfbWZ6kpCQxevRoUapUKeHk5CTeeecd8cUXX4j09HR1Hbab+R08eFDn37R+/foJIYzXRk+fPhW9evUShQoVEoUKFRK9evUSz549y6er1I9ECCHM0ydMRERERJQ3HDNLRERERFaLySwRERERWS0ms0RERERktZjMEhEREZHVYjJLRERERFaLySwRERERWS0ms0RERERktZjMEpHNu3v3LiQSCS5evGjuUNT+/vtv1K1bF05OTqhataq5wyEislhMZonI7Pr37w+JRILZs2drlP/++++QSCRmisq8wsLC4Orqips3b2L//v3Z1ouLi8PIkSPxzjvvwNHRESVLlkTHjh1z3McW9e/fH507dzZ3GERkAkxmicgiODk5Yc6cOXj27Jm5QzGajIyMXO97+/ZtvPfeewgICECRIkV01rl79y5q1KiBAwcOYO7cubhy5Qp2796Npk2bYvjw4bk+NxGRNWEyS0QWoUWLFvDz80N4eHi2daZNm6b1lfuCBQsQGBiofq7qgZs1axZ8fX1RuHBhTJ8+HZmZmZgwYQK8vLxQokQJrFixQuv4f//9N+rXrw8nJydUrFgRhw4d0th+/fp1tGvXDm5ubvD19UWfPn3w5MkT9fYmTZpgxIgRGDt2LLy9vdGyZUud16FUKjFjxgyUKFECjo6OqFq1Knbv3q3eLpFIcO7cOcyYMQMSiQTTpk3TeZzQ0FBIJBKcOXMG3bp1w7vvvouKFSti7NixOHXqlLpeTEwM3n//fbi5ucHd3R3du3fHo0ePtF7XFStWoFSpUnBzc8OwYcOgUCgwd+5c+Pn5wcfHBzNnztQ4v0QiweLFi9G2bVs4OzsjKCgImzdv1qhz5coVNGvWDM7OzihSpAj+7//+Dy9fvtRqr3nz5sHf3x9FihTB8OHDIZfL1XUyMjIwceJEFC9eHK6urqhTp45G20RERKBw4cLYs2cPKlSoADc3N7Rp0waxsbHq61u1ahX++OMPSCQSSCQSHDp0CBkZGRgxYgT8/f3h5OSEwMDAHN9/RGSZmMwSkUWQSqWYNWsWfvzxR9y/fz9Pxzpw4AAePnyII0eOYP78+Zg2bRo6dOgAT09PnD59GkOHDsXQoUPx33//aew3YcIEjBs3DhcuXED9+vXRqVMnPH36FAAQGxuLxo0bo2rVqjh79ix2796NR48eoXv37hrHWLVqFezt7XH8+HH8/PPPOuP7/vvv8e2332LevHm4fPkyWrdujU6dOuHWrVvqc1WsWBHjxo1DbGwsxo8fr3WMhIQE7N69G8OHD4erq6vW9sKFCwMAhBDo3LkzEhIScPjwYURFReH27dvo0aOHRv3bt2/jzz//xO7du7FhwwasWLEC7du3x/3793H48GHMmTMHX375pUaSDABTpkxB165dcenSJfTu3Rs9e/bEjRs3AAApKSlo06YNPD098ddff2Hz5s3Yt28fRowYoXGMgwcP4vbt2zh48CBWrVqFiIgIREREqLcPGDAAx48fx6+//orLly/jww8/RJs2bdSvl+pc8+bNw5o1a3DkyBHExMSoX7fx48eje/fu6gQ3NjYW9evXxw8//IDt27dj06ZNuHnzJtauXavxHyMishKCiMjM+vXrJ95//30hhBB169YVAwcOFEIIsW3bNvH6x1RYWJioUqWKxr7fffedCAgI0DhWQECAUCgU6rJy5cqJhg0bqp9nZmYKV1dXsWHDBiGEENHR0QKAmD17trqOXC4XJUqUEHPmzBFCCDFlyhTRqlUrjXP/999/AoC4efOmEEKIxo0bi6pVq771eosVKyZmzpypUVarVi0RGhqqfl6lShURFhaW7TFOnz4tAIitW7fmeK69e/cKqVQqYmJi1GXXrl0TAMSZM2eEEFmvq4uLi0hKSlLXad26tQgMDNR6HcPDw9XPAYihQ4dqnK9OnTpi2LBhQgghli5dKjw9PcXLly/V23ft2iXs7OxEXFycEOJVe2VmZqrrfPjhh6JHjx5CCCH+/fdfIZFIxIMHDzTO07x5czF58mQhhBArV64UAMS///6r3r5w4ULh6+urfv76e0xl5MiRolmzZkKpVGb7+hGR5WPPLBFZlDlz5mDVqlW4fv16ro9RsWJF2Nm9+njz9fVFpUqV1M+lUimKFCmC+Ph4jf3q1aunfmxvb4+aNWuqexnPnTuHgwcPws3NTf1Tvnx5AFm9mio1a9bMMbakpCQ8fPgQDRo00Chv0KCB+lz6EEIAwFtvkLtx4wZKliyJkiVLqsuCg4NRuHBhjfMFBgaiUKFC6ue+vr4IDg7Weh1zes1Uz1XHvXHjBqpUqaLRc9ygQQMolUrcvHlTXVaxYkVIpVL1c39/f/V5zp8/DyEE3n33XY3X/vDhwxqvu4uLC0qXLq3zGNnp378/Ll68iHLlymHUqFHYu3dvjvWJyDLZmzsAIqLXNWrUCK1bt8bnn3+O/v37a2yzs7NTJ3Eqr4+tVJHJZBrPJRKJzjKlUvnWeFTJolKpRMeOHTFnzhytOv7+/urHur7yz+m4KkIIg2ZuKFu2LCQSCW7cuJHjXfrZHffNclO8Zjld09vOrTqPUqmEVCrFuXPnNBJeAHBzc8vxGG++V95UvXp1REdH488//8S+ffvQvXt3tGjRAlu2bHnLFRKRJWHPLBFZnNmzZ2PHjh04ceKERnnRokURFxenkaQYc27Y18eDZmZm4ty5c+re1+rVq+PatWsIDAxEmTJlNH70TWABwN3dHcWKFcOxY8c0yk+cOIEKFSrofRwvLy+0bt0aCxcuRHJystb258+fA8jqhY2JidEYH3z9+nUkJiYadL7svDmG9tSpU+rXLDg4GBcvXtSI7/jx47Czs8O7776r1/GrVasGhUKB+Ph4rdfdz89P7zgdHBygUCi0yt3d3dGjRw8sW7YMGzduxG+//YaEhAS9j0tE5sdklogsTqVKldCrVy/8+OOPGuVNmjTB48ePMXfuXNy+fRsLFy7En3/+abTzLly4ENu2bcPff/+N4cOH49mzZxg4cCAAYPjw4UhISEDPnj1x5swZ3LlzB3v37sXAgQN1Jkk5mTBhAubMmYONGzfi5s2bmDRpEi5evIjRo0cbdJxFixZBoVCgdu3a+O2333Dr1i3cuHEDP/zwg/rr/xYtWqBy5cro1asXzp8/jzNnzqBv375o3LjxW4dE6GPz5s1YsWIF/vnnH4SFheHMmTPqG7x69eoFJycn9OvXD1evXsXBgwcxcuRI9OnTB76+vnod/91330WvXr3Qt29fbN26FdHR0fjrr78wZ84cREZG6h1nYGAgLl++jJs3b+LJkyeQy+X47rvv8Ouvv+Lvv//GP//8g82bN8PPz0998xwRWQcms0Rkkb766iutr4krVKiARYsWYeHChahSpQrOnDmj807/3Jo9ezbmzJmDKlWq4OjRo/jjjz/g7e0NAChWrBiOHz8OhUKB1q1bIyQkBKNHj4aHh4fGuFJ9jBo1CuPGjcO4ceNQqVIl7N69G9u3b0fZsmUNOk5QUBDOnz+Ppk2bYty4cQgJCUHLli2xf/9+LF68GEDW1+2///47PD090ahRI7Ro0QLvvPMONm7caNC5sjN9+nT8+uuvqFy5MlatWoV169YhODgYQNY41j179iAhIQG1atVCt27d0Lx5c/z0008GnWPlypXo27cvxo0bh3LlyqFTp044ffq0xjjgtxkyZAjKlSuHmjVromjRojh+/Djc3NwwZ84c1KxZE7Vq1cLdu3cRGRlpcHsSkXlJxNsGFREREekgkUiwbds2rqxFRGbF/34SERERkdViMktEREREVotTcxERUa5wlBoRWQL2zBIRERGR1WIyS0RERERWi8ksEREREVktJrNEREREZLWYzBIRERGR1WIyS0RERERWi8ksEREREVktJrNEREREZLWYzBIRERGR1fp/hZi6jpFBTkMAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 800x450 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def pca_col(df, name, column_name='embeddings', variance_threshold=0.90):\n",
    "    embeddings = np.array(df[column_name].tolist())\n",
    "    \n",
    "    scaler = StandardScaler()\n",
    "    embeddings_standardized = scaler.fit_transform(embeddings)\n",
    "    \n",
    "    pca = PCA()\n",
    "    pca.fit(embeddings_standardized)\n",
    "    explained_variance = pca.explained_variance_ratio_\n",
    "\n",
    "    # Components to retain (in our case 90% of variance)\n",
    "    n_components = np.argmax(np.cumsum(explained_variance) >= variance_threshold) + 1\n",
    "        \n",
    "    plt.figure(figsize=(8, 4.5))\n",
    "    plt.axhline(y=variance_threshold, color='r', linestyle='--')\n",
    "    x_value = np.argmax(np.cumsum(explained_variance) >= variance_threshold) + 1\n",
    "    plt.axvline(x=x_value, color='g', linestyle='--')\n",
    "    plt.text(x_value + 10, 0.025, f'{x_value}', color='green', size='large', weight='bold')\n",
    "    plt.text(- 50, variance_threshold + 0.025, f'{variance_threshold:.2f}', color='red', size='large', weight='bold')\n",
    "\n",
    "    plt.plot(np.cumsum(explained_variance), marker='o')\n",
    "    plt.title(f'Cumulative Explained Variance by PCA Components for {name}')\n",
    "    plt.xlabel('Number of Components')\n",
    "    plt.ylabel('Cumulative Explained Variance')\n",
    "    plt.grid(True)\n",
    "    plt.show()\n",
    "    \n",
    "    pca = PCA(n_components=n_components)\n",
    "    reduced_embeddings = pca.fit_transform(embeddings_standardized)\n",
    "    \n",
    "    # Integrate reduced embeddings into the DataFrame\n",
    "    df[f'reduced_{column_name}'] = reduced_embeddings.tolist()\n",
    "    return df\n",
    "\n",
    "# Apply PCA and add column for each dataset\n",
    "X_df_BP = pca_col(X_df_BP, 'BP')\n",
    "#x_df_CC = pca_col(x_df_CC, 'CC')\n",
    "#x_df_MF = pca_col(x_df_MF, 'MF')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Save the datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure the directory exists\n",
    "datasets_path = Path('../data/datasets')\n",
    "datasets_path.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "# Save the DataFrame to a CSV file\n",
    "X_df_BP.to_csv(datasets_path / 'X_df_BP.csv', index=False)\n",
    "y_df_BP.to_csv(datasets_path / 'y_df_BP.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MODELS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "from sklearn.model_selection import train_test_split\n",
    "from torch.utils.data import DataLoader, TensorDataset\n",
    "from sklearn.metrics import classification_report\n",
    "from torchsummary import summary\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# from here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X shape: (83064, 1024), y shape: (83064, 1487)\n",
      "y_df_CC: 0    (GO:0005575, GO:0110165, GO:0000785, GO:003299...\n",
      "1    (GO:0005575, GO:0110165, GO:0071944, GO:000588...\n",
      "2    (GO:0005575, GO:0110165, GO:0005622, GO:000582...\n",
      "3     (GO:0005575, GO:0110165, GO:0005576, GO:0005615)\n",
      "4    (GO:0005575, GO:0005737, GO:0042175, GO:003299...\n",
      "Name: GO_term, dtype: object\n",
      "y.head: [[0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]]\n",
      "type(y_df_CC): <class 'pandas.core.series.Series'>, type(y): <class 'numpy.ndarray'>\n"
     ]
    }
   ],
   "source": [
    "# Load data\n",
    "\n",
    "X = X_df_BP['embeddings'] \n",
    "X = pd.DataFrame(X.tolist())  # Convert list of lists to DataFrame\n",
    "\n",
    "# Preprocess target (MultiLabelBinarizer for multi-label classification)\n",
    "mlb = MultiLabelBinarizer()\n",
    "y = mlb.fit_transform(y_df_BP)\n",
    "\n",
    "print(f\"X shape: {X.shape}, y shape: {y.shape}\")\n",
    "print(f\"y_df_CC: {y_df_CC[:5]}\\ny.head: {y[:5]}\")\n",
    "print(f\"type(y_df_CC): {type(y_df_CC)}, type(y): {type(y)}\")\n",
    "\n",
    "# Split data\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert NumPy array to PyTorch tensor\n",
    "X_train_tensor = torch.tensor(X_train.to_numpy(), dtype=torch.float32)\n",
    "X_test_tensor = torch.tensor(X_test.to_numpy(), dtype=torch.float32)\n",
    "y_train_tensor = torch.tensor(y_train, dtype=torch.float32).to(device)\n",
    "y_test_tensor = torch.tensor(y_test, dtype=torch.float32).to(device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# X_prova = X_train_tensor[:5000]\n",
    "# y_prova = y_train_tensor[:5000]\n",
    "\n",
    "# X_prova_test = X_test_tensor[:5000]\n",
    "# y_prova_test = y_test_tensor[:5000]\n",
    "\n",
    "# X_train_tensor = X_prova\n",
    "# y_train_tensor = y_prova.to(device)\n",
    "# X_test_tensor = X_prova_test\n",
    "# y_test_tensor = y_prova_test.to(device)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NN(nn.Module):\n",
    "    def __init__(self, input_size, hidden_sizes, dropouts, output_size):\n",
    "        super(NN, self).__init__()\n",
    "        \n",
    "        layers = []\n",
    "        in_size = input_size\n",
    "        \n",
    "        for i, hidden_size in enumerate(hidden_sizes):\n",
    "            layers.append(nn.Linear(in_size, hidden_size))\n",
    "            layers.append(nn.ReLU())\n",
    "            layers.append(nn.Dropout(dropouts[i]))\n",
    "            in_size = hidden_size\n",
    "        \n",
    "        layers.append(nn.Linear(in_size, output_size))\n",
    "        layers.append(nn.Sigmoid())  # For multilabel classification\n",
    "        \n",
    "        self.model = nn.Sequential(*layers)\n",
    "    \n",
    "    def forward(self, x):\n",
    "        return self.model(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Creating/Extracting the models DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>hidden_sizes</th>\n",
       "      <th>dropouts</th>\n",
       "      <th>lr</th>\n",
       "      <th>batch_size</th>\n",
       "      <th>epochs</th>\n",
       "      <th>macro_P</th>\n",
       "      <th>macro_R</th>\n",
       "      <th>macro_F1</th>\n",
       "      <th>weighted_P</th>\n",
       "      <th>weighted_R</th>\n",
       "      <th>weighted_F1</th>\n",
       "      <th>samples_P</th>\n",
       "      <th>samples_R</th>\n",
       "      <th>samples_F1</th>\n",
       "      <th>training_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [name, hidden_sizes, dropouts, lr, batch_size, epochs, macro_P, macro_R, macro_F1, weighted_P, weighted_R, weighted_F1, samples_P, samples_R, samples_F1, training_time]\n",
       "Index: []"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if the file exists and create it if it doesn't\n",
    "datasets_path = Path('../data/datasets')\n",
    "nn_df_path = datasets_path / 'NN_BP.csv'\n",
    "\n",
    "if nn_df_path.exists():\n",
    "    \n",
    "    NN_BP = pd.read_csv(nn_df_path)\n",
    "    for c_list, c_int in zip(['hidden_sizes', 'dropouts'], ['batch_size', 'epochs']):\n",
    "        NN_BP[c_list] = NN_BP[c_list].apply(ast.literal_eval)\n",
    "        NN_BP[c_int] = NN_BP[c_int].astype(int)\n",
    "    \n",
    "    NN_BP['lr'] = NN_BP['lr'].astype(float)\n",
    "\n",
    "else:\n",
    "    if not datasets_path.exists():\n",
    "        datasets_path.mkdir(parents=True, exist_ok=True)\n",
    "    \n",
    "    NN_BP = {\n",
    "        'name': [],\n",
    "        'hidden_sizes': [],\n",
    "        'dropouts': [],\n",
    "        'lr': [],\n",
    "        'batch_size': [],\n",
    "        'epochs': [],\n",
    "        'macro_P': [],\n",
    "        'macro_R': [],\n",
    "        'macro_F1': [],\n",
    "        'weighted_P': [],\n",
    "        'weighted_R': [],\n",
    "        'weighted_F1': [],\n",
    "        'samples_P': [],\n",
    "        'samples_R': [],\n",
    "        'samples_F1': [],\n",
    "        'training_time': [],\n",
    "    }\n",
    "    NN_BP = pd.DataFrame(NN_BP)\n",
    "\n",
    "\n",
    "NN_BP.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Filling NN_BP with model parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insert_parameters(df, names, hidden_sizes_list, dropouts_list, lrs, batch_sizes, epochs_list):\n",
    "    for name, hidden_sizes, dropouts, lr, batch_size, epochs in zip(names, hidden_sizes_list, dropouts_list, lrs, batch_sizes, epochs_list):\n",
    "        exists = name in df['name'].values\n",
    "        \n",
    "        if not exists:\n",
    "            new_row = {\n",
    "                'name': name,\n",
    "                'hidden_sizes': [hidden_sizes],\n",
    "                'dropouts': [dropouts],\n",
    "                'lr': lr,\n",
    "                'batch_size': int(batch_size),\n",
    "                'epochs': int(epochs),\n",
    "            }\n",
    "            new_row_df = pd.DataFrame(new_row)\n",
    "            df = pd.concat([df, new_row_df], ignore_index=True)\n",
    "        else:\n",
    "            print(f\"{name} already exists in the DataFrame.\")\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>hidden_sizes</th>\n",
       "      <th>dropouts</th>\n",
       "      <th>lr</th>\n",
       "      <th>batch_size</th>\n",
       "      <th>epochs</th>\n",
       "      <th>macro_P</th>\n",
       "      <th>macro_R</th>\n",
       "      <th>macro_F1</th>\n",
       "      <th>weighted_P</th>\n",
       "      <th>weighted_R</th>\n",
       "      <th>weighted_F1</th>\n",
       "      <th>samples_P</th>\n",
       "      <th>samples_R</th>\n",
       "      <th>samples_F1</th>\n",
       "      <th>training_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NN1</td>\n",
       "      <td>[8192, 4096]</td>\n",
       "      <td>[0.5, 0.4]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>512.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NN2</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>2048.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NN3</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.4, 0.4]</td>\n",
       "      <td>0.005</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NN4</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.1, 0.1]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>2048.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NN5</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name  hidden_sizes    dropouts     lr  batch_size  epochs  macro_P  macro_R  \\\n",
       "0  NN1  [8192, 4096]  [0.5, 0.4]  0.001       512.0   120.0      NaN      NaN   \n",
       "1  NN2  [4096, 2048]  [0.2, 0.2]  0.003      2048.0    75.0      NaN      NaN   \n",
       "2  NN3  [4096, 2048]  [0.4, 0.4]  0.005      1024.0   100.0      NaN      NaN   \n",
       "3  NN4  [2048, 1024]  [0.1, 0.1]  0.001      2048.0    75.0      NaN      NaN   \n",
       "4  NN5  [2048, 1024]  [0.2, 0.2]  0.003      1024.0    75.0      NaN      NaN   \n",
       "\n",
       "   macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  samples_R  \\\n",
       "0       NaN         NaN         NaN          NaN        NaN        NaN   \n",
       "1       NaN         NaN         NaN          NaN        NaN        NaN   \n",
       "2       NaN         NaN         NaN          NaN        NaN        NaN   \n",
       "3       NaN         NaN         NaN          NaN        NaN        NaN   \n",
       "4       NaN         NaN         NaN          NaN        NaN        NaN   \n",
       "\n",
       "   samples_F1  training_time  \n",
       "0         NaN            NaN  \n",
       "1         NaN            NaN  \n",
       "2         NaN            NaN  \n",
       "3         NaN            NaN  \n",
       "4         NaN            NaN  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "names = ['NN1', 'NN2', 'NN3', 'NN4', 'NN5', 'NN6', 'NN7', 'NN8', 'NN9', 'NN10']\n",
    "\n",
    "\n",
    "hidden_sizes = [[8192, 4096],   #NN1\n",
    "                [4096, 2048],   #NN2\n",
    "                [4096, 2048],   #NN3\n",
    "                [2048, 1024],   #NN4\n",
    "                [2048, 1024],   #NN5\n",
    "                [2048, 1024],   #NN6\n",
    "                [4096, 2048, 1024],   #NN7\n",
    "                [4096, 2048, 2048],   #NN8\n",
    "                [2048, 2048, 2048],   #NN9\n",
    "                [2048, 4096, 2048],   #NN10\n",
    "                [4096, 4096, 2048],   #NN11        \n",
    "                [4096, 4096, 2048]]   #NN12    \n",
    "\n",
    "\n",
    "dropouts = [[0.5, 0.4],   #NN1 \n",
    "            [0.2, 0.2],   #NN2\n",
    "            [0.4, 0.4],   #NN3\n",
    "            [0.1, 0.1],   #NN4\n",
    "            [0.2, 0.2],   #NN5\n",
    "            [0.4, 0.4],   #NN6\n",
    "            [0.1, 0.2, 0.3],  #NN7\n",
    "            [0.2, 0.2, 0.2],  #NN8\n",
    "            [0.2, 0.2, 0.2],  #NN9\n",
    "            [0.3, 0.3, 0.3],  #NN10\n",
    "            [0.1, 0.2, 0.3],  #NN11        \n",
    "            [0.1, 0.2, 0.3]]  #NN12 \n",
    "\n",
    "\n",
    "lrs = [0.001, #NN1 \n",
    "       0.003, #NN2\n",
    "       0.005, #NN3\n",
    "       0.001, #NN4\n",
    "       0.003, #NN5\n",
    "       0.005, #NN6\n",
    "       0.005, #NN7\n",
    "       0.003, #NN8\n",
    "       0.003, #NN9\n",
    "       0.005, #NN10\n",
    "       0.003, #NN11        \n",
    "       0.001] #NN12 \n",
    "\n",
    "\n",
    "batch_sizes = [512, #NN1\n",
    "               2048, #NN2\n",
    "               1024, #NN3\n",
    "               2048, #NN4\n",
    "               1024, #NN5\n",
    "               1024, #NN6\n",
    "               2048, #NN7\n",
    "               1024, #NN8\n",
    "               1024, #NN9\n",
    "               2048, #NN10\n",
    "               2048, #N11        \n",
    "               1024] #N12 \n",
    "\n",
    "\n",
    "epochs = [120,  #NN1 \n",
    "          75,  #NN2\n",
    "          100, #NN3\n",
    "          75,  #NN4\n",
    "          75,  #NN5\n",
    "          100, #NN6\n",
    "          75,  #NN7\n",
    "          75,  #NN8\n",
    "          75,  #NN9\n",
    "          75,  #NN10\n",
    "          75,  #NN11\n",
    "          75]  #NN12\n",
    "\n",
    "\n",
    "\n",
    "NN_BP = insert_parameters(df = NN_BP, \n",
    "               names = names, \n",
    "               hidden_sizes_list=hidden_sizes, \n",
    "               dropouts_list=dropouts, \n",
    "               lrs=lrs, \n",
    "               batch_sizes=batch_sizes, \n",
    "               epochs_list=epochs)\n",
    "\n",
    "NN_BP.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Training the models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_model(df, name, input_size, output_size):\n",
    "\n",
    "    hidd_s = df[df['name'] == name]['hidden_sizes'].tolist()\n",
    "    dropouts = df[df['name'] == name]['dropouts'].tolist()\n",
    "    # Initialize the model\n",
    "    model = NN(\n",
    "        input_size = input_size,\n",
    "        hidden_sizes = hidd_s,\n",
    "        dropouts = dropouts, \n",
    "        output_size = output_size\n",
    "        ).to(device)\n",
    "    \n",
    "    # Loss function and optimizer\n",
    "    criterion = nn.BCELoss()  # Binary Cross-Entropy Loss for multilabel\n",
    "    optimizer = optim.Adam(model.parameters(), lr=df[df['name'] == name]['lr'])\n",
    "\n",
    "    return model, optimizer, criterion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_model(model, name, X_train, y_train, optimizer, criterion, batch_size, epochs):\n",
    "    # Determine the device (model should already be on this device)\n",
    "    device = next(model.parameters()).device\n",
    "\n",
    "    print(f\"\\nTraining {name} model:\")\n",
    "\n",
    "    # Move training data to the device\n",
    "    X_train = X_train.to(device)\n",
    "    y_train = y_train.to(device)\n",
    "\n",
    "    model.train()\n",
    "    start_time = time.time()  # Record the start time\n",
    "\n",
    "    for epoch in range(epochs):\n",
    "        for i in range(0, len(X_train), batch_size):\n",
    "            # Get the current batch\n",
    "            X_batch = X_train[i:i+batch_size]\n",
    "            y_batch = y_train[i:i+batch_size]\n",
    "\n",
    "            # Forward pass\n",
    "            outputs = model(X_batch)\n",
    "            loss = criterion(outputs, y_batch)\n",
    "\n",
    "            # Backward pass and optimization\n",
    "            optimizer.zero_grad()\n",
    "            loss.backward()\n",
    "            optimizer.step()\n",
    "        \n",
    "        \n",
    "        t = time.time() - start_time\n",
    "\n",
    "        if epoch == 0:\n",
    "            print(f\"Epoch [{epoch+1}/{epochs}], Loss: {loss.item():.4f} - time: {(t/60):.2f}min\")\n",
    "        elif epoch % 10 == 0:\n",
    "            print(f\"Epoch [{epoch}/{epochs}], Loss: {loss.item():.4f} - time: {(t/60):.2f}min\")\n",
    "\n",
    "    end_time = time.time()  # Record the end time\n",
    "\n",
    "    training_time = end_time - start_time  # Calculate the elapsed time\n",
    "    \n",
    "    return training_time / 60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "def model_evaluation(df, name, model, X, y, target_names, training_time):\n",
    "\n",
    "    model.eval()\n",
    "    with torch.no_grad():\n",
    "        y_pred = model(X)\n",
    "        probabilities = torch.sigmoid(y_pred)  # Assuming multilabel classification with sigmoid activation\n",
    "        prob = probabilities[probabilities > 0.20].float()\n",
    "        \n",
    "        y_pred = (y_pred.cpu().numpy() > 0.5).astype(int)\n",
    "\n",
    "    exists = any(df[df['name'] == name])\n",
    "    \n",
    "    if exists:\n",
    "        \n",
    "        report = classification_report(y.cpu().numpy(), y_pred, target_names=target_names, output_dict=True)\n",
    "        # Convert the classification report to a DataFrame\n",
    "        report = pd.DataFrame(report).transpose().drop(columns=['support'])\n",
    "        \n",
    "        report.loc['macro avg'] = report.loc['macro avg'].apply(lambda x: round(x, 3))\n",
    "        report.loc['weighted avg'] = report.loc['weighted avg'].apply(lambda x: round(x, 3))\n",
    "        report.loc['samples avg'] = report.loc['samples avg'].apply(lambda x: round(x, 3))\n",
    "\n",
    "\n",
    "\n",
    "        df.loc[df['name'] == name, 'macro_P'] = report.loc['macro avg']['precision']\n",
    "        df.loc[df['name'] == name, 'macro_R'] = report.loc['macro avg']['recall']\n",
    "        df.loc[df['name'] == name, 'macro_F1'] = report.loc['macro avg']['f1-score']\n",
    "\n",
    "        df.loc[df['name'] == name, 'weighted_P'] = report.loc['weighted avg']['precision']\n",
    "        df.loc[df['name'] == name, 'weighted_R'] = report.loc['weighted avg']['recall']\n",
    "        df.loc[df['name'] == name, 'weighted_F1'] = report.loc['weighted avg']['f1-score']\n",
    "        \n",
    "        df.loc[df['name'] == name, 'samples_P'] = report.loc['samples avg']['precision']\n",
    "        df.loc[df['name'] == name, 'samples_R'] = report.loc['samples avg']['recall']\n",
    "        df.loc[df['name'] == name, 'samples_F1'] = report.loc['samples avg']['f1-score']\n",
    "\n",
    "        df.loc[df['name'] == name, 'training_time'] = training_time\n",
    "\n",
    "        return df, report, prob\n",
    "    else:\n",
    "        print(f\"{name} already exists in the DataFrame.\")\n",
    "        return df, None, prob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN1 model:\n",
      "Epoch [1/120], Loss: 0.0809 - time: 0.07min\n",
      "Epoch [10/120], Loss: 0.0608 - time: 0.87min\n",
      "Epoch [20/120], Loss: 0.0501 - time: 1.66min\n",
      "Epoch [30/120], Loss: 0.0421 - time: 2.46min\n",
      "Epoch [40/120], Loss: 0.0355 - time: 3.26min\n",
      "Epoch [50/120], Loss: 0.0315 - time: 4.06min\n",
      "Epoch [60/120], Loss: 0.0287 - time: 4.87min\n",
      "Epoch [70/120], Loss: 0.0255 - time: 5.68min\n",
      "Epoch [80/120], Loss: 0.0235 - time: 6.49min\n",
      "Epoch [90/120], Loss: 0.0227 - time: 7.29min\n",
      "Epoch [100/120], Loss: 0.0209 - time: 8.10min\n",
      "Epoch [110/120], Loss: 0.0200 - time: 8.90min\n",
      "\n",
      "Training NN2 model:\n",
      "Epoch [1/75], Loss: 0.0829 - time: 0.01min\n",
      "Epoch [10/75], Loss: 0.0617 - time: 0.20min\n",
      "Epoch [20/75], Loss: 0.0520 - time: 0.39min\n",
      "Epoch [30/75], Loss: 0.0451 - time: 0.58min\n",
      "Epoch [40/75], Loss: 0.0379 - time: 0.77min\n",
      "Epoch [50/75], Loss: 0.0320 - time: 0.96min\n",
      "Epoch [60/75], Loss: 0.0287 - time: 1.15min\n",
      "Epoch [70/75], Loss: 0.0238 - time: 1.34min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN3 model:\n",
      "Epoch [1/100], Loss: 0.0812 - time: 0.02min\n",
      "Epoch [10/100], Loss: 0.0619 - time: 0.25min\n",
      "Epoch [20/100], Loss: 0.0555 - time: 0.48min\n",
      "Epoch [30/100], Loss: 0.0517 - time: 0.70min\n",
      "Epoch [40/100], Loss: 0.0482 - time: 0.93min\n",
      "Epoch [50/100], Loss: 0.0461 - time: 1.16min\n",
      "Epoch [60/100], Loss: 0.0447 - time: 1.38min\n",
      "Epoch [70/100], Loss: 0.0427 - time: 1.61min\n",
      "Epoch [80/100], Loss: 0.0411 - time: 1.83min\n",
      "Epoch [90/100], Loss: 0.0401 - time: 2.06min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN4 model:\n",
      "Epoch [1/75], Loss: 0.0874 - time: 0.00min\n",
      "Epoch [10/75], Loss: 0.0693 - time: 0.08min\n",
      "Epoch [20/75], Loss: 0.0640 - time: 0.17min\n",
      "Epoch [30/75], Loss: 0.0590 - time: 0.24min\n",
      "Epoch [40/75], Loss: 0.0546 - time: 0.32min\n",
      "Epoch [50/75], Loss: 0.0507 - time: 0.40min\n",
      "Epoch [60/75], Loss: 0.0472 - time: 0.48min\n",
      "Epoch [70/75], Loss: 0.0442 - time: 0.56min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN5 model:\n",
      "Epoch [1/75], Loss: 0.0810 - time: 0.01min\n",
      "Epoch [10/75], Loss: 0.0618 - time: 0.10min\n",
      "Epoch [20/75], Loss: 0.0545 - time: 0.19min\n",
      "Epoch [30/75], Loss: 0.0499 - time: 0.28min\n",
      "Epoch [40/75], Loss: 0.0440 - time: 0.37min\n",
      "Epoch [50/75], Loss: 0.0401 - time: 0.46min\n",
      "Epoch [60/75], Loss: 0.0375 - time: 0.55min\n",
      "Epoch [70/75], Loss: 0.0352 - time: 0.64min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN6 model:\n",
      "Epoch [1/100], Loss: 0.0811 - time: 0.01min\n",
      "Epoch [10/100], Loss: 0.0629 - time: 0.10min\n",
      "Epoch [20/100], Loss: 0.0576 - time: 0.19min\n",
      "Epoch [30/100], Loss: 0.0541 - time: 0.28min\n",
      "Epoch [40/100], Loss: 0.0513 - time: 0.37min\n",
      "Epoch [50/100], Loss: 0.0489 - time: 0.45min\n",
      "Epoch [60/100], Loss: 0.0477 - time: 0.54min\n",
      "Epoch [70/100], Loss: 0.0462 - time: 0.63min\n",
      "Epoch [80/100], Loss: 0.0452 - time: 0.72min\n",
      "Epoch [90/100], Loss: 0.0451 - time: 0.81min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN7 model:\n",
      "Epoch [1/75], Loss: 0.0820 - time: 0.01min\n",
      "Epoch [10/75], Loss: 0.0682 - time: 0.21min\n",
      "Epoch [20/75], Loss: 0.0582 - time: 0.41min\n",
      "Epoch [30/75], Loss: 0.0517 - time: 0.60min\n",
      "Epoch [40/75], Loss: 0.0474 - time: 0.80min\n",
      "Epoch [50/75], Loss: 0.0433 - time: 1.00min\n",
      "Epoch [60/75], Loss: 0.0400 - time: 1.19min\n",
      "Epoch [70/75], Loss: 0.0371 - time: 1.39min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN8 model:\n",
      "Epoch [1/75], Loss: 0.0811 - time: 0.02min\n",
      "Epoch [10/75], Loss: 0.0623 - time: 0.32min\n",
      "Epoch [20/75], Loss: 0.0540 - time: 0.61min\n",
      "Epoch [30/75], Loss: 0.0475 - time: 0.91min\n",
      "Epoch [40/75], Loss: 0.0426 - time: 1.20min\n",
      "Epoch [50/75], Loss: 0.0386 - time: 1.49min\n",
      "Epoch [60/75], Loss: 0.0357 - time: 1.78min\n",
      "Epoch [70/75], Loss: 0.0332 - time: 2.07min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN9 model:\n",
      "Epoch [1/75], Loss: 0.0811 - time: 0.02min\n",
      "Epoch [10/75], Loss: 0.0634 - time: 0.23min\n",
      "Epoch [20/75], Loss: 0.0549 - time: 0.43min\n",
      "Epoch [30/75], Loss: 0.0498 - time: 0.64min\n",
      "Epoch [40/75], Loss: 0.0451 - time: 0.85min\n",
      "Epoch [50/75], Loss: 0.0417 - time: 1.06min\n",
      "Epoch [60/75], Loss: 0.0406 - time: 1.27min\n",
      "Epoch [70/75], Loss: 0.0369 - time: 1.47min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training NN10 model:\n",
      "Epoch [1/75], Loss: 0.0829 - time: 0.02min\n",
      "Epoch [10/75], Loss: 0.0690 - time: 0.30min\n",
      "Epoch [20/75], Loss: 0.0620 - time: 0.58min\n",
      "Epoch [30/75], Loss: 0.0582 - time: 0.86min\n",
      "Epoch [40/75], Loss: 0.0551 - time: 1.14min\n",
      "Epoch [50/75], Loss: 0.0530 - time: 1.41min\n",
      "Epoch [60/75], Loss: 0.0504 - time: 1.69min\n",
      "Epoch [70/75], Loss: 0.0485 - time: 1.97min\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Vale\\miniconda3\\envs\\torch\\lib\\site-packages\\sklearn\\metrics\\_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>hidden_sizes</th>\n",
       "      <th>dropouts</th>\n",
       "      <th>lr</th>\n",
       "      <th>batch_size</th>\n",
       "      <th>epochs</th>\n",
       "      <th>macro_P</th>\n",
       "      <th>macro_R</th>\n",
       "      <th>macro_F1</th>\n",
       "      <th>weighted_P</th>\n",
       "      <th>weighted_R</th>\n",
       "      <th>weighted_F1</th>\n",
       "      <th>samples_P</th>\n",
       "      <th>samples_R</th>\n",
       "      <th>samples_F1</th>\n",
       "      <th>training_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NN1</td>\n",
       "      <td>[8192, 4096]</td>\n",
       "      <td>[0.5, 0.4]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>512.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>0.433</td>\n",
       "      <td>0.339</td>\n",
       "      <td>0.374</td>\n",
       "      <td>0.541</td>\n",
       "      <td>0.467</td>\n",
       "      <td>0.497</td>\n",
       "      <td>0.595</td>\n",
       "      <td>0.532</td>\n",
       "      <td>0.504</td>\n",
       "      <td>9.625752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NN2</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>2048.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.562</td>\n",
       "      <td>0.214</td>\n",
       "      <td>0.289</td>\n",
       "      <td>0.617</td>\n",
       "      <td>0.362</td>\n",
       "      <td>0.433</td>\n",
       "      <td>0.670</td>\n",
       "      <td>0.424</td>\n",
       "      <td>0.454</td>\n",
       "      <td>1.418368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NN3</td>\n",
       "      <td>[4096, 2048]</td>\n",
       "      <td>[0.4, 0.4]</td>\n",
       "      <td>0.005</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.588</td>\n",
       "      <td>0.225</td>\n",
       "      <td>0.298</td>\n",
       "      <td>0.631</td>\n",
       "      <td>0.386</td>\n",
       "      <td>0.445</td>\n",
       "      <td>0.650</td>\n",
       "      <td>0.450</td>\n",
       "      <td>0.474</td>\n",
       "      <td>2.263855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NN4</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.1, 0.1]</td>\n",
       "      <td>0.001</td>\n",
       "      <td>2048.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.434</td>\n",
       "      <td>0.090</td>\n",
       "      <td>0.130</td>\n",
       "      <td>0.599</td>\n",
       "      <td>0.269</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.709</td>\n",
       "      <td>0.330</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.594609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NN5</td>\n",
       "      <td>[2048, 1024]</td>\n",
       "      <td>[0.2, 0.2]</td>\n",
       "      <td>0.003</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.523</td>\n",
       "      <td>0.226</td>\n",
       "      <td>0.294</td>\n",
       "      <td>0.578</td>\n",
       "      <td>0.390</td>\n",
       "      <td>0.444</td>\n",
       "      <td>0.628</td>\n",
       "      <td>0.451</td>\n",
       "      <td>0.464</td>\n",
       "      <td>0.674865</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name  hidden_sizes    dropouts     lr  batch_size  epochs  macro_P  macro_R  \\\n",
       "0  NN1  [8192, 4096]  [0.5, 0.4]  0.001       512.0   120.0    0.433    0.339   \n",
       "1  NN2  [4096, 2048]  [0.2, 0.2]  0.003      2048.0    75.0    0.562    0.214   \n",
       "2  NN3  [4096, 2048]  [0.4, 0.4]  0.005      1024.0   100.0    0.588    0.225   \n",
       "3  NN4  [2048, 1024]  [0.1, 0.1]  0.001      2048.0    75.0    0.434    0.090   \n",
       "4  NN5  [2048, 1024]  [0.2, 0.2]  0.003      1024.0    75.0    0.523    0.226   \n",
       "\n",
       "   macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  samples_R  \\\n",
       "0     0.374       0.541       0.467        0.497      0.595      0.532   \n",
       "1     0.289       0.617       0.362        0.433      0.670      0.424   \n",
       "2     0.298       0.631       0.386        0.445      0.650      0.450   \n",
       "3     0.130       0.599       0.269        0.326      0.709      0.330   \n",
       "4     0.294       0.578       0.390        0.444      0.628      0.451   \n",
       "\n",
       "   samples_F1  training_time  \n",
       "0       0.504       9.625752  \n",
       "1       0.454       1.418368  \n",
       "2       0.474       2.263855  \n",
       "3       0.392       0.594609  \n",
       "4       0.464       0.674865  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_names = [str(cls) for cls in mlb.classes_]\n",
    "\n",
    "report_path = datasets_path / 'BP_report.csv'\n",
    "probs_path = datasets_path / 'BP_report.csv'\n",
    "\n",
    "\n",
    "if report_path.exists() and probs_path.exists():\n",
    "    report = pd.read_csv(datasets_path / 'BP_report.csv')\n",
    "    probs = pd.read_csv(datasets_path / 'BP_probs.csv')\n",
    "else:\n",
    "    report = {}\n",
    "    probs = {}\n",
    "\n",
    "for name in NN_BP['name']:\n",
    "\n",
    "    # Define model parameters\n",
    "    input_size = X_train_tensor.shape[1] \n",
    "    hidden_sizes = NN_BP[NN_BP['name'] == name]['hidden_sizes'].values[0]\n",
    "    dropouts = NN_BP[NN_BP['name'] == name]['dropouts'].values[0]\n",
    "    output_size = y_train_tensor.shape[1]\n",
    "    lr = float(NN_BP[NN_BP['name'] == name]['lr'].values[0])\n",
    "    batch_size = int(NN_BP[NN_BP['name'] == name]['batch_size'].values[0])\n",
    "    epochs = int(NN_BP[NN_BP['name'] == name]['epochs'].values[0])\n",
    "\n",
    "    # Initialize the model\n",
    "    model = NN(\n",
    "        input_size = input_size,\n",
    "        hidden_sizes = hidden_sizes,\n",
    "        dropouts = dropouts, \n",
    "        output_size = output_size\n",
    "        ).to(device)\n",
    "\n",
    "    # Loss function and optimizer\n",
    "    criterion = nn.BCELoss()  # Binary Cross-Entropy Loss for multilabel\n",
    "    optimizer = optim.Adam(model.parameters(), lr=lr)\n",
    "\n",
    "    \n",
    "\n",
    "    # model, optimizer, criterion = create_model(NN_BP, name, X_train_tensor.shape[1] , y_train_tensor.shape[1])\n",
    "    if np.isnan(NN_BP[NN_BP['name'] == name]['training_time'].values[0]):\n",
    "        # Train the model\n",
    "        training_time = train_model(model, name, X_train_tensor, y_train_tensor, optimizer, criterion, batch_size, epochs)\n",
    "\n",
    "        # Ensure X_test_tensor is on the same device as the model\n",
    "        device = next(model.parameters()).device\n",
    "        X_test_tensor = X_test_tensor.to(device)\n",
    "\n",
    "        # Evaluation\n",
    "        \n",
    "        NN_BP, report[name], probs[name] = model_evaluation(NN_BP, name, model, X_test_tensor, y_test_tensor, target_names, training_time)\n",
    "\n",
    "    # Save the DataFrame to a CSV file\n",
    "NN_BP.to_csv(datasets_path / 'NN_BP.csv', index=False)\n",
    "# Move tensors to CPU and convert to NumPy\n",
    "\n",
    "def to_cpu_numpy(data):\n",
    "    if isinstance(data, torch.Tensor):\n",
    "        return data.detach().cpu().numpy()  # Move to CPU & convert to NumPy\n",
    "    return data  # If it's not a tensor, return as is\n",
    "\n",
    "# Ensure report and probs are properly formatted\n",
    "report = pd.DataFrame({k: [to_cpu_numpy(v)] for k, v in report.items()})\n",
    "probs = pd.DataFrame({k: [to_cpu_numpy(v)] for k, v in probs.items()})\n",
    "\n",
    "\n",
    "\n",
    "        \n",
    "NN_BP.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "⚠️ Warning: report is empty! No models were trained.\n",
      "⚠️ Warning: probs is empty!\n"
     ]
    }
   ],
   "source": [
    "# Move tensors to CPU and convert to NumPy\n",
    "def to_cpu_numpy(data):\n",
    "    if isinstance(data, torch.Tensor):\n",
    "        return data.detach().cpu().numpy()  # Move to CPU & convert to NumPy\n",
    "    return data  # If it's not a tensor, return as is\n",
    "\n",
    "# Ensure report and probs are properly formatted\n",
    "if report is not None and isinstance(report, dict) and len(report) > 0:\n",
    "    report = pd.DataFrame({k: [to_cpu_numpy(v)] for k, v in report.items()})\n",
    "else:\n",
    "    print(\"⚠️ Warning: report is empty! No models were trained.\")\n",
    "    report = pd.DataFrame()\n",
    "\n",
    "if probs is not None and isinstance(probs, dict) and len(probs) > 0:\n",
    "    probs = pd.DataFrame({k: [to_cpu_numpy(v)] for k, v in probs.items()})\n",
    "else:\n",
    "    print(\"⚠️ Warning: probs is empty!\")\n",
    "    probs = pd.DataFrame()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Extracting results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(   name        hidden_sizes         dropouts     lr  batch_size  epochs  \\\n",
       " 5   NN6        [2048, 1024]       [0.4, 0.4]  0.005      1024.0   100.0   \n",
       " 2   NN3        [4096, 2048]       [0.4, 0.4]  0.005      1024.0   100.0   \n",
       " 9  NN10  [2048, 4096, 2048]  [0.3, 0.3, 0.3]  0.005      2048.0    75.0   \n",
       " 7   NN8  [4096, 2048, 2048]  [0.2, 0.2, 0.2]  0.003      1024.0    75.0   \n",
       " 1   NN2        [4096, 2048]       [0.2, 0.2]  0.003      2048.0    75.0   \n",
       " \n",
       "    training_time  \n",
       " 5       0.889524  \n",
       " 2       2.263855  \n",
       " 9       2.082515  \n",
       " 7       2.190863  \n",
       " 1       1.418368  ,\n",
       "    macro_P  macro_R  macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  \\\n",
       " 5    0.603    0.183     0.249       0.635       0.362        0.418      0.658   \n",
       " 2    0.588    0.225     0.298       0.631       0.386        0.445      0.650   \n",
       " 9    0.581    0.104     0.151       0.665       0.275        0.336      0.713   \n",
       " 7    0.566    0.214     0.288       0.620       0.363        0.433      0.665   \n",
       " 1    0.562    0.214     0.289       0.617       0.362        0.433      0.670   \n",
       " \n",
       "    samples_R  samples_F1  \n",
       " 5      0.426       0.458  \n",
       " 2      0.450       0.474  \n",
       " 9      0.342       0.400  \n",
       " 7      0.431       0.460  \n",
       " 1      0.424       0.454  )"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metrics = ['macro_P', 'macro_R', 'macro_F1', \n",
    "           'weighted_P', 'weighted_R', 'weighted_F1', \n",
    "           'samples_P', 'samples_R', 'samples_F1']\n",
    "\n",
    "\n",
    "\n",
    "BP_results = NN_BP.sort_values(by=metrics, ascending=False)[['macro_P', 'macro_R', 'macro_F1',\n",
    "                                                             'weighted_P', 'weighted_R', 'weighted_F1', \n",
    "                                                             'samples_P', 'samples_R', 'samples_F1']][:5]\n",
    "\n",
    "BP_params = NN_BP.sort_values(by=metrics, ascending=False)[['name', 'hidden_sizes', 'dropouts', 'lr', \n",
    "                                                'batch_size', 'epochs', 'training_time']][:5]\n",
    "\n",
    "BP_params.head(), BP_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(   macro_P  macro_R  macro_F1  weighted_P  weighted_R  weighted_F1  samples_P  \\\n",
       " 5    0.603    0.183     0.249       0.635       0.362        0.418      0.658   \n",
       " 2    0.588    0.225     0.298       0.631       0.386        0.445      0.650   \n",
       " 9    0.581    0.104     0.151       0.665       0.275        0.336      0.713   \n",
       " 7    0.566    0.214     0.288       0.620       0.363        0.433      0.665   \n",
       " 1    0.562    0.214     0.289       0.617       0.362        0.433      0.670   \n",
       " \n",
       "    samples_R  samples_F1 name  \n",
       " 5      0.426       0.458  NN1  \n",
       " 2      0.450       0.474  NN2  \n",
       " 9      0.342       0.400  NN3  \n",
       " 7      0.431       0.460  NN4  \n",
       " 1      0.424       0.454  NN5  ,\n",
       "   name        hidden_sizes         dropouts     lr  batch_size  epochs  \\\n",
       " 5  NN1        [2048, 1024]       [0.4, 0.4]  0.005      1024.0   100.0   \n",
       " 2  NN2        [4096, 2048]       [0.4, 0.4]  0.005      1024.0   100.0   \n",
       " 9  NN3  [2048, 4096, 2048]  [0.3, 0.3, 0.3]  0.005      2048.0    75.0   \n",
       " 7  NN4  [4096, 2048, 2048]  [0.2, 0.2, 0.2]  0.003      1024.0    75.0   \n",
       " 1  NN5        [4096, 2048]       [0.2, 0.2]  0.003      2048.0    75.0   \n",
       " \n",
       "    training_time  \n",
       " 5       0.889524  \n",
       " 2       2.263855  \n",
       " 9       2.082515  \n",
       " 7       2.190863  \n",
       " 1       1.418368  )"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_names = [f'NN{i}' for i in range(1, BP_results.shape[0] + 1)]\n",
    "\n",
    "# Add the row means to the DataFrame\n",
    "BP_results['name'] = row_names\n",
    "BP_params['name'] = row_names\n",
    "\n",
    "# Save the DataFrame to a CSV file\n",
    "BP_results.to_csv(datasets_path / 'BP_results.csv', index=False)\n",
    "BP_params.to_csv(datasets_path / 'BP_params.csv', index=False)\n",
    "\n",
    "# Print the updated DataFrame\n",
    "BP_results.head(), BP_params.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### LaTex tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\\begin{tabular}{lccccc}\n",
      "    \\toprule\n",
      "    \\textit{Metric} & \\textit{NN1} & \\textit{NN2} & \\textit{NN3} & \\textit{NN4} & \\textit{NN5} \\\\\n",
      "    \\midrule\n",
      "    Macro P & 0.60 & 0.59 & 0.58 & 0.57 & 0.56 \\\\\n",
      "    Macro R & 0.18 & 0.23 & 0.10 & 0.21 & 0.21 \\\\\n",
      "    Macro F1 & 0.25 & 0.30 & 0.15 & 0.29 & 0.29 \\\\\n",
      "    Weighted P & 0.64 & 0.63 & 0.67 & 0.62 & 0.62 \\\\\n",
      "    Weighted R & 0.36 & 0.39 & 0.28 & 0.36 & 0.36 \\\\\n",
      "    Weighted F1 & 0.42 & 0.45 & 0.34 & 0.43 & 0.43 \\\\\n",
      "    Samples P & 0.66 & 0.65 & 0.71 & 0.67 & 0.67 \\\\\n",
      "    Samples R & 0.43 & 0.45 & 0.34 & 0.43 & 0.42 \\\\\n",
      "    Samples F1 & 0.46 & 0.47 & 0.40 & 0.46 & 0.45 \\\\\n",
      "    \\bottomrule\n",
      "\\end{tabular}\n",
      "\\begin{table}[H]\n",
      "    \\centering\n",
      "    \\begin{tabular}{lcccccc}\n",
      "        \\toprule\n",
      "        \\textit{Model} & \\textit{Hidden Layer Sizes} & \\textit{Dropout} & \\textit{Learning Rate} & \\textit{Batch Size} & \\textit{Epochs} & \\textit{Regularization} \\\\\n",
      "        \\midrule\n",
      "        NN1 & [2048, 1024] & [0.4, 0.4] & 0.005 & 1024.0 & 100.0 & None \\\\\n",
      "        NN2 & [4096, 2048] & [0.4, 0.4] & 0.005 & 1024.0 & 100.0 & None \\\\\n",
      "        NN3 & [2048, 4096, 2048] & [0.3, 0.3, 0.3] & 0.005 & 2048.0 & 75.0 & None \\\\\n",
      "        NN4 & [4096, 2048, 2048] & [0.2, 0.2, 0.2] & 0.003 & 1024.0 & 75.0 & None \\\\\n",
      "        NN5 & [4096, 2048] & [0.2, 0.2] & 0.003 & 2048.0 & 75.0 & None \\\\\n",
      "        \\bottomrule\n",
      "    \\end{tabular}\n",
      "    \\caption{Model Parameters.}\n",
      "\\end{table}\n"
     ]
    }
   ],
   "source": [
    "def generate_latex_table(df):\n",
    "    metrics = ['macro_P', 'macro_R', 'macro_F1', 'weighted_P', 'weighted_R', 'weighted_F1', 'samples_P', 'samples_R', 'samples_F1']\n",
    "    metric_names = ['Macro P', 'Macro R', 'Macro F1', 'Weighted P', 'Weighted R', 'Weighted F1', 'Samples P', 'Samples R', 'Samples F1']\n",
    "    \n",
    "    # Initialize the LaTeX table\n",
    "    latex_table = \"\\\\begin{tabular}{l\" + \"c\" * len(df['name']) + \"}\\n\"\n",
    "    latex_table += \"    \\\\toprule\\n\"\n",
    "    latex_table += \"    \\\\textit{Metric} & \" + \" & \".join([f\"\\\\textit{{{name}}}\" for name in df['name']]) + \" \\\\\\\\\\n\"\n",
    "    latex_table += \"    \\\\midrule\\n\"\n",
    "    \n",
    "    for metric, metric_name in zip(metrics, metric_names):\n",
    "        values = df[metric]\n",
    "        formatted_values = [f\"{value:.2f}\" for value in values]\n",
    "        latex_table += f\"    {metric_name} & \" + \" & \".join(formatted_values) + \" \\\\\\\\\\n\"\n",
    "    \n",
    "    latex_table += \"    \\\\bottomrule\\n\"\n",
    "    latex_table += \"\\\\end{tabular}\"\n",
    "    \n",
    "    return latex_table\n",
    "\n",
    "\n",
    "def generate_parameters_latex_table(df):\n",
    "    # Initialize the LaTeX table\n",
    "    latex_table = \"\\\\begin{table}[H]\\n\"\n",
    "    latex_table += \"    \\\\centering\\n\"\n",
    "    latex_table += \"    \\\\begin{tabular}{lcccccc}\\n\"\n",
    "    latex_table += \"        \\\\toprule\\n\"\n",
    "    latex_table += \"        \\\\textit{Model} & \\\\textit{Hidden Layer Sizes} & \\\\textit{Dropout} & \\\\textit{Learning Rate} & \\\\textit{Batch Size} & \\\\textit{Epochs} & \\\\textit{Regularization} \\\\\\\\\\n\"\n",
    "    latex_table += \"        \\\\midrule\\n\"\n",
    "    \n",
    "    for index, row in df.iterrows():\n",
    "        model_name = row['name']\n",
    "        hidden_sizes = row['hidden_sizes']\n",
    "        dropouts = row['dropouts']\n",
    "        lr = row['lr']\n",
    "        batch_size = row['batch_size']\n",
    "        epochs = row['epochs']\n",
    "        regularization = 'None'  # Assuming regularization is not provided in the DataFrame\n",
    "        \n",
    "        latex_table += f\"        {model_name} & {hidden_sizes} & {dropouts} & {lr} & {batch_size} & {epochs} & {regularization} \\\\\\\\\\n\"\n",
    "    \n",
    "    latex_table += \"        \\\\bottomrule\\n\"\n",
    "    latex_table += \"    \\\\end{tabular}\\n\"\n",
    "    latex_table += \"    \\\\caption{Model Parameters.}\\n\"\n",
    "    latex_table += \"\\\\end{table}\"\n",
    "    \n",
    "    return latex_table\n",
    "\n",
    "# Generate the LaTeX table\n",
    "latex_table = generate_latex_table(BP_results)\n",
    "print(latex_table)\n",
    "\n",
    "latex_table = generate_parameters_latex_table(BP_params)\n",
    "print(latex_table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (TORCH)",
   "language": "python",
   "name": "torch"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
